<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>MGDrivE2: Simulation of Time-inhomogeneous Stochastic Processes (Seasonality) • MGDrivE2</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="MGDrivE2: Simulation of Time-inhomogeneous Stochastic Processes (Seasonality)">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MGDrivE2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/advanced_topics.html">MGDrivE2: Advanced Topics</a>
    </li>
    <li>
      <a href="../articles/epi-SEIR.html">MGDrivE2: SEIR Epidemiological Dynamics</a>
    </li>
    <li>
      <a href="../articles/epi-network.html">MGDrivE2: Metapopulation Network Epidemiological Dynamics</a>
    </li>
    <li>
      <a href="../articles/epi-node.html">MGDrivE2: One Node Epidemiological Dynamics</a>
    </li>
    <li>
      <a href="../articles/inhomogeneous.html">MGDrivE2: Simulation of Time-inhomogeneous Stochastic Processes (Seasonality)</a>
    </li>
    <li>
      <a href="../articles/lifecycle-network.html">MGDrivE2: Metapopulation Network Lifecycle Dynamics</a>
    </li>
    <li>
      <a href="../articles/lifecycle-node.html">MGDrivE2: One Node Lifecycle Dynamics</a>
    </li>
    <li>
      <a href="../articles/output-storage.html">MGDrivE2: Data Storage and Analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><script src="inhomogeneous_files/jquery-1.11.3/jquery.min.js"></script><script src="inhomogeneous_files/elevate-section-attrs-2.0/elevate-section-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>MGDrivE2: Simulation of Time-inhomogeneous Stochastic Processes (Seasonality)</h1>
            
      
      
      <div class="hidden name"><code>inhomogeneous.Rmd</code></div>

    </div>

    
    
<div id="table-of-contents" class="section level2">
<h2 class="hasAnchor">
<a href="#table-of-contents" class="anchor"></a>Table of Contents</h2>
<ol style="list-style-type: decimal">
<li><a href="#inhom_adult_mort">Time-Inhomogeneous Adult Mortality</a></li>
<li>
<a href="#inhom_rate_fns">Time-Dependent Rate Functions</a>
<ol style="list-style-type: decimal">
<li><a href="#math_consider">Mathematical Considerations</a></li>
<li><a href="#programming">Programming Time-Dependent Rates in MGDrivE2</a></li>
</ol>
</li>
<li>
<a href="#make_spn">Petri Net Setup</a>
<ol style="list-style-type: decimal">
<li><a href="#pars">Parameterization</a></li>
<li><a href="#init_pn">Initialization of the Petri Net</a></li>
<li><a href="#equilibria_haz">Equilibrium Conditions and Hazard Functions</a></li>
</ol>
</li>
<li>
<a href="#sim">Simulation of Fully Specified SPN Model</a>
<ol style="list-style-type: decimal">
<li><a href="#sim_ode">Simulation: ODE Models</a></li>
<li><a href="#sim_ctmc">Simulation: CTMC Models</a></li>
</ol>
</li>
<li><a href="#plots">Results</a></li>
<li><a href="#ref">References</a></li>
</ol>
<div id="preface" class="section level3">
<h3 class="hasAnchor">
<a href="#preface" class="anchor"></a>Preface</h3>
<p>The Stochastic Petri net (SPN) framework used in <strong>MGDrivE2</strong> allows for different functional forms of the hazard functions (or rates, if the model is being interpreted as a set of ordinary differential equations, ODEs. We use the terms hazard/rate function interchangeably). This means that time-dependent/inhomogeneous rates are relatively simple to incorporate into the model, without even changing the underlying Petri Net. We remark here that <em>exact</em> simulation of time-inhomogeneous stochastic processes is difficult, and the sampling algorithms to support such simulation have not yet been incorporated into <strong>MGDrivE2</strong> (ODE simulation will, of course, be “exact”). However, the Poisson time-step sampler provided with <strong>MGDrivE2</strong> will be correct as the stochastic time-step approaches zero, and we recommend its use for simulating inhomogeneous, continuous-time Markov chain models. Exact algorithms for simulating time-inhomogeneous processes can be found in the works of <a href="https://aip.scitation.org/doi/full/10.1063/1.2799998">Anderson</a> and <a href="https://aip.scitation.org/doi/abs/10.1063/1.4927916">Thanh</a>. Future development efforts will concentrate on better exact and approximate sampling algorithms.</p>
<p>We start by loading the <strong>MGDrivE2</strong> package, as well as the <strong>MGDrivE</strong> package for access to inheritance cubes, <strong>ggplot2</strong> for graphical analysis, and <strong>parallel</strong> for running stochastic repetitions in parallel.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true"></a><span class="co"># simulation functions</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(MGDrivE2)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true"></a><span class="co">#&gt; Loading MGDrivE2: Mosquito Gene Drive Explorer Version 2</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true"></a><span class="co"># inheritance patterns</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(MGDrivE)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true"></a><span class="co">#&gt; Loading MGDrivE: Mosquito Gene Drive Explorer</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true"></a><span class="co"># plotting</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(ggplot2)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true"></a><span class="co"># parallel sims</span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(parallel)</span></code></pre></div>
</div>
</div>
<div id="inhom_adult_mort" class="section level1">
<h1 class="hasAnchor">
<a href="#inhom_adult_mort" class="anchor"></a>Time-Inhomogeneous Adult Mortality</h1>
<p>Let us consider the case where we want to simulate a model in which the transitions made by functions <code>make_transition_male_mort()</code> and <code>make_transition_female_mort()</code>, that is, the adult male and female mosquito mortality, fire at a rate which depends on time. This time-dependence can in general be arbitrarily complex, as long as the rate function returns a non-negative number at any positive time <span class="math inline">\(t&gt;0\)</span> when it is queried. Therefore various time-inhomogeneous effects can be taken into account in <strong>MGDrivE2</strong>, from long-term climactic and seasonal changes, to rapidly oscillating changes from diurnal cycles.</p>
<p>Sample data included in <strong>MGDrivE2</strong> (<code>mu_ts</code>) provides estimated time-dependent adult mortality rates from the Comoros Islands, between Mozambique and Madagascar. These are hourly death rates over the course of an entire year, for Grande Comore, Moheli, and Anjouan.</p>
<p>First, we load the <code>mu_ts</code> object, which is a matrix of hourly death rates. Now we want to interpolate the death rates into a function that can return a death rate at any time in the interval <span class="math inline">\(0&lt;t&lt;365\)</span>. We could do this with a variety of methods, including splines and other smoothing algorithms, but for now we will use a simple step function with steps every hour, created with the function <code><a href="https://www.rdocumentation.org/packages/stats/topics/stepfun">stepfun()</a></code>.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true"></a><span class="co"># load mortality data</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true"></a><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/data">data</a></span>(mu_ts)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true"></a><span class="co"># vector of hourly timesteps</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true"></a><span class="co">#  unit is in days, so 24 steps per day, 365 days</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true"></a>tt_ts &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">nrow</a></span>(mu_ts)<span class="op">/</span><span class="dv">24</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true"></a></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true"></a><span class="co"># exaggerated amplitude of Grande Comore mortality rate, for visual effect</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true"></a>mu_ts_par &lt;-<span class="st"> </span>mu_ts[ ,<span class="dv">1</span>] <span class="op">+</span><span class="st"> </span>( (mu_ts[ ,<span class="dv">1</span>] <span class="op">-</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/mean">mean</a></span>(mu_ts[ ,<span class="dv">1</span>]) ) <span class="op">*</span><span class="st"> </span><span class="dv">50</span>)</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true"></a></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true"></a><span class="co"># approximating step function</span></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true"></a>step_approx &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/stepfun">stepfun</a></span>(<span class="dt">x =</span> tt_ts, <span class="dt">y =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(mu_ts_par[<span class="dv">1</span>], mu_ts_par),</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true"></a>                       <span class="dt">f =</span> <span class="dv">0</span>, <span class="dt">right =</span> <span class="ot">FALSE</span>)</span></code></pre></div>
<p>For clarity, we plot the approximating function evaluated at each hour over the year:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true"></a><span class="co"># long-form for ggplot2</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true"></a>df_ts &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/data.frame">data.frame</a></span>(<span class="st">"time"</span> =<span class="st"> </span>tt_ts, <span class="st">"mu"</span> =<span class="st"> </span><span class="kw">step_approx</span>(<span class="dt">v =</span> tt_ts))</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true"></a></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true"></a><span class="co"># plot</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true"></a><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="dt">data =</span> df_ts) <span class="op">+</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true"></a><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span>(<span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> time, <span class="dt">y =</span> mu), <span class="dt">alpha =</span> <span class="fl">0.75</span>, <span class="dt">color =</span> <span class="st">"dodgerblue4"</span>) <span class="op">+</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true"></a><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span>(<span class="st">"Time (days)"</span>) <span class="op">+</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true"></a><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span>(<span class="st">"Adult Mortality"</span>) <span class="op">+</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true"></a><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() <span class="op">+</span></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true"></a><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span>(<span class="st">"Grande Comore: Hourly Death Rates for One Year"</span>)</span></code></pre></div>
<p><img src="inhomogeneous_files/figure-html/unnamed-chunk-4-1.png" width="691.2"></p>
<p>We see that the data trends in a cyclic manner, consistent with proposed seasonality throughout the year. Mosquito mortality is highest in the early spring (days 50-150, approximately March to the end of May) and around the end of the year (days 325-365, the end of November through December).</p>
</div>
<div id="inhom_rate_fns" class="section level1">
<h1 class="hasAnchor">
<a href="#inhom_rate_fns" class="anchor"></a>Time-dependent Rate Functions</h1>
<p>We need to setup rate functions which correctly return time-dependent rates. First, we make a brief digression into the mathematical consequences and then we show how to implement them into <strong>MGDrivE2</strong></p>
<div id="math_consider" class="section level2">
<h2 class="hasAnchor">
<a href="#math_consider" class="anchor"></a>Mathematical Considerations</h2>
<p>Call <span class="math inline">\(\mu\)</span> a rate function that is time-homogeneous, meaning <span class="math inline">\(\mu(t) = \mu, t \in [0,\infty)\)</span>; for a continuous-time discrete event model, this means that the time between events is exponentially distributed. Another way of saying this is that the counting process (<span class="math inline">\(N(t)\)</span>) that counts how many times the event has “fired” (occurred) by time <span class="math inline">\(t\)</span> is a Poisson process, that is the number of events at time <span class="math inline">\(t\)</span> follows a Poisson distribution with parameter <span class="math inline">\(\mu \cdot t\)</span>. Consequently, if all events (transitions) have rates that are time-homogeneous functions then the entire model is a continuous-time Markov chain (CTMC). One reason we like CTMC models is that they incorporate stochasticity in a “mechanistic” way and have deep connections to classic ODE models, a mainstay of mathematical modelling. For a recent review of the relationship, using epidemic models as an example, see work by <a href="https://www.sciencedirect.com/science/article/pii/S2468042716300495">Linda Allen</a>.</p>
<p>Now, let the rate function depend on time, meaning <span class="math inline">\(\mu(t) \geq 0, t \in [0,\infty)\)</span>. We need to be careful now: while the rate function depends on <em>time</em>, it does not depend on <em>age</em>; that is, if the event is going to the grocery store, perhaps Alice goes to the grocery store at a different rate each day of the week, but if the rate does not depend <em>at all</em> on how long she has been waiting to go to the store (and has not yet), then we have a process that is time-dependent but not age-dependent. In other words, it is still a Markov process, but it is a time-inhomogeneous Markov process <em>even though</em> the inter-event times no longer follow an exponential distribution. In <strong>MGDrivE2</strong> we approximate an age-dependent (semi-Markov process) rate of egg/larvae/pupae development by using an Erlang distribution, which is constructed by summing exponential waiting times. If the rate parameter of the Erlang distribution varies with time, then we can simulate time-inhomogeneous semi-Markov processes.</p>
</div>
<div id="programming" class="section level2">
<h2 class="hasAnchor">
<a href="#programming" class="anchor"></a>Programming Time-Dependent Rates in MGDrivE2</h2>
<p>To make adult mortality rates time-dependent, we will need to replace the functions <code>make_male_mort_haz()</code> and <code>make_female_mort_haz()</code> with versions that return time-dependent rate functions. First, we observe the time-inhomogeneous rate function associated with female mortality (the function for male mortality is very similar):</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true"></a>make_female_mort_haz &lt;-<span class="st"> </span><span class="cf">function</span>(trans, u, cube, params, <span class="dt">exact =</span> <span class="ot">TRUE</span>, <span class="dt">tol =</span> <span class="fl">1e-8</span>){</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true"></a></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true"></a>  <span class="co"># rate constants</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true"></a>  muF &lt;-<span class="st"> </span>params<span class="op">$</span>muF</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true"></a></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true"></a>  <span class="co"># which places have input arcs to this transition</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true"></a>  s &lt;-<span class="st"> </span>trans<span class="op">$</span>s</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true"></a></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true"></a>  <span class="co"># weights of those arcs</span></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true"></a>  w &lt;-<span class="st"> </span>trans<span class="op">$</span>s_w</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true"></a></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true"></a>  <span class="co"># omega is dependent on genotype</span></span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true"></a>  f_gen &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/strsplit">strsplit</a></span>(<span class="dt">x =</span> u[s], <span class="dt">split =</span> <span class="st">"_"</span>, <span class="dt">fixed =</span> <span class="ot">TRUE</span>)[[<span class="dv">1</span>]][<span class="dv">2</span>]</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true"></a>  omega &lt;-<span class="st"> </span>cube<span class="op">$</span>omega[f_gen]</span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true"></a></span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true"></a>  <span class="co"># return the hazard function</span></span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true"></a>  <span class="cf">if</span>(exact){</span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true"></a></span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true"></a>    <span class="co"># EXACT hazards (check enabling degree: for discrete simulation only)</span></span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true"></a>    <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/function">return</a></span>(</span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true"></a>      <span class="cf">function</span>(trans,M){</span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true"></a>        <span class="cf">if</span>(w <span class="op">&lt;=</span><span class="st"> </span>M[s]){</span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true"></a>          <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/function">return</a></span>(muF <span class="op">*</span><span class="st"> </span>omega <span class="op">*</span><span class="st"> </span>M[s])</span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true"></a>        } <span class="cf">else</span> {</span>
<span id="cb4-25"><a href="#cb4-25" aria-hidden="true"></a>          <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/function">return</a></span>(<span class="dv">0</span>)</span>
<span id="cb4-26"><a href="#cb4-26" aria-hidden="true"></a>        }</span>
<span id="cb4-27"><a href="#cb4-27" aria-hidden="true"></a>      }</span>
<span id="cb4-28"><a href="#cb4-28" aria-hidden="true"></a>    )</span>
<span id="cb4-29"><a href="#cb4-29" aria-hidden="true"></a></span>
<span id="cb4-30"><a href="#cb4-30" aria-hidden="true"></a>  } <span class="cf">else</span> {</span>
<span id="cb4-31"><a href="#cb4-31" aria-hidden="true"></a></span>
<span id="cb4-32"><a href="#cb4-32" aria-hidden="true"></a>    <span class="co"># APPROXIMATE hazards (tolerance around zero; for continuous approximation only)</span></span>
<span id="cb4-33"><a href="#cb4-33" aria-hidden="true"></a>    <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/function">return</a></span>(</span>
<span id="cb4-34"><a href="#cb4-34" aria-hidden="true"></a>      <span class="cf">function</span>(trans,M){</span>
<span id="cb4-35"><a href="#cb4-35" aria-hidden="true"></a>        haz &lt;-<span class="st"> </span>muF <span class="op">*</span><span class="st"> </span>omega <span class="op">*</span><span class="st"> </span>M[s]</span>
<span id="cb4-36"><a href="#cb4-36" aria-hidden="true"></a>        <span class="cf">if</span>(haz <span class="op">&lt;</span><span class="st"> </span>tol){</span>
<span id="cb4-37"><a href="#cb4-37" aria-hidden="true"></a>          <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/function">return</a></span>(<span class="dv">0</span>)</span>
<span id="cb4-38"><a href="#cb4-38" aria-hidden="true"></a>        } <span class="cf">else</span> {</span>
<span id="cb4-39"><a href="#cb4-39" aria-hidden="true"></a>          <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/function">return</a></span>(haz)</span>
<span id="cb4-40"><a href="#cb4-40" aria-hidden="true"></a>        }</span>
<span id="cb4-41"><a href="#cb4-41" aria-hidden="true"></a>      }</span>
<span id="cb4-42"><a href="#cb4-42" aria-hidden="true"></a>    )</span>
<span id="cb4-43"><a href="#cb4-43" aria-hidden="true"></a></span>
<span id="cb4-44"><a href="#cb4-44" aria-hidden="true"></a>  }</span>
<span id="cb4-45"><a href="#cb4-45" aria-hidden="true"></a>  <span class="co"># end of function</span></span>
<span id="cb4-46"><a href="#cb4-46" aria-hidden="true"></a>}</span></code></pre></div>
<p>The rate of the adult female mosquito mortality event firing is <span class="math inline">\(\mu_{F} \cdot \omega \cdot M[s]\)</span>, where <span class="math inline">\(M[s]\)</span> is the number of adult female tokens. We want to make it time dependent via <span class="math inline">\(\mu_{F}(t) \cdot \omega \cdot M[s]\)</span>, where <span class="math inline">\(\mu(t)\)</span> comes from our interpolating function made with <code><a href="https://www.rdocumentation.org/packages/stats/topics/stepfun">stepfun()</a></code>. To do so, we write a new function, <code>make_female_mort_haz_inhom()</code>. Note that we preform type checking of the object <code>muF</code> to make sure it is indeed a function, and that the returned function closure evaluates it at the input time <code>muF(t)</code>.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true"></a>make_female_mort_haz_inhom &lt;-<span class="st"> </span><span class="cf">function</span>(trans, u, cube, params,</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true"></a>                                       <span class="dt">exact =</span> <span class="ot">TRUE</span>, <span class="dt">tol =</span> <span class="fl">1e-8</span>){</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true"></a></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true"></a>  <span class="co"># mortality is a time-dependent hazard</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true"></a>  muF &lt;-<span class="st"> </span>params<span class="op">$</span>muF</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true"></a>  <span class="cf">if</span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/typeof">typeof</a></span>(muF) <span class="op">!=</span><span class="st"> "closure"</span>){</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true"></a>    <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/stop">stop</a></span>(<span class="st">"Inhomogeneous hazard 'make_female_mort_haz_inhom', "</span>,</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true"></a>         <span class="st">"'muF' in 'params' list needs to be a function"</span>)</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true"></a>  }</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true"></a></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true"></a>  <span class="co"># which places have input arcs to this transition</span></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true"></a>  s &lt;-<span class="st"> </span>trans<span class="op">$</span>s</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true"></a></span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true"></a>  <span class="co"># weights of those arcs</span></span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true"></a>  w &lt;-<span class="st"> </span>trans<span class="op">$</span>s_w</span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true"></a></span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true"></a>  <span class="co"># omega is dependent on genotype</span></span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true"></a>  f_gen &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/strsplit">strsplit</a></span>(<span class="dt">x =</span> u[s], <span class="dt">split =</span> <span class="st">"_"</span>, <span class="dt">fixed =</span> <span class="ot">TRUE</span>)[[<span class="dv">1</span>]][<span class="dv">2</span>]</span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true"></a>  omega &lt;-<span class="st"> </span>cube<span class="op">$</span>omega[f_gen]</span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true"></a></span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true"></a>  <span class="co"># return the hazard function</span></span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true"></a>  <span class="cf">if</span>(exact){</span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true"></a></span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true"></a>    <span class="co"># EXACT hazards (check enabling degree: for discrete simulation only)</span></span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true"></a>    <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/function">return</a></span>(</span>
<span id="cb5-26"><a href="#cb5-26" aria-hidden="true"></a>      <span class="cf">function</span>(t,M){</span>
<span id="cb5-27"><a href="#cb5-27" aria-hidden="true"></a>        <span class="cf">if</span>(w <span class="op">&lt;=</span><span class="st"> </span>M[s]){</span>
<span id="cb5-28"><a href="#cb5-28" aria-hidden="true"></a>          <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/function">return</a></span>(<span class="kw">muF</span>(t) <span class="op">*</span><span class="st"> </span>omega <span class="op">*</span><span class="st"> </span>M[s])</span>
<span id="cb5-29"><a href="#cb5-29" aria-hidden="true"></a>        } <span class="cf">else</span> {</span>
<span id="cb5-30"><a href="#cb5-30" aria-hidden="true"></a>          <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/function">return</a></span>(<span class="dv">0</span>)</span>
<span id="cb5-31"><a href="#cb5-31" aria-hidden="true"></a>        }</span>
<span id="cb5-32"><a href="#cb5-32" aria-hidden="true"></a>      }</span>
<span id="cb5-33"><a href="#cb5-33" aria-hidden="true"></a>    )</span>
<span id="cb5-34"><a href="#cb5-34" aria-hidden="true"></a></span>
<span id="cb5-35"><a href="#cb5-35" aria-hidden="true"></a>  } <span class="cf">else</span> {</span>
<span id="cb5-36"><a href="#cb5-36" aria-hidden="true"></a></span>
<span id="cb5-37"><a href="#cb5-37" aria-hidden="true"></a>    <span class="co"># APPROXIMATE hazards (tolerance around zero; for continuous approximation only)</span></span>
<span id="cb5-38"><a href="#cb5-38" aria-hidden="true"></a>    <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/function">return</a></span>(</span>
<span id="cb5-39"><a href="#cb5-39" aria-hidden="true"></a>      <span class="cf">function</span>(t,M){</span>
<span id="cb5-40"><a href="#cb5-40" aria-hidden="true"></a>        haz &lt;-<span class="st"> </span><span class="kw">muF</span>(t) <span class="op">*</span><span class="st"> </span>omega <span class="op">*</span><span class="st"> </span>M[s]</span>
<span id="cb5-41"><a href="#cb5-41" aria-hidden="true"></a>        <span class="cf">if</span>(haz <span class="op">&lt;</span><span class="st"> </span>tol){</span>
<span id="cb5-42"><a href="#cb5-42" aria-hidden="true"></a>          <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/function">return</a></span>(<span class="dv">0</span>)</span>
<span id="cb5-43"><a href="#cb5-43" aria-hidden="true"></a>        } <span class="cf">else</span> {</span>
<span id="cb5-44"><a href="#cb5-44" aria-hidden="true"></a>          <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/function">return</a></span>(haz)</span>
<span id="cb5-45"><a href="#cb5-45" aria-hidden="true"></a>        }</span>
<span id="cb5-46"><a href="#cb5-46" aria-hidden="true"></a>      }</span>
<span id="cb5-47"><a href="#cb5-47" aria-hidden="true"></a>    )</span>
<span id="cb5-48"><a href="#cb5-48" aria-hidden="true"></a></span>
<span id="cb5-49"><a href="#cb5-49" aria-hidden="true"></a>  }</span>
<span id="cb5-50"><a href="#cb5-50" aria-hidden="true"></a>  <span class="co"># end of function</span></span>
<span id="cb5-51"><a href="#cb5-51" aria-hidden="true"></a>}</span></code></pre></div>
<p>We follow a similar procedure for the adult male mortality. The function is almost identical.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true"></a>make_male_mort_haz_inhom &lt;-<span class="st"> </span><span class="cf">function</span>(trans, u, cube, params,</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true"></a>                                     <span class="dt">exact =</span> <span class="ot">TRUE</span>, <span class="dt">tol =</span> <span class="fl">1e-8</span>){</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true"></a></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true"></a>  <span class="co"># mortality is a time-dependent hazard</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true"></a>  muM &lt;-<span class="st"> </span>params<span class="op">$</span>muM</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true"></a>  <span class="cf">if</span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/typeof">typeof</a></span>(muM) <span class="op">!=</span><span class="st"> "closure"</span>){</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true"></a>    <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/stop">stop</a></span>(<span class="st">"Inhomogeneous hazard 'make_male_mort_haz_inhom', "</span>,</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true"></a>         <span class="st">"value 'muM' in 'params' list needs to be a function"</span>)</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true"></a>  }</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true"></a></span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true"></a>  <span class="co"># which places have input arcs to this transition</span></span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true"></a>  s &lt;-<span class="st"> </span>trans<span class="op">$</span>s</span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true"></a></span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true"></a>  <span class="co"># weights of those arcs</span></span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true"></a>  w &lt;-<span class="st"> </span>trans<span class="op">$</span>s_w</span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true"></a></span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true"></a>  <span class="co"># omega is dependent on genotype</span></span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true"></a>  m_gen &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/strsplit">strsplit</a></span>(<span class="dt">x =</span> u[s], <span class="dt">split =</span> <span class="st">"_"</span>, <span class="dt">fixed =</span> <span class="ot">TRUE</span>)[[<span class="dv">1</span>]][<span class="dv">2</span>]</span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true"></a>  omega &lt;-<span class="st"> </span>cube<span class="op">$</span>omega[m_gen]</span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true"></a></span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true"></a>  <span class="co"># return the hazard function</span></span>
<span id="cb6-22"><a href="#cb6-22" aria-hidden="true"></a>  <span class="cf">if</span>(exact){</span>
<span id="cb6-23"><a href="#cb6-23" aria-hidden="true"></a></span>
<span id="cb6-24"><a href="#cb6-24" aria-hidden="true"></a>    <span class="co"># EXACT hazards (check enabling degree: for discrete simulation only)</span></span>
<span id="cb6-25"><a href="#cb6-25" aria-hidden="true"></a>    <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/function">return</a></span>(</span>
<span id="cb6-26"><a href="#cb6-26" aria-hidden="true"></a>      <span class="cf">function</span>(t,M){</span>
<span id="cb6-27"><a href="#cb6-27" aria-hidden="true"></a>        <span class="cf">if</span>(w <span class="op">&lt;=</span><span class="st"> </span>M[s]){</span>
<span id="cb6-28"><a href="#cb6-28" aria-hidden="true"></a>          <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/function">return</a></span>(<span class="kw">muM</span>(t) <span class="op">*</span><span class="st"> </span>omega <span class="op">*</span><span class="st"> </span>M[s])</span>
<span id="cb6-29"><a href="#cb6-29" aria-hidden="true"></a>        } <span class="cf">else</span> {</span>
<span id="cb6-30"><a href="#cb6-30" aria-hidden="true"></a>          <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/function">return</a></span>(<span class="dv">0</span>)</span>
<span id="cb6-31"><a href="#cb6-31" aria-hidden="true"></a>        }</span>
<span id="cb6-32"><a href="#cb6-32" aria-hidden="true"></a>      }</span>
<span id="cb6-33"><a href="#cb6-33" aria-hidden="true"></a>    )</span>
<span id="cb6-34"><a href="#cb6-34" aria-hidden="true"></a></span>
<span id="cb6-35"><a href="#cb6-35" aria-hidden="true"></a>  } <span class="cf">else</span> {</span>
<span id="cb6-36"><a href="#cb6-36" aria-hidden="true"></a></span>
<span id="cb6-37"><a href="#cb6-37" aria-hidden="true"></a>    <span class="co"># APPROXIMATE hazards (tolerance around zero; for continuous approximation only)</span></span>
<span id="cb6-38"><a href="#cb6-38" aria-hidden="true"></a>    <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/function">return</a></span>(</span>
<span id="cb6-39"><a href="#cb6-39" aria-hidden="true"></a>      <span class="cf">function</span>(t,M){</span>
<span id="cb6-40"><a href="#cb6-40" aria-hidden="true"></a>        haz &lt;-<span class="st"> </span><span class="kw">muM</span>(t) <span class="op">*</span><span class="st"> </span>omega <span class="op">*</span><span class="st"> </span>M[s]</span>
<span id="cb6-41"><a href="#cb6-41" aria-hidden="true"></a>        <span class="cf">if</span>(haz <span class="op">&lt;</span><span class="st"> </span>tol){</span>
<span id="cb6-42"><a href="#cb6-42" aria-hidden="true"></a>          <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/function">return</a></span>(<span class="dv">0</span>)</span>
<span id="cb6-43"><a href="#cb6-43" aria-hidden="true"></a>        } <span class="cf">else</span> {</span>
<span id="cb6-44"><a href="#cb6-44" aria-hidden="true"></a>          <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/function">return</a></span>(haz)</span>
<span id="cb6-45"><a href="#cb6-45" aria-hidden="true"></a>        }</span>
<span id="cb6-46"><a href="#cb6-46" aria-hidden="true"></a>      }</span>
<span id="cb6-47"><a href="#cb6-47" aria-hidden="true"></a>    )</span>
<span id="cb6-48"><a href="#cb6-48" aria-hidden="true"></a></span>
<span id="cb6-49"><a href="#cb6-49" aria-hidden="true"></a>  }</span>
<span id="cb6-50"><a href="#cb6-50" aria-hidden="true"></a>  <span class="co"># end of function</span></span>
<span id="cb6-51"><a href="#cb6-51" aria-hidden="true"></a>}</span></code></pre></div>
<p>Now that we have time-dependent hazard functions, we also need to replace the hazard generating function, so that our time-dependent functions are used. The package provided function, <code><a href="../reference/spn_hazards.html">spn_hazards()</a></code>, generates all of our standard hazard functions. Below, we replace parts of this function for time-inhomogeneous hazards: the only differences are for transition types <code>"male_mort"</code> and <code>"female_mort"</code>, where we drop in our new functions.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true"></a>spn_hazards_lifecycle_node_inhom &lt;-<span class="st"> </span><span class="cf">function</span>(spn_P, spn_T, cube, params,</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true"></a>                                             <span class="dt">log_dd =</span> <span class="ot">TRUE</span>, <span class="dt">exact =</span> <span class="ot">TRUE</span>,</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true"></a>                                             <span class="dt">tol =</span> <span class="fl">1e-12</span>, <span class="dt">verbose =</span> <span class="ot">TRUE</span>){</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true"></a></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true"></a>  <span class="cf">if</span>(tol <span class="op">&gt;</span><span class="st"> </span><span class="fl">1e-6</span> <span class="op">&amp;</span><span class="st"> </span><span class="op">!</span>exact){</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true"></a>    <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/warning">warning</a></span>(<span class="st">"warning: hazard function tolerance "</span>,tol,<span class="st">" is large; "</span>,</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true"></a>            <span class="st">"consider tolerance &lt; 1e-6 for sufficient accuracy</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true"></a>  }</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true"></a></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true"></a>  <span class="cf">if</span>(log_dd){</span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true"></a>    <span class="cf">if</span>(<span class="op">!</span>(<span class="st">"K"</span> <span class="op">%in%</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(params))){</span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true"></a>      <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/stop">stop</a></span>(<span class="st">"Specified logistic (carrying capacity) density-dependent larval "</span>,</span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true"></a>           <span class="st">"mortality, please specify parameter 'K' in params"</span>)</span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true"></a>    }</span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true"></a>  } <span class="cf">else</span> {</span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true"></a>    <span class="cf">if</span>(<span class="op">!</span>(<span class="st">"gamma"</span> <span class="op">%in%</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(params))){</span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true"></a>      <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/stop">stop</a></span>(<span class="st">"Specified Lotka-Volterra (carrying capacity) density-dependent "</span>,</span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true"></a>           <span class="st">"larval mortality, please specify parameter 'gamma' in params"</span>)</span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true"></a>    }</span>
<span id="cb7-20"><a href="#cb7-20" aria-hidden="true"></a>  }</span>
<span id="cb7-21"><a href="#cb7-21" aria-hidden="true"></a></span>
<span id="cb7-22"><a href="#cb7-22" aria-hidden="true"></a>  <span class="co"># transitions and places</span></span>
<span id="cb7-23"><a href="#cb7-23" aria-hidden="true"></a>  v &lt;-<span class="st"> </span>spn_T<span class="op">$</span>v</span>
<span id="cb7-24"><a href="#cb7-24" aria-hidden="true"></a>  u &lt;-<span class="st"> </span>spn_P<span class="op">$</span>u</span>
<span id="cb7-25"><a href="#cb7-25" aria-hidden="true"></a>  n &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(v)</span>
<span id="cb7-26"><a href="#cb7-26" aria-hidden="true"></a></span>
<span id="cb7-27"><a href="#cb7-27" aria-hidden="true"></a>  <span class="co"># get male and larvae indices</span></span>
<span id="cb7-28"><a href="#cb7-28" aria-hidden="true"></a>  l_ix &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/vector">as.vector</a></span>(spn_P<span class="op">$</span>ix[[<span class="dv">1</span>]]<span class="op">$</span>larvae)</span>
<span id="cb7-29"><a href="#cb7-29" aria-hidden="true"></a>  m_ix &lt;-<span class="st"> </span>spn_P<span class="op">$</span>ix[[<span class="dv">1</span>]]<span class="op">$</span>males</span>
<span id="cb7-30"><a href="#cb7-30" aria-hidden="true"></a></span>
<span id="cb7-31"><a href="#cb7-31" aria-hidden="true"></a>  <span class="co"># the hazard functions</span></span>
<span id="cb7-32"><a href="#cb7-32" aria-hidden="true"></a>  h &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/setNames">setNames</a></span>(<span class="dt">object =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/vector">vector</a></span>(<span class="st">"list"</span>,n), <span class="dt">nm =</span> v)</span>
<span id="cb7-33"><a href="#cb7-33" aria-hidden="true"></a></span>
<span id="cb7-34"><a href="#cb7-34" aria-hidden="true"></a></span>
<span id="cb7-35"><a href="#cb7-35" aria-hidden="true"></a>  <span class="cf">if</span>(verbose){</span>
<span id="cb7-36"><a href="#cb7-36" aria-hidden="true"></a>    pb &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/txtProgressBar">txtProgressBar</a></span>(<span class="dt">min =</span> <span class="dv">1</span>,<span class="dt">max =</span> n,<span class="dt">style =</span> <span class="dv">3</span>)</span>
<span id="cb7-37"><a href="#cb7-37" aria-hidden="true"></a>    pp &lt;-<span class="st"> </span><span class="dv">1</span></span>
<span id="cb7-38"><a href="#cb7-38" aria-hidden="true"></a></span>
<span id="cb7-39"><a href="#cb7-39" aria-hidden="true"></a>    <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cat">cat</a></span>(<span class="st">" --- generating hazard functions for SPN --- </span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb7-40"><a href="#cb7-40" aria-hidden="true"></a>  }</span>
<span id="cb7-41"><a href="#cb7-41" aria-hidden="true"></a></span>
<span id="cb7-42"><a href="#cb7-42" aria-hidden="true"></a>  <span class="co"># make the hazards</span></span>
<span id="cb7-43"><a href="#cb7-43" aria-hidden="true"></a>  <span class="cf">for</span>(t <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>n){</span>
<span id="cb7-44"><a href="#cb7-44" aria-hidden="true"></a></span>
<span id="cb7-45"><a href="#cb7-45" aria-hidden="true"></a>    type &lt;-<span class="st"> </span>spn_T<span class="op">$</span>T[[t]]<span class="op">$</span>class</span>
<span id="cb7-46"><a href="#cb7-46" aria-hidden="true"></a></span>
<span id="cb7-47"><a href="#cb7-47" aria-hidden="true"></a>    <span class="co"># make the correct type of hazard</span></span>
<span id="cb7-48"><a href="#cb7-48" aria-hidden="true"></a>    <span class="cf">if</span>(type <span class="op">==</span><span class="st"> "oviposit"</span>){</span>
<span id="cb7-49"><a href="#cb7-49" aria-hidden="true"></a>      h[[t]] &lt;-<span class="st"> </span>MGDrivE2<span class="op">:::</span><span class="kw">make_oviposit_haz</span>(<span class="dt">trans =</span> spn_T<span class="op">$</span>T[[t]], <span class="dt">u =</span> u,</span>
<span id="cb7-50"><a href="#cb7-50" aria-hidden="true"></a>                                             <span class="dt">cube =</span> cube, <span class="dt">params =</span> params,</span>
<span id="cb7-51"><a href="#cb7-51" aria-hidden="true"></a>                                             <span class="dt">exact =</span> exact, <span class="dt">tol =</span> tol)</span>
<span id="cb7-52"><a href="#cb7-52" aria-hidden="true"></a>    } <span class="cf">else</span> <span class="cf">if</span>(type <span class="op">==</span><span class="st"> "egg_adv"</span>){</span>
<span id="cb7-53"><a href="#cb7-53" aria-hidden="true"></a>      h[[t]] &lt;-<span class="st"> </span>MGDrivE2<span class="op">:::</span><span class="kw">make_egg_adv_haz</span>(<span class="dt">trans =</span> spn_T<span class="op">$</span>T[[t]], <span class="dt">u =</span> u,</span>
<span id="cb7-54"><a href="#cb7-54" aria-hidden="true"></a>                                            <span class="dt">cube =</span> cube, <span class="dt">params =</span> params,</span>
<span id="cb7-55"><a href="#cb7-55" aria-hidden="true"></a>                                            <span class="dt">exact =</span> exact, <span class="dt">tol =</span> tol)</span>
<span id="cb7-56"><a href="#cb7-56" aria-hidden="true"></a>    } <span class="cf">else</span> <span class="cf">if</span>(type <span class="op">==</span><span class="st"> "egg_mort"</span>){</span>
<span id="cb7-57"><a href="#cb7-57" aria-hidden="true"></a>      h[[t]] &lt;-<span class="st"> </span>MGDrivE2<span class="op">:::</span><span class="kw">make_egg_mort_haz</span>(<span class="dt">trans =</span> spn_T<span class="op">$</span>T[[t]], <span class="dt">u =</span> u,</span>
<span id="cb7-58"><a href="#cb7-58" aria-hidden="true"></a>                                             <span class="dt">cube =</span> cube, <span class="dt">params =</span> params,</span>
<span id="cb7-59"><a href="#cb7-59" aria-hidden="true"></a>                                             <span class="dt">exact =</span> exact, <span class="dt">tol =</span> tol)</span>
<span id="cb7-60"><a href="#cb7-60" aria-hidden="true"></a>    } <span class="cf">else</span> <span class="cf">if</span>(type <span class="op">==</span><span class="st"> "larvae_adv"</span>){</span>
<span id="cb7-61"><a href="#cb7-61" aria-hidden="true"></a>      h[[t]] &lt;-<span class="st"> </span>MGDrivE2<span class="op">:::</span><span class="kw">make_larvae_adv_haz</span>(<span class="dt">trans =</span> spn_T<span class="op">$</span>T[[t]], <span class="dt">u =</span> u,</span>
<span id="cb7-62"><a href="#cb7-62" aria-hidden="true"></a>                                               <span class="dt">cube =</span> cube, <span class="dt">params =</span> params,</span>
<span id="cb7-63"><a href="#cb7-63" aria-hidden="true"></a>                                               <span class="dt">exact =</span> exact,<span class="dt">tol =</span> tol)</span>
<span id="cb7-64"><a href="#cb7-64" aria-hidden="true"></a>    } <span class="cf">else</span> <span class="cf">if</span>(type <span class="op">==</span><span class="st"> "larvae_mort"</span>){</span>
<span id="cb7-65"><a href="#cb7-65" aria-hidden="true"></a>      <span class="cf">if</span>(log_dd){</span>
<span id="cb7-66"><a href="#cb7-66" aria-hidden="true"></a>        h[[t]] &lt;-<span class="st"> </span>MGDrivE2<span class="op">:::</span><span class="kw">make_larvae_mort_haz_log</span>(<span class="dt">trans =</span> spn_T<span class="op">$</span>T[[t]], <span class="dt">u =</span> u,</span>
<span id="cb7-67"><a href="#cb7-67" aria-hidden="true"></a>                                                      <span class="dt">l_ix =</span> l_ix, <span class="dt">node =</span> <span class="dv">1</span>,</span>
<span id="cb7-68"><a href="#cb7-68" aria-hidden="true"></a>                                                      <span class="dt">cube =</span> cube, <span class="dt">params =</span> params,</span>
<span id="cb7-69"><a href="#cb7-69" aria-hidden="true"></a>                                                      <span class="dt">exact =</span> exact, <span class="dt">tol =</span> tol)</span>
<span id="cb7-70"><a href="#cb7-70" aria-hidden="true"></a>      } <span class="cf">else</span> {</span>
<span id="cb7-71"><a href="#cb7-71" aria-hidden="true"></a>        h[[t]] &lt;-<span class="st"> </span>MGDrivE2<span class="op">:::</span><span class="kw">make_larvae_mort_haz_lk</span>(<span class="dt">trans =</span> spn_T<span class="op">$</span>T[[t]], <span class="dt">u =</span> u,</span>
<span id="cb7-72"><a href="#cb7-72" aria-hidden="true"></a>                                                     <span class="dt">l_ix =</span> l_ix, <span class="dt">node =</span> <span class="dv">1</span>,</span>
<span id="cb7-73"><a href="#cb7-73" aria-hidden="true"></a>                                                     <span class="dt">cube =</span> cube, <span class="dt">params =</span> params,</span>
<span id="cb7-74"><a href="#cb7-74" aria-hidden="true"></a>                                                     <span class="dt">exact =</span> exact, <span class="dt">tol =</span> tol)</span>
<span id="cb7-75"><a href="#cb7-75" aria-hidden="true"></a>      }</span>
<span id="cb7-76"><a href="#cb7-76" aria-hidden="true"></a>    } <span class="cf">else</span> <span class="cf">if</span>(type <span class="op">==</span><span class="st"> "pupae_adv"</span>){</span>
<span id="cb7-77"><a href="#cb7-77" aria-hidden="true"></a>      h[[t]] &lt;-<span class="st"> </span>MGDrivE2<span class="op">:::</span><span class="kw">make_pupae_adv_haz</span>(<span class="dt">trans =</span> spn_T<span class="op">$</span>T[[t]], <span class="dt">u =</span> u,</span>
<span id="cb7-78"><a href="#cb7-78" aria-hidden="true"></a>                                              <span class="dt">cube =</span> cube, <span class="dt">params =</span> params,</span>
<span id="cb7-79"><a href="#cb7-79" aria-hidden="true"></a>                                              <span class="dt">exact =</span> exact, <span class="dt">tol =</span> tol)</span>
<span id="cb7-80"><a href="#cb7-80" aria-hidden="true"></a>    } <span class="cf">else</span> <span class="cf">if</span>(type <span class="op">==</span><span class="st"> "pupae_mort"</span>){</span>
<span id="cb7-81"><a href="#cb7-81" aria-hidden="true"></a>      h[[t]] &lt;-<span class="st"> </span>MGDrivE2<span class="op">:::</span><span class="kw">make_pupae_mort_haz</span>(<span class="dt">trans =</span> spn_T<span class="op">$</span>T[[t]], <span class="dt">u =</span> u,</span>
<span id="cb7-82"><a href="#cb7-82" aria-hidden="true"></a>                                               <span class="dt">cube =</span> cube, <span class="dt">params =</span> params,</span>
<span id="cb7-83"><a href="#cb7-83" aria-hidden="true"></a>                                               <span class="dt">exact =</span> exact, <span class="dt">tol =</span> tol)</span>
<span id="cb7-84"><a href="#cb7-84" aria-hidden="true"></a>    } <span class="cf">else</span> <span class="cf">if</span>(type <span class="op">==</span><span class="st"> "pupae_2m"</span>){</span>
<span id="cb7-85"><a href="#cb7-85" aria-hidden="true"></a>      h[[t]] &lt;-<span class="st"> </span>MGDrivE2<span class="op">:::</span><span class="kw">make_pupae_2male_haz</span>(<span class="dt">trans =</span> spn_T<span class="op">$</span>T[[t]], <span class="dt">u =</span> u,</span>
<span id="cb7-86"><a href="#cb7-86" aria-hidden="true"></a>                                                <span class="dt">cube =</span> cube, <span class="dt">params =</span> params,</span>
<span id="cb7-87"><a href="#cb7-87" aria-hidden="true"></a>                                                <span class="dt">exact =</span> exact, <span class="dt">tol =</span> tol)</span>
<span id="cb7-88"><a href="#cb7-88" aria-hidden="true"></a>    } <span class="cf">else</span> <span class="cf">if</span>(type <span class="op">==</span><span class="st"> "pupae_2f"</span>){</span>
<span id="cb7-89"><a href="#cb7-89" aria-hidden="true"></a>      h[[t]] &lt;-<span class="st"> </span>MGDrivE2<span class="op">:::</span><span class="kw">make_pupae_2female_haz</span>(<span class="dt">trans =</span> spn_T<span class="op">$</span>T[[t]], <span class="dt">u =</span> u,</span>
<span id="cb7-90"><a href="#cb7-90" aria-hidden="true"></a>                                                  <span class="dt">m_ix =</span> m_ix, <span class="dt">cube =</span> cube,</span>
<span id="cb7-91"><a href="#cb7-91" aria-hidden="true"></a>                                                  <span class="dt">params =</span> params, <span class="dt">exact =</span> exact,</span>
<span id="cb7-92"><a href="#cb7-92" aria-hidden="true"></a>                                                  <span class="dt">tol =</span> tol)</span>
<span id="cb7-93"><a href="#cb7-93" aria-hidden="true"></a>    } <span class="cf">else</span> <span class="cf">if</span>(type <span class="op">==</span><span class="st"> "male_mort"</span>){</span>
<span id="cb7-94"><a href="#cb7-94" aria-hidden="true"></a>      <span class="co"># inhomogeneous male mortality</span></span>
<span id="cb7-95"><a href="#cb7-95" aria-hidden="true"></a>      h[[t]] &lt;-<span class="st"> </span><span class="kw">make_male_mort_haz_inhom</span>(<span class="dt">trans =</span> spn_T<span class="op">$</span>T[[t]], <span class="dt">u =</span> u ,<span class="dt">cube =</span> cube,</span>
<span id="cb7-96"><a href="#cb7-96" aria-hidden="true"></a>                                         <span class="dt">params =</span> params, <span class="dt">exact =</span> exact, <span class="dt">tol =</span> tol)</span>
<span id="cb7-97"><a href="#cb7-97" aria-hidden="true"></a>    } <span class="cf">else</span> <span class="cf">if</span>(type <span class="op">==</span><span class="st"> "female_mort"</span>){</span>
<span id="cb7-98"><a href="#cb7-98" aria-hidden="true"></a>      <span class="co"># inhomogeneous female mortality</span></span>
<span id="cb7-99"><a href="#cb7-99" aria-hidden="true"></a>      h[[t]] &lt;-<span class="st"> </span><span class="kw">make_female_mort_haz_inhom</span>(<span class="dt">trans =</span> spn_T<span class="op">$</span>T[[t]], <span class="dt">u =</span> u, <span class="dt">cube =</span> cube,</span>
<span id="cb7-100"><a href="#cb7-100" aria-hidden="true"></a>                                           <span class="dt">params =</span> params, <span class="dt">exact =</span> exact, <span class="dt">tol =</span> tol)</span>
<span id="cb7-101"><a href="#cb7-101" aria-hidden="true"></a>    } <span class="cf">else</span> <span class="cf">if</span>(type <span class="op">==</span><span class="st"> "pupae_2unmated"</span>){</span>
<span id="cb7-102"><a href="#cb7-102" aria-hidden="true"></a>      h[[t]] &lt;-<span class="st"> </span>MGDrivE2<span class="op">:::</span><span class="kw">make_pupae_2unmated_haz</span>(<span class="dt">trans =</span> spn_T<span class="op">$</span>T[[t]], <span class="dt">u =</span> u,</span>
<span id="cb7-103"><a href="#cb7-103" aria-hidden="true"></a>                                                   <span class="dt">m_ix =</span> m_ix, <span class="dt">cube =</span> cube,</span>
<span id="cb7-104"><a href="#cb7-104" aria-hidden="true"></a>                                                   <span class="dt">params =</span> params, <span class="dt">exact =</span> exact,</span>
<span id="cb7-105"><a href="#cb7-105" aria-hidden="true"></a>                                                   <span class="dt">tol =</span> tol)</span>
<span id="cb7-106"><a href="#cb7-106" aria-hidden="true"></a>    } <span class="cf">else</span> <span class="cf">if</span>(type <span class="op">==</span><span class="st"> "female_unmated_mate"</span>){</span>
<span id="cb7-107"><a href="#cb7-107" aria-hidden="true"></a>      h[[t]] &lt;-<span class="st"> </span>MGDrivE2<span class="op">:::</span><span class="kw">make_unmated_2female_haz</span>(<span class="dt">trans =</span> spn_T<span class="op">$</span>T[[t]], <span class="dt">u =</span> u,</span>
<span id="cb7-108"><a href="#cb7-108" aria-hidden="true"></a>                                                    <span class="dt">m_ix =</span> m_ix, <span class="dt">cube =</span> cube,</span>
<span id="cb7-109"><a href="#cb7-109" aria-hidden="true"></a>                                                    <span class="dt">params =</span> params, <span class="dt">exact =</span> exact,</span>
<span id="cb7-110"><a href="#cb7-110" aria-hidden="true"></a>                                                    <span class="dt">tol =</span> tol)</span>
<span id="cb7-111"><a href="#cb7-111" aria-hidden="true"></a>    } <span class="cf">else</span> <span class="cf">if</span>(type <span class="op">==</span><span class="st"> "female_unmated_mort"</span>){</span>
<span id="cb7-112"><a href="#cb7-112" aria-hidden="true"></a>      h[[t]] &lt;-<span class="st"> </span><span class="kw">make_female_mort_haz_inhom</span>(<span class="dt">trans =</span> spn_T<span class="op">$</span>T[[t]], <span class="dt">u =</span> u, <span class="dt">cube =</span> cube,</span>
<span id="cb7-113"><a href="#cb7-113" aria-hidden="true"></a>                                           <span class="dt">params =</span> params, <span class="dt">exact =</span> exact, <span class="dt">tol =</span> tol)</span>
<span id="cb7-114"><a href="#cb7-114" aria-hidden="true"></a>    } <span class="cf">else</span> {</span>
<span id="cb7-115"><a href="#cb7-115" aria-hidden="true"></a>      <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/stop">stop</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(<span class="st">"error in making hazard function for unknown class type: "</span>,type))</span>
<span id="cb7-116"><a href="#cb7-116" aria-hidden="true"></a>    }</span>
<span id="cb7-117"><a href="#cb7-117" aria-hidden="true"></a></span>
<span id="cb7-118"><a href="#cb7-118" aria-hidden="true"></a>    <span class="cf">if</span>(verbose){<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/txtProgressBar">setTxtProgressBar</a></span>(pb,t)}</span>
<span id="cb7-119"><a href="#cb7-119" aria-hidden="true"></a>  } <span class="co"># end loop</span></span>
<span id="cb7-120"><a href="#cb7-120" aria-hidden="true"></a></span>
<span id="cb7-121"><a href="#cb7-121" aria-hidden="true"></a>  <span class="cf">if</span>(verbose){</span>
<span id="cb7-122"><a href="#cb7-122" aria-hidden="true"></a>    <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/connections">close</a></span>(pb)</span>
<span id="cb7-123"><a href="#cb7-123" aria-hidden="true"></a>    <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cat">cat</a></span>(<span class="st">" --- done generating hazard functions for SPN --- </span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="cb7-124"><a href="#cb7-124" aria-hidden="true"></a>  }</span>
<span id="cb7-125"><a href="#cb7-125" aria-hidden="true"></a></span>
<span id="cb7-126"><a href="#cb7-126" aria-hidden="true"></a></span>
<span id="cb7-127"><a href="#cb7-127" aria-hidden="true"></a>  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/function">return</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="st">"hazards"</span>=h, <span class="st">"flag"</span>=exact))</span>
<span id="cb7-128"><a href="#cb7-128" aria-hidden="true"></a>} <span class="co"># end function</span></span></code></pre></div>
</div>
</div>
<div id="make_spn" class="section level1">
<h1 class="hasAnchor">
<a href="#make_spn" class="anchor"></a>Petri Net Setup</h1>
<p>Much of the following setup is from the <a href="lifecycle-node.html">“MGDrivE2: One Node Lifecycle Dynamics”</a> vignette, with small adaptations for the time-inhomogeneous simulations.</p>
<div id="pars" class="section level2">
<h2 class="hasAnchor">
<a href="#pars" class="anchor"></a>Parameterization</h2>
<p>In a familiar procedure, we define parameters for the simulation. Note that we use the mean of the time-dependent mortality as the equilibrium value, we will use this value to run a set of time-homogeneous simulations to compare results.</p>
<p>We also set the simulation time for 1 year, as our mortality data spans a year. The data is stored daily, and we run a basic Mendelian inheritance pattern.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true"></a><span class="co"># number of adult female mosquitoes</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true"></a>NF &lt;-<span class="st"> </span><span class="dv">500</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true"></a></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true"></a><span class="co"># entomological parameters</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true"></a>theta &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true"></a>  <span class="dt">qE =</span> <span class="dv">1</span><span class="op">/</span><span class="dv">4</span>,</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true"></a>  <span class="dt">nE =</span> <span class="dv">2</span>,</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true"></a>  <span class="dt">qL =</span> <span class="dv">1</span><span class="op">/</span><span class="dv">3</span>,</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true"></a>  <span class="dt">nL =</span> <span class="dv">3</span>,</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true"></a>  <span class="dt">qP =</span> <span class="dv">1</span><span class="op">/</span><span class="dv">6</span>,</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true"></a>  <span class="dt">nP =</span> <span class="dv">2</span>,</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true"></a>  <span class="dt">muE =</span> <span class="fl">0.05</span>,</span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true"></a>  <span class="dt">muL =</span> <span class="fl">0.15</span>,</span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true"></a>  <span class="dt">muP =</span> <span class="fl">0.05</span>,</span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true"></a>  <span class="dt">muF =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/mean">mean</a></span>(mu_ts[,<span class="dv">1</span>]),</span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true"></a>  <span class="dt">muM =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/mean">mean</a></span>(mu_ts[,<span class="dv">1</span>]),</span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true"></a>  <span class="dt">beta =</span> <span class="dv">16</span>,</span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true"></a>  <span class="dt">nu =</span> <span class="dv">1</span><span class="op">/</span>(<span class="dv">4</span><span class="op">/</span><span class="dv">24</span>)</span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true"></a>)</span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true"></a></span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true"></a><span class="co"># simulation parameters</span></span>
<span id="cb8-22"><a href="#cb8-22" aria-hidden="true"></a>tmax &lt;-<span class="st"> </span><span class="dv">365</span></span>
<span id="cb8-23"><a href="#cb8-23" aria-hidden="true"></a>dt &lt;-<span class="st"> </span><span class="dv">1</span></span>
<span id="cb8-24"><a href="#cb8-24" aria-hidden="true"></a></span>
<span id="cb8-25"><a href="#cb8-25" aria-hidden="true"></a><span class="co"># basic inheritance pattern</span></span>
<span id="cb8-26"><a href="#cb8-26" aria-hidden="true"></a>cube &lt;-<span class="st"> </span>MGDrivE<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/MGDrivE/topics/cubeMendelian">cubeMendelian</a></span>()</span></code></pre></div>
</div>
<div id="init_pn" class="section level2">
<h2 class="hasAnchor">
<a href="#init_pn" class="anchor"></a>Initialization of the Petri Net</h2>
<p>With parameters defined, we can setup the places and transition in our Petri Net. These structural components will be the same for all of the simulations, time dependent and independent.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true"></a><span class="co"># Places and transitions for one node</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true"></a>SPN_P &lt;-<span class="st"> </span><span class="kw"><a href="../reference/spn_P_lifecycle_node.html">spn_P_lifecycle_node</a></span>(<span class="dt">params =</span> theta, <span class="dt">cube =</span> cube)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true"></a>SPN_T &lt;-<span class="st"> </span><span class="kw"><a href="../reference/spn_T_lifecycle_node.html">spn_T_lifecycle_node</a></span>(<span class="dt">spn_P =</span> SPN_P, <span class="dt">params =</span> theta, <span class="dt">cube =</span> cube)</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true"></a></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true"></a><span class="co"># Stoichiometry matrix</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true"></a>S &lt;-<span class="st"> </span><span class="kw"><a href="../reference/spn_S.html">spn_S</a></span>(<span class="dt">spn_P =</span> SPN_P, <span class="dt">spn_T =</span> SPN_T)</span></code></pre></div>
</div>
<div id="equilibria_haz" class="section level2">
<h2 class="hasAnchor">
<a href="#equilibria_haz" class="anchor"></a>Equilibrium Conditions and Hazard Functions</h2>
<p>Since equilibrium is calculated at the average adult mortality over the year, we use the parameters defined above (<code>theta</code>). However, the parameters returned by the <code><a href="../reference/equilibrium_lifeycle.html">equilibrium_lifeycle()</a></code> function, namely <code>initialCons$params$muF</code> and <code>muM</code> (adult mortality rates), need to be replaces with our time-inhomogeneous <a href="#inhom_adult_mort">step function</a>.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true"></a><span class="co"># calculate equilibrium and setup initial conditions</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true"></a><span class="co">#  outputs required parameters in the named list "params"</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true"></a><span class="co">#  outputs intial equilibrium for adv users, "init</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true"></a><span class="co">#  outputs properly filled initial markings, "M0"</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true"></a>initialCons &lt;-<span class="st"> </span><span class="kw"><a href="../reference/equilibrium_lifeycle.html">equilibrium_lifeycle</a></span>(<span class="dt">params =</span> theta, <span class="dt">NF =</span> NF, <span class="dt">log_dd =</span> <span class="ot">TRUE</span>,</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true"></a>                                    <span class="dt">spn_P =</span> SPN_P, <span class="dt">cube =</span> cube)</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true"></a></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true"></a><span class="co"># store homogenous params</span></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true"></a>theta_hom &lt;-<span class="st"> </span>initialCons<span class="op">$</span>params</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true"></a></span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true"></a><span class="co"># update params for inhomogeneous hazards (these are function closures)</span></span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true"></a>initialCons<span class="op">$</span>params<span class="op">$</span>muF &lt;-<span class="st"> </span>step_approx</span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true"></a>initialCons<span class="op">$</span>params<span class="op">$</span>muM &lt;-<span class="st"> </span>step_approx</span></code></pre></div>
<p>Now we are ready to make the vectors of hazards. We need four vectors to compare four different model formulations: time-homogeneous and inhomogeneous, for ODE and CTMC simulations.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true"></a><span class="co"># ODE (inhomogeneous) hazards</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true"></a>approx_hazards &lt;-<span class="st"> </span><span class="kw">spn_hazards_lifecycle_node_inhom</span>(<span class="dt">spn_P =</span> SPN_P, <span class="dt">spn_T =</span> SPN_T,</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true"></a>                                                   <span class="dt">cube =</span> cube,</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true"></a>                                                   <span class="dt">params =</span> initialCons<span class="op">$</span>params,</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true"></a>                                                   <span class="dt">exact =</span> <span class="ot">FALSE</span>, <span class="dt">tol =</span> <span class="fl">1e-8</span>,</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true"></a>                                                   <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true"></a></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true"></a><span class="co"># ODE (homogeneous) hazards</span></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true"></a>approx_hazards_hom &lt;-<span class="st"> </span><span class="kw"><a href="../reference/spn_hazards.html">spn_hazards</a></span>(<span class="dt">spn_P =</span> SPN_P, <span class="dt">spn_T =</span> SPN_T, <span class="dt">cube =</span> cube,</span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true"></a>                                  <span class="dt">params =</span> theta_hom, <span class="dt">type =</span> <span class="st">"life"</span>, <span class="dt">log_dd =</span> <span class="ot">TRUE</span>,</span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true"></a>                                  <span class="dt">exact =</span> <span class="ot">FALSE</span>, <span class="dt">tol =</span> <span class="fl">1e-8</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true"></a></span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true"></a><span class="co"># CTMC (inhomogeneous) hazards</span></span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true"></a>exact_hazards &lt;-<span class="st"> </span><span class="kw">spn_hazards_lifecycle_node_inhom</span>(<span class="dt">spn_P =</span> SPN_P, <span class="dt">spn_T =</span> SPN_T,</span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true"></a>                                                  <span class="dt">cube =</span> cube,</span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true"></a>                                                  <span class="dt">params =</span> initialCons<span class="op">$</span>params,</span>
<span id="cb11-17"><a href="#cb11-17" aria-hidden="true"></a>                                                  <span class="dt">exact =</span> <span class="ot">TRUE</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</span>
<span id="cb11-18"><a href="#cb11-18" aria-hidden="true"></a></span>
<span id="cb11-19"><a href="#cb11-19" aria-hidden="true"></a><span class="co"># CTMC (homogeneous) hazards</span></span>
<span id="cb11-20"><a href="#cb11-20" aria-hidden="true"></a>exact_hazards_hom &lt;-<span class="st"> </span><span class="kw"><a href="../reference/spn_hazards.html">spn_hazards</a></span>(<span class="dt">spn_P =</span> SPN_P, <span class="dt">spn_T =</span> SPN_T, <span class="dt">cube =</span> cube,</span>
<span id="cb11-21"><a href="#cb11-21" aria-hidden="true"></a>                                  <span class="dt">params =</span> theta_hom, <span class="dt">type =</span> <span class="st">"life"</span>, <span class="dt">log_dd =</span> <span class="ot">TRUE</span>,</span>
<span id="cb11-22"><a href="#cb11-22" aria-hidden="true"></a>                                  <span class="dt">exact =</span> <span class="ot">TRUE</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</span></code></pre></div>
</div>
</div>
<div id="sim" class="section level1">
<h1 class="hasAnchor">
<a href="#sim" class="anchor"></a>Simulation of Fully Specified SPN Model</h1>
<p>Finally, before we run simulations, we set up a release scheme to generate interesting dynamics. We will release 50 adult females and males, with homozygous recessive alleles, every week for five weeks, starting at day 25. Remember, it is important that <strong>the event names match a place name</strong> in the simulation. This format is used in <strong>MGDrivE2</strong> for consistency with solvers in <code>deSolve</code>.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true"></a><span class="co"># releases</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true"></a>r_times &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/seq">seq</a></span>(<span class="dt">from =</span> <span class="dv">25</span>, <span class="dt">length.out =</span> <span class="dv">5</span>, <span class="dt">by =</span> <span class="dv">7</span>)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true"></a>r_size &lt;-<span class="st"> </span><span class="dv">50</span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true"></a>events_f &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/data.frame">data.frame</a></span>(<span class="st">"var"</span> =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(<span class="st">"F_"</span>, cube<span class="op">$</span>releaseType, <span class="st">"_"</span>, cube<span class="op">$</span>wildType),</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true"></a>                       <span class="st">"time"</span> =<span class="st"> </span>r_times,</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true"></a>                       <span class="st">"value"</span> =<span class="st"> </span>r_size,</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true"></a>                       <span class="st">"method"</span> =<span class="st"> "add"</span>,</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true"></a>                       <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true"></a></span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true"></a>events_m &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/data.frame">data.frame</a></span>(<span class="st">"var"</span> =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(<span class="st">"M_"</span>, cube<span class="op">$</span>releaseType),</span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true"></a>                       <span class="st">"time"</span> =<span class="st"> </span>r_times,</span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true"></a>                       <span class="st">"value"</span> =<span class="st"> </span>r_size,</span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true"></a>                       <span class="st">"method"</span> =<span class="st"> "add"</span>,</span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true"></a>                       <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)</span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true"></a></span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true"></a>events &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cbind">rbind</a></span>(events_f,events_m)</span></code></pre></div>
<div id="sim_ode" class="section level2">
<h2 class="hasAnchor">
<a href="#sim_ode" class="anchor"></a>Simulation: ODE Models</h2>
<p>Initially, we should numerically integrate our model, providing exact results for both time homogeneous and inhomogeneous versions. We recommend (and use!) the excellent ODE integration routines from <code>deSolve</code> to solve our deterministic models. Note that we specify the Runge-Kutta 4/5th order method (<code>method = "ode45"</code>) in the call to <code><a href="https://www.rdocumentation.org/packages/deSolve/topics/ode">deSolve::ode</a></code> rather than the default method (<code>method = "lsoda"</code>). This is because the default solvers behave badly with time-dependent rates. We clean up the output using functions from the <code>MGDrivE2</code> package and put it in a <code>data.frame</code> for <a href="#plots">plotting</a> below.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true"></a><span class="co"># ODE (inhomogeneous) simulation</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true"></a>ODE_out_inhom &lt;-<span class="st"> </span><span class="kw"><a href="../reference/sim_trajectory_R.html">sim_trajectory_R</a></span>(<span class="dt">x0 =</span> initialCons<span class="op">$</span>M0, <span class="dt">t0 =</span> <span class="dv">0</span>, <span class="dt">tt =</span> tmax, <span class="dt">dt =</span> dt,</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true"></a>                                  <span class="dt">S =</span> S, <span class="dt">hazards =</span> approx_hazards, <span class="dt">sampler =</span> <span class="st">"ode"</span>,</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true"></a>                                  <span class="dt">method =</span> <span class="st">"ode45"</span>, <span class="dt">events =</span> events, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true"></a></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true"></a><span class="co"># summarize</span></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true"></a>ODE_female_inhom &lt;-<span class="st"> </span><span class="kw"><a href="../reference/summarize_females.html">summarize_females</a></span>(<span class="dt">out =</span> ODE_out_inhom<span class="op">$</span>state, <span class="dt">spn_P =</span> SPN_P)</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true"></a>ODE_male_inhom &lt;-<span class="st"> </span><span class="kw"><a href="../reference/summarize_males.html">summarize_males</a></span>(<span class="dt">out =</span> ODE_out_inhom<span class="op">$</span>state)</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true"></a></span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true"></a><span class="co"># combine for plotting later</span></span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true"></a>ODE_out_inhom_melt &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cbind">rbind</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cbind">cbind</a></span>(ODE_female_inhom, <span class="st">"sex"</span> =<span class="st"> "F"</span>),</span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true"></a>                            <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cbind">cbind</a></span>(ODE_male_inhom, <span class="st">"sex"</span> =<span class="st"> "M"</span>) )</span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true"></a></span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true"></a></span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true"></a><span class="co"># ODE (homogeneous) simulation</span></span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true"></a>ODE_out_hom &lt;-<span class="st"> </span><span class="kw"><a href="../reference/sim_trajectory_R.html">sim_trajectory_R</a></span>(<span class="dt">x0 =</span> initialCons<span class="op">$</span>M0, <span class="dt">t0 =</span> <span class="dv">0</span>, <span class="dt">tt =</span> tmax, <span class="dt">dt =</span> dt,</span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true"></a>                                <span class="dt">S =</span> S, <span class="dt">hazards =</span> approx_hazards_hom, <span class="dt">sampler =</span> <span class="st">"ode"</span>,</span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true"></a>                                <span class="dt">method =</span> <span class="st">"ode45"</span>, <span class="dt">events =</span> events, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</span>
<span id="cb13-19"><a href="#cb13-19" aria-hidden="true"></a></span>
<span id="cb13-20"><a href="#cb13-20" aria-hidden="true"></a><span class="co"># summarize</span></span>
<span id="cb13-21"><a href="#cb13-21" aria-hidden="true"></a>ODE_female_hom &lt;-<span class="st"> </span><span class="kw"><a href="../reference/summarize_females.html">summarize_females</a></span>(<span class="dt">out =</span> ODE_out_hom<span class="op">$</span>state, <span class="dt">spn_P =</span> SPN_P)</span>
<span id="cb13-22"><a href="#cb13-22" aria-hidden="true"></a>ODE_male_hom &lt;-<span class="st"> </span><span class="kw"><a href="../reference/summarize_males.html">summarize_males</a></span>(<span class="dt">out =</span> ODE_out_hom<span class="op">$</span>state)</span>
<span id="cb13-23"><a href="#cb13-23" aria-hidden="true"></a></span>
<span id="cb13-24"><a href="#cb13-24" aria-hidden="true"></a><span class="co"># combine for plotting later</span></span>
<span id="cb13-25"><a href="#cb13-25" aria-hidden="true"></a>ODE_out_hom_melt &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cbind">rbind</a></span>(</span>
<span id="cb13-26"><a href="#cb13-26" aria-hidden="true"></a>  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cbind">cbind</a></span>(ODE_female_hom, <span class="st">"sex"</span> =<span class="st"> "F"</span>),</span>
<span id="cb13-27"><a href="#cb13-27" aria-hidden="true"></a>  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cbind">cbind</a></span>(ODE_male_hom, <span class="st">"sex"</span> =<span class="st"> "M"</span>)</span>
<span id="cb13-28"><a href="#cb13-28" aria-hidden="true"></a>)</span>
<span id="cb13-29"><a href="#cb13-29" aria-hidden="true"></a></span>
<span id="cb13-30"><a href="#cb13-30" aria-hidden="true"></a>ODE_out_inhom_melt<span class="op">$</span>death &lt;-<span class="st"> "inhomogeneous"</span></span>
<span id="cb13-31"><a href="#cb13-31" aria-hidden="true"></a>ODE_out_hom_melt<span class="op">$</span>death &lt;-<span class="st"> "homogeneous"</span></span>
<span id="cb13-32"><a href="#cb13-32" aria-hidden="true"></a></span>
<span id="cb13-33"><a href="#cb13-33" aria-hidden="true"></a><span class="co"># plot everything together</span></span>
<span id="cb13-34"><a href="#cb13-34" aria-hidden="true"></a><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="dt">data =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cbind">rbind</a></span>(ODE_out_inhom_melt, ODE_out_hom_melt) ) <span class="op">+</span></span>
<span id="cb13-35"><a href="#cb13-35" aria-hidden="true"></a><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_path</a></span>(<span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> time, <span class="dt">y =</span> value,</span>
<span id="cb13-36"><a href="#cb13-36" aria-hidden="true"></a>                <span class="dt">color =</span> death)) <span class="op">+</span></span>
<span id="cb13-37"><a href="#cb13-37" aria-hidden="true"></a><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html">facet_grid</a></span>(genotype <span class="op">~</span><span class="st"> </span>sex, <span class="dt">scales =</span> <span class="st">"free_y"</span>) <span class="op">+</span></span>
<span id="cb13-38"><a href="#cb13-38" aria-hidden="true"></a><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span>(<span class="dt">values =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"inhomogeneous"</span> =<span class="st"> "firebrick3"</span>,</span>
<span id="cb13-39"><a href="#cb13-39" aria-hidden="true"></a>                                <span class="st">"homogeneous"</span> =<span class="st"> "dodgerblue3"</span>)) <span class="op">+</span></span>
<span id="cb13-40"><a href="#cb13-40" aria-hidden="true"></a><span class="st">  </span><span class="co"># final formatting</span></span>
<span id="cb13-41"><a href="#cb13-41" aria-hidden="true"></a><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() <span class="op">+</span></span>
<span id="cb13-42"><a href="#cb13-42" aria-hidden="true"></a><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span>(<span class="st">"Constant vs Seasonally Time-Varying Mortality Rates"</span>) <span class="op">+</span></span>
<span id="cb13-43"><a href="#cb13-43" aria-hidden="true"></a><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/guides.html">guides</a></span>(<span class="dt">color =</span> <span class="ot">FALSE</span>)</span></code></pre></div>
<p><img src="inhomogeneous_files/figure-html/unnamed-chunk-14-1.png" width="691.2"></p>
</div>
<div id="sim_ctmc" class="section level2">
<h2 class="hasAnchor">
<a href="#sim_ctmc" class="anchor"></a>Simulation: CTMC Models</h2>
<p>We run 10 realizations of the stochastic simulations, to get a better sense of the overall variability in trajectories. We can run these repetitions in two ways: using the built-in repetitions provided by <strong>MGDrivE2</strong>, or running a loop over <code><a href="../reference/sim_trajectory_R.html">sim_trajectory_R()</a></code> to generate repetitions. Both of these are valuable, but in different circumstances. The built-in repetitions provided by <strong>MGDrivE2</strong> are deliberately single-threaded (as are all analysis routines), and more efficient than running single repetitions. This is useful for parameter sweeps, where several repetitions of each parameter needs to be run. However, for very large repetitions, it is not practical to run all of the repetitions on a single-thread. In this instance, and when running many parameter combinations, it is better to setup a parallel cluster, and evaluate multiple trajectories at once. For demonstration purposes, we will setup both methods.</p>
<p>We need a few extra parameters for the stochastic realizations, namely the number of realizations and the time-step. We will run 10 repetitions, using a time step equal to one hour.</p>
<p>First we simulate the time-inhomogeneous Markov model, using the built-in repetition wrapper. Please note that this vignette does not run the following simulation code to combat excessive build times; however the code can be run locally after building the Petri net and hazards as done previously in this vignette.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true"></a><span class="co"># parameters for stochastic simulations</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true"></a>num_rep &lt;-<span class="st"> </span><span class="dv">10</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true"></a>dt_stoch &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">/</span><span class="dv">24</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true"></a></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true"></a><span class="co"># run inhomogeneous simulations</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true"></a>PTS_out_inhom &lt;-<span class="st"> </span><span class="kw"><a href="../reference/sim_trajectory_R.html">sim_trajectory_R</a></span>(<span class="dt">x0 =</span> initialCons<span class="op">$</span>M0, <span class="dt">t0 =</span> <span class="dv">0</span>, <span class="dt">tt =</span> tmax, <span class="dt">dt =</span> dt,</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true"></a>                                  <span class="dt">dt_stoch =</span> dt_stoch, <span class="dt">S =</span> S, <span class="dt">num_reps =</span> num_rep,</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true"></a>                                  <span class="dt">hazards =</span> exact_hazards, <span class="dt">sampler =</span> <span class="st">"tau"</span>,</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true"></a>                                  <span class="dt">events =</span> events, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true"></a></span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true"></a><span class="co"># summarize females/males, add sex for plotting later</span></span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true"></a>PTS_female &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cbind">cbind</a></span>(<span class="kw"><a href="../reference/summarize_females.html">summarize_females</a></span>(<span class="dt">out =</span> PTS_out_inhom<span class="op">$</span>state, <span class="dt">spn_P =</span> SPN_P),</span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true"></a>                    <span class="st">"sex"</span> =<span class="st"> "F"</span>)</span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true"></a>PTS_male &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cbind">cbind</a></span>(<span class="kw"><a href="../reference/summarize_males.html">summarize_males</a></span>(<span class="dt">out =</span> PTS_out_inhom<span class="op">$</span>state), <span class="st">"sex"</span> =<span class="st"> "M"</span>)</span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true"></a></span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true"></a><span class="co"># reductions for plotting</span></span>
<span id="cb14-17"><a href="#cb14-17" aria-hidden="true"></a>PTS_inhom_par_unroll &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cbind">rbind</a></span>(PTS_female, PTS_male)</span>
<span id="cb14-18"><a href="#cb14-18" aria-hidden="true"></a></span>
<span id="cb14-19"><a href="#cb14-19" aria-hidden="true"></a><span class="co"># summarize_* functions return long-form data for ggplot2</span></span>
<span id="cb14-20"><a href="#cb14-20" aria-hidden="true"></a><span class="co">#  therefore, we take the first rep of data, the sim_time by num_genotypes</span></span>
<span id="cb14-21"><a href="#cb14-21" aria-hidden="true"></a><span class="co">#  Then, we reorganize the data into an array, and take the mean over every</span></span>
<span id="cb14-22"><a href="#cb14-22" aria-hidden="true"></a><span class="co">#  repetition</span></span>
<span id="cb14-23"><a href="#cb14-23" aria-hidden="true"></a>tot_time &lt;-<span class="st"> </span>tmax <span class="op">+</span><span class="st"> </span><span class="dv">1</span></span>
<span id="cb14-24"><a href="#cb14-24" aria-hidden="true"></a>num_geno &lt;-<span class="st"> </span>cube<span class="op">$</span>genotypesN</span>
<span id="cb14-25"><a href="#cb14-25" aria-hidden="true"></a>PTS_inhom_par_unroll_mean &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cbind">rbind</a></span>(</span>
<span id="cb14-26"><a href="#cb14-26" aria-hidden="true"></a>  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cbind">cbind</a></span>(PTS_female[<span class="dv">1</span><span class="op">:</span>(tot_time <span class="op">*</span><span class="st"> </span>num_geno), <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"genotype"</span>, <span class="st">"sex"</span>, <span class="st">"time"</span>)],</span>
<span id="cb14-27"><a href="#cb14-27" aria-hidden="true"></a>        <span class="st">"Mean"</span> =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/vector">as.vector</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colSums">rowMeans</a></span>(<span class="dt">x =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/array">array</a></span>(<span class="dt">data =</span> PTS_female<span class="op">$</span>value,</span>
<span id="cb14-28"><a href="#cb14-28" aria-hidden="true"></a>                                              <span class="dt">dim =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(tot_time, num_geno, num_rep)),</span>
<span id="cb14-29"><a href="#cb14-29" aria-hidden="true"></a>                                    <span class="dt">dims =</span> <span class="dv">2</span>)) ),</span>
<span id="cb14-30"><a href="#cb14-30" aria-hidden="true"></a>  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cbind">cbind</a></span>(PTS_male[<span class="dv">1</span><span class="op">:</span>(tot_time <span class="op">*</span><span class="st"> </span>num_geno), <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"genotype"</span>, <span class="st">"sex"</span>, <span class="st">"time"</span>)],</span>
<span id="cb14-31"><a href="#cb14-31" aria-hidden="true"></a>        <span class="st">"Mean"</span> =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/vector">as.vector</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colSums">rowMeans</a></span>(<span class="dt">x =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/array">array</a></span>(<span class="dt">data =</span> PTS_male<span class="op">$</span>value,</span>
<span id="cb14-32"><a href="#cb14-32" aria-hidden="true"></a>                                              <span class="dt">dim =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(tot_time, num_geno, num_rep)),</span>
<span id="cb14-33"><a href="#cb14-33" aria-hidden="true"></a>                                    <span class="dt">dims =</span> <span class="dv">2</span>)) )</span>
<span id="cb14-34"><a href="#cb14-34" aria-hidden="true"></a>  )</span></code></pre></div>
<p>For the time-homogeneous simulation, we will demonstrate how to simulate many trajectories in parallel.</p>
<p>There are many packages available for parallel evaluation in <code>R</code>. The default package within <code>R</code> is the <code>parallel</code> package. Other useful packages include <a href="https://cran.r-project.org/package=foreach">foreach</a>, <a href="https://cran.r-project.org/package=doSNOW">doSNOW</a>, and <a href="https://cran.r-project.org/package=future">future</a>. Because <code>parallel</code> is a base package, we use it to setup a socket cluster (this cluster works on Windows and nix systems) for this example. We then use <code><a href="https://www.rdocumentation.org/packages/parallel/topics/RngStream">parallel::clusterSetRNGStream</a></code> to set independent, reproducible streams of random numbers for each core in the cluster.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true"></a><span class="co"># parameters for stochastic simulations</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true"></a>num_core &lt;-<span class="st"> </span><span class="dv">1</span></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true"></a>M0 &lt;-<span class="st"> </span>initialCons<span class="op">$</span>M0</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true"></a></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true"></a><span class="co"># setup the cluster</span></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true"></a>cl &lt;-<span class="st"> </span>parallel<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/parallel/topics/makeCluster">makePSOCKcluster</a></span>(<span class="dt">names =</span> num_core)</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true"></a></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true"></a><span class="co"># set parallel seed</span></span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true"></a>parallel<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/parallel/topics/RngStream">clusterSetRNGStream</a></span>(<span class="dt">cl =</span> cl, <span class="dt">iseed =</span> 18438L)</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true"></a></span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true"></a><span class="co"># export required objects to each socket</span></span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true"></a>parallel<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/parallel/topics/clusterApply">clusterExport</a></span>(<span class="dt">cl=</span>cl, <span class="dt">varlist=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"M0"</span>, <span class="st">"tmax"</span>, <span class="st">"dt"</span>, <span class="st">"dt_stoch"</span>, <span class="st">"S"</span>,</span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true"></a>                                         <span class="st">"exact_hazards_hom"</span>, <span class="st">"events"</span>, <span class="st">"SPN_P"</span>))</span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true"></a></span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true"></a><span class="co"># load MGDrivE2 on each socket</span></span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true"></a>parallel<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/parallel/topics/clusterApply">clusterEvalQ</a></span>(<span class="dt">cl=</span>cl, <span class="dt">expr=</span>{<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(MGDrivE2)})</span></code></pre></div>
<p>Finally, we evaluate our function on each cluster using the load-balanced apply statement, <code><a href="https://www.rdocumentation.org/packages/parallel/topics/clusterApply">parallel::clusterApplyLB()</a></code>. This will run each repetition, analyze it, and store it in a list for further analysis. In this example, load-balancing is not particularly important, as these simulations are short and small, thus will have fairly consistent run times. Especially for parameter sweeps, some combinations will create significantly more complicated, and thus longer-running, dynamics, making load-balancing an important consideration.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true"></a><span class="co"># run homogeneous simulations</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true"></a>PTS_hom_par &lt;-<span class="st"> </span>parallel<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/parallel/topics/clusterApply">clusterApplyLB</a></span>(<span class="dt">cl=</span>cl, <span class="dt">x=</span><span class="dv">1</span><span class="op">:</span>num_rep, <span class="dt">fun=</span><span class="cf">function</span>(x){</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true"></a>  <span class="co"># run trajectory</span></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true"></a>  PTS_out_hom &lt;-<span class="st"> </span><span class="kw"><a href="../reference/sim_trajectory_R.html">sim_trajectory_R</a></span>(<span class="dt">x0 =</span> M0, <span class="dt">t0 =</span> <span class="dv">0</span>, <span class="dt">tt =</span> tmax, <span class="dt">dt =</span> dt,</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true"></a>                                  <span class="dt">dt_stoch =</span> dt_stoch, <span class="dt">S =</span> S, <span class="dt">hazards =</span> exact_hazards_hom,</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true"></a>                                  <span class="dt">sampler =</span> <span class="st">"tau"</span>,<span class="dt">events =</span> events, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true"></a></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true"></a>  <span class="co"># summarize females/males for plotting  later</span></span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true"></a>  f_sum &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cbind">cbind</a></span>(<span class="kw"><a href="../reference/summarize_females.html">summarize_females</a></span>(<span class="dt">out =</span> PTS_out_hom<span class="op">$</span>state, <span class="dt">spn_P =</span> SPN_P),</span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true"></a>                 <span class="st">"rep"</span>=x, <span class="st">"sex"</span>=<span class="st">"F"</span>)</span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true"></a>  m_sum &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cbind">cbind</a></span>(<span class="kw"><a href="../reference/summarize_males.html">summarize_males</a></span>(<span class="dt">out =</span> PTS_out_hom<span class="op">$</span>state), <span class="st">"rep"</span>=x, <span class="st">"sex"</span>=<span class="st">"M"</span>)</span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true"></a></span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true"></a>  <span class="co"># return single dataframe</span></span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true"></a>  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/function">return</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cbind">rbind</a></span>(f_sum, m_sum))</span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true"></a>})</span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true"></a></span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true"></a><span class="co"># stop cluster</span></span>
<span id="cb16-18"><a href="#cb16-18" aria-hidden="true"></a>parallel<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/parallel/topics/makeCluster">stopCluster</a></span>(cl)</span>
<span id="cb16-19"><a href="#cb16-19" aria-hidden="true"></a></span>
<span id="cb16-20"><a href="#cb16-20" aria-hidden="true"></a><span class="co"># summarize data</span></span>
<span id="cb16-21"><a href="#cb16-21" aria-hidden="true"></a><span class="co">#  the Reduce() call generates a mean of the populations</span></span>
<span id="cb16-22"><a href="#cb16-22" aria-hidden="true"></a>PTS_hom_par_unroll &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/do.call">do.call</a></span>(<span class="dt">what =</span> rbind, <span class="dt">args =</span> PTS_hom_par)</span>
<span id="cb16-23"><a href="#cb16-23" aria-hidden="true"></a>PTS_hom_par_unroll_mean &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cbind">cbind</a></span>(PTS_hom_par[[<span class="dv">1</span>]][ ,<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"genotype"</span>, <span class="st">"sex"</span>, <span class="st">"time"</span>)],</span>
<span id="cb16-24"><a href="#cb16-24" aria-hidden="true"></a>                                 <span class="st">"Mean"</span> =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/funprog">Reduce</a></span>(<span class="dt">f =</span> <span class="st">"+"</span>, <span class="dt">x =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">lapply</a></span>(<span class="dt">X =</span> PTS_hom_par,</span>
<span id="cb16-25"><a href="#cb16-25" aria-hidden="true"></a>                                                                     <span class="dt">FUN =</span> <span class="cf">function</span>(x){x[,<span class="dv">3</span>]})</span>
<span id="cb16-26"><a href="#cb16-26" aria-hidden="true"></a>                                                 )<span class="op">/</span>num_rep)</span></code></pre></div>
</div>
<div id="plots" class="section level2">
<h2 class="hasAnchor">
<a href="#plots" class="anchor"></a>Results</h2>
<p>Finally we combine the data into one plot. The ODE trajectories are plotted as dashed lines, the mean of the stochastic simulations as solid lines, and each individual stochastic trajectory is plotted as a faint curve. Lines in red are the time-inhomogeneous model runs and blue are the time-homogeneous simulations.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true"></a><span class="co"># add mortality for plotting</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true"></a>PTS_inhom_par_unroll<span class="op">$</span>death &lt;-<span class="st"> "inhomogeneous"</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true"></a>PTS_hom_par_unroll<span class="op">$</span>death &lt;-<span class="st"> "homogeneous"</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true"></a>ODE_out_inhom_melt<span class="op">$</span>death &lt;-<span class="st"> "inhomogeneous"</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true"></a>ODE_out_hom_melt<span class="op">$</span>death &lt;-<span class="st"> "homogeneous"</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true"></a></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true"></a><span class="co"># plot everything together</span></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true"></a><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="dt">data =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cbind">rbind</a></span>(PTS_inhom_par_unroll, PTS_hom_par_unroll) ) <span class="op">+</span></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true"></a><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_path</a></span>(<span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> time, <span class="dt">y =</span> value, <span class="dt">group =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/interaction">interaction</a></span>(rep, death),</span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true"></a>                <span class="dt">color =</span> death), <span class="dt">alpha =</span> <span class="fl">0.15</span>) <span class="op">+</span></span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true"></a><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html">facet_grid</a></span>(genotype <span class="op">~</span><span class="st"> </span>sex, <span class="dt">scales =</span> <span class="st">"free_y"</span>) <span class="op">+</span></span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true"></a><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span>(<span class="dt">values =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"inhomogeneous"</span> =<span class="st"> "firebrick3"</span>,</span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true"></a>                                <span class="st">"homogeneous"</span> =<span class="st"> "dodgerblue3"</span>)) <span class="op">+</span></span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true"></a><span class="st">  </span><span class="co"># inhomogeneous mean</span></span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true"></a><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span>(<span class="dt">data =</span> PTS_inhom_par_unroll_mean, <span class="dt">mapping =</span> <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> time, <span class="dt">y =</span> Mean),</span>
<span id="cb17-16"><a href="#cb17-16" aria-hidden="true"></a>            <span class="dt">color =</span> <span class="st">"firebrick4"</span>) <span class="op">+</span></span>
<span id="cb17-17"><a href="#cb17-17" aria-hidden="true"></a><span class="st">  </span><span class="co"># homogeneous mean</span></span>
<span id="cb17-18"><a href="#cb17-18" aria-hidden="true"></a><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span>(<span class="dt">data =</span> PTS_hom_par_unroll_mean, <span class="dt">mapping =</span> <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> time, <span class="dt">y =</span> Mean),</span>
<span id="cb17-19"><a href="#cb17-19" aria-hidden="true"></a>            <span class="dt">color =</span> <span class="st">"dodgerblue4"</span>) <span class="op">+</span></span>
<span id="cb17-20"><a href="#cb17-20" aria-hidden="true"></a><span class="st">  </span><span class="co"># ODE inhomogeneous</span></span>
<span id="cb17-21"><a href="#cb17-21" aria-hidden="true"></a><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span>(<span class="dt">data =</span> ODE_out_inhom_melt, <span class="dt">mapping =</span> <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> time, <span class="dt">y =</span> value),</span>
<span id="cb17-22"><a href="#cb17-22" aria-hidden="true"></a>            <span class="dt">linetype =</span> <span class="dv">2</span>, <span class="dt">color =</span> <span class="st">"firebrick4"</span>) <span class="op">+</span></span>
<span id="cb17-23"><a href="#cb17-23" aria-hidden="true"></a><span class="st">  </span><span class="co"># ODE homogeneous</span></span>
<span id="cb17-24"><a href="#cb17-24" aria-hidden="true"></a><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span>(<span class="dt">data =</span> ODE_out_hom_melt, <span class="dt">mapping =</span> <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> time, <span class="dt">y =</span> value),</span>
<span id="cb17-25"><a href="#cb17-25" aria-hidden="true"></a>            <span class="dt">linetype =</span> <span class="dv">2</span>, <span class="dt">color =</span> <span class="st">"dodgerblue4"</span>) <span class="op">+</span></span>
<span id="cb17-26"><a href="#cb17-26" aria-hidden="true"></a><span class="st">  </span><span class="co"># final formatting</span></span>
<span id="cb17-27"><a href="#cb17-27" aria-hidden="true"></a><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() <span class="op">+</span></span>
<span id="cb17-28"><a href="#cb17-28" aria-hidden="true"></a><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span>(<span class="st">"Constant vs Seasonally Time-Varying Mortality Rates"</span>) <span class="op">+</span></span>
<span id="cb17-29"><a href="#cb17-29" aria-hidden="true"></a><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/guides.html">guides</a></span>(<span class="dt">color =</span> <span class="ot">FALSE</span>)</span></code></pre></div>
<p>There is a clear difference in population size over the course of a year. Since we used the average mortality rate to calculate the equilibrium, we see that the time-inhomogeneous simulations center around our time-homogeneous ones. Comparing the dashed-lines (ODE trajectories) to the solid-lines (SPN averages), we see that our stochastic realizations closely follow their deterministic solutions. However, the faint lines indicate significant variation is possible.</p>
<div id="ref" class="section level3">
<h3 class="hasAnchor">
<a href="#ref" class="anchor"></a>References</h3>
<ul>
<li>Anderson, David F. “A modified next reaction method for simulating chemical systems with time dependent propensities and delays.” The Journal of chemical physics 127.21 (2007): 214107.</li>
<li>Thanh, Vo Hong, and Corrado Priami. “Simulation of biochemical reactions with time-dependent rates by the rejection-based algorithm.” The Journal of chemical physics 143.5 (2015): 08B601_1.</li>
<li>Allen, Linda JS. “A primer on stochastic epidemic models: Formulation, numerical simulation, and analysis.” Infectious Disease Modelling 2.2 (2017): 128-142.</li>
</ul>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#table-of-contents">Table of Contents</a></li>
      <li><a href="#inhom_adult_mort">Time-Inhomogeneous Adult Mortality</a></li>
      <li>
<a href="#inhom_rate_fns">Time-dependent Rate Functions</a>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#math_consider">Mathematical Considerations</a></li>
      <li><a href="#programming">Programming Time-Dependent Rates in MGDrivE2</a></li>
      </ul>
</li>
      <li>
<a href="#make_spn">Petri Net Setup</a>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#pars">Parameterization</a></li>
      <li><a href="#init_pn">Initialization of the Petri Net</a></li>
      <li><a href="#equilibria_haz">Equilibrium Conditions and Hazard Functions</a></li>
      </ul>
</li>
      <li>
<a href="#sim">Simulation of Fully Specified SPN Model</a>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#sim_ode">Simulation: ODE Models</a></li>
      <li><a href="#sim_ctmc">Simulation: CTMC Models</a></li>
      <li><a href="#plots">Results</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="http://www.slwu89.com">Sean L. Wu</a>, <a href="https://github.com/GilChrist19">Jared B. Bennett</a>, <a href="https://www.marshalllab.com">John M. Marshall</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
