<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>MGDrivE Mathematical Description • MGDrivE</title>
<!-- mathjax math --><script src="https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.js" integrity="sha256-MASABpB4tYktI2Oitl4t+78w/lyA+D7b/s9GEP0JOGI=" crossorigin="anonymous"></script><script>
  window.MathJax = {
    chtml: {
      fontURL: "https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2"
    }
  };
</script><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.1/jquery-3.6.1.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="MGDrivE Mathematical Description">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MGDrivE</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.6.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><a class="dropdown-item" href="../articles/mgdrive_examples.html">Example Simulations</a></li>
    <li><a class="dropdown-item" href="../articles/mgdrive_math.html">Mathematical Description</a></li>
    <li><a class="dropdown-item" href="../articles/mgdrive_run.html">Detailed Run Example</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/MarshallLab/MGDrivE" aria-label="View on GitHub"><span class="fa fa-lg fa-github"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>MGDrivE Mathematical Description</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/MarshallLab/MGDrivE/blob/HEAD/vignettes/mgdrive_math.Rmd" class="external-link"><code>vignettes/mgdrive_math.Rmd</code></a></small>
      <div class="d-none name"><code>mgdrive_math.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Recent developments of CRISPR-Cas9 based homing endonuclease gene
drive systems, either for the suppression or replacement of mosquito
populations, have generated much interest in their use for control of
mosquito-borne diseases (such as dengue, malaria, Chikungunya and Zika).
This is because genetic control of pathogen transmission may complement
or even substitute traditional vector-control interventions, which have
had limited success in bringing the spread of these diseases to a
halt.</p>
<p>Despite excitement for the use of gene drives for mosquito control,
current modeling efforts have analyzed only a handful of these new
approaches (usually studying just one per framework). Moreover, these
models usually consider well-mixed populations with no explicit spatial
dynamics. To this end, we are developing <strong>MGDrivE</strong>
(Mosquito Gene DRIVe Explorer), in cooperation with the ‘UCI Malaria
Elimination Initiative’, as a flexible modeling framework to evaluate a
variety of drive systems in spatial networks of mosquito populations.
This framework provides a reliable testbed to evaluate and optimize the
efficacy of gene drive mosquito releases. What separates
<strong>MGDrivE</strong> from other models is the incorporation of
mathematical and computational mechanisms to simulate a wide array of
inheritance-based technologies within the same, coherent set of
equations.</p>
<p>We do this by treating the population dynamics, genetic inheritance
operations, and migration between habitats as separate processes coupled
together through the use of mathematical tensor operations. This way we
can conveniently swap inheritance patterns whilst still making use of
the same set of population dynamics equations. This is a crucial
advantage of our system, as it allows other research groups to test
their ideas without developing new models and without the need to spend
time adapting other frameworks to suit their needs.</p>
<div class="section level3">
<h3 id="brief-description">Brief Description<a class="anchor" aria-label="anchor" href="#brief-description"></a>
</h3>
<p><strong>MGDrivE</strong> is based on the idea that we can decouple
the genotype inheritance process from the population dynamics equations.
This allows the system to be treated and developed in three
semi-independent modules that come together to form the system.</p>
<div class="figure" style="text-align: center">
<img src="diagram.png" alt="Fig. 1: **MGDrivE** Conceptual Model Framework" width="80%"><p class="caption">
Fig. 1: <strong>MGDrivE</strong> Conceptual Model Framework
</p>
</div>
<div class="section level4">
<h4 id="previous-work">Previous Work:<a class="anchor" aria-label="anchor" href="#previous-work"></a>
</h4>
<p>The original version of this model was based on work by Deredec et
al. (2011) and Hancock &amp; Godfray (2007), and adapted to accommodate
CRISPR homing dynamics in a previous publication by our team (Marshall
et al. 2017). As described, we extended this framework to be able to
handle a variable number of genotypes, and migration across spatial
scenarios. We did this by adapting the equations to work in a
tensor-oriented manner, where each genotype can have different processes
affecting their particular strain (death rates, mating fitness,
sex-ratio bias, et cetera).</p>
</div>
<div class="section level4">
<h4 id="notation-and-conventions">Notation and Conventions<a class="anchor" aria-label="anchor" href="#notation-and-conventions"></a>
</h4>
<p>Before beginning the full description of the model we will define
some of the conventions we followed for the notation of the written
description of the system.</p>
<ul>
<li>Overlines are used to denote the dimension of a tensor</li>
<li>Subscript brackets are used to indicate an element in time. For
example: <span class="math inline">\(L_{[t-1]}\)</span> is the larval
population at time: <span class="math inline">\(t-1\)</span>.</li>
<li>Parentheses are used to indicate the parameter(s) of a function. For
example: <span class="math inline">\(\overline{O(T_{e}+T_{l})}\)</span>
represents the function <span class="math inline">\(O\)</span> evaluated
with the parameter: <span class="math inline">\(T_{e}+T_{l}\)</span>
</li>
<li>Matrices follow a ‘row-first’ indexing order (<span class="math inline">\(i\)</span>: row, <span class="math inline">\(j\)</span>: column)</li>
</ul>
<p>In the case of one dimensional tensors, each slot represents a
genotype of the population. For example, the male population is stored
in the following way: <span class="math display">\[ \overline{Am} =
\left(\begin{array}{c} g_1 \\ g_2 \\ g_3 \\ \vdots \\ g_n
\end{array}\right) _{i} \]</span></p>
<p>All the processes that affect mosquitoes in a genotype-specific way
are defined and stored in this way within the framework.</p>
<p>There are two tensors of squared dimensionality in the model: the
adult females matrix, and the genotype-specific male-mating ability,
<span class="math inline">\(\overline{\overline{\eta}}\)</span>. In the
case of the former, the rows represent the females’ genotype, whilst the
columns represent the genotype of the male they mated with: <span class="math display">\[ \overline{\overline{Af}} =
\left(\begin{array}{ccccc} g_{11} &amp; g_{12} &amp; g_{13} &amp; \cdots
&amp; g_{1n}\\ g_{21} &amp; g_{22} &amp; g_{23} &amp; \cdots &amp;
g_{2n}\\ g_{31} &amp; g_{32} &amp; g_{33} &amp; \cdots &amp; g_{3n}\\
\vdots &amp; \vdots &amp; \vdots &amp; \ddots &amp; \vdots\\ g_{n1}
&amp; g_{n2} &amp; g_{n3} &amp; \cdots &amp; g_{nn} \end{array}\right)
_{ij} \]</span></p>
<p>The genotype-specific male-mating ability, on the other hand, stores
the mothers’ genotype in the rows, the male mate genotype as columns,
and each value indicates the ability of a specific male genotype to mate
with a specific female genotype. This allows assortative mating to be
modeled.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="mathematical-framework">Mathematical Framework<a class="anchor" aria-label="anchor" href="#mathematical-framework"></a>
</h2>
<div class="section level3">
<h3 id="inheritance-cube">Inheritance Cube<a class="anchor" aria-label="anchor" href="#inheritance-cube"></a>
</h3>
<p>To model an arbitrary number of genotypes efficiently in the same
mathematical framework we use a 3-dimensional array structure (cube)
where each axis represents the following information:</p>
<ul>
<li>x: female adult mate genotype</li>
<li>y: male adult mate genotype</li>
<li>z: proportion of the offspring that inherits a given genotype
(slice)</li>
</ul>
<p>The cube structure gives us the flexibility to apply tensor
operations to the elements within our equations, so that we can
calculate the stratified population dynamics rapidly; and within a
readable, flexible computational framework. This becomes apparent when
we define the equation we use for the computation of eggs laid at any
given point in time:</p>
<p><span class="math display">\[ \overline{O(T_x)} = \sum_{j=1}^{n}
\Bigg( \bigg( (\beta*\overline{s} * \overline{ \overline{Af_{[t-T_x]}}})
* \overline{\overline{\overline{Ih}}} \bigg) *
\Lambda  \Bigg)^{\top}_{ij} \]</span></p>
<p>In this equation, the matrix containing the number of mated adult
females <span class="math inline">\((\overline{\overline{Af}})\)</span>
is multiplied element-wise with each one of the layers containing the
eggs genotypes proportions expected from this cross <span class="math inline">\((\overline{\overline{\overline{Ih}}})\)</span>.</p>
<p>The resulting matrix is then multiplied by a binary ‘viability mask’
<span class="math inline">\((\Lambda)\)</span> that filters out
female-parent to offspring genetic combinations that are not viable due
to biological impediments (such as cytoplasmic incompatibility).</p>
<p>The summation of the transposed resulting matrix returns us the total
fraction of eggs resulting from all the male to female genotype crosses
(<span class="math inline">\(\overline{O(T_x)}\)</span>).</p>
<p>Note: For inheritance operations to be consistent within the
framework the summation of each element in the z-axis (this is, the
proportions of each one of the offspring’s genotypes) must be equal to
one.</p>
<div class="section level4">
<h4 id="drive-specific-cubes">Drive-specific Cubes<a class="anchor" aria-label="anchor" href="#drive-specific-cubes"></a>
</h4>
<p>An inheritance cube in an array object that specifies inheritance
probabilities (offspring genotype probability) stratified by male and
female parent genotypes. <strong>MGDrivE</strong> provides the following
cubes to model different gene drive systems:</p>
<ul>
<li><a href="../reference/cubeOneLocusTA.html">1 Locus
Maternal-Toxin/Zygotic-Antidote System</a></li>
<li><a href="../reference/cubeTwoLocusTA.html">2 Locus
Maternal-Toxin/Zygotic-Antidote System</a></li>
<li><a href="../reference/cubeAlleleSail.html">Allele Sail</a></li>
<li><a href="../reference/cubeASmidler.html">Split Drive with
Parental-Specific Impacts</a></li>
<li><a href="../reference/cubeCLEAVRMF.html">DsX-based Cleave and Rescue
(ClvR) - Akbari Lab</a></li>
<li><a href="../reference/cubeCLEAVRX.html">X-Linked Cleave and Rescue
(ClvR)</a></li>
<li><a href="../reference/cubeClvR.html">1 Locus Cleave and Rescue
(ClvR)</a></li>
<li><a href="../reference/cubeClvR2.html">2 Locus Cleave and Rescue
(ClvR)</a></li>
<li><a href="../reference/cubeConfinableHoming.html">2 Locus Confinable
Homing</a></li>
<li><a href="../reference/cubeConfinableHomingX.html">X-Linked
Confinable Homing</a></li>
<li><a href="../reference/cubeHoming1RA.html">Homing Drive with 1
Resistance Allele</a></li>
<li><a href="../reference/cubeHomingDrive.html">CRISPR (Clustered
Regularly Interspaced Short Palindromic Repeats) with 2 Resistance
Alleles and Maternal Deposition</a></li>
<li><a href="../reference/cubeHomingDriveSM.html">Small-Molecule Induced
CRISPR</a></li>
<li><a href="../reference/cubeXHomingDeposition.html">X-Linked
CRISPR</a></li>
<li><a href="../reference/cubeECHACR.html">Autosomal ECHACR</a></li>
<li><a href="../reference/cubeECHACRX.html">X-Linked ECHACR</a></li>
<li><a href="../reference/cubeImmunizingReversalMF.html">Autosomal
Immunizing Reversal System</a></li>
<li><a href="../reference/cubeImmunizingReversalX.html">X-Linked
Autosomal Immunizing Reversal System</a></li>
<li><a href="../reference/cubeKillerRescue.html">Killer-Rescue
System</a></li>
<li><a href="../reference/cubeMEDEA.html">MEDEA (Maternal Effect
Dominant Embryonic Arrest)</a></li>
<li><a href="../reference/cubeReciprocalTranslocations.html">Reciprocal
Translocations</a></li>
<li><a href="../reference/cubeRIDL.html">RIDL (Release of Insects with
Dominant Lethality)</a></li>
<li><a href="../reference/cubeXShredderMF.html">Autosomal
X-Shredder</a></li>
<li><a href="../reference/cubeXShredderY.html">Y-Linked
X-Shredder</a></li>
<li><a href="../reference/cubeMendelian.html">Mendelian</a></li>
<li><a href="../reference/cubeSplitDrive.html">Autosomal
Split-Drive</a></li>
<li><a href="../reference/cubeSplitDriveX.html">X-Linked
Split-Drive</a></li>
<li><a href="../reference/cubeSplitDriveY.html">Y-Linked
Split-Drive</a></li>
<li><a href="../reference/cubeTGD.html">trans-Complementing Gene
Drive</a></li>
<li><a href="../reference/cubeTGDX.html">X-Linked trans-Complementing
Gene Drive</a></li>
<li><a href="../reference/cubeWolbachia.html">Wolbachia</a></li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="population-dynamics">Population Dynamics<a class="anchor" aria-label="anchor" href="#population-dynamics"></a>
</h3>
<p>During the three aquatic stages, a density-independent mortality
process takes place at each stage (<em>st</em>): <span class="math display">\[ \theta_{st}=(1-\mu_{st})^{T_{st}} \]</span>
Along with a density dependent process dependent on the number of larvae
in the environment: <span class="math display">\[
F(L[t])=\Bigg(\frac{\alpha}{\alpha+\sum{\overline{L[t]}}}\Bigg)^{1/T_l}
\]</span> where <span class="math inline">\(\alpha\)</span> represents
the strength of the density-dependent process. This parameter is
calculated with: <span class="math display">\[ \alpha=\Bigg( \frac{1/2 *
\beta * \theta_e * Ad_{eq}}{R_m-1} \Bigg) * \Bigg( \frac{1-(\theta_l /
R_m)}{1-(\theta_l / R_m)^{1/T_l}} \Bigg) \]</span> in which <span class="math inline">\(\beta\)</span> is the species’ fertility in the
absence of gene-drive, <span class="math inline">\(Ad_{eq}\)</span> is
the adult mosquito population equilibrium size, and <span class="math inline">\(R_{m}\)</span> is the population growth in the
absence of density-dependent mortality. This population growth is
calculated with the average generation time (<span class="math inline">\(g\)</span>), the adult mortality rate (<span class="math inline">\(\mu_{ad}\)</span>), and the daily population
growth rate (<span class="math inline">\(r_{m}\)</span>): <span class="math display">\[
g=T_{e}+T_{l}+T_{p}+\frac{1}{\mu_{ad}}\\R_{m}=(r_{m})^{g} \]</span></p>
<div class="section level4">
<h4 id="larval-stages">Larval Stages<a class="anchor" aria-label="anchor" href="#larval-stages"></a>
</h4>
<p>The computation of the larval stage in the population is crucial to
the model because the density dependent processes necessary for
equilibrium trajectories occur here. This calculation is performed with
the following equation: <span class="math display">\[
D(\theta_l,T_x) = \left\{
         \begin{array}{ll}
             \theta_{l[0]}^{'}=\theta_l                                 &amp;
\quad i = 0 \\
             \theta_{l[i+1]}^{'} = \theta_{l[i]}^{'}
*F(\overline{L_{[t-i-T_x]}})   &amp; \quad i \leq T_l
         \end{array}
         \right.
\]</span></p>
<p>In addition to this, we need the larval mortality (<span class="math inline">\(\mu_{l}\)</span>): <span class="math display">\[
    %L_{eq}=&amp;\alpha*\lfloor R_{m} -1\rfloor
    %&amp;
    \mu_{l}=1-\Bigg( \frac{R_{m} * \mu_{ad}}{1/2 * \beta * (1-\mu_{m})}
\Bigg)^{\frac{1}{T_{e}+T_{l}+T_{p}}}
\]</span></p>
<p>With these mortality processes, we are now able to calculate the
larval population: <span class="math display">\[
    \overline{L_{[t]}}=
        \overline{L_{[t-1]}} * (1-\mu_{l}) * F(\overline{L_{[t-1]})}\\
        +\overline{O(T_{e})}* \theta_{e} \\
        %+\overline{\beta}* \theta_{e} *
(\overline{\overline{Af_{(t-T_{e})}}} \circ
\overline{\overline{\overline{Ih}}})\\
        - \overline{O(T_{e}+T_{l})} * \theta_{e} * D(\theta_{l},0)
        %\prod_{i=1}^{T_{l}} F(\overline{L_{[t-i]}})
        %\theta_{l}
\]</span> where the first term accounts for larvae surviving one day to
the other; the second therm accounts for the eggs that have hatched
within the same period of time; and the last term computes the number of
larvae that have transformed into pupae.</p>
</div>
<div class="section level4">
<h4 id="adult-stages">Adult Stages<a class="anchor" aria-label="anchor" href="#adult-stages"></a>
</h4>
<p>We are ultimately interested in calculating how many adults of each
genotype exist at any given point in time. For this, we first calculate
the number of eggs that are laid and survive to the adult stages with
the equation: <span class="math display">\[
    \overline{E^{'}}= \overline{O(T_{e}+T_{l}+T_{p})} \\
    * \bigg(\overline{\xi_{m}} * (\theta_{e} * \theta_{p}) *
(1-\mu_{ad}) * D(\theta_{l},T_{p}) \bigg)
\]</span></p>
<p>With this information we can calculate the current number of male
adults in the population by computing the following equation: <span class="math display">\[
    \overline{Am_{[t]}}=
        \overline{Am_{[t-1]}} * (1-\mu_{ad})*\overline{\omega_{m}}\\
        + (1-\overline{\phi}) *  \overline{E^{'}}\\
        + \overline{\nu m_{[t-1]}}
\]</span> in which the first term represents the number of males
surviving from one day to the next; the second term, the fraction of
males that survive to adulthood (<span class="math inline">\(\overline{E'}\)</span>) and emerge as males
(<span class="math inline">\(1-\phi\)</span>); the last term is used to
add males into the population as part of gene-drive release
campaigns.</p>
<p>Female adult populations are calculated in a similar way: <span class="math display">\[
  \overline{\overline{Af_{[t]}}}=
    \overline{\overline{Af_{[t-1]}}} * (1-\mu_{ad}) *
\overline{\omega_{f}}\\
    +  \bigg( \overline{\phi} * \overline{E^{'}}+\overline{\nu
f_{[t-1]}}\bigg)^{\top} * \bigg(
\frac{\overline{\overline{\eta}}*\overline{Am_{[t-1]}}}{\sum{\overline{Am_{[t-1]}}}}
\bigg)%\overline{\overline{Mf}}
\]</span> where we first compute the surviving female adults from one
day to the next; and then we calculate the mating composition of the
female fraction emerging from pupa stage. To do this, we obtain the
surviving fraction of eggs that survive to adulthood (<span class="math inline">\(\overline{E'}\)</span>) and emerge as females
(<span class="math inline">\(\phi\)</span>), we then add the new females
added as a result of gene-drive releases (<span class="math inline">\(\overline{\nu f_{[t-1]}}\)</span>). After doing
this, we calculate the proportion of males that are allocated to each
female genotype, taking into account their respective mating fitnesses
(<span class="math inline">\(\overline{\overline{\eta}}\)</span>) so
that we can introduce the new adult females into the population
pool.</p>
</div>
</div>
<div class="section level3">
<h3 id="gene-drive-releases-and-effects">Gene Drive Releases and Effects<a class="anchor" aria-label="anchor" href="#gene-drive-releases-and-effects"></a>
</h3>
<p>As was briefly mentioned before, we are including the option to
release both male and/or female individuals into the populations.
Another important thing to emphasize is that we allow flexible releases
sizes and schedules. Our model handles releases internally as lists of
population compositions, so it is possible to have releases performed at
irregular intervals and with different numbers of mosquito genetic
compositions as long as no new genotypes are introduced (which have not
been previously defined in the inheritance cube). <span class="math display">\[
  \overline{\nu} = \bigg\{
    \left(\begin{array}{c} g_1 \\ g_2 \\ g_3 \\ \vdots \\ g_n
\end{array}\right)_{t=1} ,
    \left(\begin{array}{c} g_1 \\ g_2 \\ g_3 \\ \vdots \\ g_n
\end{array}\right)_{t=2} ,
    \cdots ,
    \left(\begin{array}{c} g_1 \\ g_2 \\ g_3 \\ \vdots \\ g_n
\end{array}\right)_{t=x}
    \bigg\}
\]</span></p>
<p>So far, however, we have not described the way in which the effects
of these gene-drives are included into the mosquito populations
dynamics. This is done through the use of various modifiers included in
the equations:</p>
<ul>
<li>
<span class="math inline">\(\overline{\omega}\)</span>: Relative
increase in mortality (zero being full mortality effects and one no
mortality effect)</li>
<li>
<span class="math inline">\(\overline{\phi}\)</span>: Relative shift
in the sex of the pupating mosquitoes (zero biases the sex ratio towards
males, whilst 1 biases the ratio towards females).</li>
<li>
<span class="math inline">\(\overline{\overline{\eta}}\)</span>:
Standardized mating fitness (zero being complete fitness ineptitude, and
one being regular mating skills).</li>
<li>
<span class="math inline">\(\overline{\beta}\)</span>: Fecundity
(average number of eggs laid per day).</li>
<li>
<span class="math inline">\(\overline{\xi}\)</span>: Pupation
success (zero being full mortality and one full pupation success).</li>
</ul>
</div>
<div class="section level3">
<h3 id="migration">Migration<a class="anchor" aria-label="anchor" href="#migration"></a>
</h3>
<p>To simulate migration within our framework we are considering patches
(or nodes) of fully-mixed populations in a network structure. This
allows us to handle mosquito movement across spatially-distributed
populations with a transitions matrix, which is calculated with the
tensor outer product of the genotypes population tensors and the
transitions matrix of the network as follows: <span class="math display">\[
    \overline{Am_{(t)}^{i}}=
        \sum{\overline{A_{m}^j} \otimes \overline{\overline{\tau
m_{[t-1]}}}} \\
        \overline{\overline{Af_{(t)}^{i}}}=
      \sum{\overline{\overline{A_{f}^j}} \otimes
\overline{\overline{\tau f_{[t-1]}}}}
\]</span></p>
<p>In these equations the new population of the patch <span class="math inline">\(i\)</span> is calculated by summing the migrating
mosquitoes of all the <span class="math inline">\(j\)</span> patches
across the network defined by the transitions matrix <span class="math inline">\(\tau\)</span>, which stores the mosquito migration
probabilities from patch to patch. It is worth noting that the migration
probabilities matrices can be different for males and females; and that
there’s no inherent need for them to be static (the migration
probabilities may vary over time to accommodate wind changes due to
seasonality).</p>
</div>
</div>
<div class="section level2">
<h2 id="stochasticity">Stochasticity<a class="anchor" aria-label="anchor" href="#stochasticity"></a>
</h2>
<p><strong>MGDrivE</strong> allows all inheritance, migration, and
population dynamics processes to be simulated stochastically; this
accounts for the inherent probabilistic nature of the processes
governing the interactions and life-cycles of organisms. In the next
section, we will describe all the stochastic processes that can be
activated in the program. It should be noted that all of these can be
turned on and off independently from one another as required by the
researcher.</p>
<div class="section level4">
<h4 id="mosquito-biology">Mosquito Biology<a class="anchor" aria-label="anchor" href="#mosquito-biology"></a>
</h4>
<p><strong>Oviposition</strong></p>
<p>Stochastic egg laying by female/male pairs is separated into two
steps: calculating the number of eggs laid by the females and then
distributing laid eggs according to their genotypes. The number of eggs
laid follows a Poisson distribution conditioned on the number of
female/male pairs and the fertility of each female.</p>
<p><span class="math display">\[ Poisson( \lambda =
numFemales*Fertility) \]</span></p>
<p>Multinomial sampling, conditioned on the number of offspring and the
relative viability of each genotype, determines the genotypes of the
offspring.</p>
<p><span class="math display">\[ Multinomial \left(numOffspring, p_1,
p_2\dots p_b \right)=\frac{numOffspring!}{p_1!\,p_2\,\dots
p_n}p_1^{n_1}p_2^{n_2}\dots p_n^{n_n} \]</span></p>
<p><strong>Sex Determination</strong></p>
<p>Sex of the offspring is determined by multinomial sampling. This is
conditioned on the number of eggs that live to hatching and a
probability of being female, allowing the user to design systems that
skew the sex ratio of the offspring through reproductive mechanisms.</p>
<p><span class="math display">\[ Multinomial(numHatchingEggs,
p_{female}, p_{female}) \]</span></p>
<p><strong>Mating</strong></p>
<p>Stochastic mating is determined by multinomial sampling conditioned
on the number of males and their fitness. It is assumed that females
mate only once in their life, therefore each female will sample from the
available males and be done, while the males are free to potentially
mate with multiple females. The males’ ability to mate is modulated with
a fitness term, thereby allowing some genotypes to be less fit than
others (as seen often with lab releases).</p>
<p><span class="math display">\[ Multinomial(numFemales, p_1f_1, p_2f_2,
\dots p_nf_n) \]</span></p>
<p><strong>Other Stochastic Processes</strong></p>
<p>All remaining stochastic processes (larval survival, hatching
,pupating, surviving to adult hood) are determined by binomial sampling
conditioned on factors affecting the current life stage. These factors
are determined empirically from mosquito population data.</p>
<p><strong>Migration</strong></p>
<p>Variance of stochastic movement (not used in diffusion model of
migration).</p>
</div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<ul>
<li>Deredec A, Godfray HCJ, Burt A (2011). “Requirements for effective
malaria control with homing endonuclease genes.” <em>Proceedings of the
National Academy of Sciences of the United States of America</em>,
108(43), E874–80. ISSN 1091-6490, doi: 10.1073/pnas.1110717108 , <a href="https://www.pnas.org/content/108/43/E874" class="external-link uri">https://www.pnas.org/content/108/43/E874</a>.</li>
<li>Hancock PA, Godfray HCJ (2007). “Application of the lumped age-class
technique to studying the dynamics of malaria-mosquito-human
interactions.” <em>Malaria Journal</em>, 6, 98. ISSN 1475-2875, doi:
10.1186/1475-2875-6-98 , <a href="https://malariajournal.biomedcentral.com/articles/10.1186/1475-2875-6-98" class="external-link uri">https://malariajournal.biomedcentral.com/articles/10.1186/1475-2875-6-98</a>.</li>
<li>Marshall J, Buchman A, C. HMS, Akbari OS (2017). “Overcoming evolved
resistance to population-suppressing homing-based gene drives.”
<em>Nature Scientific Reports</em>, 1–46. ISSN 2045-2322, doi: <a href="https://doi.org/10.1101/088427" class="external-link uri">https://doi.org/10.1101/088427</a> , <a href="https://www.nature.com/articles/s41598-017-02744-7" class="external-link uri">https://www.nature.com/articles/s41598-017-02744-7</a>
</li>
</ul>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://chipdelmal.github.io" class="external-link">Héctor Manuel Sánchez Castellanos</a>, <a href="https://github.com/GilChrist19" class="external-link">Jared Bennett</a>, <a href="https://slwu89.github.io" class="external-link">Sean Wu</a>, <a href="https://www.marshalllab.com" class="external-link">John M. Marshall</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
