<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>MGDrivE Run Example • MGDrivE</title>
<!-- mathjax math --><script src="https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.js" integrity="sha256-MASABpB4tYktI2Oitl4t+78w/lyA+D7b/s9GEP0JOGI=" crossorigin="anonymous"></script><script>
  window.MathJax = {
    chtml: {
      fontURL: "https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2"
    }
  };
</script><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.1/jquery-3.6.1.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="MGDrivE Run Example">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MGDrivE</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.6.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><a class="dropdown-item" href="../articles/mgdrive_examples.html">Example Simulations</a></li>
    <li><a class="dropdown-item" href="../articles/mgdrive_math.html">Mathematical Description</a></li>
    <li><a class="dropdown-item" href="../articles/mgdrive_run.html">Detailed Run Example</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/MarshallLab/MGDrivE" aria-label="View on GitHub"><span class="fa fa-lg fa-github"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>MGDrivE Run Example</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/MarshallLab/MGDrivE/blob/HEAD/vignettes/mgdrive_run.Rmd" class="external-link"><code>vignettes/mgdrive_run.Rmd</code></a></small>
      <div class="d-none name"><code>mgdrive_run.Rmd</code></div>
    </div>

    
    
<style type="text/css">
pre {
  white-space: pre !important;
  overflow-y: scroll !important;
  max-height: 25vh !important;
}

# reference
#  https://stackoverflow.com/questions/41135085/how-to-make-vertical-scrollbar-appear-in-rmarkdown-code-chunks-html-view
</style>
<div class="section level2">
<h2 id="mgdrive-one-run">MGDrivE, One Run<a class="anchor" aria-label="anchor" href="#mgdrive-one-run"></a>
</h2>
<p>The other <a href="mgdrive_examples.html">vignette</a> has many
examples of calculating movement matrices, different simulations, and
some theory for comparison. However, it does not go through each step of
setting up, running, and analyzing a simulation. This vignette is
designed to go through every step of a simulation, exploring why each
step is performed, some thoughts on how they can be modified, and
looking at the output (if any).</p>
<div class="section level3">
<h3 id="mendelian-inheritance-stochastic-multiple-populations">Mendelian Inheritance, Stochastic, Multiple Populations<a class="anchor" aria-label="anchor" href="#mendelian-inheritance-stochastic-multiple-populations"></a>
</h3>
<p>For our example, we will take the Stochastic, Small Migration example
from the other vignette as a starting point. We will expand it to 5
populations, so that that code is a bit longer and the file structure
becomes more obvious.</p>
<div class="section level4">
<h4 id="full-code">Full Code<a class="anchor" aria-label="anchor" href="#full-code"></a>
</h4>
<p>For reference, here is the full code for our example. It will be
explained piece-by-piece below.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">####################</span></span>
<span><span class="co"># Load libraries</span></span>
<span><span class="co">####################</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://marshalllab.github.io/MGDrivE/" class="external-link">MGDrivE</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">####################</span></span>
<span><span class="co"># Output Folder</span></span>
<span><span class="co">####################</span></span>
<span><span class="va">outFolder</span> <span class="op">&lt;-</span> <span class="st">"mgdrive"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span>path <span class="op">=</span> <span class="va">outFolder</span><span class="op">)</span></span>
<span></span>
<span><span class="co">####################</span></span>
<span><span class="co"># Simulation Parameters</span></span>
<span><span class="co">####################</span></span>
<span><span class="co"># days to run the simulation</span></span>
<span><span class="va">tMax</span> <span class="op">&lt;-</span> <span class="fl">365</span></span>
<span></span>
<span><span class="co"># number of Monte Carlo iterations</span></span>
<span><span class="va">nRep</span> <span class="op">&lt;-</span> <span class="fl">5</span></span>
<span></span>
<span><span class="co"># each Monte Carlo iteration gets its own folder</span></span>
<span><span class="va">folderNames</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outFolder</span>,</span>
<span>                        <span class="fu"><a href="https://rdrr.io/r/base/formatc.html" class="external-link">formatC</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="va">nRep</span>, width <span class="op">=</span> <span class="fl">3</span>, format <span class="op">=</span> <span class="st">"d"</span>, flag <span class="op">=</span> <span class="st">"0"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># entomological parameters</span></span>
<span><span class="va">bioParameters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>betaK<span class="op">=</span><span class="fl">20</span>,tEgg<span class="op">=</span><span class="fl">5</span>,tLarva<span class="op">=</span><span class="fl">6</span>,tPupa<span class="op">=</span><span class="fl">4</span>,popGrowth<span class="op">=</span><span class="fl">1.175</span>,muAd<span class="op">=</span><span class="fl">0.09</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># a 5-node network with 5% per day migration rate</span></span>
<span><span class="va">sitesNumber</span> <span class="op">&lt;-</span> <span class="fl">5</span></span>
<span><span class="va">adultPopEquilibrium</span> <span class="op">&lt;-</span> <span class="fl">500</span></span>
<span></span>
<span><span class="co"># auxiliary function</span></span>
<span><span class="va">triDiag</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">upper</span>, <span class="va">lower</span><span class="op">)</span><span class="op">{</span></span>
<span>  </span>
<span>  <span class="co"># return matrix</span></span>
<span>  <span class="va">retMat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fl">0</span>, nrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">upper</span><span class="op">)</span> <span class="op">+</span> <span class="fl">1</span>, ncol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">upper</span><span class="op">)</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co"># set index values for upper/lower triangles</span></span>
<span>  <span class="va">indx</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">upper</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co"># set forward/backward migration using matrix access</span></span>
<span>  <span class="va">retMat</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">indx</span><span class="op">+</span><span class="fl">1</span>,<span class="va">indx</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">lower</span></span>
<span>  <span class="va">retMat</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">indx</span>,<span class="va">indx</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">upper</span></span>
<span>  </span>
<span>  <span class="co"># set stay probs</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">retMat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">retMat</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">retMat</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># fill movement matrix</span></span>
<span><span class="co">#  Remember, rows need to sum to 1.</span></span>
<span><span class="va">moveMat</span> <span class="op">&lt;-</span> <span class="fu">triDiag</span><span class="op">(</span>upper <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep.int</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fl">0.05</span>, times <span class="op">=</span> <span class="va">sitesNumber</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span>,</span>
<span>                   lower <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep.int</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fl">0.05</span>, times <span class="op">=</span> <span class="va">sitesNumber</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># batch migration is disabled by setting the probability to 0</span></span>
<span><span class="va">batchMigration</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/basicBatchMigration.html">basicBatchMigration</a></span><span class="op">(</span>batchProbs<span class="op">=</span><span class="fl">0</span>,</span>
<span>                                     sexProbs<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">.5</span>,<span class="fl">.5</span><span class="op">)</span>,</span>
<span>                                     numPatches<span class="op">=</span><span class="va">sitesNumber</span><span class="op">)</span></span>
<span></span>
<span><span class="co">####################</span></span>
<span><span class="co"># Basic Inheritance pattern</span></span>
<span><span class="co">####################</span></span>
<span><span class="co"># Mendelian cube with standard (default) paraameters</span></span>
<span><span class="va">cube</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cubeMendelian.html">cubeMendelian</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">####################</span></span>
<span><span class="co"># Setup releases and batch migration</span></span>
<span><span class="co">####################</span></span>
<span><span class="co"># set up the empty release vector</span></span>
<span><span class="co">#  MGDrivE pulls things out by name</span></span>
<span><span class="va">patchReleases</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">replicate</a></span><span class="op">(</span>n<span class="op">=</span><span class="va">sitesNumber</span>,</span>
<span>                           expr<span class="op">=</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>maleReleases<span class="op">=</span><span class="cn">NULL</span>,femaleReleases<span class="op">=</span><span class="cn">NULL</span>,</span>
<span>                                      eggReleases<span class="op">=</span><span class="cn">NULL</span>,matedFemaleReleases<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span><span class="op">}</span>,</span>
<span>                           simplify<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># choose release parameters</span></span>
<span><span class="co">#  Releases start at time 25, occur once a week, for 10 weeks.</span></span>
<span><span class="co">#  There are 100 mosquitoes released every time.</span></span>
<span><span class="va">releasesParameters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>releasesStart<span class="op">=</span><span class="fl">25</span>,</span>
<span>                          releasesNumber<span class="op">=</span><span class="fl">10</span>,</span>
<span>                          releasesInterval<span class="op">=</span><span class="fl">7</span>,</span>
<span>                          releaseProportion<span class="op">=</span><span class="fl">100</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># generate male release vector</span></span>
<span><span class="va">maleReleasesVector</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generateReleaseVector.html">generateReleaseVector</a></span><span class="op">(</span>driveCube<span class="op">=</span><span class="va">cube</span>, </span>
<span>                                           releasesParameters<span class="op">=</span><span class="va">releasesParameters</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># generate female release vector</span></span>
<span><span class="va">femaleReleasesVector</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generateReleaseVector.html">generateReleaseVector</a></span><span class="op">(</span>driveCube<span class="op">=</span><span class="va">cube</span>,</span>
<span>                                             releasesParameters<span class="op">=</span><span class="va">releasesParameters</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># put releases into the proper place in the release list</span></span>
<span><span class="co">#  This performs the releases in the first patch only</span></span>
<span><span class="va">patchReleases</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">maleReleases</span> <span class="op">&lt;-</span> <span class="va">maleReleasesVector</span></span>
<span><span class="va">patchReleases</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">femaleReleases</span> <span class="op">&lt;-</span> <span class="va">femaleReleasesVector</span></span>
<span></span>
<span></span>
<span><span class="co">####################</span></span>
<span><span class="co"># Combine parameters and run!</span></span>
<span><span class="co">####################</span></span>
<span><span class="co"># setup parameters for the network. This builds a list of parameters required for</span></span>
<span><span class="co">#  every population in the network.</span></span>
<span><span class="va">netPar</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/parameterizeMGDrivE.html">parameterizeMGDrivE</a></span><span class="op">(</span>runID<span class="op">=</span><span class="fl">1</span>, simTime<span class="op">=</span><span class="va">tMax</span>, nPatch<span class="op">=</span><span class="va">sitesNumber</span>,</span>
<span>                             beta<span class="op">=</span><span class="va">bioParameters</span><span class="op">$</span><span class="va">betaK</span>, muAd<span class="op">=</span><span class="va">bioParameters</span><span class="op">$</span><span class="va">muAd</span>,</span>
<span>                             popGrowth<span class="op">=</span><span class="va">bioParameters</span><span class="op">$</span><span class="va">popGrowth</span>, tEgg<span class="op">=</span><span class="va">bioParameters</span><span class="op">$</span><span class="va">tEgg</span>,</span>
<span>                             tLarva<span class="op">=</span><span class="va">bioParameters</span><span class="op">$</span><span class="va">tLarva</span>, tPupa<span class="op">=</span><span class="va">bioParameters</span><span class="op">$</span><span class="va">tPupa</span>,</span>
<span>                             AdPopEQ<span class="op">=</span><span class="va">adultPopEquilibrium</span>, inheritanceCube <span class="op">=</span> <span class="va">cube</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># set MGDrivE to run stochastic</span></span>
<span><span class="fu"><a href="../reference/setupMGDrivE.html">setupMGDrivE</a></span><span class="op">(</span>stochasticityON <span class="op">=</span> <span class="cn">TRUE</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># build network prior to run</span></span>
<span><span class="va">MGDrivESim</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/Network.html">Network</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span>params<span class="op">=</span><span class="va">netPar</span>,</span>
<span>                         driveCube<span class="op">=</span><span class="va">cube</span>,</span>
<span>                         patchReleases<span class="op">=</span><span class="va">patchReleases</span>,</span>
<span>                         migrationMale<span class="op">=</span><span class="va">moveMat</span>,</span>
<span>                         migrationFemale<span class="op">=</span><span class="va">moveMat</span>,</span>
<span>                         migrationBatch<span class="op">=</span><span class="va">batchMigration</span>,</span>
<span>                         directory<span class="op">=</span><span class="va">folderNames</span>,</span>
<span>                         verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co"># run simulation</span></span>
<span><span class="va">MGDrivESim</span><span class="op">$</span><span class="fu">multRun</span><span class="op">(</span>verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">####################</span></span>
<span><span class="co"># Post Analysis</span></span>
<span><span class="co">####################</span></span>
<span><span class="co"># First, split output by patch</span></span>
<span><span class="co"># Second, aggregate females by their mate choice</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">nRep</span><span class="op">)</span><span class="op">{</span></span>
<span> <span class="fu"><a href="../reference/splitOutput.html">splitOutput</a></span><span class="op">(</span>readDir <span class="op">=</span> <span class="va">folderNames</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, remFile <span class="op">=</span> <span class="cn">TRUE</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span> <span class="fu"><a href="../reference/aggregateFemales.html">aggregateFemales</a></span><span class="op">(</span>readDir <span class="op">=</span> <span class="va">folderNames</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, genotypes <span class="op">=</span> <span class="va">cube</span><span class="op">$</span><span class="va">genotypesID</span>,</span>
<span>                  remFile <span class="op">=</span> <span class="cn">TRUE</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># plot output of first run to see effect</span></span>
<span><span class="fu"><a href="../reference/plotMGDrivESingle.html">plotMGDrivESingle</a></span><span class="op">(</span>readDir<span class="op">=</span><span class="va">folderNames</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,totalPop <span class="op">=</span> <span class="cn">TRUE</span>,lwd<span class="op">=</span><span class="fl">3.5</span>,alpha<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot all 5 repetitions together</span></span>
<span><span class="fu"><a href="../reference/plotMGDrivEMult.html">plotMGDrivEMult</a></span><span class="op">(</span>readDir<span class="op">=</span><span class="va">outFolder</span>,lwd<span class="op">=</span><span class="fl">0.35</span>,alpha<span class="op">=</span><span class="fl">0.75</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="load-libraries">Load Libraries<a class="anchor" aria-label="anchor" href="#load-libraries"></a>
</h4>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">####################</span></span>
<span><span class="co"># Load libraries</span></span>
<span><span class="co">####################</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://marshalllab.github.io/MGDrivE/" class="external-link">MGDrivE</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading MGDrivE: Mosquito Gene Drive Explorer</span></span></code></pre></div>
<p>Step one, load <em>MGDrivE</em>.</p>
</div>
<div class="section level4">
<h4 id="setup-folder-directory">Setup Folder Directory<a class="anchor" aria-label="anchor" href="#setup-folder-directory"></a>
</h4>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">####################</span></span>
<span><span class="co"># Output Folder</span></span>
<span><span class="co">####################</span></span>
<span><span class="va">outFolder</span> <span class="op">&lt;-</span> <span class="st">"mgdrive"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span>path <span class="op">=</span> <span class="va">outFolder</span><span class="op">)</span></span></code></pre></div>
<p>Here, we setup the folder for simulation data. Since this is a local
folder, it doesn’t exist, and needs to be created. If you already have a
folder prepared, make sure it is empty. <em>MGDrivE</em> gives a warning
if there are other things in the folder (to prevent overwriting of
data), but will run fine if the folder is not empty.</p>
</div>
<div class="section level4">
<h4 id="simulation-parameters">Simulation Parameters<a class="anchor" aria-label="anchor" href="#simulation-parameters"></a>
</h4>
<p>Next, there are several parameters that need to be set for a
simulation. The run time and the number of repetitions are crucial. Once
the number of repetitions has been chosen, the folder names can be
created.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">####################</span></span>
<span><span class="co"># Simulation Parameters</span></span>
<span><span class="co">####################</span></span>
<span><span class="co"># days to run the simulation</span></span>
<span><span class="va">tMax</span> <span class="op">&lt;-</span> <span class="fl">365</span></span>
<span></span>
<span><span class="co"># number of Monte Carlo iterations</span></span>
<span><span class="va">nRep</span> <span class="op">&lt;-</span> <span class="fl">5</span></span>
<span></span>
<span><span class="co"># each Monte Carlo iteration gets its own folder</span></span>
<span><span class="va">folderNames</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outFolder</span>,</span>
<span>                        <span class="fu"><a href="https://rdrr.io/r/base/formatc.html" class="external-link">formatC</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="va">nRep</span>, width <span class="op">=</span> <span class="fl">3</span>, format <span class="op">=</span> <span class="st">"d"</span>, flag <span class="op">=</span> <span class="st">"0"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">folderNames</span></span>
<span><span class="co">#&gt; [1] "mgdrive/001" "mgdrive/002" "mgdrive/003" "mgdrive/004" "mgdrive/005"</span></span></code></pre></div>
<p>Here, folder names reflect their purpose - holding each repetition of
our stochastic simulation. The precise names don’t matter later on, but
they do need to be unique (or data gets overwritten, <em>MGDrivE</em>
will not throw a warning about this), and in a location that your user
account has permission to create directories.</p>
<p>The biological parameters were chosen for <em>Aedes aegypti</em> in
our original simulations. These parameters represent:</p>
<ul>
<li>
<code>betaK</code>: Daily number of eggs laid per female
mosquito</li>
<li>
<code>tEgg</code>: Number of days spent in the egg stage</li>
<li>
<code>tLarva</code>: Number of days spent in the larval stage</li>
<li>
<code>tPupa</code>: Number of days spent in the pupa stage</li>
<li>
<code>popGrowth</code>: Population growth per generation<br>
</li>
<li>
<code>muAD</code>: Daily death rate for adult mosquitoes</li>
</ul>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># entomological parameters</span></span>
<span><span class="va">bioParameters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>betaK<span class="op">=</span><span class="fl">20</span>,tEgg<span class="op">=</span><span class="fl">5</span>,tLarva<span class="op">=</span><span class="fl">6</span>,tPupa<span class="op">=</span><span class="fl">4</span>,popGrowth<span class="op">=</span><span class="fl">1.175</span>,muAd<span class="op">=</span><span class="fl">0.09</span><span class="op">)</span></span></code></pre></div>
<p>Next, we set the number of populations in the simulation. Here, we
chose 5 populations, each containing 500 mosquitoes. After that, we
setup the landscape.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># a 5-node network with 5% per day migration rate</span></span>
<span><span class="va">sitesNumber</span> <span class="op">&lt;-</span> <span class="fl">5</span></span>
<span><span class="va">adultPopEquilibrium</span> <span class="op">&lt;-</span> <span class="fl">500</span></span>
<span><span class="va">patchPops</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="va">adultPopEquilibrium</span>,<span class="va">sitesNumber</span><span class="op">)</span> </span>
<span><span class="co"># patchPops is optional. If all populations are the same size, parameterizeMGDrivE</span></span>
<span><span class="co">#  can take a single number. However, if you desire different population sizes, it </span></span>
<span><span class="co">#  must be a vector of length equal to the number of sites</span></span>
<span></span>
<span><span class="co"># landscape</span></span>
<span><span class="co"># auxiliary function</span></span>
<span><span class="va">triDiag</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">upper</span>, <span class="va">lower</span><span class="op">)</span><span class="op">{</span></span>
<span>  </span>
<span>  <span class="co"># return matrix</span></span>
<span>  <span class="va">retMat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fl">0</span>, nrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">upper</span><span class="op">)</span> <span class="op">+</span> <span class="fl">1</span>, ncol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">upper</span><span class="op">)</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co"># set index values for upper/lower triangles</span></span>
<span>  <span class="va">indx</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">upper</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co"># set forward/backward migration using matrix access</span></span>
<span>  <span class="va">retMat</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">indx</span><span class="op">+</span><span class="fl">1</span>,<span class="va">indx</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">lower</span></span>
<span>  <span class="va">retMat</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">indx</span>,<span class="va">indx</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">upper</span></span>
<span>  </span>
<span>  <span class="co"># set stay probs</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">retMat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">retMat</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">retMat</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># fill movement matrix</span></span>
<span><span class="co">#  Remember, rows need to sum to 1.</span></span>
<span><span class="va">moveMat</span> <span class="op">&lt;-</span> <span class="fu">triDiag</span><span class="op">(</span>upper <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep.int</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fl">0.05</span>, times <span class="op">=</span> <span class="va">sitesNumber</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span>,</span>
<span>                   lower <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep.int</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fl">0.05</span>, times <span class="op">=</span> <span class="va">sitesNumber</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">moveMat</span></span>
<span><span class="co">#&gt;      [,1] [,2] [,3] [,4] [,5]</span></span>
<span><span class="co">#&gt; [1,] 0.95 0.05 0.00 0.00 0.00</span></span>
<span><span class="co">#&gt; [2,] 0.05 0.90 0.05 0.00 0.00</span></span>
<span><span class="co">#&gt; [3,] 0.00 0.05 0.90 0.05 0.00</span></span>
<span><span class="co">#&gt; [4,] 0.00 0.00 0.05 0.90 0.05</span></span>
<span><span class="co">#&gt; [5,] 0.00 0.00 0.00 0.05 0.95</span></span></code></pre></div>
<p>Our landscape will be a line, with each population only connected to
the populations on either side. We set a 5% per day migration rate.
There are several examples in the other <a href="mgdrive_examples.html">vignette</a> for setting up different
landscapes.</p>
<p>Finally, there is the possibility of large-group movement facilitated
by humans (using <em>Aedes</em> for example, this could be boats or
water containers in truck beds, etc.). <em>MGDrivE</em> handles this
through <code><a href="../reference/basicBatchMigration.html">basicBatchMigration()</a></code>. For simplicity, we are
ignoring this possibility (by setting the probability of it to
zero).</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># batch migration is disabled by setting the probability to 0</span></span>
<span><span class="va">batchMigration</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/basicBatchMigration.html">basicBatchMigration</a></span><span class="op">(</span>batchProbs<span class="op">=</span><span class="fl">0</span>, numPatches<span class="op">=</span><span class="va">sitesNumber</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="inheritance-pattern">Inheritance Pattern<a class="anchor" aria-label="anchor" href="#inheritance-pattern"></a>
</h4>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">####################</span></span>
<span><span class="co"># Basic Inheritance pattern</span></span>
<span><span class="co">####################</span></span>
<span><span class="co"># Mendelian cube with standard (default) paraameters</span></span>
<span><span class="va">cube</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cubeMendelian.html">cubeMendelian</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>For this experiment, we use simple Mendelian inheritance.
<em>MGDrivE</em> has several other inheritance patterns, which
include:</p>
<ul>
<li>
<code><a href="../reference/cubeReciprocalTranslocations.html">cubeReciprocalTranslocations()</a></code>: Reciprocal
Translocations<br>
</li>
<li>
<code><a href="../reference/cubeHomingDrive.html">cubeHomingDrive()</a></code>: CRISPR-based homing drive, with two
resistance alleles<br>
</li>
<li>
<code><a href="../reference/cubeMEDEA.html">cubeMEDEA()</a></code>: Maternal Effect Dominant Embryonic
Arrest<br>
</li>
<li>See <a href="mgdrive_math.html">MGDrivE Mathematical Description</a>
for a complete list</li>
</ul>
<p>While we use the default parameters, there are several fitness costs
that can be applied. Every cost is applied in a genotype-specific
manner. The optional fitness costs are:</p>
<ul>
<li>
<span class="math inline">\(\eta\)</span>: Male mating fitness
against each female<br>
</li>
<li>
<span class="math inline">\(\phi\)</span>: Sex-ratio at
emergence<br>
</li>
<li>
<span class="math inline">\(\omega\)</span>: Multiplicative modifier
of adult mortality (males and females)<br>
</li>
<li>
<span class="math inline">\(\xi_F\)</span>: Female pupatory success
(applies after <span class="math inline">\(\phi\)</span>)<br>
</li>
<li>
<span class="math inline">\(\xi_M\)</span>: Male pupatory success
(applies after <span class="math inline">\(\phi\)</span>)<br>
</li>
<li>
<span class="math inline">\(s\)</span>: Fractional increase/decrease
in fertility</li>
</ul>
</div>
<div class="section level4">
<h4 id="releases">Releases<a class="anchor" aria-label="anchor" href="#releases"></a>
</h4>
<p><em>MGDrivE</em> uses a list specifying releases of males or females
to parameterize the release schedule. The initial list is
<code>NULL</code>, and only if there are releases to be performed is
anything added.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">####################</span></span>
<span><span class="co"># Setup releases and batch migration</span></span>
<span><span class="co">####################</span></span>
<span><span class="co"># set up the empty release vector</span></span>
<span><span class="co">#  MGDrivE pulls things out by name</span></span>
<span><span class="va">patchReleases</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">replicate</a></span><span class="op">(</span>n<span class="op">=</span><span class="va">sitesNumber</span>,</span>
<span>                           expr<span class="op">=</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>maleReleases<span class="op">=</span><span class="cn">NULL</span>,femaleReleases<span class="op">=</span><span class="cn">NULL</span>,</span>
<span>                                      eggReleases<span class="op">=</span><span class="cn">NULL</span>,matedFemaleReleases<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span><span class="op">}</span>,</span>
<span>                           simplify<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># choose release parameters</span></span>
<span><span class="co">#  Releases start at time 25, occur once a week, for 10 weeks.</span></span>
<span><span class="co">#  There are 100 mosquitoes released every time.</span></span>
<span><span class="va">releasesParameters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>releasesStart<span class="op">=</span><span class="fl">25</span>,</span>
<span>                          releasesNumber<span class="op">=</span><span class="fl">10</span>,</span>
<span>                          releasesInterval<span class="op">=</span><span class="fl">7</span>,</span>
<span>                          releaseProportion<span class="op">=</span><span class="fl">100</span><span class="op">)</span></span></code></pre></div>
<p>Here, we are performing 10 releases of 100 individuals. These
releases start on day 25 and occur every 7 days. We now generate what
this looks like for a male release.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># generate male release vector</span></span>
<span><span class="va">maleReleasesVector</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generateReleaseVector.html">generateReleaseVector</a></span><span class="op">(</span>driveCube<span class="op">=</span><span class="va">cube</span>, </span>
<span>                                           releasesParameters<span class="op">=</span><span class="va">releasesParameters</span><span class="op">)</span></span>
<span><span class="va">maleReleasesVector</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">#&gt; $nRelease</span></span>
<span><span class="co">#&gt;      [,1] [,2]</span></span>
<span><span class="co">#&gt; [1,]    3  100</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $tRelease</span></span>
<span><span class="co">#&gt; [1] 25</span></span></code></pre></div>
<p>The release vector generated is 10 elements long, because we have 10
releases, and each element contains the number and numeric genotype of
individuals to release and the time of that release. We now generate the
same thing for female releases, and insert the releases into our release
list so that releases are only performed in the first patch.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># generate female release vector</span></span>
<span><span class="va">femaleReleasesVector</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generateReleaseVector.html">generateReleaseVector</a></span><span class="op">(</span>driveCube<span class="op">=</span><span class="va">cube</span>,</span>
<span>                                             releasesParameters<span class="op">=</span><span class="va">releasesParameters</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># put releases into the proper place in the release list</span></span>
<span><span class="va">patchReleases</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">maleReleases</span> <span class="op">&lt;-</span> <span class="va">maleReleasesVector</span></span>
<span><span class="va">patchReleases</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">femaleReleases</span> <span class="op">&lt;-</span> <span class="va">femaleReleasesVector</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="setup-and-run">Setup and Run<a class="anchor" aria-label="anchor" href="#setup-and-run"></a>
</h4>
<p>The final steps before running our simulation! All of the parameters
we have setup are combined using <code><a href="../reference/parameterizeMGDrivE.html">parameterizeMGDrivE()</a></code>.
Additionally, this function calculates equilibrium values for aquatic
populations, death rates, and density dependence for every population.
See the documentation for all of the possible options.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">####################</span></span>
<span><span class="co"># Combine parameters and run!</span></span>
<span><span class="co">####################</span></span>
<span><span class="co"># setup parameters for the network. This builds a list of parameters required for</span></span>
<span><span class="co">#  every population in the network.</span></span>
<span><span class="va">netPar</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/parameterizeMGDrivE.html">parameterizeMGDrivE</a></span><span class="op">(</span>runID<span class="op">=</span><span class="fl">1</span>, simTime<span class="op">=</span><span class="va">tMax</span>, nPatch<span class="op">=</span><span class="va">sitesNumber</span>,</span>
<span>                             beta<span class="op">=</span><span class="va">bioParameters</span><span class="op">$</span><span class="va">betaK</span>, muAd<span class="op">=</span><span class="va">bioParameters</span><span class="op">$</span><span class="va">muAd</span>,</span>
<span>                             popGrowth<span class="op">=</span><span class="va">bioParameters</span><span class="op">$</span><span class="va">popGrowth</span>, tEgg<span class="op">=</span><span class="va">bioParameters</span><span class="op">$</span><span class="va">tEgg</span>,</span>
<span>                             tLarva<span class="op">=</span><span class="va">bioParameters</span><span class="op">$</span><span class="va">tLarva</span>, tPupa<span class="op">=</span><span class="va">bioParameters</span><span class="op">$</span><span class="va">tPupa</span>,</span>
<span>                             AdPopEQ<span class="op">=</span><span class="va">patchPops</span>, inheritanceCube <span class="op">=</span> <span class="va">cube</span><span class="op">)</span></span></code></pre></div>
<p><em>MGDrivE</em> can run in either a deterministic or stochastic
setting, so we set it to stochastic for these simulations.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># set MGDrivE to run stochastic</span></span>
<span><span class="fu"><a href="../reference/setupMGDrivE.html">setupMGDrivE</a></span><span class="op">(</span>stochasticityON <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; initializing MGDrivE</span></span></code></pre></div>
<p>Finally, we setup the complete network to run our simulation.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># build network prior to run</span></span>
<span><span class="va">MGDrivESim</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/Network.html">Network</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span>params<span class="op">=</span><span class="va">netPar</span>,</span>
<span>                         driveCube<span class="op">=</span><span class="va">cube</span>,</span>
<span>                         patchReleases<span class="op">=</span><span class="va">patchReleases</span>,</span>
<span>                         migrationMale<span class="op">=</span><span class="va">moveMat</span>,</span>
<span>                         migrationFemale<span class="op">=</span><span class="va">moveMat</span>,</span>
<span>                         migrationBatch<span class="op">=</span><span class="va">batchMigration</span>,</span>
<span>                         directory<span class="op">=</span><span class="va">folderNames</span>,</span>
<span>                         verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; initializing patch:  1  of  5 </span></span>
<span><span class="co">#&gt; initializing patch:  2  of  5 </span></span>
<span><span class="co">#&gt; initializing patch:  3  of  5 </span></span>
<span><span class="co">#&gt; initializing patch:  4  of  5 </span></span>
<span><span class="co">#&gt; initializing patch:  5  of  5</span></span>
<span></span>
<span><span class="co"># list folders to show that they have been created</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span>path <span class="op">=</span> <span class="va">outFolder</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "001" "002" "003" "004" "005"</span></span></code></pre></div>
<p>The <code>Network</code> object represents the metapopulation of
<em>MGDrivE</em>. It can be setup to hold 1 patch, and therefore be a
single population, or several patches, and be a true metapopulation.
Documentation can be found by running <code><a href="../reference/Network.html">?Network</a></code>. The
different notation is an attribute of the <code>R6</code> structure that
<em>MGDrivE</em> is built on (Shoutout to <a href="https://github.com/r-lib/R6" class="external-link">R6</a>). The <code>params</code>
argument takes the network parameters created by
<code><a href="../reference/parameterizeMGDrivE.html">parameterizeMGDrivE()</a></code>. The inheritance pattern is supplied
in the <code>driveCube</code> option, which we chose to be Mendelian.
Migration can be different for males and females, so even though here we
use the same migration matrix, users can supply different ones if
desired. The <code>migrationBatch</code> option must always be filled
in, but can be set to zero, as done here, if it isn’t desired or for
deterministic simulations (where it doesn’t make sense). Finally, the
<code>directory</code> parameter takes the vector of folder names, and
creates them for use during the simulation.</p>
<p>Finally, <em>MGDrivE</em> can be run once using
<code>MGDrivESim$oneRun()</code> or several times using
<code>MGDrivESim$multRun()</code>. The benefit of using the built-in
multiple repetition function is that memory is not released between
runs, just cleared, so setup-time is reduced. Additionally, each of
these functions can produce a progress bar, which is suppressed in all
of our vignettes.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># run simulation</span></span>
<span><span class="va">MGDrivESim</span><span class="op">$</span><span class="fu">multRun</span><span class="op">(</span>verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="post-processing">Post Processing<a class="anchor" aria-label="anchor" href="#post-processing"></a>
</h4>
<p><em>MGDrivE</em> outputs 2 *.csv files every simulation. These are
stored as <code>{F|M}_RunXXX.csv</code> for female and male adults. In
them are the population counts for every genotype at every day.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># first and last repetitions</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span>path <span class="op">=</span> <span class="va">outFolder</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="co">#&gt; [1] "001"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span>path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span>path <span class="op">=</span> <span class="va">outFolder</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "F_Run001.csv" "M_Run001.csv"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span>path <span class="op">=</span> <span class="va">outFolder</span><span class="op">)</span><span class="op">[</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt; [1] "005"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span>path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span>path <span class="op">=</span> <span class="va">outFolder</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">5</span><span class="op">]</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "F_Run005.csv" "M_Run005.csv"</span></span></code></pre></div>
<p>This is the raw output from <em>MGDrivE</em>. The files contain all
of the information from the simulation.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># read in male and female files</span></span>
<span><span class="va">mMat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span>file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span>path <span class="op">=</span> <span class="va">outFolder</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                             recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>,</span>
<span>                           header <span class="op">=</span> <span class="cn">TRUE</span>, sep <span class="op">=</span> <span class="st">","</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">fMat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span>file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span>path <span class="op">=</span> <span class="va">outFolder</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                             recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>                           header <span class="op">=</span> <span class="cn">TRUE</span>, sep <span class="op">=</span> <span class="st">","</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># look at male file header</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">mMat</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Time"  "Patch" "AA"    "Aa"    "aa"</span></span></code></pre></div>
<p>The header for male files (<code>M_RunXXX.csv</code>) is shown above.
They always contains the time, patch, and every possible genotype in the
simulation. For the Mendelian inheritance here, there are three possible
genotypes, AA, Aa, aa. The Time and Patch columns will always be the
same, but the number of genotypes changes depending on the inheritance
pattern used. For example, CRISPR-based homing inheritance has 10, so
there would be 12 columns in the male files. Below, we see how each
patch is designated in the file.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">mMat</span>, n <span class="op">=</span> <span class="fl">2</span><span class="op">*</span><span class="va">sitesNumber</span><span class="op">)</span></span>
<span><span class="co">#&gt;       Time Patch  AA Aa aa</span></span>
<span><span class="co">#&gt;  [1,]    0     1 250  0  0</span></span>
<span><span class="co">#&gt;  [2,]    0     2 250  0  0</span></span>
<span><span class="co">#&gt;  [3,]    0     3 250  0  0</span></span>
<span><span class="co">#&gt;  [4,]    0     4 250  0  0</span></span>
<span><span class="co">#&gt;  [5,]    0     5 250  0  0</span></span>
<span><span class="co">#&gt;  [6,]    2     1 241  0  0</span></span>
<span><span class="co">#&gt;  [7,]    2     2 241  0  0</span></span>
<span><span class="co">#&gt;  [8,]    2     3 254  0  0</span></span>
<span><span class="co">#&gt;  [9,]    2     4 243  0  0</span></span>
<span><span class="co">#&gt; [10,]    2     5 267  0  0</span></span></code></pre></div>
<p>Notice how each patch is printed for every time point, then the time
increases and every patch is printed again. Time = 1 is at equilibrium,
where we start the simulations, then the patches begin to diverge,
staying around equilibrium but no longer being exactly the same.</p>
<p>The female output files (<code>F_RunXXX.csv</code>) are slightly more
complicated. Since <em>MGDrivE</em> keeps track of females and their
mates, the output maintains that as well. Thus, the header in the female
files will be <span class="math inline">\(2 +
\text{numGenotypes}^2\)</span> long. For Mendelian inheritance, this
looks like:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># look at female file header</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">fMat</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "Time"  "Patch" "AAAA"  "AAAa"  "AAaa"  "AaAA"  "AaAa"  "Aaaa"  "aaAA" </span></span>
<span><span class="co">#&gt; [10] "aaAa"  "aaaa"</span></span></code></pre></div>
<p>The Time and Patch columns are the same as in male files, but the
“genotypes” are now a combination of the female genotype and her mate’s
genotype. The length of the header is 11, consistent with having 3
genotypes, times the number of mate genotypes, plus the Time and Patch
columns. The header “genotypes” are printed female genotype first, male
(mate) second. So, the <em>aaAA</em> column represents females that are
<em>aa</em> mated to males that are <em>AA</em>.<br>
The Time/Patch pattern is the same as in the male files.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">fMat</span>, n <span class="op">=</span> <span class="fl">2</span><span class="op">*</span><span class="va">sitesNumber</span><span class="op">)</span></span>
<span><span class="co">#&gt;       Time Patch AAAA AAAa AAaa AaAA AaAa Aaaa aaAA aaAa aaaa</span></span>
<span><span class="co">#&gt;  [1,]    0     1  250    0    0    0    0    0    0    0    0</span></span>
<span><span class="co">#&gt;  [2,]    0     2  250    0    0    0    0    0    0    0    0</span></span>
<span><span class="co">#&gt;  [3,]    0     3  250    0    0    0    0    0    0    0    0</span></span>
<span><span class="co">#&gt;  [4,]    0     4  250    0    0    0    0    0    0    0    0</span></span>
<span><span class="co">#&gt;  [5,]    0     5  250    0    0    0    0    0    0    0    0</span></span>
<span><span class="co">#&gt;  [6,]    2     1  245    0    0    0    0    0    0    0    0</span></span>
<span><span class="co">#&gt;  [7,]    2     2  263    0    0    0    0    0    0    0    0</span></span>
<span><span class="co">#&gt;  [8,]    2     3  245    0    0    0    0    0    0    0    0</span></span>
<span><span class="co">#&gt;  [9,]    2     4  255    0    0    0    0    0    0    0    0</span></span>
<span><span class="co">#&gt; [10,]    2     5  255    0    0    0    0    0    0    0    0</span></span></code></pre></div>
<p>There are several things that we can do from here. We can work with
the raw output directly, but this gets difficult as the simulations get
larger. Therefor, we have provided functions to split the raw output by
patches, generating a new file for every patch.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># First, split output by patch</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">nRep</span><span class="op">)</span><span class="op">{</span></span>
<span> <span class="fu"><a href="../reference/splitOutput.html">splitOutput</a></span><span class="op">(</span>readDir <span class="op">=</span> <span class="va">folderNames</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, remFile <span class="op">=</span> <span class="cn">TRUE</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><code>splitOutput</code> splits the raw output by patch, generating
one *.csv for every patch in our simulation. We loop over each
repetition folder, removing the original files, and replacing them with
individual patch files. Inside each repetition folder, we now see the
files:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># first and last repetitions</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span>path <span class="op">=</span> <span class="va">outFolder</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="co">#&gt; [1] "001"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span>path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span>path <span class="op">=</span> <span class="va">outFolder</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "F_Run001_Patch001.csv" "F_Run001_Patch002.csv" "F_Run001_Patch003.csv"</span></span>
<span><span class="co">#&gt;  [4] "F_Run001_Patch004.csv" "F_Run001_Patch005.csv" "M_Run001_Patch001.csv"</span></span>
<span><span class="co">#&gt;  [7] "M_Run001_Patch002.csv" "M_Run001_Patch003.csv" "M_Run001_Patch004.csv"</span></span>
<span><span class="co">#&gt; [10] "M_Run001_Patch005.csv"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span>path <span class="op">=</span> <span class="va">outFolder</span><span class="op">)</span><span class="op">[</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt; [1] "005"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span>path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span>path <span class="op">=</span> <span class="va">outFolder</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">5</span><span class="op">]</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "F_Run005_Patch001.csv" "F_Run005_Patch002.csv" "F_Run005_Patch003.csv"</span></span>
<span><span class="co">#&gt;  [4] "F_Run005_Patch004.csv" "F_Run005_Patch005.csv" "M_Run005_Patch001.csv"</span></span>
<span><span class="co">#&gt;  [7] "M_Run005_Patch002.csv" "M_Run005_Patch003.csv" "M_Run005_Patch004.csv"</span></span>
<span><span class="co">#&gt; [10] "M_Run005_Patch005.csv"</span></span>
<span></span>
<span><span class="co"># read in examples of new male/female files</span></span>
<span><span class="va">twoFiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span>path <span class="op">=</span> <span class="va">outFolder</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                       recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">6</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># read in male and female files</span></span>
<span><span class="va">mMat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span>file <span class="op">=</span> <span class="va">twoFiles</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, header <span class="op">=</span> <span class="cn">TRUE</span>, sep <span class="op">=</span> <span class="st">","</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">fMat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span>file <span class="op">=</span> <span class="va">twoFiles</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, header <span class="op">=</span> <span class="cn">TRUE</span>, sep <span class="op">=</span> <span class="st">","</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We again look at a male file, examining the header and the first few
rows.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">mMat</span>, n <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">#&gt;      Time  AA Aa aa</span></span>
<span><span class="co">#&gt; [1,]    0 250  0  0</span></span>
<span><span class="co">#&gt; [2,]    2 241  0  0</span></span>
<span><span class="co">#&gt; [3,]    3 241  0  0</span></span>
<span><span class="co">#&gt; [4,]    4 242  0  0</span></span>
<span><span class="co">#&gt; [5,]    5 246  0  0</span></span></code></pre></div>
<p>The Patch column has been removed, it is now part of the file name,
and the Time column increases every row. This is the first repetition,
first male file.</p>
<p>Female files are still a little more complicated. They have been
split by patch, but still retain the mate information.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">fMat</span>, n <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">#&gt;      Time AAAA AAAa AAaa AaAA AaAa Aaaa aaAA aaAa aaaa</span></span>
<span><span class="co">#&gt; [1,]    0  250    0    0    0    0    0    0    0    0</span></span>
<span><span class="co">#&gt; [2,]    2  245    0    0    0    0    0    0    0    0</span></span>
<span><span class="co">#&gt; [3,]    3  247    0    0    0    0    0    0    0    0</span></span>
<span><span class="co">#&gt; [4,]    4  255    0    0    0    0    0    0    0    0</span></span>
<span><span class="co">#&gt; [5,]    5  259    0    0    0    0    0    0    0    0</span></span></code></pre></div>
<p>The Patch column is gone, the Time column counts as expected, and we
still have all of the mate information.</p>
<p>In the event that we don’t want/need the mate information, we can
remove it using the <code><a href="../reference/aggregateFemales.html">aggregateFemales()</a></code> function.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Second, aggregate females by their mate choice</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">nRep</span><span class="op">)</span><span class="op">{</span></span>
<span> <span class="fu"><a href="../reference/aggregateFemales.html">aggregateFemales</a></span><span class="op">(</span>readDir <span class="op">=</span> <span class="va">folderNames</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, genotypes <span class="op">=</span> <span class="va">cube</span><span class="op">$</span><span class="va">genotypesID</span>,</span>
<span>                  remFile <span class="op">=</span> <span class="cn">TRUE</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>This collects all of the females for each genotype, summing over the
possible mates.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># changed female files in first repetition</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span>path <span class="op">=</span> <span class="va">outFolder</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="co">#&gt; [1] "001"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span>path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span>path <span class="op">=</span> <span class="va">outFolder</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "F_Aggregate_Run001_Patch001.csv" "F_Aggregate_Run001_Patch002.csv"</span></span>
<span><span class="co">#&gt;  [3] "F_Aggregate_Run001_Patch003.csv" "F_Aggregate_Run001_Patch004.csv"</span></span>
<span><span class="co">#&gt;  [5] "F_Aggregate_Run001_Patch005.csv" "M_Run001_Patch001.csv"          </span></span>
<span><span class="co">#&gt;  [7] "M_Run001_Patch002.csv"           "M_Run001_Patch003.csv"          </span></span>
<span><span class="co">#&gt;  [9] "M_Run001_Patch004.csv"           "M_Run001_Patch005.csv"</span></span>
<span></span>
<span><span class="co"># read in examples of new female file</span></span>
<span><span class="va">fMat2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span>file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span>path <span class="op">=</span> <span class="va">outFolder</span>,</span>
<span>                                              recursive <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                              full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>                            header <span class="op">=</span> <span class="cn">TRUE</span>, sep <span class="op">=</span> <span class="st">","</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We see that all of the female files have been updated to denote that
they are aggregated over their mate genotypes. Internally, the files now
resemble the male files.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">fMat2</span>, n <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">#&gt;      Time  AA Aa aa</span></span>
<span><span class="co">#&gt; [1,]    0 250  0  0</span></span>
<span><span class="co">#&gt; [2,]    2 245  0  0</span></span>
<span><span class="co">#&gt; [3,]    3 247  0  0</span></span>
<span><span class="co">#&gt; [4,]    4 255  0  0</span></span>
<span><span class="co">#&gt; [5,]    5 259  0  0</span></span></code></pre></div>
<p>This doesn’t show much change, but if we compare the female file
split by patch to the aggregated female file at the time of the release,
we can see that the genotypes have been combined.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># show non-aggregated female file split by patch</span></span>
<span><span class="co">#  This is for patch number 1</span></span>
<span><span class="va">fMat</span><span class="op">[</span><span class="op">(</span><span class="va">releasesParameters</span><span class="op">$</span><span class="va">releasesStart</span><span class="op">-</span><span class="fl">2</span><span class="op">)</span><span class="op">:</span><span class="op">(</span><span class="va">releasesParameters</span><span class="op">$</span><span class="va">releasesStart</span><span class="op">+</span><span class="fl">2</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="co">#&gt;      Time AAAA AAAa AAaa AaAA AaAa Aaaa aaAA aaAa aaaa</span></span>
<span><span class="co">#&gt; [1,]   23  234    0    0    0    0    0    0    0    0</span></span>
<span><span class="co">#&gt; [2,]   24  231    0    0    0    0    0    0    0    0</span></span>
<span><span class="co">#&gt; [3,]   25  231    0    6    0    0    0   67    0   30</span></span>
<span><span class="co">#&gt; [4,]   26  225    0   10    0    0    0   57    0   29</span></span>
<span><span class="co">#&gt; [5,]   27  218    0   16    0    0    0   50    0   25</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"\n"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># show aggregated female file</span></span>
<span><span class="co">#  This is for patch number 1</span></span>
<span><span class="va">fMat2</span><span class="op">[</span><span class="op">(</span><span class="va">releasesParameters</span><span class="op">$</span><span class="va">releasesStart</span><span class="op">-</span><span class="fl">2</span><span class="op">)</span><span class="op">:</span><span class="op">(</span><span class="va">releasesParameters</span><span class="op">$</span><span class="va">releasesStart</span><span class="op">+</span><span class="fl">2</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="co">#&gt;      Time  AA Aa aa</span></span>
<span><span class="co">#&gt; [1,]   23 234  0  0</span></span>
<span><span class="co">#&gt; [2,]   24 231  0  0</span></span>
<span><span class="co">#&gt; [3,]   25 237  0 97</span></span>
<span><span class="co">#&gt; [4,]   26 235  0 86</span></span>
<span><span class="co">#&gt; [5,]   27 234  0 75</span></span></code></pre></div>
<p>Looking closely, we see the “AAAA” column combined with the “AAaa”
column in the first output sum to the value in the “AA” column in the
second.<br>
The same goes for the “aaAA” and “aaaa” columns in the first and the
“aa” column in the second.</p>
</div>
<div class="section level4">
<h4 id="plotting">Plotting<a class="anchor" aria-label="anchor" href="#plotting"></a>
</h4>
<p><em>MGDrivE</em> comes with two built-in plot functions,
<code><a href="../reference/plotMGDrivESingle.html">plotMGDrivESingle()</a></code> and
<code><a href="../reference/plotMGDrivEMult.html">plotMGDrivEMult()</a></code>.<br><code><a href="../reference/plotMGDrivESingle.html">plotMGDrivESingle()</a></code> takes the name of one repetition
folder, and plots all of the patches in it for both males and
females.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot the first repetition</span></span>
<span><span class="fu"><a href="../reference/plotMGDrivESingle.html">plotMGDrivESingle</a></span><span class="op">(</span>readDir<span class="op">=</span><span class="va">folderNames</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,totalPop <span class="op">=</span> <span class="cn">TRUE</span>,lwd<span class="op">=</span><span class="fl">3.5</span>,alpha<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="mgdrive_run_files/figure-html/unnamed-chunk-31-1.png" alt="Example of how to plot a single repetition from a set of simulations." width="672"></p>
<p>The left Y-axis is population size, the right Y-axis is patch number,
and the X-axis is time. We can see the releases in the first patch,
equal for males and females, and follow their migration down the plots
from patch 1 to patch 5. The total population remains around 250
individuals for males and females, which gives the expected equilibrium
value of 500.<br>
We can also plot all of our repetitions using
<code><a href="../reference/plotMGDrivEMult.html">plotMGDrivEMult()</a></code>.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot all 5 repetitions together</span></span>
<span><span class="fu"><a href="../reference/plotMGDrivEMult.html">plotMGDrivEMult</a></span><span class="op">(</span>readDir<span class="op">=</span><span class="va">outFolder</span>,lwd<span class="op">=</span><span class="fl">0.35</span>,alpha<span class="op">=</span><span class="fl">0.75</span><span class="op">)</span></span></code></pre></div>
<p><img src="mgdrive_run_files/figure-html/unnamed-chunk-32-1.png" alt="Example of how to plot all repetitions from a set of simulations." width="672"></p>
<p><code><a href="../reference/plotMGDrivEMult.html">plotMGDrivEMult()</a></code> takes the name of the main directory,
and finds all of the repetitions inside of it. It then plots males and
females by patch, the same as <code><a href="../reference/plotMGDrivESingle.html">plotMGDrivESingle()</a></code>, but
overlays a trace of every repetition for each patch.</p>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://chipdelmal.github.io" class="external-link">Héctor Manuel Sánchez Castellanos</a>, <a href="https://github.com/GilChrist19" class="external-link">Jared Bennett</a>, <a href="https://slwu89.github.io" class="external-link">Sean Wu</a>, <a href="https://www.marshalllab.com" class="external-link">John M. Marshall</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
