<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>MGDrivE2: One Node Lifecycle Dynamics • MGDrivE2</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="MGDrivE2: One Node Lifecycle Dynamics">
<meta property="og:description" content="MGDrivE2">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MGDrivE2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/advanced_topics.html">MGDrivE2: Advanced Topics</a>
    </li>
    <li>
      <a href="../articles/epi-SEIR.html">MGDrivE2: SEIR Epidemiological Dynamics</a>
    </li>
    <li>
      <a href="../articles/epi-network.html">MGDrivE2: Metapopulation Network Epidemiological Dynamics</a>
    </li>
    <li>
      <a href="../articles/epi-node.html">MGDrivE2: One Node Epidemiological Dynamics</a>
    </li>
    <li>
      <a href="../articles/inhomogeneous.html">MGDrivE2: Simulation of Time-inhomogeneous Stochastic Processes (Seasonality)</a>
    </li>
    <li>
      <a href="../articles/lifecycle-network.html">MGDrivE2: Metapopulation Network Lifecycle Dynamics</a>
    </li>
    <li>
      <a href="../articles/lifecycle-node.html">MGDrivE2: One Node Lifecycle Dynamics</a>
    </li>
    <li>
      <a href="../articles/output-storage.html">MGDrivE2: Data Storage and Analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/MarshallLab/MGDrivE/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="lifecycle-node_files/header-attrs-2.5/header-attrs.js"></script><link href="lifecycle-node_files/anchor-sections-1.0/anchor-sections.css" rel="stylesheet">
<script src="lifecycle-node_files/anchor-sections-1.0/anchor-sections.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>MGDrivE2: One Node Lifecycle Dynamics</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/MarshallLab/MGDrivE/blob/master/vignettes/lifecycle-node.Rmd"><code>vignettes/lifecycle-node.Rmd</code></a></small>
      <div class="hidden name"><code>lifecycle-node.Rmd</code></div>

    </div>

    
    
<div id="table-of-contents" class="section level2">
<h2 class="hasAnchor">
<a href="#table-of-contents" class="anchor"></a>Table of Contents</h2>
<ol style="list-style-type: decimal">
<li>
<a href="#ex_log">Example: Logistic Density Dependence</a>
<ol style="list-style-type: decimal">
<li><a href="#pars_log">Parameterization</a></li>
<li><a href="#init_pn_log">Initialization of the Petri Net</a></li>
<li><a href="#equilibria_haz_log">Equilibrium Conditions and Hazard Functions</a></li>
<li>
<a href="#sim_log">Simulation of Fully Specified SPN Model</a>
<ol style="list-style-type: decimal">
<li><a href="#det_ode_soln_log">Deterministic: ODE Solutions</a></li>
<li><a href="#stoch_tau_soln_log">Stochastic: Tau-leaping Solutions</a></li>
</ol>
</li>
</ol>
</li>
<li>
<a href="#ex_lk">Example: Lotka-Volterra Density Dependence</a>
<ol style="list-style-type: decimal">
<li><a href="#equilibria_haz_lk">Equilibrium Conditions and Hazard Functions</a></li>
<li>
<a href="#sim_lk">Simulation of Fully Specified SPN Model</a>
<ol style="list-style-type: decimal">
<li><a href="#det_ode_soln_lk">Deterministic: ODE Solutions</a></li>
<li><a href="#stoch_sde_soln_lk">Stochastic: CLE Solutions</a></li>
</ol>
</li>
</ol>
</li>
</ol>
<div id="preface" class="section level3">
<h3 class="hasAnchor">
<a href="#preface" class="anchor"></a>Preface</h3>
<p>In this vignette, we show how to run <strong>MGDrivE2</strong> simulations using a variety of sampling techniques, including deterministic integration with ODE solvers from the <code>deSolve</code> package. We demonstrate a logistic (carrying capacity) form of larval mortality as well as the Lotka-Volterra form. While the behavior of the two types of density-dependent mortality will be nearly identical, because they are both quadratic in the number of larvae, this is meant to demonstrate the relative ease at which new density-dependent (and in general, any hazard, see <a href="inhomogeneous.html">“MGDrivE2: Simulation of Time-inhomogeneous Stochastic Processes (Seasonality)”</a> for a more complex example) functions can be swapped as necessary for the simulation task at hand.</p>
</div>
</div>
<div id="ex_log" class="section level1">
<h1 class="hasAnchor">
<a href="#ex_log" class="anchor"></a>Example: Logistic Density Dependence</h1>
<p>We start by loading the <strong>MGDrivE2</strong> package, as well as the <strong>MGDrivE</strong> package for access to inheritance cubes and <strong>ggplot2</strong> for graphical analysis. We will use the basic cube to simulate Mendelian inheritance for this example.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># simulation functions</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://marshalllab.github.io/MGDrivE/">MGDrivE2</a></span><span class="op">)</span>
<span class="co">#&gt; Loading MGDrivE2: Mosquito Gene Drive Explorer Version 2</span>
<span class="co"># inheritance patterns</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://marshalllab.github.io/MGDrivE/">MGDrivE</a></span><span class="op">)</span>
<span class="co">#&gt; Loading MGDrivE: Mosquito Gene Drive Explorer</span>
<span class="co"># plotting</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>

<span class="co"># basic inheritance pattern</span>
<span class="va">cube</span> <span class="op">&lt;-</span> <span class="fu">MGDrivE</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MGDrivE/man/cubeMendelian.html">cubeMendelian</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div id="pars_log" class="section level2">
<h2 class="hasAnchor">
<a href="#pars_log" class="anchor"></a>Parameterization</h2>
<p>Next, we set entomological parameters that dictate life-history characteristics, as well as the equilibrium number of adult female mosquitoes. The density dependent per-capita larval mortality is thus: <span class="math inline">\(\mu_{L}(1 + \frac{L}{K})\)</span>, where <span class="math inline">\(L\)</span> is the current larval population size. Note that when the population is zero there is no effect of density dependence.</p>
<p>The parameters <code>qE</code>, <code>qL</code>, and <code>qP</code> are the mean dwell times in the egg, larval, and pupae stages, respectively. The Erlang distributed aquatic stages allow users to select the mean and variance of the dwell time distribution for each stage. For example, <span class="math inline">\(\frac{1}{q_{E}}\)</span> will give the mean dwell time in the egg stage, and variance <span class="math inline">\(\frac{1}{n_{E} \cdot q_{E}^{2}}\)</span>. A table of (case-sensitive) biological parameters the user needs to specify is given below (Note that all parameters must be specified as <strong>rates per day</strong>). The equilibrium constant for larval density dependence (<code>K</code>) is returned when equilibrium is calculated.</p>
<p>The <code>nu</code> parameter gives the rate at which unmated female mosquitoes mate when males become available again. Ordinarily, female mosquitoes are assumed to mate upon emergence if males are present. However if male mosquitoes become locally extinct and are later replenished either by currently developing egg batches or releases (or for metapopulation simulations, inbound male migration), then there will be some time where newly emerging adult females are unable to find mates. These unmated female mosquitoes experience mortality at the same rate as mated females, <code>muF</code>. Because they are unmated, they do not produce eggs, nor do they blood-feed (which is done in order to provision developing egg batches). While blood-feeding behavior doesn’t matter for lifecycle simulations, this means that for the epidemiological simulations we don’t need to worry about unmated females becoming infected, because they will not be biologically primed for this behavior until after they have mated and need to produce egg batches. Once males become available, they will mate at rate <code>nu</code>.</p>
<table class="table">
<thead><tr class="header">
<th align="center">Parameter</th>
<th align="left">Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center"><code>qE</code></td>
<td align="left">inverse of mean duration of egg stage</td>
</tr>
<tr class="even">
<td align="center"><code>nE</code></td>
<td align="left">shape parameter of Erlang-distributed egg stage</td>
</tr>
<tr class="odd">
<td align="center"><code>qL</code></td>
<td align="left">inverse of mean duration of larval stage</td>
</tr>
<tr class="even">
<td align="center"><code>nL</code></td>
<td align="left">shape parameter of Erlang-distributed larval stage</td>
</tr>
<tr class="odd">
<td align="center"><code>qP</code></td>
<td align="left">inverse of mean duration of pupal stage</td>
</tr>
<tr class="even">
<td align="center"><code>nP</code></td>
<td align="left">shape parameter of Erlang-distributed pupal stage</td>
</tr>
<tr class="odd">
<td align="center"><code>muE</code></td>
<td align="left">density-independent egg mortality</td>
</tr>
<tr class="even">
<td align="center"><code>muL</code></td>
<td align="left">density-independent larval mortality</td>
</tr>
<tr class="odd">
<td align="center"><code>muP</code></td>
<td align="left">density-independent pupal mortality</td>
</tr>
<tr class="even">
<td align="center"><code>muF</code></td>
<td align="left">density-independent adult female mortality</td>
</tr>
<tr class="odd">
<td align="center"><code>muM</code></td>
<td align="left">density-independent adult male mortality</td>
</tr>
<tr class="even">
<td align="center"><code>beta</code></td>
<td align="left">rate of egg production in adult females</td>
</tr>
<tr class="odd">
<td align="center"><code>nu</code></td>
<td align="left">mating rate of unmated females</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># number of adult female mosquitoes</span>
<span class="va">NF</span> <span class="op">&lt;-</span> <span class="fl">500</span>

<span class="co"># entomological parameters</span>
<span class="va">theta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>
  qE <span class="op">=</span> <span class="fl">1</span><span class="op">/</span><span class="fl">4</span>,
  nE <span class="op">=</span> <span class="fl">2</span>,
  qL <span class="op">=</span> <span class="fl">1</span><span class="op">/</span><span class="fl">3</span>,
  nL <span class="op">=</span> <span class="fl">3</span>,
  qP <span class="op">=</span> <span class="fl">1</span><span class="op">/</span><span class="fl">6</span>,
  nP <span class="op">=</span> <span class="fl">2</span>,
  muE <span class="op">=</span> <span class="fl">0.05</span>,
  muL <span class="op">=</span> <span class="fl">0.15</span>,
  muP <span class="op">=</span> <span class="fl">0.05</span>,
  muF <span class="op">=</span> <span class="fl">0.09</span>,
  muM <span class="op">=</span> <span class="fl">0.09</span>,
  beta <span class="op">=</span> <span class="fl">16</span>,
  nu <span class="op">=</span> <span class="fl">1</span><span class="op">/</span><span class="op">(</span><span class="fl">4</span><span class="op">/</span><span class="fl">24</span><span class="op">)</span>
<span class="op">)</span></code></pre></div>
</div>
<div id="init_pn_log" class="section level2">
<h2 class="hasAnchor">
<a href="#init_pn_log" class="anchor"></a>Initialization of the Petri Net</h2>
<p>The next thing we need to do is set up the structural properties of the SPN model (technically just a Petri Net, PN, at this point). First, we set all of the “places” in the model. As this is a compartmental model, “places” are simply distinct life stages (and locations for metapopulation versions). Thus, the “places” include all aquatic stages (number of bins in each egg, larval, and pupal stages), adult males, and adult females (further separated by mate genotype). All of these stages are unique for each genotype as well, so the number of “places” (i.e., compartments) grows with the number of genotypes and Erlang-distributed stage length.</p>
<p>Once the places for the SPN are setup, we define the transitions between each place. Transitions are how objects move from one compartment to another. These are not the rates that objects change compartments, only whether or not such a move is possible. The rate of movement from one “place” to another are calculated later by the <code><a href="../reference/spn_hazards.html">spn_hazards()</a></code> function.</p>
<p>Finally, as not all transitions apply to all “places”, we create a summary of possible transitions to and from each “place”. This is handled by the <code><a href="../reference/spn_S.html">spn_S()</a></code> function.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># "Places"</span>
<span class="co">#  These are defined by Erlang-distributed life stages and genotypes</span>
<span class="va">SPN_P</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spn_P_lifecycle_node.html">spn_P_lifecycle_node</a></span><span class="op">(</span>params <span class="op">=</span> <span class="va">theta</span>,cube <span class="op">=</span> <span class="va">cube</span><span class="op">)</span>

<span class="co"># Transitions</span>
<span class="co"># This is a list of viable transitions from one "place" to another</span>
<span class="va">SPN_T</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spn_T_lifecycle_node.html">spn_T_lifecycle_node</a></span><span class="op">(</span>spn_P <span class="op">=</span> <span class="va">SPN_P</span>,params <span class="op">=</span> <span class="va">theta</span>,cube <span class="op">=</span> <span class="va">cube</span><span class="op">)</span>

<span class="co"># Stoichiometry matrix</span>
<span class="co">#  A sparse matrix representing the effect of each transition on each place</span>
<span class="va">S</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spn_S.html">spn_S</a></span><span class="op">(</span>spn_P <span class="op">=</span> <span class="va">SPN_P</span>, spn_T <span class="op">=</span> <span class="va">SPN_T</span><span class="op">)</span></code></pre></div>
</div>
<div id="equilibria_haz_log" class="section level2">
<h2 class="hasAnchor">
<a href="#equilibria_haz_log" class="anchor"></a>Equilibrium Conditions and Hazard Functions</h2>
<p>Once the “places” are defined, we use <code><a href="../reference/equilibrium_lifeycle.html">equilibrium_lifeycle()</a></code> to get initial conditions (<span class="math inline">\(M(0)\)</span>), solved at the dynamic equilibrium for <code>NF</code> number of adult females, as well as the density-dependent constants for larval competition, either <code>K</code> (logistic dependence) or <code>gamma</code> (Lotka-Volterra dependence).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># calculate equilibrium and setup initial conditions</span>
<span class="co">#  outputs required parameters in the named list "params"</span>
<span class="co">#  outputs intial equilibrium for adv users, "init</span>
<span class="co">#  outputs properly filled initial markings, "M0", vectorized over:</span>
<span class="co">#    all patches</span>
<span class="co">#    any genotypes( includes XX/XY inheritance patterns)</span>
<span class="co">#    allows initial ratios to vary</span>
<span class="co">#    properly mates males/females</span>
<span class="va">initialCons</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/equilibrium_lifeycle.html">equilibrium_lifeycle</a></span><span class="op">(</span>params <span class="op">=</span> <span class="va">theta</span>, NF <span class="op">=</span> <span class="va">NF</span>, phi <span class="op">=</span> <span class="fl">0.5</span>,
                                    log <span class="op">=</span> <span class="cn">TRUE</span>, spn_P <span class="op">=</span> <span class="va">SPN_P</span>, cube <span class="op">=</span> <span class="va">cube</span><span class="op">)</span></code></pre></div>
<p>The transitions in our SPN define possible movement from one compartment to another, but they do not define the rate that these changes occur. The rate of change from one compartment to another is called a “hazard”, calculated by the <code><a href="../reference/spn_hazards.html">spn_hazards()</a></code> function. This calculates the rate of change from one compartment to another. Below, we make both a set of exact hazards, for sampling algorithms with integer valued state spaces, and a set of approximate hazards, for algorithms using a continuous approximation.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># approximate hazards for continous approximation</span>
<span class="va">approx_hazards</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spn_hazards.html">spn_hazards</a></span><span class="op">(</span>spn_P <span class="op">=</span> <span class="va">SPN_P</span>, spn_T <span class="op">=</span> <span class="va">SPN_T</span>, cube <span class="op">=</span> <span class="va">cube</span>,
                              params <span class="op">=</span> <span class="va">initialCons</span><span class="op">$</span><span class="va">params</span>, type <span class="op">=</span> <span class="st">"life"</span>,
                              log <span class="op">=</span> <span class="cn">TRUE</span>, exact <span class="op">=</span> <span class="cn">FALSE</span>, tol <span class="op">=</span> <span class="fl">1e-8</span>,
                              verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="co"># exact hazards for integer-valued state space</span>
<span class="va">exact_hazards</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spn_hazards.html">spn_hazards</a></span><span class="op">(</span>spn_P <span class="op">=</span> <span class="va">SPN_P</span>, spn_T <span class="op">=</span> <span class="va">SPN_T</span>, cube <span class="op">=</span> <span class="va">cube</span>,
                             params <span class="op">=</span> <span class="va">initialCons</span><span class="op">$</span><span class="va">params</span>, type <span class="op">=</span> <span class="st">"life"</span>,
                             log <span class="op">=</span> <span class="cn">TRUE</span>, exact <span class="op">=</span> <span class="cn">TRUE</span>, tol <span class="op">=</span> <span class="cn">NaN</span>,
                             verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
</div>
<div id="sim_log" class="section level2">
<h2 class="hasAnchor">
<a href="#sim_log" class="anchor"></a>Simulation of Fully Specified SPN Model</h2>
<p>Finally, before we run simulations, we set up a release scheme to generate interesting dynamics. We will release 50 adult females with homozygous recessive alleles 5 times, every 10 days, starting at day 20. Releases are possible at every life-stage, so long as the event name matches the place in the simulation. It is critically important that <strong>the event names match a place name</strong> in the simulation. The simulation function checks this and will throw an error if the event name does not exist as a place in the simulation. This format is used in <strong>MGDrivE2</strong> for consistency with solvers in <code>deSolve</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># releases</span>
<span class="va">r_times</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span>from <span class="op">=</span> <span class="fl">20</span>, length.out <span class="op">=</span> <span class="fl">5</span>, by <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>
<span class="va">r_size</span> <span class="op">&lt;-</span> <span class="fl">50</span>
<span class="va">events</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="st">"var"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"F_"</span>, <span class="va">cube</span><span class="op">$</span><span class="va">releaseType</span>, <span class="st">"_"</span>, <span class="va">cube</span><span class="op">$</span><span class="va">wildType</span><span class="op">)</span>,
                     <span class="st">"time"</span> <span class="op">=</span> <span class="va">r_times</span>,
                     <span class="st">"value"</span> <span class="op">=</span> <span class="va">r_size</span>,
                     <span class="st">"method"</span> <span class="op">=</span> <span class="st">"add"</span>,
                     stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<div id="det_ode_soln_log" class="section level3">
<h3 class="hasAnchor">
<a href="#det_ode_soln_log" class="anchor"></a>Deterministic: ODE Solutions</h3>
<p>The first simulation we will run is a mean-field deterministic approximation to the full stochastic model. Formally, this is a moment-closure technique where we model the 1st moment and ignore contributions from all higher order moments, so we can expect that the ODEs will approximate the mean of stochastic simulations well only when nonlinear effects and influence of higher order moments are small. For this continuous-state approximation, we need to use the approximate hazards, created by setting the following parameters <code>exact = FALSE, tol = 1e-8</code> (above) in the function <code><a href="../reference/spn_hazards.html">spn_hazards()</a></code>. When used in ODE approximation, the interpretation of hazard functions simplifies to simple rate functions.</p>
<p>Internally, <strong>MGDrivE2</strong> uses the high quality numerical solvers in from <code>deSolve</code> to integrate a mean-field approximation to the stochastic model. Effectively, it evaluates the ODEs over the interval between time points where the user requested model output, calculating the rate of that hazard, given the current simulation state, multiplied with the stoichiometry matrix. The default method used is <code>lsoda</code>, though for highly variable rates and inhomogeneous systems, consider using <code>rk4</code>.</p>
<p>Now that the deterministic model is fully specified, let’s run the simulation for 200 days, outputting values of the state variables every day.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># max simulation time</span>
<span class="va">tmax</span> <span class="op">&lt;-</span> <span class="fl">125</span>
<span class="co"># time-step for output return, not the time-step of the sampling algorithm</span>
<span class="va">dt</span> <span class="op">&lt;-</span> <span class="fl">1</span>

<span class="co"># run deterministic simulation</span>
<span class="va">ODE_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_trajectory_R.html">sim_trajectory_R</a></span><span class="op">(</span>x0 <span class="op">=</span> <span class="va">initialCons</span><span class="op">$</span><span class="va">M0</span>, tmax <span class="op">=</span> <span class="va">tmax</span>, dt <span class="op">=</span> <span class="va">dt</span>, S <span class="op">=</span> <span class="va">S</span>,
                            hazards <span class="op">=</span> <span class="va">approx_hazards</span>, sampler <span class="op">=</span> <span class="st">"ode"</span>, method <span class="op">=</span> <span class="st">"lsoda"</span>,
                            events <span class="op">=</span> <span class="va">events</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<p><strong>MGDrivE2</strong> has several helper functions for summarizing simulation data in a nice format for plotting. Here, we will look at the adult male and female data, though there are many more options explored below.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># summarize females by genotype</span>
<span class="va">ODE_out_f</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarize_females.html">summarize_females</a></span><span class="op">(</span>out <span class="op">=</span> <span class="va">ODE_out</span><span class="op">$</span><span class="va">state</span>, spn_P <span class="op">=</span> <span class="va">SPN_P</span><span class="op">)</span>

<span class="co"># summarize males by genotype</span>
<span class="va">ODE_out_m</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarize_males.html">summarize_males</a></span><span class="op">(</span>out <span class="op">=</span> <span class="va">ODE_out</span><span class="op">$</span><span class="va">state</span><span class="op">)</span>

<span class="co"># add sex for plotting</span>
<span class="va">ODE_out_f</span><span class="op">$</span><span class="va">sex</span> <span class="op">&lt;-</span> <span class="st">"Female"</span>
<span class="va">ODE_out_m</span><span class="op">$</span><span class="va">sex</span> <span class="op">&lt;-</span> <span class="st">"Male"</span>

<span class="co"># plot</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span><span class="va">ODE_out_f</span>, <span class="va">ODE_out_m</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">time</span>, y <span class="op">=</span> <span class="va">value</span>, color <span class="op">=</span> <span class="va">genotype</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span>facets <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html">vars</a></span><span class="op">(</span><span class="va">sex</span><span class="op">)</span>, scales <span class="op">=</span> <span class="st">"fixed"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"SPN: ODE Solution"</span><span class="op">)</span></code></pre></div>
<p><img src="lifecycle-node_files/figure-html/unnamed-chunk-9-1.png" width="691.2"></p>
<p>We see the initial equilibrium at 500 females, as was set above, and 500 males, because we defined an equal sex breakdown in the population. At time <code>t = 25</code>, we see the first release of <code>aa</code> females into the population, and then the 4 subsequent releases. Notice that releases only appear in the females, nothing was released into the males. Since there are no fitness costs, we see the <code>aa</code> individuals come to equilibrium predominantly as heterozygotes, i.e. <code>Aa</code>, with a very small fraction of homozygotes, and the majority of the population remaining <code>AA</code>.</p>
</div>
<div id="stoch_tau_soln_log" class="section level3">
<h3 class="hasAnchor">
<a href="#stoch_tau_soln_log" class="anchor"></a>Stochastic: Tau-leaping Solutions</h3>
<p>For stochastic simulations, we generally rely on tau-leaping; this method retains the integer valued state space, thus necessitating the use of exact hazards. The theoretical basis of tau-leaping depends upon realizing that each transition can be represented as a Poisson process; tau-leaping is therefore an Euler-method for solving continuous-time Markov chains where the “tau-leaping” refers to a fixed time-step of size “tau”, where the number of each transition that fires over such a time-step (<span class="math inline">\(\Delta t\)</span>) is approximated by a Poisson random variable with a linear approximation to the cumulative hazard. This is provided as the <code>dt_stoch</code> an argument to the function <code><a href="../reference/sim_trajectory_R.html">sim_trajectory_R()</a></code>.</p>
<p>We set <code>tau</code> as our sampler type, use the same release events, conditions, and stoichiometry matrix as above, set the exact hazards instead of the approximate, and finally run the simulation and plot the output.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># delta t</span>
<span class="va">dt_stoch</span> <span class="op">&lt;-</span> <span class="fl">0.1</span>

<span class="co"># tau sampling</span>
<span class="va">PTS_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_trajectory_R.html">sim_trajectory_R</a></span><span class="op">(</span>x0 <span class="op">=</span> <span class="va">initialCons</span><span class="op">$</span><span class="va">M0</span>, tmax <span class="op">=</span> <span class="va">tmax</span>, dt <span class="op">=</span> <span class="va">dt</span>,
                            dt_stoch <span class="op">=</span> <span class="va">dt_stoch</span>, S <span class="op">=</span> <span class="va">S</span>, hazards <span class="op">=</span> <span class="va">exact_hazards</span>,
                            sampler <span class="op">=</span> <span class="st">"tau"</span>, events <span class="op">=</span> <span class="va">events</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="co"># summarize females/males</span>
<span class="va">PTS_out_f</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarize_females.html">summarize_females</a></span><span class="op">(</span>out <span class="op">=</span> <span class="va">PTS_out</span><span class="op">$</span><span class="va">state</span>, spn_P <span class="op">=</span> <span class="va">SPN_P</span><span class="op">)</span>
<span class="va">PTS_out_m</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarize_males.html">summarize_males</a></span><span class="op">(</span>out <span class="op">=</span> <span class="va">PTS_out</span><span class="op">$</span><span class="va">state</span><span class="op">)</span>

<span class="co"># add sex for plotting</span>
<span class="va">PTS_out_f</span><span class="op">$</span><span class="va">sex</span> <span class="op">&lt;-</span> <span class="st">"Female"</span>
<span class="va">PTS_out_m</span><span class="op">$</span><span class="va">sex</span> <span class="op">&lt;-</span> <span class="st">"Male"</span>

<span class="co"># plot adults</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span><span class="va">PTS_out_f</span>, <span class="va">PTS_out_m</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">time</span>, y <span class="op">=</span> <span class="va">value</span>, color <span class="op">=</span> <span class="va">genotype</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span>facets <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html">vars</a></span><span class="op">(</span><span class="va">sex</span><span class="op">)</span>, scales <span class="op">=</span> <span class="st">"fixed"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"SPN: Tau-leaping Approximation"</span><span class="op">)</span></code></pre></div>
<p><img src="lifecycle-node_files/figure-html/unnamed-chunk-10-1.png" width="691.2"></p>
<p>As Tau-leaping is a stochastic approximation of our system, we should see some variation around the solution provided by the ODE simulation above. These simulations begin at the proper equilibrium amount, 500 individuals of each sex, with releases performed in females and an equilibrium shortly thereafter. The plots are heuristically similar to the ODE solution, which is good, but notice the slow trend of <code>Aa</code> individuals towards extinction. Since there is such a small amount of <code>a</code> alleles in the population, they are steadily being lost through stochastic drift. If we carried the simulations out longer, it is possible that the <code>a</code> allele will be completely lost from the population.</p>
</div>
</div>
</div>
<div id="ex_lk" class="section level1">
<h1 class="hasAnchor">
<a href="#ex_lk" class="anchor"></a>Example: Lotka-Volterra Density Dependence</h1>
<p>Now we will show an example with the Lotka-Volterra style density dependence in larval stages. Please note that we do not need to recreate the structural pieces of the <a href="#init_pn_log">Petri Net</a> (places and transitions), or update any of the <a href="#pars_log">parameters</a>, because they remain the same; we simply need to calculate a new equilibrium point and generate new hazards.</p>
<div id="equilibria_haz_lk" class="section level2">
<h2 class="hasAnchor">
<a href="#equilibria_haz_lk" class="anchor"></a>Equilibrium Conditions and Hazard Functions</h2>
<p>To use Lotka-Volterra density dependence, we need to recalculate the equilibrium and then rebuild the hazard functions. The per-capital larval mortality is <span class="math inline">\(\mu_{L} + \gamma L\)</span>, noted by <code>log = FALSE</code> in the options to <code><a href="../reference/equilibrium_lifeycle.html">equilibrium_lifeycle()</a></code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># using the same parameters as above, along with the SPN_P object already created</span>
<span class="co"># calculate equilibrium for lotka-volterra dynamics</span>
<span class="va">initialCons</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/equilibrium_lifeycle.html">equilibrium_lifeycle</a></span><span class="op">(</span>params <span class="op">=</span> <span class="va">theta</span>, NF <span class="op">=</span> <span class="va">NF</span>, phi <span class="op">=</span> <span class="fl">0.5</span>,
                                    log <span class="op">=</span> <span class="cn">FALSE</span>, spn_P <span class="op">=</span> <span class="va">SPN_P</span>,cube<span class="op">=</span><span class="va">cube</span><span class="op">)</span></code></pre></div>
<p>We now follow the same steps as before to make the vector of hazards (setting <code>log = FALSE</code>).</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># approximate hazards for continous approximation</span>
<span class="va">approx_hazards</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spn_hazards.html">spn_hazards</a></span><span class="op">(</span>spn_P <span class="op">=</span> <span class="va">SPN_P</span>, spn_T <span class="op">=</span> <span class="va">SPN_T</span>, cube <span class="op">=</span> <span class="va">cube</span>,
                              params <span class="op">=</span> <span class="va">initialCons</span><span class="op">$</span><span class="va">params</span>, type <span class="op">=</span> <span class="st">"life"</span>,
                              exact <span class="op">=</span> <span class="cn">FALSE</span>, tol <span class="op">=</span> <span class="fl">1e-8</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span>,
                              log <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="co"># exact hazards for integer-valued state space</span>
<span class="va">exact_hazards</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spn_hazards.html">spn_hazards</a></span><span class="op">(</span>spn_P <span class="op">=</span> <span class="va">SPN_P</span>, spn_T <span class="op">=</span> <span class="va">SPN_T</span>, cube <span class="op">=</span> <span class="va">cube</span>,
                             params <span class="op">=</span> <span class="va">initialCons</span><span class="op">$</span><span class="va">params</span>, type <span class="op">=</span> <span class="st">"life"</span>,
                             exact <span class="op">=</span> <span class="cn">TRUE</span>, tol <span class="op">=</span> <span class="cn">NaN</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span>,
                             log <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
</div>
<div id="sim_lk" class="section level2">
<h2 class="hasAnchor">
<a href="#sim_lk" class="anchor"></a>Simulation of Fully Specified SPN Model</h2>
<p>We will use the releases defined <a href="#sim_log">above</a>, and again begin with an ODE solution to our system, before testing a different stochastic sampler.</p>
</div>
<div id="det_ode_soln_lk" class="section level2">
<h2 class="hasAnchor">
<a href="#det_ode_soln_lk" class="anchor"></a>Deterministic: ODE Solutions</h2>
<p>As before, we run an ODE simulation, but this time we will plot the aquatic stages in two ways: looking at their genotypes over time, and looking at the progress through Erlang-distributed stages.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># deterministic simulation</span>
<span class="va">ODE_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_trajectory_R.html">sim_trajectory_R</a></span><span class="op">(</span>x0 <span class="op">=</span> <span class="va">initialCons</span><span class="op">$</span><span class="va">M0</span>, tmax <span class="op">=</span> <span class="va">tmax</span>, dt <span class="op">=</span> <span class="va">dt</span>,
                            S <span class="op">=</span> <span class="va">S</span>, hazards <span class="op">=</span> <span class="va">approx_hazards</span>, sampler <span class="op">=</span> <span class="st">"ode"</span>,
                            events <span class="op">=</span> <span class="va">events</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="co"># summarize aquatic stages by genotype</span>
<span class="va">ODE_out_e</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarize_eggs_geno.html">summarize_eggs_geno</a></span><span class="op">(</span>out <span class="op">=</span> <span class="va">ODE_out</span><span class="op">$</span><span class="va">state</span>, spn_P <span class="op">=</span> <span class="va">SPN_P</span><span class="op">)</span>
<span class="va">ODE_out_l</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarize_larvae_geno.html">summarize_larvae_geno</a></span><span class="op">(</span>out <span class="op">=</span> <span class="va">ODE_out</span><span class="op">$</span><span class="va">state</span>, spn_P <span class="op">=</span> <span class="va">SPN_P</span><span class="op">)</span>
<span class="va">ODE_out_p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarize_pupae_geno.html">summarize_pupae_geno</a></span><span class="op">(</span>out <span class="op">=</span> <span class="va">ODE_out</span><span class="op">$</span><span class="va">state</span>, spn_P <span class="op">=</span> <span class="va">SPN_P</span><span class="op">)</span>

<span class="co"># add stage name</span>
<span class="va">ODE_out_e</span><span class="op">$</span><span class="va">stage</span> <span class="op">&lt;-</span> <span class="st">"Egg"</span>
<span class="va">ODE_out_l</span><span class="op">$</span><span class="va">stage</span> <span class="op">&lt;-</span> <span class="st">"Larvae"</span>
<span class="va">ODE_out_p</span><span class="op">$</span><span class="va">stage</span> <span class="op">&lt;-</span> <span class="st">"Pupae"</span>

<span class="co"># plot by genotype</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span><span class="va">ODE_out_e</span>, <span class="va">ODE_out_l</span>,<span class="va">ODE_out_p</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">time</span>, y <span class="op">=</span> <span class="va">value</span>, color <span class="op">=</span> <span class="va">genotype</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span>facets <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html">vars</a></span><span class="op">(</span><span class="va">stage</span><span class="op">)</span>, scales <span class="op">=</span> <span class="st">"free_y"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"SPN: ODE Solution - Genotypes"</span><span class="op">)</span></code></pre></div>
<p><img src="lifecycle-node_files/figure-html/unnamed-chunk-13-1.png" width="691.2"></p>
<p>Notice the vastly different equilibrium number for the aquatic stages. These values are parameterized from the adult female value, and the death rates at each stage. Notice that the “releases” here appear to be <code>Aa</code> individuals - this is because our released females (<code>aa</code>) mated with wild-type males (<code>AA</code>), and produced heterozygous (<code>Aa</code>) offspring. Also notice how the releases are smoothed out from Egg to Larvae and Larvae to Pupae stages, due to the death in each stage. We can also see a small dip in the pupal stage, stemming from the over-compensation of density-dependent mortality in the larval stage, before returning to equilibrium.</p>
<p>We can also explore the aquatic stages by dwell time instead of genotype.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># summarize aquatic stages by Erlang stage</span>
<span class="va">ODE_out_e</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarize_eggs_stage.html">summarize_eggs_stage</a></span><span class="op">(</span>out <span class="op">=</span> <span class="va">ODE_out</span><span class="op">$</span><span class="va">state</span>, spn_P <span class="op">=</span> <span class="va">SPN_P</span><span class="op">)</span>
<span class="va">ODE_out_l</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarize_larvae_stage.html">summarize_larvae_stage</a></span><span class="op">(</span>out <span class="op">=</span> <span class="va">ODE_out</span><span class="op">$</span><span class="va">state</span>, spn_P <span class="op">=</span> <span class="va">SPN_P</span><span class="op">)</span>
<span class="va">ODE_out_p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarize_pupae_stage.html">summarize_pupae_stage</a></span><span class="op">(</span>out <span class="op">=</span> <span class="va">ODE_out</span><span class="op">$</span><span class="va">state</span>, spn_P <span class="op">=</span> <span class="va">SPN_P</span><span class="op">)</span>

<span class="co"># add stage name</span>
<span class="va">ODE_out_e</span><span class="op">$</span><span class="va">stage</span> <span class="op">&lt;-</span> <span class="st">"Egg"</span>
<span class="va">ODE_out_l</span><span class="op">$</span><span class="va">stage</span> <span class="op">&lt;-</span> <span class="st">"Larvae"</span>
<span class="va">ODE_out_p</span><span class="op">$</span><span class="va">stage</span> <span class="op">&lt;-</span> <span class="st">"Pupae"</span>

<span class="co"># plot by Erlang stage</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span><span class="va">ODE_out_e</span>, <span class="va">ODE_out_l</span>,<span class="va">ODE_out_p</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">time</span>, y <span class="op">=</span> <span class="va">value</span>, color <span class="op">=</span> <span class="va">`Erlang-stage`</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span>facets <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html">vars</a></span><span class="op">(</span><span class="va">stage</span><span class="op">)</span>, scales <span class="op">=</span> <span class="st">"free_y"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"SPN: ODE Solution - Erlang Dwell Stage"</span><span class="op">)</span></code></pre></div>
<p><img src="lifecycle-node_files/figure-html/unnamed-chunk-14-1.png" width="691.2"></p>
<p>This plot shows how individuals are moving through each aquatic stage. There are significantly more individuals in the first Erlang-stage, compared to the second Erlang-stage, and we clearly see how the large increase in eggs is attenuated by death through the larval stage, and the effect of density-dependence actually lowering the amount of larvae that make it to the pupal stage.</p>
<p>Finally, for comparison with the logistic density-dependent dynamics above, and a different stochastic sampler below, we plot the adult stages.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># summarize females/males</span>
<span class="va">ODE_out_f</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarize_females.html">summarize_females</a></span><span class="op">(</span>out <span class="op">=</span> <span class="va">ODE_out</span><span class="op">$</span><span class="va">state</span>, spn_P <span class="op">=</span> <span class="va">SPN_P</span><span class="op">)</span>
<span class="va">ODE_out_m</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarize_males.html">summarize_males</a></span><span class="op">(</span>out <span class="op">=</span> <span class="va">ODE_out</span><span class="op">$</span><span class="va">state</span><span class="op">)</span>

<span class="co"># add sex for plotting</span>
<span class="va">ODE_out_f</span><span class="op">$</span><span class="va">sex</span> <span class="op">&lt;-</span> <span class="st">"Female"</span>
<span class="va">ODE_out_m</span><span class="op">$</span><span class="va">sex</span> <span class="op">&lt;-</span> <span class="st">"Male"</span>

<span class="co"># plot adults</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span><span class="va">ODE_out_f</span>, <span class="va">ODE_out_m</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">time</span>, y <span class="op">=</span> <span class="va">value</span>, color <span class="op">=</span> <span class="va">genotype</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span>facets <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html">vars</a></span><span class="op">(</span><span class="va">sex</span><span class="op">)</span>, scales <span class="op">=</span> <span class="st">"fixed"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"SPN: ODE Solution - Adult Stages"</span><span class="op">)</span></code></pre></div>
<p><img src="lifecycle-node_files/figure-html/unnamed-chunk-15-1.png" width="691.2"></p>
<p>Notice the similarity with the logistic density-dependent solution. Since there are no fitness effects or complex dynamics happening here, the solutions are effectively the same.</p>
<div id="stoch_sde_soln_lk" class="section level3">
<h3 class="hasAnchor">
<a href="#stoch_sde_soln_lk" class="anchor"></a>Stochastic: CLE Solutions</h3>
<p>As an alternative to tau-leaping, which is a discrete stochastic approximation, we investigate a continuous stochastic approximation (Chemical Langevin equation) to the full stochastic model. Theoretically, the CLE approximation is a second-order approximation using continuous state; it is the Fokker-Planck approximation to the Master equation (Kolmogorov Forwards Equations) of the integer valued stochastic process. Because we rely on a relatively simple Euler-Maruyama scheme to solve the system of stochastic differential equations, much like tau-leaping, the user must provide a <span class="math inline">\(\Delta t\)</span> to the <code><a href="../reference/sim_trajectory_R.html">sim_trajectory_R()</a></code> function, and specify the <code>cle</code> sampler.</p>
<p>Note the change in the sampler, to <code>cle</code>, and the use of approximate hazards.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># chemical langevin sampler</span>
<span class="va">CLE_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_trajectory_R.html">sim_trajectory_R</a></span><span class="op">(</span>x0 <span class="op">=</span> <span class="va">initialCons</span><span class="op">$</span><span class="va">M0</span>, tmax <span class="op">=</span> <span class="va">tmax</span>, dt <span class="op">=</span> <span class="va">dt</span>,
                            dt_stoch <span class="op">=</span> <span class="va">dt_stoch</span>, S <span class="op">=</span> <span class="va">S</span>, hazards <span class="op">=</span> <span class="va">approx_hazards</span>,
                            sampler <span class="op">=</span> <span class="st">"cle"</span>, events <span class="op">=</span> <span class="va">events</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="co"># summarize females/males</span>
<span class="va">CLE_out_f</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarize_females.html">summarize_females</a></span><span class="op">(</span>out <span class="op">=</span> <span class="va">CLE_out</span><span class="op">$</span><span class="va">state</span>, spn_P <span class="op">=</span> <span class="va">SPN_P</span><span class="op">)</span>
<span class="va">CLE_out_m</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarize_males.html">summarize_males</a></span><span class="op">(</span>out <span class="op">=</span> <span class="va">CLE_out</span><span class="op">$</span><span class="va">state</span><span class="op">)</span>

<span class="co"># add sex for plotting</span>
<span class="va">CLE_out_f</span><span class="op">$</span><span class="va">sex</span> <span class="op">&lt;-</span> <span class="st">"Female"</span>
<span class="va">CLE_out_m</span><span class="op">$</span><span class="va">sex</span> <span class="op">&lt;-</span> <span class="st">"Male"</span>

<span class="co"># plot adults</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span><span class="va">CLE_out_f</span>, <span class="va">CLE_out_m</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">time</span>, y <span class="op">=</span> <span class="va">value</span>, color <span class="op">=</span> <span class="va">genotype</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span>facets <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html">vars</a></span><span class="op">(</span><span class="va">sex</span><span class="op">)</span>, scales <span class="op">=</span> <span class="st">"fixed"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"SPN: Chemical Langevin Approximation"</span><span class="op">)</span></code></pre></div>
<p><img src="lifecycle-node_files/figure-html/unnamed-chunk-16-1.png" width="691.2"></p>
<p>Looking at this final set of plots, we immediately see the similarity with the ODE solution above. This is an indication that the stochastic time-step is sufficient for an accurate approximation.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="http://www.slwu89.com">Sean L. Wu</a>, <a href="https://github.com/GilChrist19">Jared B. Bennett</a>, <a href="https://www.marshalllab.com">John M. Marshall</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
