<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>MGDrivE2: Metapopulation Network Lifecycle Dynamics • MGDrivE2</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="MGDrivE2: Metapopulation Network Lifecycle Dynamics">
<meta property="og:description" content="MGDrivE2">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MGDrivE2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/advanced_topics.html">MGDrivE2: Advanced Topics</a>
    </li>
    <li>
      <a href="../articles/epi-SEIR.html">MGDrivE2: SEIR Epidemiological Dynamics</a>
    </li>
    <li>
      <a href="../articles/epi-network.html">MGDrivE2: Metapopulation Network Epidemiological Dynamics</a>
    </li>
    <li>
      <a href="../articles/epi-node.html">MGDrivE2: One Node Epidemiological Dynamics</a>
    </li>
    <li>
      <a href="../articles/inhomogeneous.html">MGDrivE2: Simulation of Time-inhomogeneous Stochastic Processes (Seasonality)</a>
    </li>
    <li>
      <a href="../articles/lifecycle-network.html">MGDrivE2: Metapopulation Network Lifecycle Dynamics</a>
    </li>
    <li>
      <a href="../articles/lifecycle-node.html">MGDrivE2: One Node Lifecycle Dynamics</a>
    </li>
    <li>
      <a href="../articles/output-storage.html">MGDrivE2: Data Storage and Analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/MarshallLab/MGDrivE/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="lifecycle-network_files/header-attrs-2.5/header-attrs.js"></script><link href="lifecycle-network_files/anchor-sections-1.0/anchor-sections.css" rel="stylesheet">
<script src="lifecycle-network_files/anchor-sections-1.0/anchor-sections.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>MGDrivE2: Metapopulation Network Lifecycle Dynamics</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/MarshallLab/MGDrivE/blob/master/vignettes/lifecycle-network.Rmd"><code>vignettes/lifecycle-network.Rmd</code></a></small>
      <div class="hidden name"><code>lifecycle-network.Rmd</code></div>

    </div>

    
    
<div id="table-of-contents" class="section level2">
<h2 class="hasAnchor">
<a href="#table-of-contents" class="anchor"></a>Table of Contents</h2>
<ol style="list-style-type: decimal">
<li><a href="#pars">Parameterization</a></li>
<li><a href="#init_pn">Initialization of the Petri Net</a></li>
<li><a href="#equilibria_haz">Equilibrium Conditions and Hazard Functions</a></li>
<li>
<a href="#sim">Simulation of Fully Specified SPN Model</a>
<ol style="list-style-type: decimal">
<li><a href="#soln_ode">Deterministic: ODE Solutions</a></li>
<li><a href="#soln_tau">Stochastic: Tau-leaping Solutions</a></li>
</ol>
</li>
</ol>
<div id="preface" class="section level3">
<h3 class="hasAnchor">
<a href="#preface" class="anchor"></a>Preface</h3>
<p>This vignette describes how to use the facilities provided by <strong>MGDrivE2</strong> to run simulations of gene drive dynamics on metapopulation networks; a network of mosquito breeding sites where edges represent the directed per-capita rate of migration between nodes. The movement rates are allowed to be asymmetric and to differ between male and female mosquitoes.</p>
<p>We start by loading the <strong>MGDrivE2</strong> package, as well as the <strong>MGDrivE</strong> package for access to inheritance cubes, <strong>ggplot2</strong> for graphical analysis, and <strong>Matrix</strong> for sparse matrices used in migration. We will use the basic cube to simulate Mendelian inheritance for this example.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># simulation functions</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://marshalllab.github.io/MGDrivE/">MGDrivE2</a></span><span class="op">)</span>
<span class="co">#&gt; Loading MGDrivE2: Mosquito Gene Drive Explorer Version 2</span>
<span class="co"># inheritance patterns</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://marshalllab.github.io/MGDrivE/">MGDrivE</a></span><span class="op">)</span>
<span class="co">#&gt; Loading MGDrivE: Mosquito Gene Drive Explorer</span>
<span class="co"># plotting</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="co"># sparse migration</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://Matrix.R-forge.R-project.org/">Matrix</a></span><span class="op">)</span>

<span class="co"># basic inheritance pattern</span>
<span class="va">cube</span> <span class="op">&lt;-</span> <span class="fu">MGDrivE</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MGDrivE/man/cubeMendelian.html">cubeMendelian</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="pars" class="section level2">
<h2 class="hasAnchor">
<a href="#pars" class="anchor"></a>Parameterization</h2>
<p>We specify biological parameter values and equilibrium population sizes. We will use the same parameter values as in the vignette <a href="lifecycle-node.html">“MGDrivE2: One Node Lifecycle Dynamics”</a>, additionally specifying the total simulation time (<code>tmax</code>) and the timestep to save output (<code>dt</code>).</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># entomological parameters</span>
<span class="va">theta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>
  qE <span class="op">=</span> <span class="fl">1</span><span class="op">/</span><span class="fl">4</span>,
  nE <span class="op">=</span> <span class="fl">2</span>,
  qL <span class="op">=</span> <span class="fl">1</span><span class="op">/</span><span class="fl">3</span>,
  nL <span class="op">=</span> <span class="fl">3</span>,
  qP <span class="op">=</span> <span class="fl">1</span><span class="op">/</span><span class="fl">6</span>,
  nP <span class="op">=</span> <span class="fl">2</span>,
  muE <span class="op">=</span> <span class="fl">0.05</span>,
  muL <span class="op">=</span> <span class="fl">0.15</span>,
  muP <span class="op">=</span> <span class="fl">0.05</span>,
  muF <span class="op">=</span> <span class="fl">0.09</span>,
  muM <span class="op">=</span> <span class="fl">0.09</span>,
  beta <span class="op">=</span> <span class="fl">16</span>,
  nu <span class="op">=</span> <span class="fl">1</span><span class="op">/</span><span class="op">(</span><span class="fl">4</span><span class="op">/</span><span class="fl">24</span><span class="op">)</span>
<span class="op">)</span>

<span class="co"># simulation parameters</span>
<span class="va">tmax</span> <span class="op">&lt;-</span> <span class="fl">125</span>
<span class="va">dt</span> <span class="op">&lt;-</span> <span class="fl">1</span></code></pre></div>
</div>
<div id="init_pn" class="section level2">
<h2 class="hasAnchor">
<a href="#init_pn" class="anchor"></a>Initialization of the Petri Net</h2>
<p>In order to setup the places and transition of a metapopulation model, we need to provide data describing the edges between nodes so that the network can be constructed. First, we make a very simple 2 node network. We will make a binary matrix <code>adj</code> that indices whether or not 2 nodes are connected by an edge; the absence of an edge means that movement between those two nodes is not possible. We use the <code>sparseMatrix</code> class from the <code>Matrix</code> package because, for larger network topologies, the sparse matrix representation may have significant benefits. It should be specified with the diagonal empty, because we do not consider “migration” to your current location.</p>
<p>Next, we can setup all of the places in the model, provided the entomological parameters above, the inheritance pattern, and the size of the network.</p>
<p>Then, we create all of the possible transitions between states, given the places in the model and the network adjacency matrix.</p>
<p>Finally, as not all transitions apply to all “places”, we create a summary of possible transitions to and from each “place”. This is handled by the <code><a href="../reference/spn_S.html">spn_S()</a></code> function.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># adjacency matrix</span>
<span class="co">#  specify where individuals can migrate</span>
<span class="va">adj</span> <span class="op">&lt;-</span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/sparseMatrix.html">sparseMatrix</a></span><span class="op">(</span>i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>,j <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span>
<span class="va">n</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">adj</span><span class="op">)</span>

<span class="co"># Places and transitions</span>
<span class="va">SPN_P</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spn_P_lifecycle_network.html">spn_P_lifecycle_network</a></span><span class="op">(</span>num_nodes <span class="op">=</span> <span class="va">n</span>, params <span class="op">=</span> <span class="va">theta</span>,cube <span class="op">=</span> <span class="va">cube</span><span class="op">)</span>
<span class="va">SPN_T</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spn_T_lifecycle_network.html">spn_T_lifecycle_network</a></span><span class="op">(</span>spn_P <span class="op">=</span> <span class="va">SPN_P</span>, params <span class="op">=</span> <span class="va">theta</span>,cube <span class="op">=</span> <span class="va">cube</span>,m_move <span class="op">=</span> <span class="va">adj</span><span class="op">)</span>

<span class="co"># Stoichiometry matrix</span>
<span class="va">S</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spn_S.html">spn_S</a></span><span class="op">(</span>spn_P <span class="op">=</span> <span class="va">SPN_P</span>, spn_T <span class="op">=</span> <span class="va">SPN_T</span><span class="op">)</span></code></pre></div>
</div>
<div id="equilibria_haz" class="section level2">
<h2 class="hasAnchor">
<a href="#equilibria_haz" class="anchor"></a>Equilibrium Conditions and Hazard Functions</h2>
<p>Now that the structural properties of the SPN model have been set up, we will calculate equilibrium population sizes, <code>M0</code>. We will use the same parameter values as in the vignette <a href="lifecycle-node.html">“MGDrivE2 One Node Lifecycle Dynamics”</a>. Note that when setting <code>M0</code> we only set the wild-type populations; the equilibrium calculations assume a wildtype population in equilibrium at <span class="math inline">\(t=0\)</span>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># now that we have a network size, set adult females in each node</span>
<span class="va">NF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fl">500</span>, times <span class="op">=</span> <span class="va">n</span><span class="op">)</span>

<span class="co"># calculate equilibrium and setup initial conditions</span>
<span class="co">#  outputs required parameters in the named list "params"</span>
<span class="co">#  outputs intial equilibrium for adv users, "init</span>
<span class="co">#  outputs properly filled initial markings, "M0"</span>
<span class="va">initialCons</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/equilibrium_lifeycle.html">equilibrium_lifeycle</a></span><span class="op">(</span>params <span class="op">=</span> <span class="va">theta</span>, NF <span class="op">=</span> <span class="va">NF</span>, phi <span class="op">=</span> <span class="fl">0.5</span>,
                                    log_dd <span class="op">=</span> <span class="cn">TRUE</span>, spn_P <span class="op">=</span> <span class="va">SPN_P</span>, cube <span class="op">=</span> <span class="va">cube</span><span class="op">)</span></code></pre></div>
<p>Before making the hazard functions and finishing the SPN, we need to calculate movement rates (because in <strong>MGDrivE2</strong>, the movement matrix is a matrix of per-capita movement <em>rates</em> rather than probabilities; matrix exponentiation would give corresponding probabilities over a time interval). We use the helper function <code><a href="../reference/calc_move_rate.html">calc_move_rate()</a></code>, which calculates total per-capita rate of movement out of a node based on mortality rate and the probability to leave that node before dying.</p>
<p>The vector <code>move_rates</code> is the total rate at which a mosquito leaves a node, and the <code>sparseMatrix</code> object <code>move_probs</code> is the conditional probability of where it goes, given it has chosen to leave; we make movement uniform (although it won’t matter in the 2-node case, leaving node 1 means they will always go to node 2). Finally, we attach these objects to the list of parameters <code>initialCons$params</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># calculate movement rates and movement probabilities</span>
<span class="va">gam</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calc_move_rate.html">calc_move_rate</a></span><span class="op">(</span>mu <span class="op">=</span> <span class="va">theta</span><span class="op">$</span><span class="va">muF</span>, P <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span>

<span class="va">move_rates</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">gam</span>, times <span class="op">=</span> <span class="va">n</span><span class="op">)</span>
<span class="va">move_probs</span> <span class="op">&lt;-</span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/sparseMatrix.html">sparseMatrix</a></span><span class="op">(</span>i <span class="op">=</span> <span class="op">{</span><span class="op">}</span>, j <span class="op">=</span> <span class="op">{</span><span class="op">}</span>,x <span class="op">=</span> <span class="fl">0L</span>,dims <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">adj</span><span class="op">)</span><span class="op">)</span>

<span class="co"># uniform movement probabilities</span>
<span class="va">rowprobs</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html">rowSums</a></span><span class="op">(</span><span class="va">adj</span><span class="op">)</span>
<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">move_probs</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
  <span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu">Matrix</span><span class="fu">::</span><span class="fu">which</span><span class="op">(</span><span class="va">adj</span><span class="op">[</span><span class="va">i</span>,<span class="op">]</span><span class="op">)</span>
  <span class="va">move_probs</span><span class="op">[</span><span class="va">i</span>,<span class="va">cols</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="va">rowprobs</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>,<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">cols</span><span class="op">)</span><span class="op">)</span>
<span class="op">}</span>

<span class="co"># put rates and probs into the parameter list</span>
<span class="va">initialCons</span><span class="op">$</span><span class="va">params</span><span class="op">$</span><span class="va">mosquito_move_rates</span> <span class="op">&lt;-</span> <span class="va">move_rates</span>
<span class="va">initialCons</span><span class="op">$</span><span class="va">params</span><span class="op">$</span><span class="va">mosquito_move_probs</span> <span class="op">&lt;-</span> <span class="va">move_probs</span></code></pre></div>
<p>Now that all the necessary parameters have been added to the named list <code>initialCons$params</code>, we can generate the hazard functions. By specifying <code>log_dd = TRUE</code>, we use logistic density dependence for these simulations.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># approximate hazards for continous approximation</span>
<span class="va">approx_hazards</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spn_hazards.html">spn_hazards</a></span><span class="op">(</span>spn_P <span class="op">=</span> <span class="va">SPN_P</span>, spn_T <span class="op">=</span> <span class="va">SPN_T</span>, cube <span class="op">=</span> <span class="va">cube</span>,
                              params <span class="op">=</span> <span class="va">initialCons</span><span class="op">$</span><span class="va">params</span>, type <span class="op">=</span> <span class="st">"life"</span>,
                              log_dd <span class="op">=</span> <span class="cn">TRUE</span>, exact <span class="op">=</span> <span class="cn">FALSE</span>, tol <span class="op">=</span> <span class="fl">1e-6</span>,
                              verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="co"># exact hazards for integer-valued state space</span>
<span class="va">exact_hazards</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spn_hazards.html">spn_hazards</a></span><span class="op">(</span>spn_P <span class="op">=</span> <span class="va">SPN_P</span>, spn_T <span class="op">=</span> <span class="va">SPN_T</span>, cube <span class="op">=</span> <span class="va">cube</span>,
                             params <span class="op">=</span> <span class="va">initialCons</span><span class="op">$</span><span class="va">params</span>, type <span class="op">=</span> <span class="st">"life"</span>,
                             log_dd <span class="op">=</span> <span class="cn">TRUE</span>, exact <span class="op">=</span> <span class="cn">TRUE</span>, tol <span class="op">=</span> <span class="cn">NaN</span>,
                             verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
</div>
<div id="sim" class="section level2">
<h2 class="hasAnchor">
<a href="#sim" class="anchor"></a>Simulation of Fully Specified SPN Model</h2>
<p>Now that the structural elements of the Petri Net have been built, and we have a vector of hazard functions, we can simulate dynamics on the network. First, we will make a <code>data.frame</code> to hold information about the releases; We will release 50 adult females with homozygous recessive alleles 5 times, every 10 days, starting at day 20, and only in node 1. It is critically important that <strong>the event names match a place name</strong> in the simulation. The simulation function checks this and will throw an error if the event name does not exist as a place in the simulation. This format is used in <strong>MGDrivE2</strong> for consistency with solvers in <code>deSolve</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># releases</span>
<span class="va">r_times</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span>from <span class="op">=</span> <span class="fl">20</span>, length.out <span class="op">=</span> <span class="fl">5</span>, by <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>
<span class="va">r_size</span> <span class="op">&lt;-</span> <span class="fl">50</span>
<span class="va">events</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="st">"var"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"F_"</span>, <span class="va">cube</span><span class="op">$</span><span class="va">releaseType</span>, <span class="st">"_"</span>, <span class="va">cube</span><span class="op">$</span><span class="va">wildType</span>, <span class="st">"_1"</span><span class="op">)</span>,
                     <span class="st">"time"</span> <span class="op">=</span> <span class="va">r_times</span>,
                     <span class="st">"value"</span> <span class="op">=</span> <span class="va">r_size</span>,
                     <span class="st">"method"</span> <span class="op">=</span> <span class="st">"add"</span>,
                     stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<div id="soln_ode" class="section level3">
<h3 class="hasAnchor">
<a href="#soln_ode" class="anchor"></a>Deterministic: ODE Solutions</h3>
<p>First we run a mean-field approximation of the stochastic model. Internally, <strong>MGDrivE2</strong> uses the high quality numerical solvers in <code>deSolve</code> to integrate a mean-field approximation to the stochastic model. We also plot the adult dynamics for both nodes so we can see how the releases spread through the network.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># run deterministic simulation</span>
<span class="va">ODE_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_trajectory_R.html">sim_trajectory_R</a></span><span class="op">(</span>x0 <span class="op">=</span> <span class="va">initialCons</span><span class="op">$</span><span class="va">M0</span>, tmax <span class="op">=</span> <span class="va">tmax</span>, dt <span class="op">=</span> <span class="va">dt</span>, S <span class="op">=</span> <span class="va">S</span>,
                            hazards <span class="op">=</span> <span class="va">approx_hazards</span>, sampler <span class="op">=</span> <span class="st">"ode"</span>, method <span class="op">=</span> <span class="st">"lsoda"</span>,
                            events <span class="op">=</span> <span class="va">events</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="co"># summarize females/males by genotype</span>
<span class="va">ODE_female</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarize_females.html">summarize_females</a></span><span class="op">(</span>out <span class="op">=</span> <span class="va">ODE_out</span><span class="op">$</span><span class="va">state</span>, spn_P <span class="op">=</span> <span class="va">SPN_P</span><span class="op">)</span>
<span class="va">ODE_male</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarize_males.html">summarize_males</a></span><span class="op">(</span>out <span class="op">=</span> <span class="va">ODE_out</span><span class="op">$</span><span class="va">state</span><span class="op">)</span>

<span class="co"># add sex for plotting</span>
<span class="va">ODE_female</span><span class="op">$</span><span class="va">sex</span> <span class="op">&lt;-</span> <span class="st">"Female"</span>
<span class="va">ODE_male</span><span class="op">$</span><span class="va">sex</span> <span class="op">&lt;-</span> <span class="st">"Male"</span>

<span class="co"># plot</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span><span class="va">ODE_female</span>, <span class="va">ODE_male</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">time</span>, y <span class="op">=</span> <span class="va">value</span>, color <span class="op">=</span> <span class="va">genotype</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html">facet_grid</a></span><span class="op">(</span><span class="va">node</span> <span class="op">~</span> <span class="va">sex</span>, scales <span class="op">=</span> <span class="st">"fixed"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"SPN: ODE solution"</span><span class="op">)</span></code></pre></div>
<p><img src="lifecycle-network_files/figure-html/unnamed-chunk-9-1.png" width="691.2"></p>
<p>From the plots, we see the initial equilibria in both patches is 500 individuals, the same for each sex, as specified above. We see the releases in node 1 females, and how they quickly spread into the population, becoming evenly spread between females and males. In node 2, we see a much slower introgression of <code>a</code> alleles, all as heterozygous <code>Aa</code>, due to the very low migration rate between nodes. If run long enough, we would eventually see all genotype frequencies reach their equilibrium values in both nodes.</p>
</div>
<div id="soln_tau" class="section level3">
<h3 class="hasAnchor">
<a href="#soln_tau" class="anchor"></a>Stochastic: Tau-leaping Solutions</h3>
<p>As a further example, we run a single stochastic realization of the same network, using a tau-leaping method with <span class="math inline">\(\Delta t = 0.1\)</span>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># delta t</span>
<span class="va">dt_stoch</span> <span class="op">&lt;-</span> <span class="fl">0.1</span>

<span class="co"># tau leaping simulation</span>
<span class="va">PTS_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_trajectory_R.html">sim_trajectory_R</a></span><span class="op">(</span>x0 <span class="op">=</span> <span class="va">initialCons</span><span class="op">$</span><span class="va">M0</span>, tmax <span class="op">=</span> <span class="va">tmax</span>, dt <span class="op">=</span> <span class="va">dt</span>,
                            dt_stoch <span class="op">=</span> <span class="va">dt_stoch</span>, S <span class="op">=</span> <span class="va">S</span>, hazards <span class="op">=</span> <span class="va">exact_hazards</span>,
                            sampler <span class="op">=</span> <span class="st">"tau"</span>, events <span class="op">=</span> <span class="va">events</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="co"># summarize females/males by genotype</span>
<span class="va">PTS_female</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarize_females.html">summarize_females</a></span><span class="op">(</span>out <span class="op">=</span> <span class="va">PTS_out</span><span class="op">$</span><span class="va">state</span>, spn_P <span class="op">=</span> <span class="va">SPN_P</span><span class="op">)</span>
<span class="va">PTS_male</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarize_males.html">summarize_males</a></span><span class="op">(</span>out <span class="op">=</span> <span class="va">PTS_out</span><span class="op">$</span><span class="va">state</span><span class="op">)</span>

<span class="co"># add sex for plotting</span>
<span class="va">PTS_female</span><span class="op">$</span><span class="va">sex</span> <span class="op">&lt;-</span> <span class="st">"Female"</span>
<span class="va">PTS_male</span><span class="op">$</span><span class="va">sex</span> <span class="op">&lt;-</span> <span class="st">"Male"</span>

<span class="co"># plot</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span><span class="va">PTS_female</span>, <span class="va">PTS_male</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_path.html">geom_line</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">time</span>, y <span class="op">=</span> <span class="va">value</span>, color <span class="op">=</span> <span class="va">genotype</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html">facet_grid</a></span><span class="op">(</span><span class="va">node</span> <span class="op">~</span> <span class="va">sex</span>, scales <span class="op">=</span> <span class="st">"fixed"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"SPN: Tau-leaping Approximation"</span><span class="op">)</span></code></pre></div>
<p><img src="lifecycle-network_files/figure-html/unnamed-chunk-10-1.png" width="691.2"></p>
<p>We see a heuristically similar figure as above. This is only 1 stochastic realization, but it closely follows the dynamics from the ODE solver, indicating that a time-step of 0.1 provides and accurate simulation at this level.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="http://www.slwu89.com">Sean L. Wu</a>, <a href="https://github.com/GilChrist19">Jared B. Bennett</a>, <a href="https://www.marshalllab.com">John M. Marshall</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
