<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Inheritance Cube: 2-Locus ClvR (Cleave and Rescue) — cubeClvR2 • MGDrivE</title><!-- mathjax math --><script src="https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.js" integrity="sha256-MASABpB4tYktI2Oitl4t+78w/lyA+D7b/s9GEP0JOGI=" crossorigin="anonymous"></script><script>
  window.MathJax = {
    chtml: {
      fontURL: "https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2"
    }
  };
</script><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.1/jquery-3.6.1.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Inheritance Cube: 2-Locus ClvR (Cleave and Rescue) — cubeClvR2"><meta name="description" content="Based on the Cleave-and-Rescue system of Oberhofer,
this is a 3-locus Cas9-based toxin-antidote system. The first locus carries the
Cas9, the second locus carries the gRNAs, and a recoded copy of an essential gene.
The third locus is the targeted essential gene. This gene can be completely
haplosufficient (hSuf = 1) or completely haploinsufficient (hSuf = 0).
It is assumed that having 2 copies of the gene (be it wild-type at the second
locus or recoded at the first) confers complete viability. Additionally, loci
1 and 2 can be linked, given crM and crF, imitating the original
2-locus ClvR system.
For this construct, the first locus will have 2 alleles, the second will have 2
alleles, and the third will have 3 alleles:
Locus 1
W: Wild-type
C: Cas9

Locus 2
W: Wild-type
G: gRNAs and recoded essential gene

Locus 3
W: Wild-type
R: Functional resistant
B: Non-functional resistant


"><meta property="og:description" content="Based on the Cleave-and-Rescue system of Oberhofer,
this is a 3-locus Cas9-based toxin-antidote system. The first locus carries the
Cas9, the second locus carries the gRNAs, and a recoded copy of an essential gene.
The third locus is the targeted essential gene. This gene can be completely
haplosufficient (hSuf = 1) or completely haploinsufficient (hSuf = 0).
It is assumed that having 2 copies of the gene (be it wild-type at the second
locus or recoded at the first) confers complete viability. Additionally, loci
1 and 2 can be linked, given crM and crF, imitating the original
2-locus ClvR system.
For this construct, the first locus will have 2 alleles, the second will have 2
alleles, and the third will have 3 alleles:
Locus 1
W: Wild-type
C: Cas9

Locus 2
W: Wild-type
G: gRNAs and recoded essential gene

Locus 3
W: Wild-type
R: Functional resistant
B: Non-functional resistant


"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MGDrivE</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.6.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes"><li><a class="dropdown-item" href="../articles/mgdrive_examples.html">Example Simulations</a></li>
    <li><a class="dropdown-item" href="../articles/mgdrive_math.html">Mathematical Description</a></li>
    <li><a class="dropdown-item" href="../articles/mgdrive_run.html">Detailed Run Example</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/MarshallLab/MGDrivE" aria-label="View on GitHub"><span class="fa fa-lg fa-github"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Inheritance Cube: 2-Locus ClvR (Cleave and Rescue)</h1>
      <small class="dont-index">Source: <a href="https://github.com/MarshallLab/MGDrivE/blob/HEAD/R/Cube-ClvR2.R" class="external-link"><code>R/Cube-ClvR2.R</code></a></small>
      <div class="d-none name"><code>cubeClvR2.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Based on the Cleave-and-Rescue system of <a href="https://doi.org/10.1101/2020.07.09.196253" class="external-link">Oberhofer</a>,
this is a 3-locus Cas9-based toxin-antidote system. The first locus carries the
Cas9, the second locus carries the gRNAs, and a recoded copy of an essential gene.
The third locus is the targeted essential gene. This gene can be completely
haplosufficient (<code>hSuf</code> = 1) or completely haploinsufficient (<code>hSuf</code> = 0).
It is assumed that having 2 copies of the gene (be it wild-type at the second
locus or recoded at the first) confers complete viability. Additionally, loci
1 and 2 can be linked, given <code>crM</code> and <code>crF</code>, imitating the original
2-locus ClvR system.
For this construct, the first locus will have 2 alleles, the second will have 2
alleles, and the third will have 3 alleles:</p><ul><li><p>Locus 1</p><ul><li><p>W: Wild-type</p></li>
<li><p>C: Cas9</p></li>
</ul></li>
<li><p>Locus 2</p><ul><li><p>W: Wild-type</p></li>
<li><p>G: gRNAs and recoded essential gene</p></li>
</ul></li>
<li><p>Locus 3</p><ul><li><p>W: Wild-type</p></li>
<li><p>R: Functional resistant</p></li>
<li><p>B: Non-functional resistant</p></li>
</ul></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">cubeClvR2</span><span class="op">(</span></span>
<span>  cF <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  crF <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  ccF <span class="op">=</span> <span class="va">cF</span>,</span>
<span>  ccrF <span class="op">=</span> <span class="va">crF</span>,</span>
<span>  cM <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  crM <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  ccM <span class="op">=</span> <span class="va">cM</span>,</span>
<span>  ccrM <span class="op">=</span> <span class="va">crM</span>,</span>
<span>  dW <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  drW <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  ddW <span class="op">=</span> <span class="va">dW</span>,</span>
<span>  ddrW <span class="op">=</span> <span class="va">drW</span>,</span>
<span>  hSuf <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  crossF <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  crossM <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  eta <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  phi <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  omega <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  xiF <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  xiM <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  s <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-cf">cF<a class="anchor" aria-label="anchor" href="#arg-cf"></a></dt>
<dd><p>Female cutting rate, one ClvR allele</p></dd>


<dt id="arg-crf">crF<a class="anchor" aria-label="anchor" href="#arg-crf"></a></dt>
<dd><p>Female functional resistance rate, one ClvR allele</p></dd>


<dt id="arg-ccf">ccF<a class="anchor" aria-label="anchor" href="#arg-ccf"></a></dt>
<dd><p>Female cutting rate, two ClvR alleles</p></dd>


<dt id="arg-ccrf">ccrF<a class="anchor" aria-label="anchor" href="#arg-ccrf"></a></dt>
<dd><p>Female functional resistance rate, two ClvR alleles</p></dd>


<dt id="arg-cm">cM<a class="anchor" aria-label="anchor" href="#arg-cm"></a></dt>
<dd><p>Male cutting rate, one ClvR allele</p></dd>


<dt id="arg-crm">crM<a class="anchor" aria-label="anchor" href="#arg-crm"></a></dt>
<dd><p>Male functional resistance rate, one ClvR allele</p></dd>


<dt id="arg-ccm">ccM<a class="anchor" aria-label="anchor" href="#arg-ccm"></a></dt>
<dd><p>Male cutting rate, two ClvR alleles</p></dd>


<dt id="arg-ccrm">ccrM<a class="anchor" aria-label="anchor" href="#arg-ccrm"></a></dt>
<dd><p>Male functional resistance rate, two ClvR alleles</p></dd>


<dt id="arg-dw">dW<a class="anchor" aria-label="anchor" href="#arg-dw"></a></dt>
<dd><p>Female deposition cutting rate</p></dd>


<dt id="arg-drw">drW<a class="anchor" aria-label="anchor" href="#arg-drw"></a></dt>
<dd><p>Female deposition functional resistance rate</p></dd>


<dt id="arg-ddw">ddW<a class="anchor" aria-label="anchor" href="#arg-ddw"></a></dt>
<dd><p>Female deposition (HH) cutting rate</p></dd>


<dt id="arg-ddrw">ddrW<a class="anchor" aria-label="anchor" href="#arg-ddrw"></a></dt>
<dd><p>Female deposition (HH) functional resistance rate</p></dd>


<dt id="arg-hsuf">hSuf<a class="anchor" aria-label="anchor" href="#arg-hsuf"></a></dt>
<dd><p>Haplosufficiency level, default is completely sufficient</p></dd>


<dt id="arg-crossf">crossF<a class="anchor" aria-label="anchor" href="#arg-crossf"></a></dt>
<dd><p>Crossover rate in females, 0 is completely linked, 0.5 is unlinked, 1.0 is complete divergence</p></dd>


<dt id="arg-crossm">crossM<a class="anchor" aria-label="anchor" href="#arg-crossm"></a></dt>
<dd><p>Crossover rate in males, 0 is completely linked, 0.5 is unlinked, 1.0 is complete divergence</p></dd>


<dt id="arg-eta">eta<a class="anchor" aria-label="anchor" href="#arg-eta"></a></dt>
<dd><p>Genotype-specific mating fitness</p></dd>


<dt id="arg-phi">phi<a class="anchor" aria-label="anchor" href="#arg-phi"></a></dt>
<dd><p>Genotype-specific sex ratio at emergence</p></dd>


<dt id="arg-omega">omega<a class="anchor" aria-label="anchor" href="#arg-omega"></a></dt>
<dd><p>Genotype-specific multiplicative modifier of adult mortality</p></dd>


<dt id="arg-xif">xiF<a class="anchor" aria-label="anchor" href="#arg-xif"></a></dt>
<dd><p>Genotype-specific female pupatory success</p></dd>


<dt id="arg-xim">xiM<a class="anchor" aria-label="anchor" href="#arg-xim"></a></dt>
<dd><p>Genotype-specific male pupatory success</p></dd>


<dt id="arg-s">s<a class="anchor" aria-label="anchor" href="#arg-s"></a></dt>
<dd><p>Genotype-specific fractional reduction(increase) in fertility</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Named list containing the inheritance cube, transition matrix, genotypes, wild-type allele,
and all genotype-specific parameters.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Female deposition is implemented incorrectly. Right now, it is performed on
male alleles prior to zygote formation - it should happen post-zygote formation.
Since this construct doesn't have HDR, this should be fine. <br>
Additionally, it is assumed that deposition requries loaded Cas9-RNP complexes
from the mother, having Cas9 and no maternal gRNA, even in the presence of
paternal gRNA, will not result in maternal deposition mediated cleavage.</p>
<p>Copy-number dependent rates are based on Cas9, not gRNA. The assumption is that
RNA is easier to produce, and therefore won't limit cleavage by Cas9.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://chipdelmal.github.io" class="external-link">Héctor Manuel Sánchez Castellanos</a>, <a href="https://github.com/GilChrist19" class="external-link">Jared Bennett</a>, <a href="https://slwu89.github.io" class="external-link">Sean Wu</a>, <a href="https://www.marshalllab.com" class="external-link">John M. Marshall</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

