<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>MGDrivE Mathematical Description • MGDrivE</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="MGDrivE Mathematical Description">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MGDrivE</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.5.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/mgdrive_examples.html">Example Simulations</a>
    </li>
    <li>
      <a href="../articles/mgdrive_math.html">Mathematical Description</a>
    </li>
    <li>
      <a href="../articles/mgdrive_run.html">Detailed Run Example</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/MarshallLab/MGDrivE">
    <span class="fa fa-lg fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>MGDrivE Mathematical Description</h1>
            
      
      
      <div class="hidden name"><code>mgdrive_math.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>Recent developments of CRISPR-Cas9 based homing endonuclease gene drive systems, either for the suppression or replacement of mosquito populations, have generated much interest in their use for control of mosquito-borne diseases (such as dengue, malaria, Chikungunya and Zika). This is because genetic control of pathogen transmission may complement or even substitute traditional vector-control interventions, which have had limited success in bringing the spread of these diseases to a halt.</p>
<p>Despite excitement for the use of gene drives for mosquito control, current modeling efforts have analyzed only a handful of these new approaches (usually studying just one per framework). Moreover, these models usually consider well-mixed populations with no explicit spatial dynamics. To this end, we are developing <strong>MGDrivE</strong> (Mosquito Gene DRIVe Explorer), in cooperation with the ‘UCI Malaria Elimination Initiative’, as a flexible modeling framework to evaluate a variety of drive systems in spatial networks of mosquito populations. This framework provides a reliable testbed to evaluate and optimize the efficacy of gene drive mosquito releases. What separates <strong>MGDrivE</strong> from other models is the incorporation of mathematical and computational mechanisms to simulate a wide array of inheritance-based technologies within the same, coherent set of equations.</p>
<p>We do this by treating the population dynamics, genetic inheritance operations, and migration between habitats as separate processes coupled together through the use of mathematical tensor operations. This way we can conveniently swap inheritance patterns whilst still making use of the same set of population dynamics equations. This is a crucial advantage of our system, as it allows other research groups to test their ideas without developing new models and without the need to spend time adapting other frameworks to suit their needs.</p>
<div id="brief-description" class="section level2">
<h2 class="hasAnchor">
<a href="#brief-description" class="anchor"></a>Brief Description</h2>
<p><strong>MGDrivE</strong> is based on the idea that we can decouple the genotype inheritance process from the population dynamics equations. This allows the system to be treated and developed in three semi-independent modules that come together to form the system.</p>
<div class="figure" style="text-align: center">
<img src="diagram.png" alt="Fig. 1: **MGDrivE** Conceptual Model Framework" width="80%"><p class="caption">
Fig. 1: <strong>MGDrivE</strong> Conceptual Model Framework
</p>
</div>
<div id="previous-work" class="section level3">
<h3 class="hasAnchor">
<a href="#previous-work" class="anchor"></a>Previous Work:</h3>
<p>The original version of this model was based on work by Deredec et al. (2011) and Hancock &amp; Godfray (2007), and adapted to accommodate CRISPR homing dynamics in a previous publication by our team (Marshall et al. 2017). As described, we extended this framework to be able to handle a variable number of genotypes, and migration across spatial scenarios. We did this by adapting the equations to work in a tensor-oriented manner, where each genotype can have different processes affecting their particular strain (death rates, mating fitness, sex-ratio bias, et cetera).</p>
</div>
<div id="notation-and-conventions" class="section level3">
<h3 class="hasAnchor">
<a href="#notation-and-conventions" class="anchor"></a>Notation and Conventions</h3>
<p>Before beginning the full description of the model we will define some of the conventions we followed for the notation of the written description of the system.</p>
<ul>
<li>Overlines are used to denote the dimension of a tensor</li>
<li>Subscript brackets are used to indicate an element in time. For example: <span class="math inline">\(L_{[t-1]}\)</span> is the larval population at time: <span class="math inline">\(t-1\)</span>.</li>
<li>Parentheses are used to indicate the parameter(s) of a function. For example: <span class="math inline">\(\overline{O(T_{e}+T_{l})}\)</span> represents the function <span class="math inline">\(O\)</span> evaluated with the parameter: <span class="math inline">\(T_{e}+T_{l}\)</span>
</li>
<li>Matrices follow a ‘row-first’ indexing order (<span class="math inline">\(i\)</span>: row, <span class="math inline">\(j\)</span>: column)</li>
</ul>
<p>In the case of one dimensional tensors, each slot represents a genotype of the population. For example, the male population is stored in the following way: <span class="math display">\[ \overline{Am} = \left(\begin{array}{c} g_1 \\ g_2 \\ g_3 \\ \vdots \\ g_n \end{array}\right) _{i} \]</span></p>
<p>All the processes that affect mosquitoes in a genotype-specific way are defined and stored in this way within the framework.</p>
<p>There are two tensors of squared dimensionality in the model: the adult females matrix, and the genotype-specific male-mating ability, <span class="math inline">\(\overline{\overline{\eta}}\)</span>. In the case of the former, the rows represent the females’ genotype, whilst the columns represent the genotype of the male they mated with: <span class="math display">\[ \overline{\overline{Af}} = \left(\begin{array}{ccccc} g_{11} &amp; g_{12} &amp; g_{13} &amp; \cdots &amp; g_{1n}\\ g_{21} &amp; g_{22} &amp; g_{23} &amp; \cdots &amp; g_{2n}\\ g_{31} &amp; g_{32} &amp; g_{33} &amp; \cdots &amp; g_{3n}\\ \vdots &amp; \vdots &amp; \vdots &amp; \ddots &amp; \vdots\\ g_{n1} &amp; g_{n2} &amp; g_{n3} &amp; \cdots &amp; g_{nn} \end{array}\right) _{ij} \]</span></p>
<p>The genotype-specific male-mating ability, on the other hand, stores the mothers’ genotype in the rows, the male mate genotype as columns, and each value indicates the ability of a specific male genotype to mate with a specific female genotype. This allows assortative mating to be modeled.</p>
</div>
</div>
</div>
<div id="mathematical-framework" class="section level1">
<h1 class="hasAnchor">
<a href="#mathematical-framework" class="anchor"></a>Mathematical Framework</h1>
<div id="inheritance-cube" class="section level2">
<h2 class="hasAnchor">
<a href="#inheritance-cube" class="anchor"></a>Inheritance Cube</h2>
<p>To model an arbitrary number of genotypes efficiently in the same mathematical framework we use a 3-dimensional array structure (cube) where each axis represents the following information:</p>
<ul>
<li>x: female adult mate genotype</li>
<li>y: male adult mate genotype</li>
<li>z: proportion of the offspring that inherits a given genotype (slice)</li>
</ul>
<p>The cube structure gives us the flexibility to apply tensor operations to the elements within our equations, so that we can calculate the stratified population dynamics rapidly; and within a readable, flexible computational framework. This becomes apparent when we define the equation we use for the computation of eggs laid at any given point in time:</p>
<p><span class="math display">\[ \overline{O(T_x)} = \sum_{j=1}^{n} \Bigg( \bigg( (\beta*\overline{s} * \overline{ \overline{Af_{[t-T_x]}}}) * \overline{\overline{\overline{Ih}}} \bigg) * \Lambda  \Bigg)^{\top}_{ij} \]</span></p>
<p>In this equation, the matrix containing the number of mated adult females <span class="math inline">\((\overline{\overline{Af}})\)</span> is multiplied element-wise with each one of the layers containing the eggs genotypes proportions expected from this cross <span class="math inline">\((\overline{\overline{\overline{Ih}}})\)</span>.</p>
<p>The resulting matrix is then multiplied by a binary ‘viability mask’ <span class="math inline">\((\Lambda)\)</span> that filters out female-parent to offspring genetic combinations that are not viable due to biological impediments (such as cytoplasmic incompatibility).</p>
<p>The summation of the transposed resulting matrix returns us the total fraction of eggs resulting from all the male to female genotype crosses (<span class="math inline">\(\overline{O(T_x)}\)</span>).</p>
<p>Note: For inheritance operations to be consistent within the framework the summation of each element in the z-axis (this is, the proportions of each one of the offspring’s genotypes) must be equal to one.</p>
<div id="drive-specific-cubes" class="section level3">
<h3 class="hasAnchor">
<a href="#drive-specific-cubes" class="anchor"></a>Drive-specific Cubes</h3>
<p>An inheritance cube in an array object that specifies inheritance probabilities (offspring genotype probability) stratified by male and female parent genotypes. <strong>MGDrivE</strong> provides the following cubes to model different gene drive systems:</p>
<ul>
<li><a href="../reference/cubeOneLocusTA.html">1 Locus Maternal-Toxin/Zygotic-Antidote System</a></li>
<li><a href="../reference/cubeTwoLocusTA.html">2 Locus Maternal-Toxin/Zygotic-Antidote System</a></li>
<li><a href="../reference/cubeHoming1RA.html">Homing Drive with 1 Resistance Allele</a></li>
<li><a href="../reference/cubeHomingDrive.html">CRISPR (Clustered Regularly Interspaced Short Palindromic Repeats) with 2 Resistance Alleles and maternal deposition</a></li>
<li><a href="../reference/cubeKillerRescue.html">Killer-Rescue System</a></li>
<li><a href="../reference/cubeMEDEA.html">MEDEA (Maternal Effect Dominant Embryonic Arrest)</a></li>
<li><a href="../reference/cubeReciprocalTranslocations.html">Reciprocal Translocations</a></li>
<li><a href="../reference/cubeRIDL.html">RIDL (Release of Insects with Dominant Lethality)</a></li>
<li><a href="../reference/cubeMendelian.html">Mendelian</a></li>
<li><a href="../reference/cubeSplitDrive.html">Split-Drive</a></li>
<li><a href="../reference/cubeTGD.html">trans-Vomplementing Gene Drive</a></li>
<li><a href="../reference/cubeTGDX.html">X-Linked trans-Complementing Gene Drive</a></li>
<li><a href="../reference/cubeWolbachia.html">Wolbachia</a></li>
</ul>
</div>
</div>
<div id="population-dynamics" class="section level2">
<h2 class="hasAnchor">
<a href="#population-dynamics" class="anchor"></a>Population Dynamics</h2>
<p>During the three aquatic stages, a density-independent mortality process takes place at each stage (<em>st</em>): <span class="math display">\[ \theta_{st}=(1-\mu_{st})^{T_{st}} \]</span> Along with a density dependent process dependent on the number of larvae in the environment: <span class="math display">\[ F(L[t])=\Bigg(\frac{\alpha}{\alpha+\sum{\overline{L[t]}}}\Bigg)^{1/T_l} \]</span> where <span class="math inline">\(\alpha\)</span> represents the strength of the density-dependent process. This parameter is calculated with: <span class="math display">\[ \alpha=\Bigg( \frac{1/2 * \beta * \theta_e * Ad_{eq}}{R_m-1} \Bigg) * \Bigg( \frac{1-(\theta_l / R_m)}{1-(\theta_l / R_m)^{1/T_l}} \Bigg) \]</span> in which <span class="math inline">\(\beta\)</span> is the species’ fertility in the absence of gene-drive, <span class="math inline">\(Ad_{eq}\)</span> is the adult mosquito population equilibrium size, and <span class="math inline">\(R_{m}\)</span> is the population growth in the absence of density-dependent mortality. This population growth is calculated with the average generation time (<span class="math inline">\(g\)</span>), the adult mortality rate (<span class="math inline">\(\mu_{ad}\)</span>), and the daily population growth rate (<span class="math inline">\(r_{m}\)</span>): <span class="math display">\[ g=T_{e}+T_{l}+T_{p}+\frac{1}{\mu_{ad}}\\R_{m}=(r_{m})^{g} \]</span></p>
<div id="larval-stages" class="section level3">
<h3 class="hasAnchor">
<a href="#larval-stages" class="anchor"></a>Larval Stages</h3>
<p>The computation of the larval stage in the population is crucial to the model because the density dependent processes necessary for equilibrium trajectories occur here. This calculation is performed with the following equation: <span class="math display">\[
D(\theta_l,T_x) = \left\{
         \begin{array}{ll}
             \theta_{l[0]}^{'}=\theta_l                                 &amp; \quad i = 0 \\
             \theta_{l[i+1]}^{'} = \theta_{l[i]}^{'} *F(\overline{L_{[t-i-T_x]}})   &amp; \quad i \leq T_l
         \end{array}
     \right.
\]</span></p>
<p>In addition to this, we need the larval mortality (<span class="math inline">\(\mu_{l}\)</span>): <span class="math display">\[
    %L_{eq}=&amp;\alpha*\lfloor R_{m} -1\rfloor
    %&amp;
    \mu_{l}=1-\Bigg( \frac{R_{m} * \mu_{ad}}{1/2 * \beta * (1-\mu_{m})} \Bigg)^{\frac{1}{T_{e}+T_{l}+T_{p}}}
\]</span></p>
<p>With these mortality processes, we are now able to calculate the larval population: <span class="math display">\[
    \overline{L_{[t]}}=
        \overline{L_{[t-1]}} * (1-\mu_{l}) * F(\overline{L_{[t-1]})}\\
        +\overline{O(T_{e})}* \theta_{e} \\
        %+\overline{\beta}* \theta_{e} * (\overline{\overline{Af_{(t-T_{e})}}} \circ \overline{\overline{\overline{Ih}}})\\
        - \overline{O(T_{e}+T_{l})} * \theta_{e} * D(\theta_{l},0)
        %\prod_{i=1}^{T_{l}} F(\overline{L_{[t-i]}})
        %\theta_{l}
\]</span> where the first term accounts for larvae surviving one day to the other; the second therm accounts for the eggs that have hatched within the same period of time; and the last term computes the number of larvae that have transformed into pupae.</p>
</div>
<div id="adult-stages" class="section level3">
<h3 class="hasAnchor">
<a href="#adult-stages" class="anchor"></a>Adult Stages</h3>
<p>We are ultimately interested in calculating how many adults of each genotype exist at any given point in time. For this, we first calculate the number of eggs that are laid and survive to the adult stages with the equation: <span class="math display">\[
    \overline{E^{'}}= \overline{O(T_{e}+T_{l}+T_{p})} \\
    * \bigg(\overline{\xi_{m}} * (\theta_{e} * \theta_{p}) * (1-\mu_{ad}) * D(\theta_{l},T_{p}) \bigg)
\]</span></p>
<p>With this information we can calculate the current number of male adults in the population by computing the following equation: <span class="math display">\[
    \overline{Am_{[t]}}=
        \overline{Am_{[t-1]}} * (1-\mu_{ad})*\overline{\omega_{m}}\\
        + (1-\overline{\phi}) *  \overline{E^{'}}\\
        + \overline{\nu m_{[t-1]}}
\]</span> in which the first term represents the number of males surviving from one day to the next; the second term, the fraction of males that survive to adulthood (<span class="math inline">\(\overline{E'}\)</span>) and emerge as males (<span class="math inline">\(1-\phi\)</span>); the last term is used to add males into the population as part of gene-drive release campaigns.</p>
<p>Female adult populations are calculated in a similar way: <span class="math display">\[
  \overline{\overline{Af_{[t]}}}=
    \overline{\overline{Af_{[t-1]}}} * (1-\mu_{ad}) * \overline{\omega_{f}}\\
    +  \bigg( \overline{\phi} * \overline{E^{'}}+\overline{\nu f_{[t-1]}}\bigg)^{\top} * \bigg( \frac{\overline{\overline{\eta}}*\overline{Am_{[t-1]}}}{\sum{\overline{Am_{[t-1]}}}} \bigg)%\overline{\overline{Mf}}
\]</span> where we first compute the surviving female adults from one day to the next; and then we calculate the mating composition of the female fraction emerging from pupa stage. To do this, we obtain the surviving fraction of eggs that survive to adulthood (<span class="math inline">\(\overline{E'}\)</span>) and emerge as females (<span class="math inline">\(\phi\)</span>), we then add the new females added as a result of gene-drive releases (<span class="math inline">\(\overline{\nu f_{[t-1]}}\)</span>). After doing this, we calculate the proportion of males that are allocated to each female genotype, taking into account their respective mating fitnesses (<span class="math inline">\(\overline{\overline{\eta}}\)</span>) so that we can introduce the new adult females into the population pool.</p>
</div>
</div>
<div id="gene-drive-releases-and-effects" class="section level2">
<h2 class="hasAnchor">
<a href="#gene-drive-releases-and-effects" class="anchor"></a>Gene Drive Releases and Effects</h2>
<p>As was briefly mentioned before, we are including the option to release both male and/or female individuals into the populations. Another important thing to emphasize is that we allow flexible releases sizes and schedules. Our model handles releases internally as lists of population compositions, so it is possible to have releases performed at irregular intervals and with different numbers of mosquito genetic compositions as long as no new genotypes are introduced (which have not been previously defined in the inheritance cube). <span class="math display">\[
  \overline{\nu} = \bigg\{
    \left(\begin{array}{c} g_1 \\ g_2 \\ g_3 \\ \vdots \\ g_n \end{array}\right)_{t=1} ,
    \left(\begin{array}{c} g_1 \\ g_2 \\ g_3 \\ \vdots \\ g_n \end{array}\right)_{t=2} ,
    \cdots ,
    \left(\begin{array}{c} g_1 \\ g_2 \\ g_3 \\ \vdots \\ g_n \end{array}\right)_{t=x}
    \bigg\}
\]</span></p>
<p>So far, however, we have not described the way in which the effects of these gene-drives are included into the mosquito populations dynamics. This is done through the use of various modifiers included in the equations:</p>
<ul>
<li>
<span class="math inline">\(\overline{\omega}\)</span>: Relative increase in mortality (zero being full mortality effects and one no mortality effect)</li>
<li>
<span class="math inline">\(\overline{\phi}\)</span>: Relative shift in the sex of the pupating mosquitoes (zero biases the sex ratio towards males, whilst 1 biases the ratio towards females).</li>
<li>
<span class="math inline">\(\overline{\overline{\eta}}\)</span>: Standardized mating fitness (zero being complete fitness ineptitude, and one being regular mating skills).</li>
<li>
<span class="math inline">\(\overline{\beta}\)</span>: Fecundity (average number of eggs laid per day).</li>
<li>
<span class="math inline">\(\overline{\xi}\)</span>: Pupation success (zero being full mortality and one full pupation success).</li>
</ul>
</div>
<div id="migration" class="section level2">
<h2 class="hasAnchor">
<a href="#migration" class="anchor"></a>Migration</h2>
<p>To simulate migration within our framework we are considering patches (or nodes) of fully-mixed populations in a network structure. This allows us to handle mosquito movement across spatially-distributed populations with a transitions matrix, which is calculated with the tensor outer product of the genotypes population tensors and the transitions matrix of the network as follows: <span class="math display">\[
    \overline{Am_{(t)}^{i}}=
        \sum{\overline{A_{m}^j} \otimes \overline{\overline{\tau m_{[t-1]}}}} \\
        \overline{\overline{Af_{(t)}^{i}}}=
      \sum{\overline{\overline{A_{f}^j}} \otimes \overline{\overline{\tau f_{[t-1]}}}}
\]</span></p>
<p>In these equations the new population of the patch <span class="math inline">\(i\)</span> is calculated by summing the migrating mosquitoes of all the <span class="math inline">\(j\)</span> patches across the network defined by the transitions matrix <span class="math inline">\(\tau\)</span>, which stores the mosquito migration probabilities from patch to patch. It is worth noting that the migration probabilities matrices can be different for males and females; and that there’s no inherent need for them to be static (the migration probabilities may vary over time to accommodate wind changes due to seasonality).</p>
</div>
</div>
<div id="stochasticity" class="section level1">
<h1 class="hasAnchor">
<a href="#stochasticity" class="anchor"></a>Stochasticity</h1>
<p><strong>MGDrivE</strong> allows all inheritance, migration, and population dynamics processes to be simulated stochastically; this accounts for the inherent probabilistic nature of the processes governing the interactions and life-cycles of organisms. In the next section, we will describe all the stochastic processes that can be activated in the program. It should be noted that all of these can be turned on and off independently from one another as required by the researcher.</p>
<div id="mosquito-biology" class="section level3">
<h3 class="hasAnchor">
<a href="#mosquito-biology" class="anchor"></a>Mosquito Biology</h3>
<p><strong>Oviposition</strong></p>
<p>Stochastic egg laying by female/male pairs is separated into two steps: calculating the number of eggs laid by the females and then distributing laid eggs according to their genotypes. The number of eggs laid follows a Poisson distribution conditioned on the number of female/male pairs and the fertility of each female.</p>
<p><span class="math display">\[ Poisson( \lambda = numFemales*Fertility) \]</span></p>
<p>Multinomial sampling, conditioned on the number of offspring and the relative viability of each genotype, determines the genotypes of the offspring.</p>
<p><span class="math display">\[ Multinomial \left(numOffspring, p_1, p_2\dots p_b \right)=\frac{numOffspring!}{p_1!\,p_2\,\dots p_n}p_1^{n_1}p_2^{n_2}\dots p_n^{n_n} \]</span></p>
<p><strong>Sex Determination</strong></p>
<p>Sex of the offspring is determined by multinomial sampling. This is conditioned on the number of eggs that live to hatching and a probability of being female, allowing the user to design systems that skew the sex ratio of the offspring through reproductive mechanisms.</p>
<p><span class="math display">\[ Multinomial(numHatchingEggs, p_{female}, p_{female}) \]</span></p>
<p><strong>Mating</strong></p>
<p>Stochastic mating is determined by multinomial sampling conditioned on the number of males and their fitness. It is assumed that females mate only once in their life, therefore each female will sample from the available males and be done, while the males are free to potentially mate with multiple females. The males’ ability to mate is modulated with a fitness term, thereby allowing some genotypes to be less fit than others (as seen often with lab releases).</p>
<p><span class="math display">\[ Multinomial(numFemales, p_1f_1, p_2f_2, \dots p_nf_n) \]</span></p>
<p><strong>Other Stochastic Processes</strong></p>
<p>All remaining stochastic processes (larval survival, hatching ,pupating, surviving to adult hood) are determined by binomial sampling conditioned on factors affecting the current life stage. These factors are determined empirically from mosquito population data.</p>
<p><strong>Migration</strong></p>
<p>Variance of stochastic movement (not used in diffusion model of migration). It affects the concentration of probability in the Dirchlet simplex, small values lead to high variance and large values lead to low variance.</p>
</div>
</div>
<div id="references" class="section level1">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<ul>
<li>Deredec A, Godfray HCJ, Burt A (2011). “Requirements for effective malaria control with homing endonuclease genes.” <em>Proceedings of the National Academy of Sciences of the United States of America</em>, 108(43), E874–80. ISSN 1091-6490, doi: 10.1073/pnas.1110717108 , <a href="https://www.pnas.org/content/108/43/E874" class="uri">https://www.pnas.org/content/108/43/E874</a>.</li>
<li>Hancock PA, Godfray HCJ (2007). “Application of the lumped age-class technique to studying the dynamics of malaria-mosquito-human interactions.” <em>Malaria Journal</em>, 6, 98. ISSN 1475-2875, doi: 10.1186/1475-2875-6-98 , <a href="https://malariajournal.biomedcentral.com/articles/10.1186/1475-2875-6-98" class="uri">https://malariajournal.biomedcentral.com/articles/10.1186/1475-2875-6-98</a>.</li>
<li>Marshall J, Buchman A, C. HMS, Akbari OS (2017). “Overcoming evolved resistance to population-suppressing homing-based gene drives.” <em>Nature Scientific Reports</em>, 1–46. ISSN 2045-2322, doi: <a href="https://doi.org/10.1101/088427" class="uri">https://doi.org/10.1101/088427</a> , <a href="https://www.nature.com/articles/s41598-017-02744-7" class="uri">https://www.nature.com/articles/s41598-017-02744-7</a>
</li>
</ul>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#introduction">Introduction</a><ul class="nav nav-pills nav-stacked">
<li><a href="#brief-description">Brief Description</a></li>
      </ul>
</li>
      <li>
<a href="#mathematical-framework">Mathematical Framework</a><ul class="nav nav-pills nav-stacked">
<li><a href="#inheritance-cube">Inheritance Cube</a></li>
      <li><a href="#population-dynamics">Population Dynamics</a></li>
      <li><a href="#gene-drive-releases-and-effects">Gene Drive Releases and Effects</a></li>
      <li><a href="#migration">Migration</a></li>
      </ul>
</li>
      <li><a href="#stochasticity">Stochasticity</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://chipdelmal.github.io">Héctor Manuel Sánchez Castellanos</a>, <a href="https://github.com/GilChrist19">Jared Bennett</a>, <a href="https://slwu89.github.io">Sean Wu</a>, <a href="https://www.marshalllab.com">John M. Marshall</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
