<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>MGDrivE Examples • MGDrivE</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="MGDrivE Examples">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MGDrivE</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/mgdrive_examples.html">Example Simulations</a>
    </li>
    <li>
      <a href="../articles/mgdrive_math.html">Mathematical Description</a>
    </li>
    <li>
      <a href="../articles/mgdrive_run.html">Detailed Run Example</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/MarshallLab/MGDrivE">
    <span class="fa fa-lg fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>MGDrivE Examples</h1>
            
      
      
      <div class="hidden name"><code>mgdrive_examples.Rmd</code></div>

    </div>

    
    
<style type="text/css">
pre {
  white-space: pre !important;
  overflow-y: scroll !important;
  max-height: 25vh !important;
}

# reference
#  https://stackoverflow.com/questions/41135085/how-to-make-vertical-scrollbar-appear-in-rmarkdown-code-chunks-html-view
</style>
<div id="landscape-setup" class="section level2">
<h2 class="hasAnchor">
<a href="#landscape-setup" class="anchor"></a>Landscape Setup</h2>
<p><em>MGDrivE</em> is capable of running in one population, on a simple network, or on networks derived from real locations and parameterized using local topology and climate. Topology and climate analysis are out of the scope of this vignette, but we will show how to setup simple, single-node examples to real landscapes parameterized only by distance. The key to this flexibility is that the landscape module of <em>MGDrivE</em> only requires a matrix of daily movement rates between nodes. Nodes are listed along the X/Y-axis, with the diagonal representing the proportion of mosquitoes that don’t leave a node at each day.</p>
<p>Each example produces a <code>moveMat</code> object - this is the matrix of daily movement rates. It is used for the <code>migrationMale</code> and/or <code>migrationFemale</code> matrices in the constructor of the <code>Network</code> object.</p>
<div id="single-population" class="section level3">
<h3 class="hasAnchor">
<a href="#single-population" class="anchor"></a>Single Population</h3>
<p>Each node on our network represents an independent population. Thus, it is important that <em>MGDrivE</em> run on a on a single node. A single population can be provided by parameterizing <em>MGDrivE</em> with a 1-by-1 matrix, where there is a 100% chance of individuals staying at that location (i.e., there is no where else to move, thus they have to remain).</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1"><span class="co"># setup movement matrix for 1 node</span></a>
<a class="sourceLine" id="cb1-2" title="2">moveMat &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="dt">data =</span> <span class="dv">1</span>, <span class="dt">nrow =</span> <span class="dv">1</span>, <span class="dt">ncol =</span> <span class="dv">1</span>)</a>
<a class="sourceLine" id="cb1-3" title="3">moveMat</a>
<a class="sourceLine" id="cb1-4" title="4"><span class="co">#&gt;      [,1]</span></a>
<a class="sourceLine" id="cb1-5" title="5"><span class="co">#&gt; [1,]    1</span></a></code></pre></div>
</div>
<div id="two-populations" class="section level3">
<h3 class="hasAnchor">
<a href="#two-populations" class="anchor"></a>Two Populations</h3>
<p>Extending our single population example, running <em>MGDrivE</em> on 2 populations is also simple. However, movement rates are calculated using the rows, i.e., the movement of critters from node 1 to any other node is parameterized by row 1 in the movement matrix. Thus, the rows of the movement matrix must be normalized.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" title="1"><span class="co"># setup movement matrix for 2 nodes</span></a>
<a class="sourceLine" id="cb2-2" title="2"></a>
<a class="sourceLine" id="cb2-3" title="3"><span class="co">####################</span></a>
<a class="sourceLine" id="cb2-4" title="4"><span class="co"># 2 nodes, no migration</span></a>
<a class="sourceLine" id="cb2-5" title="5"><span class="co">####################</span></a>
<a class="sourceLine" id="cb2-6" title="6">moveMat &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="dt">data =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">1</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">1</span>), <span class="dt">nrow =</span> <span class="dv">2</span>, <span class="dt">ncol =</span> <span class="dv">2</span>, <span class="dt">byrow =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb2-7" title="7">moveMat</a>
<a class="sourceLine" id="cb2-8" title="8"><span class="co">#&gt;      [,1] [,2]</span></a>
<a class="sourceLine" id="cb2-9" title="9"><span class="co">#&gt; [1,]    1    0</span></a>
<a class="sourceLine" id="cb2-10" title="10"><span class="co">#&gt; [2,]    0    1</span></a>
<a class="sourceLine" id="cb2-11" title="11"></a>
<a class="sourceLine" id="cb2-12" title="12"><span class="co">####################</span></a>
<a class="sourceLine" id="cb2-13" title="13"><span class="co"># 2 nodes, with migration</span></a>
<a class="sourceLine" id="cb2-14" title="14"><span class="co">####################</span></a>
<a class="sourceLine" id="cb2-15" title="15"><span class="co"># 5% migration per day from population 1</span></a>
<a class="sourceLine" id="cb2-16" title="16"><span class="co"># 10% migraton per day from population 2</span></a>
<a class="sourceLine" id="cb2-17" title="17"><span class="co">#  Notice that the rows sum to 1</span></a>
<a class="sourceLine" id="cb2-18" title="18">moveMat &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="dt">data =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="fl">0.95</span>, <span class="fl">0.05</span>, <span class="fl">0.01</span>, <span class="fl">0.90</span>),</a>
<a class="sourceLine" id="cb2-19" title="19">                  <span class="dt">nrow =</span> <span class="dv">2</span>, <span class="dt">ncol =</span> <span class="dv">2</span>, <span class="dt">byrow =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb2-20" title="20">moveMat</a>
<a class="sourceLine" id="cb2-21" title="21"><span class="co">#&gt;      [,1] [,2]</span></a>
<a class="sourceLine" id="cb2-22" title="22"><span class="co">#&gt; [1,] 0.95 0.05</span></a>
<a class="sourceLine" id="cb2-23" title="23"><span class="co">#&gt; [2,] 0.01 0.90</span></a></code></pre></div>
</div>
<div id="n-populations" class="section level3">
<h3 class="hasAnchor">
<a href="#n-populations" class="anchor"></a>n Populations</h3>
<p><em>MGDrivE</em> can run on arbitrary networks, provided that the rows of the movement matrix are normalized. Below are two more examples, one completely random, and one where there is migration to the population on either side only, as if all populations were in a line.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" title="1"><span class="co"># setup random movement matrix for 5 nodes</span></a>
<a class="sourceLine" id="cb3-2" title="2"></a>
<a class="sourceLine" id="cb3-3" title="3"><span class="co">####################</span></a>
<a class="sourceLine" id="cb3-4" title="4"><span class="co"># 5 nodes</span></a>
<a class="sourceLine" id="cb3-5" title="5"><span class="co">####################</span></a>
<a class="sourceLine" id="cb3-6" title="6">nNodes &lt;-<span class="st"> </span><span class="dv">5</span></a>
<a class="sourceLine" id="cb3-7" title="7"></a>
<a class="sourceLine" id="cb3-8" title="8"><span class="co"># fill with random data</span></a>
<a class="sourceLine" id="cb3-9" title="9">moveMat &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="dt">data =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Uniform">runif</a></span>(<span class="dt">n =</span> nNodes<span class="op">*</span>nNodes), <span class="dt">nrow =</span> nNodes, <span class="dt">ncol =</span> nNodes)</a>
<a class="sourceLine" id="cb3-10" title="10"></a>
<a class="sourceLine" id="cb3-11" title="11"><span class="co"># normalize</span></a>
<a class="sourceLine" id="cb3-12" title="12">moveMat &lt;-<span class="st"> </span>moveMat<span class="op">/</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colSums">rowSums</a></span>(<span class="dt">x =</span> moveMat)</a>
<a class="sourceLine" id="cb3-13" title="13">moveMat</a>
<a class="sourceLine" id="cb3-14" title="14"><span class="co">#&gt;           [,1]       [,2]       [,3]       [,4]       [,5]</span></a>
<a class="sourceLine" id="cb3-15" title="15"><span class="co">#&gt; [1,] 0.2448828 0.13895585 0.12253207 0.25163078 0.24199852</span></a>
<a class="sourceLine" id="cb3-16" title="16"><span class="co">#&gt; [2,] 0.2669949 0.20987141 0.20489206 0.27871981 0.03952180</span></a>
<a class="sourceLine" id="cb3-17" title="17"><span class="co">#&gt; [3,] 0.1162291 0.05470122 0.37966140 0.04772306 0.40168521</span></a>
<a class="sourceLine" id="cb3-18" title="18"><span class="co">#&gt; [4,] 0.2624233 0.20761107 0.08071609 0.15010016 0.29914933</span></a>
<a class="sourceLine" id="cb3-19" title="19"><span class="co">#&gt; [5,] 0.2617368 0.28756166 0.18854677 0.22853249 0.03362227</span></a></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" title="1"><span class="co"># setup line with 10 nodes</span></a>
<a class="sourceLine" id="cb4-2" title="2"></a>
<a class="sourceLine" id="cb4-3" title="3"><span class="co">####################</span></a>
<a class="sourceLine" id="cb4-4" title="4"><span class="co"># 10 nodes in a line</span></a>
<a class="sourceLine" id="cb4-5" title="5"><span class="co">####################</span></a>
<a class="sourceLine" id="cb4-6" title="6">nNodes &lt;-<span class="st"> </span><span class="dv">10</span></a>
<a class="sourceLine" id="cb4-7" title="7"></a>
<a class="sourceLine" id="cb4-8" title="8"><span class="co"># define function for use</span></a>
<a class="sourceLine" id="cb4-9" title="9">triDiag &lt;-<span class="st"> </span><span class="cf">function</span>(upper, lower){</a>
<a class="sourceLine" id="cb4-10" title="10">  </a>
<a class="sourceLine" id="cb4-11" title="11">  <span class="co"># return matrix</span></a>
<a class="sourceLine" id="cb4-12" title="12">  retMat &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="dt">data =</span> <span class="dv">0</span>, <span class="dt">nrow =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(upper) <span class="op">+</span><span class="st"> </span><span class="dv">1</span>, <span class="dt">ncol =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(upper) <span class="op">+</span><span class="st"> </span><span class="dv">1</span>)</a>
<a class="sourceLine" id="cb4-13" title="13">  </a>
<a class="sourceLine" id="cb4-14" title="14">  <span class="co"># set index values for upper/lower triangles</span></a>
<a class="sourceLine" id="cb4-15" title="15">  indx &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(upper)</a>
<a class="sourceLine" id="cb4-16" title="16">  </a>
<a class="sourceLine" id="cb4-17" title="17">  <span class="co"># set forward/backward migration using matrix access</span></a>
<a class="sourceLine" id="cb4-18" title="18">  retMat[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cbind">cbind</a></span>(indx<span class="op">+</span><span class="dv">1</span>,indx)] &lt;-<span class="st"> </span>lower</a>
<a class="sourceLine" id="cb4-19" title="19">  retMat[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cbind">cbind</a></span>(indx,indx<span class="op">+</span><span class="dv">1</span>)] &lt;-<span class="st"> </span>upper</a>
<a class="sourceLine" id="cb4-20" title="20">  </a>
<a class="sourceLine" id="cb4-21" title="21">  <span class="co"># set stay probs</span></a>
<a class="sourceLine" id="cb4-22" title="22">  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/diag">diag</a></span>(<span class="dt">x =</span> retMat) &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">-</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colSums">rowSums</a></span>(<span class="dt">x =</span> retMat)</a>
<a class="sourceLine" id="cb4-23" title="23">  </a>
<a class="sourceLine" id="cb4-24" title="24">  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/function">return</a></span>(retMat)</a>
<a class="sourceLine" id="cb4-25" title="25">}</a>
<a class="sourceLine" id="cb4-26" title="26"></a>
<a class="sourceLine" id="cb4-27" title="27"><span class="co"># fill movement matrix</span></a>
<a class="sourceLine" id="cb4-28" title="28"><span class="co">#  Remember, rows need to sum to 1.</span></a>
<a class="sourceLine" id="cb4-29" title="29">moveMat &lt;-<span class="st"> </span><span class="kw">triDiag</span>(<span class="dt">upper =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep.int</a></span>(<span class="dt">x =</span> <span class="fl">0.05</span>, <span class="dt">times =</span> nNodes<span class="dv">-1</span>),</a>
<a class="sourceLine" id="cb4-30" title="30">                   <span class="dt">lower =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep.int</a></span>(<span class="dt">x =</span> <span class="fl">0.05</span>, <span class="dt">times =</span> nNodes<span class="dv">-1</span>))</a>
<a class="sourceLine" id="cb4-31" title="31"></a>
<a class="sourceLine" id="cb4-32" title="32">moveMat</a>
<a class="sourceLine" id="cb4-33" title="33"><span class="co">#&gt;       [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]</span></a>
<a class="sourceLine" id="cb4-34" title="34"><span class="co">#&gt;  [1,] 0.95 0.05 0.00 0.00 0.00 0.00 0.00 0.00 0.00  0.00</span></a>
<a class="sourceLine" id="cb4-35" title="35"><span class="co">#&gt;  [2,] 0.05 0.90 0.05 0.00 0.00 0.00 0.00 0.00 0.00  0.00</span></a>
<a class="sourceLine" id="cb4-36" title="36"><span class="co">#&gt;  [3,] 0.00 0.05 0.90 0.05 0.00 0.00 0.00 0.00 0.00  0.00</span></a>
<a class="sourceLine" id="cb4-37" title="37"><span class="co">#&gt;  [4,] 0.00 0.00 0.05 0.90 0.05 0.00 0.00 0.00 0.00  0.00</span></a>
<a class="sourceLine" id="cb4-38" title="38"><span class="co">#&gt;  [5,] 0.00 0.00 0.00 0.05 0.90 0.05 0.00 0.00 0.00  0.00</span></a>
<a class="sourceLine" id="cb4-39" title="39"><span class="co">#&gt;  [6,] 0.00 0.00 0.00 0.00 0.05 0.90 0.05 0.00 0.00  0.00</span></a>
<a class="sourceLine" id="cb4-40" title="40"><span class="co">#&gt;  [7,] 0.00 0.00 0.00 0.00 0.00 0.05 0.90 0.05 0.00  0.00</span></a>
<a class="sourceLine" id="cb4-41" title="41"><span class="co">#&gt;  [8,] 0.00 0.00 0.00 0.00 0.00 0.00 0.05 0.90 0.05  0.00</span></a>
<a class="sourceLine" id="cb4-42" title="42"><span class="co">#&gt;  [9,] 0.00 0.00 0.00 0.00 0.00 0.00 0.00 0.05 0.90  0.05</span></a>
<a class="sourceLine" id="cb4-43" title="43"><span class="co">#&gt; [10,] 0.00 0.00 0.00 0.00 0.00 0.00 0.00 0.00 0.05  0.95</span></a></code></pre></div>
</div>
<div id="realistic-location" class="section level3">
<h3 class="hasAnchor">
<a href="#realistic-location" class="anchor"></a>Realistic Location</h3>
<p>Here we show how to make a landscape from a set of coordinates, each of which will represent a node the user wishes to simulate in the metapopulation model. In this example we show a simple setup with 3 nodes.</p>
<p>The first step is to calculate the distance between points using built-in functions. Examples include <code><a href="../reference/calcCos.html">calcCos()</a></code>, <code><a href="../reference/calcHaversine.html">calcHaversine()</a></code>, or most accurately, <code><a href="../reference/calcVinEll.html">calcVinEll()</a></code>.</p>
<p>The second step that we do here is to put an exponential kernel over those distances, using function <code><a href="../reference/calcExpKernel.html">calcExpKernel()</a></code>. We do this because it is hypothesized that mosquitoes follow a leptokurtic movement pattern. This also ensures that our rows are normalized. However, this is not necessary for all scenarios, one must only remember to normalize the rows.</p>
<p>We also provide functions to calculate a zero-inflated exponential kernel (<code><a href="../reference/calcHurdleExpKernel.html">calcHurdleExpKernel()</a></code>), gamma kernel (<code><a href="../reference/calcGammaKernel.html">calcGammaKernel()</a></code>), and log-normal kernel (<code><a href="../reference/calcLognormalKernel.html">calcLognormalKernel()</a></code>).</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" title="1"><span class="co"># realistic landscape</span></a>
<a class="sourceLine" id="cb5-2" title="2"></a>
<a class="sourceLine" id="cb5-3" title="3"><span class="co"># matrix of coordinates as latitude/longitude pairs</span></a>
<a class="sourceLine" id="cb5-4" title="4">lat_longs &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="dt">data =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="fl">37.873507</span>, <span class="fl">-122.268181</span>,</a>
<a class="sourceLine" id="cb5-5" title="5">                             <span class="fl">37.873578</span>, <span class="fl">-122.254430</span>,</a>
<a class="sourceLine" id="cb5-6" title="6">                             <span class="fl">37.869806</span>, <span class="fl">-122.267639</span>), </a>
<a class="sourceLine" id="cb5-7" title="7">                    <span class="dt">nrow =</span> <span class="dv">3</span>, <span class="dt">ncol =</span> <span class="dv">2</span>, <span class="dt">byrow =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb5-8" title="8">                    <span class="dt">dimnames =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="ot">NULL</span>, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">'Lat'</span>,<span class="st">'Lon'</span>)))</a>
<a class="sourceLine" id="cb5-9" title="9"></a>
<a class="sourceLine" id="cb5-10" title="10"><span class="co"># calculate distance matrix between points</span></a>
<a class="sourceLine" id="cb5-11" title="11"><span class="co"># dmat &lt;- MGDrivE::calcHaversine(latLongs = lat_longs)</span></a>
<a class="sourceLine" id="cb5-12" title="12"><span class="co"># dmat &lt;- MGDrivE::calcVinSph(latLongs = lat_longs)</span></a>
<a class="sourceLine" id="cb5-13" title="13">distMat &lt;-<span class="st"> </span>MGDrivE<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/MGDrivE/topics/calcVinEll">calcVinEll</a></span>(<span class="dt">latLongs =</span> lat_longs)</a>
<a class="sourceLine" id="cb5-14" title="14"></a>
<a class="sourceLine" id="cb5-15" title="15"></a>
<a class="sourceLine" id="cb5-16" title="16"><span class="co"># rate parameter such that about 90% of movement is within 1036 meters</span></a>
<a class="sourceLine" id="cb5-17" title="17">moveMat &lt;-<span class="st"> </span>MGDrivE<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/MGDrivE/topics/calcExpKernel">calcExpKernel</a></span>(<span class="dt">distMat =</span> distMat, <span class="dt">r =</span> <span class="dv">1</span><span class="op">/</span><span class="dv">450</span>)</a>
<a class="sourceLine" id="cb5-18" title="18">moveMat</a>
<a class="sourceLine" id="cb5-19" title="19"><span class="co">#&gt;            [,1]       [,2]       [,3]</span></a>
<a class="sourceLine" id="cb5-20" title="20"><span class="co">#&gt; [1,] 0.68171244 0.04633899 0.27194857</span></a>
<a class="sourceLine" id="cb5-21" title="21"><span class="co">#&gt; [2,] 0.06003669 0.88322499 0.05673833</span></a>
<a class="sourceLine" id="cb5-22" title="22"><span class="co">#&gt; [3,] 0.27264267 0.04390495 0.68345239</span></a></code></pre></div>
</div>
</div>
<div id="inheritance-simulations" class="section level2">
<h2 class="hasAnchor">
<a href="#inheritance-simulations" class="anchor"></a>Inheritance Simulations</h2>
<p>It is important to know that a model produces results consistent with accepted theory. It is also important that other members of the community (i.e., you, the user!) understand and can run their own explorations with a model. Thus, we use this section to provide small, simple examples illustrating how <em>MGDrivE</em> is setup and run, and connecting these simulations to the underlying theory.</p>
<div id="mendelian-inheritance-simulations-single-population" class="section level3">
<h3 class="hasAnchor">
<a href="#mendelian-inheritance-simulations-single-population" class="anchor"></a>Mendelian Inheritance Simulations, Single Population</h3>
<p>A useful benchmark of population models that include genetic inheritance is to study basic Mendelian inheritance. <em>MGDrivE</em> provides a 1-locus, 2-allele Mendelian inheritance pattern in the function <code><a href="../reference/cubeMendelian.html">cubeMendelian()</a></code>.</p>
<div id="deterministic" class="section level4">
<h4 class="hasAnchor">
<a href="#deterministic" class="anchor"></a>Deterministic</h4>
<p>First, we show an example running a deterministic <em>MGDrivE</em> simulation under Mendelian inheritance in a single node; that is, a non-spatial model. The population is 100% <em>AA</em> (homozygous dominant) individuals at time 0. At day 25, 10 female and 10 male <em>aa</em> (homozygous recessive) individuals enter the population.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" title="1"><span class="co">####################</span></a>
<a class="sourceLine" id="cb6-2" title="2"><span class="co"># Load libraries</span></a>
<a class="sourceLine" id="cb6-3" title="3"><span class="co">####################</span></a>
<a class="sourceLine" id="cb6-4" title="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(MGDrivE)</a>
<a class="sourceLine" id="cb6-5" title="5"><span class="co">#&gt; Loading MGDrivE: Mosquito Gene Drive Explorer</span></a>
<a class="sourceLine" id="cb6-6" title="6"></a>
<a class="sourceLine" id="cb6-7" title="7"><span class="co">####################</span></a>
<a class="sourceLine" id="cb6-8" title="8"><span class="co"># Output Folder</span></a>
<a class="sourceLine" id="cb6-9" title="9"><span class="co">####################</span></a>
<a class="sourceLine" id="cb6-10" title="10">outFolder &lt;-<span class="st"> "mgdrive"</span></a>
<a class="sourceLine" id="cb6-11" title="11"></a>
<a class="sourceLine" id="cb6-12" title="12"><span class="co">####################</span></a>
<a class="sourceLine" id="cb6-13" title="13"><span class="co"># Simulation Parameters</span></a>
<a class="sourceLine" id="cb6-14" title="14"><span class="co">####################</span></a>
<a class="sourceLine" id="cb6-15" title="15"><span class="co"># days to run the simulation</span></a>
<a class="sourceLine" id="cb6-16" title="16">tMax &lt;-<span class="st"> </span><span class="dv">365</span></a>
<a class="sourceLine" id="cb6-17" title="17"></a>
<a class="sourceLine" id="cb6-18" title="18"><span class="co"># entomological parameters</span></a>
<a class="sourceLine" id="cb6-19" title="19">bioParameters &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">betaK=</span><span class="dv">20</span>,<span class="dt">tEgg=</span><span class="dv">5</span>,<span class="dt">tLarva=</span><span class="dv">6</span>,<span class="dt">tPupa=</span><span class="dv">4</span>,<span class="dt">popGrowth=</span><span class="fl">1.175</span>,<span class="dt">muAd=</span><span class="fl">0.09</span>)</a>
<a class="sourceLine" id="cb6-20" title="20"></a>
<a class="sourceLine" id="cb6-21" title="21"><span class="co"># a 1-node network where mosquitoes do not leave</span></a>
<a class="sourceLine" id="cb6-22" title="22">moveMat &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">as.matrix</a></span>(<span class="dv">1</span>)</a>
<a class="sourceLine" id="cb6-23" title="23"></a>
<a class="sourceLine" id="cb6-24" title="24"><span class="co"># parameters of the population equilibrium</span></a>
<a class="sourceLine" id="cb6-25" title="25">adultPopEquilibrium &lt;-<span class="st"> </span><span class="dv">500</span></a>
<a class="sourceLine" id="cb6-26" title="26">sitesNumber &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">nrow</a></span>(moveMat)</a>
<a class="sourceLine" id="cb6-27" title="27">patchPops &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>(adultPopEquilibrium,sitesNumber)</a>
<a class="sourceLine" id="cb6-28" title="28"></a>
<a class="sourceLine" id="cb6-29" title="29"><span class="co">####################</span></a>
<a class="sourceLine" id="cb6-30" title="30"><span class="co"># Basic Inheritance pattern</span></a>
<a class="sourceLine" id="cb6-31" title="31"><span class="co">####################</span></a>
<a class="sourceLine" id="cb6-32" title="32"><span class="co"># Mendelian cube with standard (default) paraameters</span></a>
<a class="sourceLine" id="cb6-33" title="33">cube &lt;-<span class="st"> </span><span class="kw"><a href="../reference/cubeMendelian.html">cubeMendelian</a></span>()</a>
<a class="sourceLine" id="cb6-34" title="34"></a>
<a class="sourceLine" id="cb6-35" title="35"></a>
<a class="sourceLine" id="cb6-36" title="36"><span class="co">####################</span></a>
<a class="sourceLine" id="cb6-37" title="37"><span class="co"># Setup releases and batch migration</span></a>
<a class="sourceLine" id="cb6-38" title="38"><span class="co">####################</span></a>
<a class="sourceLine" id="cb6-39" title="39"><span class="co"># set up the empty release vector</span></a>
<a class="sourceLine" id="cb6-40" title="40"><span class="co">#  MGDrivE pulls things out by name</span></a>
<a class="sourceLine" id="cb6-41" title="41">patchReleases &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">replicate</a></span>(<span class="dt">n=</span>sitesNumber,</a>
<a class="sourceLine" id="cb6-42" title="42">                           <span class="dt">expr=</span>{<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">maleReleases=</span><span class="ot">NULL</span>,<span class="dt">femaleReleases=</span><span class="ot">NULL</span>)},</a>
<a class="sourceLine" id="cb6-43" title="43">                           <span class="dt">simplify=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb6-44" title="44"></a>
<a class="sourceLine" id="cb6-45" title="45"><span class="co"># choose release parameters</span></a>
<a class="sourceLine" id="cb6-46" title="46"><span class="co">#  Releases start at time 25, occur every day, for 1 day.</span></a>
<a class="sourceLine" id="cb6-47" title="47"><span class="co">#  There are 10 mosquitoes relaeses every time.</span></a>
<a class="sourceLine" id="cb6-48" title="48">releasesParameters &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">releasesStart=</span><span class="dv">25</span>,</a>
<a class="sourceLine" id="cb6-49" title="49">                           <span class="dt">releasesNumber=</span><span class="dv">1</span>,</a>
<a class="sourceLine" id="cb6-50" title="50">                           <span class="dt">releasesInterval=</span><span class="dv">0</span>,</a>
<a class="sourceLine" id="cb6-51" title="51">                           <span class="dt">releaseProportion=</span><span class="dv">10</span>)</a>
<a class="sourceLine" id="cb6-52" title="52"></a>
<a class="sourceLine" id="cb6-53" title="53"><span class="co"># generate male release vector</span></a>
<a class="sourceLine" id="cb6-54" title="54">maleReleasesVector &lt;-<span class="st"> </span><span class="kw"><a href="../reference/generateReleaseVector.html">generateReleaseVector</a></span>(<span class="dt">driveCube=</span>cube, </a>
<a class="sourceLine" id="cb6-55" title="55">                                            <span class="dt">releasesParameters=</span>releasesParameters,</a>
<a class="sourceLine" id="cb6-56" title="56">                                            <span class="dt">sex=</span><span class="st">"M"</span>)</a>
<a class="sourceLine" id="cb6-57" title="57"></a>
<a class="sourceLine" id="cb6-58" title="58"><span class="co"># generate female release vector</span></a>
<a class="sourceLine" id="cb6-59" title="59">femaleReleasesVector &lt;-<span class="st"> </span><span class="kw"><a href="../reference/generateReleaseVector.html">generateReleaseVector</a></span>(<span class="dt">driveCube=</span>cube,</a>
<a class="sourceLine" id="cb6-60" title="60">                                              <span class="dt">releasesParameters=</span>releasesParameters,</a>
<a class="sourceLine" id="cb6-61" title="61">                                              <span class="dt">sex=</span><span class="st">"F"</span>)</a>
<a class="sourceLine" id="cb6-62" title="62"></a>
<a class="sourceLine" id="cb6-63" title="63"><span class="co"># put releases into the proper place in the release list</span></a>
<a class="sourceLine" id="cb6-64" title="64">patchReleases[[<span class="dv">1</span>]]<span class="op">$</span>maleReleases &lt;-<span class="st"> </span>maleReleasesVector</a>
<a class="sourceLine" id="cb6-65" title="65">patchReleases[[<span class="dv">1</span>]]<span class="op">$</span>femaleReleases &lt;-<span class="st"> </span>femaleReleasesVector</a>
<a class="sourceLine" id="cb6-66" title="66"></a>
<a class="sourceLine" id="cb6-67" title="67"></a>
<a class="sourceLine" id="cb6-68" title="68"><span class="co"># batch migration is disabled by setting the probability to 0</span></a>
<a class="sourceLine" id="cb6-69" title="69"><span class="co"># This is required because of the stochastic simulations, but doesn't make sense </span></a>
<a class="sourceLine" id="cb6-70" title="70"><span class="co">#  in a deterministic simulation.</span></a>
<a class="sourceLine" id="cb6-71" title="71">batchMigration &lt;-<span class="st"> </span><span class="kw"><a href="../reference/basicBatchMigration.html">basicBatchMigration</a></span>(<span class="dt">batchProbs=</span><span class="dv">0</span>,</a>
<a class="sourceLine" id="cb6-72" title="72">                                      <span class="dt">sexProbs=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(.<span class="dv">5</span>,.<span class="dv">5</span>),</a>
<a class="sourceLine" id="cb6-73" title="73">                                      <span class="dt">numPatches=</span>sitesNumber)</a>
<a class="sourceLine" id="cb6-74" title="74"></a>
<a class="sourceLine" id="cb6-75" title="75"></a>
<a class="sourceLine" id="cb6-76" title="76"><span class="co">####################</span></a>
<a class="sourceLine" id="cb6-77" title="77"><span class="co"># Combine parameters and run!</span></a>
<a class="sourceLine" id="cb6-78" title="78"><span class="co">####################</span></a>
<a class="sourceLine" id="cb6-79" title="79"><span class="co"># set MGDrivE to run deterministic</span></a>
<a class="sourceLine" id="cb6-80" title="80"><span class="kw"><a href="../reference/setupMGDrivE.html">setupMGDrivE</a></span>(<span class="dt">stochasticityON =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb6-81" title="81"><span class="co">#&gt; initializing MGDrivE</span></a>
<a class="sourceLine" id="cb6-82" title="82"></a>
<a class="sourceLine" id="cb6-83" title="83"><span class="co"># setup parameters for the network. This builds a list of parameters required for</span></a>
<a class="sourceLine" id="cb6-84" title="84"><span class="co">#  every population in the network. In ths case, we havee a network of 1 population.</span></a>
<a class="sourceLine" id="cb6-85" title="85">netPar &lt;-<span class="st"> </span><span class="kw"><a href="../reference/parameterizeMGDrivE.html">parameterizeMGDrivE</a></span>(<span class="dt">runID=</span><span class="dv">1</span>, <span class="dt">simTime=</span>tMax, <span class="dt">nPatch=</span>sitesNumber,</a>
<a class="sourceLine" id="cb6-86" title="86">                              <span class="dt">beta=</span>bioParameters<span class="op">$</span>betaK, <span class="dt">muAd=</span>bioParameters<span class="op">$</span>muAd,</a>
<a class="sourceLine" id="cb6-87" title="87">                              <span class="dt">popGrowth=</span>bioParameters<span class="op">$</span>popGrowth, <span class="dt">tEgg=</span>bioParameters<span class="op">$</span>tEgg,</a>
<a class="sourceLine" id="cb6-88" title="88">                              <span class="dt">tLarva=</span>bioParameters<span class="op">$</span>tLarva, <span class="dt">tPupa=</span>bioParameters<span class="op">$</span>tPupa,</a>
<a class="sourceLine" id="cb6-89" title="89">                              <span class="dt">AdPopEQ=</span>patchPops)</a>
<a class="sourceLine" id="cb6-90" title="90"></a>
<a class="sourceLine" id="cb6-91" title="91"><span class="co"># build network prior to run</span></a>
<a class="sourceLine" id="cb6-92" title="92">MGDrivESim &lt;-<span class="st"> </span>Network<span class="op">$</span><span class="kw">new</span>(<span class="dt">params=</span>netPar,</a>
<a class="sourceLine" id="cb6-93" title="93">                          <span class="dt">driveCube=</span>cube,</a>
<a class="sourceLine" id="cb6-94" title="94">                          <span class="dt">patchReleases=</span>patchReleases,</a>
<a class="sourceLine" id="cb6-95" title="95">                          <span class="dt">migrationMale=</span>moveMat,</a>
<a class="sourceLine" id="cb6-96" title="96">                          <span class="dt">migrationFemale=</span>moveMat,</a>
<a class="sourceLine" id="cb6-97" title="97">                          <span class="dt">migrationBatch=</span>batchMigration,</a>
<a class="sourceLine" id="cb6-98" title="98">                          <span class="dt">directory=</span>outFolder)</a>
<a class="sourceLine" id="cb6-99" title="99"><span class="co">#&gt; initializing patch:  1  of  1</span></a>
<a class="sourceLine" id="cb6-100" title="100"><span class="co"># run simulation</span></a>
<a class="sourceLine" id="cb6-101" title="101">MGDrivESim<span class="op">$</span><span class="kw">oneRun</span>(<span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb6-102" title="102"></a>
<a class="sourceLine" id="cb6-103" title="103"></a>
<a class="sourceLine" id="cb6-104" title="104"><span class="co">####################</span></a>
<a class="sourceLine" id="cb6-105" title="105"><span class="co"># Post Analysis</span></a>
<a class="sourceLine" id="cb6-106" title="106"><span class="co">####################</span></a>
<a class="sourceLine" id="cb6-107" title="107"><span class="co"># split output by patch</span></a>
<a class="sourceLine" id="cb6-108" title="108"><span class="co">#  Required for plotting later</span></a>
<a class="sourceLine" id="cb6-109" title="109"><span class="kw"><a href="../reference/splitOutput.html">splitOutput</a></span>(<span class="dt">readDir =</span> outFolder, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb6-110" title="110"></a>
<a class="sourceLine" id="cb6-111" title="111"><span class="co"># aggregate females by their mate choice</span></a>
<a class="sourceLine" id="cb6-112" title="112"><span class="co">#  This reduces the female file to have the same columns as the male file</span></a>
<a class="sourceLine" id="cb6-113" title="113"><span class="kw"><a href="../reference/aggregateFemales.html">aggregateFemales</a></span>(<span class="dt">readDir =</span> outFolder, <span class="dt">genotypes =</span> cube<span class="op">$</span>genotypesID,</a>
<a class="sourceLine" id="cb6-114" title="114">                 <span class="dt">remFile =</span> <span class="ot">TRUE</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb6-115" title="115"></a>
<a class="sourceLine" id="cb6-116" title="116"><span class="co"># plot output to see effect</span></a>
<a class="sourceLine" id="cb6-117" title="117"><span class="kw"><a href="../reference/plotMGDrivESingle.html">plotMGDrivESingle</a></span>(<span class="dt">readDir=</span>outFolder,<span class="dt">totalPop =</span> <span class="ot">TRUE</span>,<span class="dt">lwd=</span><span class="fl">3.5</span>,<span class="dt">alpha=</span><span class="dv">1</span>)</a></code></pre></div>
<p><img src="mgdrive_examples_files/figure-html/unnamed-chunk-8-1.png" width="672"></p>
<p>In the figure, we see males and females of every genotype plotted by time, along with the total population. With an equilibrium population size of 500 individuals, and no sex biasing, we have an equal division of males and females. The release occurs equally in both sexes at time t = 25, as expected, and we see an increase in the total population due to the release before recovering to the equilibrium population size.</p>
<p>Since this is basic Mendelian inheritance, we should be able to calculate the expected allele and genotype frequencies assuming Hardy-Weinberg equilibrium (HWE). As we released 20 <em>aa</em> individuals into a population of 500 <em>AA</em> individuals, we can calculate the expected allele frequencies:</p>
<ul>
<li>
<em>a</em>: <span class="math inline">\(\frac{1}{26}\)</span><br>
</li>
<li>
<em>A</em>: <span class="math inline">\(\frac{25}{26}\)</span>
</li>
</ul>
<p>This leads to the following expected genotype frequencies:</p>
<ul>
<li>
<em>AA</em>: <span class="math inline">\(\big(\frac{25}{26}\big)^2 =\)</span> 0.9246<br>
</li>
<li>
<em>Aa</em>: <span class="math inline">\(2\cdot \big(\frac{25}{26}\big)\cdot\big(\frac{1}{26}\big) =\)</span> 0.0740<br>
</li>
<li>
<em>aa</em>: <span class="math inline">\(\big(\frac{1}{26}\big)^2 =\)</span> 0.0015</li>
</ul>
<p>Which, in a population of size 500, implies the following number of individuals (remember, this is a deterministic simulation, so “individuals” will not be whole numbers):</p>
<ul>
<li>
<em>AA</em>: 462.28</li>
<li>
<em>Aa</em>: 36.98</li>
<li>
<em>aa</em>: 0.74</li>
</ul>
<p>However, when we look at the equilibrium genotypes from the simulation, the populations are:</p>
<ul>
<li>
<em>AA</em>: 485.48</li>
<li>
<em>Aa</em>: 14.41</li>
<li>
<em>aa</em>: 0.11</li>
</ul>
<p>These are significantly different from the expected genotype numbers according to HWE. This is because of the assumptions in how HWE is calculated, specifically, ignoring maturation of child stages prior to reproduction by assuming discrete generations, and infinite population size. In contrast, <em>MGDrivE</em> has overlapping generations and tracks offspring development prior to adulthood. Additionally, while the release performed here is small, the population is far from “infinite”, and the release represents 4% of the population. Therefore, the effective population size is significantly larger than 500. Ignoring the effect of constant daily mortality on the distribution of adults, implying individual lifespans follow a geometric distribution and reducing the proportion of released individuals to wild-type as adults die off, we can calculate the effective population size using the following parameters;</p>
<ul>
<li>Adult Population Size: 500<br>
</li>
<li>Release Amount: 20<br>
</li>
<li>Daily Pupation Amount: 45<br>
</li>
<li>Maturation Time: 15 days</li>
</ul>
<p>From these parameters, an approximate effective population size is calculated by combining the <em>Adult Population Size</em>, plus the <em>Release Amount</em>, plus the <em>Maturation Time</em> times the <em>Daily Pupation Amount</em>, which estimates a population size of 1195. Using this estimate of our effective population size, the expected allele frequencies are:</p>
<ul>
<li>
<em>a</em>: <span class="math inline">\(\frac{4}{239}\)</span><br>
</li>
<li>
<em>A</em>: <span class="math inline">\(\frac{235}{239}\)</span>
</li>
</ul>
<p>Our new expected genotype frequencies:</p>
<ul>
<li>
<em>AA</em>: <span class="math inline">\(\big(\frac{235}{239}\big)^2 =\)</span> 0.96681<br>
</li>
<li>
<em>Aa</em>: <span class="math inline">\(2\cdot \big(\frac{235}{239}\big)\cdot\big(\frac{4}{239}\big) =\)</span> 0.03291<br>
</li>
<li>
<em>aa</em>: <span class="math inline">\(\big(\frac{4}{239}\big)^2 =\)</span> 0.00028</li>
</ul>
<p>Since our population will return to equilibrium after some time, 500 individuals, we expect the following number of individuals of each genotype:</p>
<ul>
<li>
<em>AA</em>: 483.40</li>
<li>
<em>Aa</em>: 16.46</li>
<li>
<em>aa</em>: 0.14</li>
</ul>
<p>This is significantly closer to the simulated population. As expected, it is slightly high, which stems from ignoring the geometric distribution of our adult population.</p>
</div>
<div id="deterministic-with-fitness-cost" class="section level4">
<h4 class="hasAnchor">
<a href="#deterministic-with-fitness-cost" class="anchor"></a>Deterministic, With Fitness Cost</h4>
<p>Now that we have shown a basic Mendelian simulation, we will expand it a little by including a fitness cost on certain genotypes. In the following simulation, the homozygotes, <em>AA</em> and <em>aa</em>, are only 60% as fit as the heterozygote over their adult lifetime. This is an example of a heterozygote advantage, and we should see the population become primarily heterozygous. We will increase the size and number of releases, to speed up the equilibration process, but everything else will remain the same, ie., one population and a deterministic simulation.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" title="1"><span class="co">####################</span></a>
<a class="sourceLine" id="cb7-2" title="2"><span class="co"># Load libraries</span></a>
<a class="sourceLine" id="cb7-3" title="3"><span class="co">####################</span></a>
<a class="sourceLine" id="cb7-4" title="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(MGDrivE)</a>
<a class="sourceLine" id="cb7-5" title="5"></a>
<a class="sourceLine" id="cb7-6" title="6"><span class="co">####################</span></a>
<a class="sourceLine" id="cb7-7" title="7"><span class="co"># Output Folder</span></a>
<a class="sourceLine" id="cb7-8" title="8"><span class="co">####################</span></a>
<a class="sourceLine" id="cb7-9" title="9">outFolder &lt;-<span class="st"> "mgdrive"</span></a>
<a class="sourceLine" id="cb7-10" title="10"></a>
<a class="sourceLine" id="cb7-11" title="11"><span class="co">####################</span></a>
<a class="sourceLine" id="cb7-12" title="12"><span class="co"># Simulation Parameters</span></a>
<a class="sourceLine" id="cb7-13" title="13"><span class="co">####################</span></a>
<a class="sourceLine" id="cb7-14" title="14"><span class="co"># days to run the simulation</span></a>
<a class="sourceLine" id="cb7-15" title="15">tMax &lt;-<span class="st"> </span><span class="dv">365</span><span class="op">*</span><span class="dv">2</span></a>
<a class="sourceLine" id="cb7-16" title="16"></a>
<a class="sourceLine" id="cb7-17" title="17"><span class="co"># entomological parameters</span></a>
<a class="sourceLine" id="cb7-18" title="18">bioParameters &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">betaK=</span><span class="dv">20</span>,<span class="dt">tEgg=</span><span class="dv">5</span>,<span class="dt">tLarva=</span><span class="dv">6</span>,<span class="dt">tPupa=</span><span class="dv">4</span>,<span class="dt">popGrowth=</span><span class="fl">1.175</span>,<span class="dt">muAd=</span><span class="fl">0.09</span>)</a>
<a class="sourceLine" id="cb7-19" title="19"></a>
<a class="sourceLine" id="cb7-20" title="20"><span class="co"># a 1-node network where mosquitoes do not leave</span></a>
<a class="sourceLine" id="cb7-21" title="21">moveMat &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">as.matrix</a></span>(<span class="dv">1</span>)</a>
<a class="sourceLine" id="cb7-22" title="22"></a>
<a class="sourceLine" id="cb7-23" title="23"><span class="co"># parameters of the population equilibrium</span></a>
<a class="sourceLine" id="cb7-24" title="24">adultPopEquilibrium &lt;-<span class="st"> </span><span class="dv">500</span></a>
<a class="sourceLine" id="cb7-25" title="25">sitesNumber &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">nrow</a></span>(moveMat)</a>
<a class="sourceLine" id="cb7-26" title="26">patchPops &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>(adultPopEquilibrium,sitesNumber)</a>
<a class="sourceLine" id="cb7-27" title="27"></a>
<a class="sourceLine" id="cb7-28" title="28"><span class="co">####################</span></a>
<a class="sourceLine" id="cb7-29" title="29"><span class="co"># Basic Inheritance pattern</span></a>
<a class="sourceLine" id="cb7-30" title="30"><span class="co">####################</span></a>
<a class="sourceLine" id="cb7-31" title="31"><span class="co"># Mendelian cube with standard (default) paraameters</span></a>
<a class="sourceLine" id="cb7-32" title="32"></a>
<a class="sourceLine" id="cb7-33" title="33"><span class="co"># This time, lets put fitness cost on the homozygotes, giving the heterozygote </span></a>
<a class="sourceLine" id="cb7-34" title="34"><span class="co">#  an advantage</span></a>
<a class="sourceLine" id="cb7-35" title="35"></a>
<a class="sourceLine" id="cb7-36" title="36"><span class="co"># These genotypes correspond to ones in the cube. Look at a base cube first, </span></a>
<a class="sourceLine" id="cb7-37" title="37"><span class="co">#  then set this.</span></a>
<a class="sourceLine" id="cb7-38" title="38"><span class="co"># homozygotes are 60% as fit as heterozygote over their entire lifetime</span></a>
<a class="sourceLine" id="cb7-39" title="39"><span class="co">#  Since omega is the adult daily death rate, we use the built-in function to </span></a>
<a class="sourceLine" id="cb7-40" title="40"><span class="co">#  calculate our desired lifetime cost as applied daily</span></a>
<a class="sourceLine" id="cb7-41" title="41">dayOmega &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getOmega.html">getOmega</a></span>(<span class="dt">mu =</span> bioParameters<span class="op">$</span>muAd, <span class="dt">lifespanReduction =</span> <span class="fl">0.60</span>)</a>
<a class="sourceLine" id="cb7-42" title="42">omegaNew &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"AA"</span>=dayOmega, <span class="st">"aa"</span>=dayOmega)</a>
<a class="sourceLine" id="cb7-43" title="43"></a>
<a class="sourceLine" id="cb7-44" title="44"><span class="co"># setup cube</span></a>
<a class="sourceLine" id="cb7-45" title="45">cube &lt;-<span class="st"> </span><span class="kw"><a href="../reference/cubeMendelian.html">cubeMendelian</a></span>(<span class="dt">omega =</span> omegaNew)</a>
<a class="sourceLine" id="cb7-46" title="46"></a>
<a class="sourceLine" id="cb7-47" title="47"></a>
<a class="sourceLine" id="cb7-48" title="48"><span class="co">####################</span></a>
<a class="sourceLine" id="cb7-49" title="49"><span class="co"># Setup releases and batch migration</span></a>
<a class="sourceLine" id="cb7-50" title="50"><span class="co">####################</span></a>
<a class="sourceLine" id="cb7-51" title="51"><span class="co"># set up the empty release vector</span></a>
<a class="sourceLine" id="cb7-52" title="52"><span class="co">#  MGDrivE pulls things out by name</span></a>
<a class="sourceLine" id="cb7-53" title="53">patchReleases &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">replicate</a></span>(<span class="dt">n=</span>sitesNumber,</a>
<a class="sourceLine" id="cb7-54" title="54">                           <span class="dt">expr=</span>{<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">maleReleases=</span><span class="ot">NULL</span>,<span class="dt">femaleReleases=</span><span class="ot">NULL</span>)},</a>
<a class="sourceLine" id="cb7-55" title="55">                           <span class="dt">simplify=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb7-56" title="56"></a>
<a class="sourceLine" id="cb7-57" title="57"><span class="co"># choose release parameters</span></a>
<a class="sourceLine" id="cb7-58" title="58"><span class="co">#  Releases start at time 25, occur every day, for 5 days.</span></a>
<a class="sourceLine" id="cb7-59" title="59"><span class="co">#  There are 50 mosquitoes relaeses every time.</span></a>
<a class="sourceLine" id="cb7-60" title="60">releasesParameters &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">releasesStart=</span><span class="dv">100</span>,</a>
<a class="sourceLine" id="cb7-61" title="61">                           <span class="dt">releasesNumber=</span><span class="dv">5</span>,</a>
<a class="sourceLine" id="cb7-62" title="62">                           <span class="dt">releasesInterval=</span><span class="dv">0</span>,</a>
<a class="sourceLine" id="cb7-63" title="63">                           <span class="dt">releaseProportion=</span><span class="dv">50</span>)</a>
<a class="sourceLine" id="cb7-64" title="64"></a>
<a class="sourceLine" id="cb7-65" title="65"><span class="co"># generate male release vector</span></a>
<a class="sourceLine" id="cb7-66" title="66">maleReleasesVector &lt;-<span class="st"> </span><span class="kw"><a href="../reference/generateReleaseVector.html">generateReleaseVector</a></span>(<span class="dt">driveCube=</span>cube, </a>
<a class="sourceLine" id="cb7-67" title="67">                                            <span class="dt">releasesParameters=</span>releasesParameters,</a>
<a class="sourceLine" id="cb7-68" title="68">                                            <span class="dt">sex=</span><span class="st">"M"</span>)</a>
<a class="sourceLine" id="cb7-69" title="69"></a>
<a class="sourceLine" id="cb7-70" title="70"><span class="co"># generate female release vector</span></a>
<a class="sourceLine" id="cb7-71" title="71">femaleReleasesVector &lt;-<span class="st"> </span><span class="kw"><a href="../reference/generateReleaseVector.html">generateReleaseVector</a></span>(<span class="dt">driveCube=</span>cube,</a>
<a class="sourceLine" id="cb7-72" title="72">                                              <span class="dt">releasesParameters=</span>releasesParameters,</a>
<a class="sourceLine" id="cb7-73" title="73">                                              <span class="dt">sex=</span><span class="st">"F"</span>)</a>
<a class="sourceLine" id="cb7-74" title="74"></a>
<a class="sourceLine" id="cb7-75" title="75"><span class="co"># put releases into the proper place in the release list</span></a>
<a class="sourceLine" id="cb7-76" title="76">patchReleases[[<span class="dv">1</span>]]<span class="op">$</span>maleReleases &lt;-<span class="st"> </span>maleReleasesVector</a>
<a class="sourceLine" id="cb7-77" title="77">patchReleases[[<span class="dv">1</span>]]<span class="op">$</span>femaleReleases &lt;-<span class="st"> </span>femaleReleasesVector</a>
<a class="sourceLine" id="cb7-78" title="78"></a>
<a class="sourceLine" id="cb7-79" title="79"></a>
<a class="sourceLine" id="cb7-80" title="80"><span class="co"># batch migration is disabled by setting the probability to 0</span></a>
<a class="sourceLine" id="cb7-81" title="81"><span class="co"># This is required because of the stochastic simulations, but doesn't make sense </span></a>
<a class="sourceLine" id="cb7-82" title="82"><span class="co">#  in a deterministic simulation.</span></a>
<a class="sourceLine" id="cb7-83" title="83">batchMigration &lt;-<span class="st"> </span><span class="kw"><a href="../reference/basicBatchMigration.html">basicBatchMigration</a></span>(<span class="dt">batchProbs=</span><span class="dv">0</span>,</a>
<a class="sourceLine" id="cb7-84" title="84">                                      <span class="dt">sexProbs=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(.<span class="dv">5</span>,.<span class="dv">5</span>),</a>
<a class="sourceLine" id="cb7-85" title="85">                                      <span class="dt">numPatches=</span>sitesNumber)</a>
<a class="sourceLine" id="cb7-86" title="86"></a>
<a class="sourceLine" id="cb7-87" title="87"></a>
<a class="sourceLine" id="cb7-88" title="88"><span class="co">####################</span></a>
<a class="sourceLine" id="cb7-89" title="89"><span class="co"># Combine parameters and run!</span></a>
<a class="sourceLine" id="cb7-90" title="90"><span class="co">####################</span></a>
<a class="sourceLine" id="cb7-91" title="91"><span class="co"># set MGDrivE to run deterministic</span></a>
<a class="sourceLine" id="cb7-92" title="92"><span class="kw"><a href="../reference/setupMGDrivE.html">setupMGDrivE</a></span>(<span class="dt">stochasticityON =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb7-93" title="93"><span class="co">#&gt; initializing MGDrivE</span></a>
<a class="sourceLine" id="cb7-94" title="94"></a>
<a class="sourceLine" id="cb7-95" title="95"><span class="co"># setup parameters for the network. This builds a list of parameters required for</span></a>
<a class="sourceLine" id="cb7-96" title="96"><span class="co">#  every population in the network. In ths case, we havee a network of 1 population.</span></a>
<a class="sourceLine" id="cb7-97" title="97">netPar &lt;-<span class="st"> </span><span class="kw"><a href="../reference/parameterizeMGDrivE.html">parameterizeMGDrivE</a></span>(<span class="dt">runID=</span><span class="dv">1</span>, <span class="dt">simTime=</span>tMax, <span class="dt">nPatch=</span>sitesNumber,</a>
<a class="sourceLine" id="cb7-98" title="98">                              <span class="dt">beta=</span>bioParameters<span class="op">$</span>betaK, <span class="dt">muAd=</span>bioParameters<span class="op">$</span>muAd,</a>
<a class="sourceLine" id="cb7-99" title="99">                              <span class="dt">popGrowth=</span>bioParameters<span class="op">$</span>popGrowth, <span class="dt">tEgg=</span>bioParameters<span class="op">$</span>tEgg,</a>
<a class="sourceLine" id="cb7-100" title="100">                              <span class="dt">tLarva=</span>bioParameters<span class="op">$</span>tLarva, <span class="dt">tPupa=</span>bioParameters<span class="op">$</span>tPupa,</a>
<a class="sourceLine" id="cb7-101" title="101">                              <span class="dt">AdPopEQ=</span>patchPops)</a>
<a class="sourceLine" id="cb7-102" title="102"></a>
<a class="sourceLine" id="cb7-103" title="103"><span class="co"># build network prior to run</span></a>
<a class="sourceLine" id="cb7-104" title="104">MGDrivESim &lt;-<span class="st"> </span>Network<span class="op">$</span><span class="kw">new</span>(<span class="dt">params=</span>netPar,</a>
<a class="sourceLine" id="cb7-105" title="105">                          <span class="dt">driveCube=</span>cube,</a>
<a class="sourceLine" id="cb7-106" title="106">                          <span class="dt">patchReleases=</span>patchReleases,</a>
<a class="sourceLine" id="cb7-107" title="107">                          <span class="dt">migrationMale=</span>moveMat,</a>
<a class="sourceLine" id="cb7-108" title="108">                          <span class="dt">migrationFemale=</span>moveMat,</a>
<a class="sourceLine" id="cb7-109" title="109">                          <span class="dt">migrationBatch=</span>batchMigration,</a>
<a class="sourceLine" id="cb7-110" title="110">                          <span class="dt">directory=</span>outFolder)</a>
<a class="sourceLine" id="cb7-111" title="111"><span class="co">#&gt; initializing patch:  1  of  1</span></a>
<a class="sourceLine" id="cb7-112" title="112"><span class="co"># run simulation</span></a>
<a class="sourceLine" id="cb7-113" title="113">MGDrivESim<span class="op">$</span><span class="kw">oneRun</span>(<span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb7-114" title="114"></a>
<a class="sourceLine" id="cb7-115" title="115"></a>
<a class="sourceLine" id="cb7-116" title="116"><span class="co">####################</span></a>
<a class="sourceLine" id="cb7-117" title="117"><span class="co"># Post Analysis</span></a>
<a class="sourceLine" id="cb7-118" title="118"><span class="co">####################</span></a>
<a class="sourceLine" id="cb7-119" title="119"><span class="co"># split output by patch</span></a>
<a class="sourceLine" id="cb7-120" title="120"><span class="co">#  Required for plotting later</span></a>
<a class="sourceLine" id="cb7-121" title="121"><span class="kw"><a href="../reference/splitOutput.html">splitOutput</a></span>(<span class="dt">readDir=</span>outFolder, <span class="dt">verbose=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb7-122" title="122"></a>
<a class="sourceLine" id="cb7-123" title="123"><span class="co"># aggregate females by their mate choice</span></a>
<a class="sourceLine" id="cb7-124" title="124"><span class="co">#  This reduces the female file to have the same columns as the male file</span></a>
<a class="sourceLine" id="cb7-125" title="125"><span class="kw"><a href="../reference/aggregateFemales.html">aggregateFemales</a></span>(<span class="dt">readDir=</span>outFolder,<span class="dt">genotypes=</span>cube<span class="op">$</span>genotypesID,<span class="dt">remFile=</span><span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb7-126" title="126">                 <span class="dt">verbose=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb7-127" title="127"></a>
<a class="sourceLine" id="cb7-128" title="128"><span class="co"># plot output to see effect</span></a>
<a class="sourceLine" id="cb7-129" title="129"><span class="kw"><a href="../reference/plotMGDrivESingle.html">plotMGDrivESingle</a></span>(<span class="dt">readDir=</span>outFolder,<span class="dt">totalPop=</span><span class="ot">TRUE</span>,<span class="dt">lwd=</span><span class="fl">3.5</span>,<span class="dt">alpha=</span><span class="dv">1</span>)</a></code></pre></div>
<p><img src="mgdrive_examples_files/figure-html/unnamed-chunk-11-1.png" width="672"></p>
<p>Again, we plot the females on the left, the males on the right, with the population size on the Y-axis and simulation time along the X-axis. We see the total population size starts at 250 males and females, a total of 500 individuals, but quickly drops to 274.9283493. This is because of the fitness cost applied to the <em>AA</em> individuals that make up the population. As we applied an 60% fitness cost, we expect the total population to be 60% of the desired population. 60% of 500 individuals, assuming HWE applies, is 300, which is close to the simulated population size prior to the releases. This discrepancy can be accounted for by the presence of density-dependence during larval maturation.</p>
<p>After releases of <em>aa</em> individuals are performed, we see the population quickly drive towards heterozygous individuals, and the population size marginally recover. The population partially recovers because there is no fitness cost on the heterozygotes, but does not fully recover because heterozygotes create homozygotes at each generation. At equilibrium, given that heterozygotes produce homozygotes at a rate of 50% per generation, our expected genotype amounts (assuming all HWE assumptions apply) are:</p>
<ul>
<li>
<em>AA</em>: <span class="math inline">\(500 \cdot \frac{1}{4} \cdot 0.60 = 75\)</span><br>
</li>
<li>
<em>Aa</em>: <span class="math inline">\(500 \cdot \frac{1}{2} \cdot 1.00 = 250\)</span><br>
</li>
<li>
<em>aa</em>: <span class="math inline">\(500 \cdot \frac{1}{4} \cdot 0.60 = 75\)</span><br>
</li>
<li>Total: <em>AA</em> + <em>Aa</em> + <em>aa</em> = 400</li>
</ul>
<p>Checking at the end of our simulation, we find that the empirical genotype amounts are:</p>
<ul>
<li>
<em>AA</em>: 72.58</li>
<li>
<em>Aa</em>: 241.03</li>
<li>
<em>aa</em>: 72.02</li>
<li>Total: 385.63</li>
</ul>
<p>This closely matches our simulated population, differing because HWE assumes discrete generations.</p>
</div>
<div id="stochastic" class="section level4">
<h4 class="hasAnchor">
<a href="#stochastic" class="anchor"></a>Stochastic</h4>
<p>The processes of birth, death, mating, and inheritance are inherently probabilistic, meaning that in order to understand the full spectrum of outcomes a model may produce, it is necessary to simulate from correctly-specified stochastic models. We show an example of the same modeled system here when the stochastic simulation is used; we run an ensemble of simulations, then plot the trajectories to give a sense of the expected variability in possible model trajectories.</p>
<p>We now use the same simulation parameters as the previous example but run the stochastic version of the model. To view variability in sampled trajectories, we run 50 repetitions. For realistic applications, large ensembles of simulations (&gt;100) should be run and statistically analyzed to characterize quantities of interest.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" title="1"><span class="co">####################</span></a>
<a class="sourceLine" id="cb8-2" title="2"><span class="co"># Load libraries</span></a>
<a class="sourceLine" id="cb8-3" title="3"><span class="co">####################</span></a>
<a class="sourceLine" id="cb8-4" title="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(MGDrivE)</a>
<a class="sourceLine" id="cb8-5" title="5"></a>
<a class="sourceLine" id="cb8-6" title="6"><span class="co">####################</span></a>
<a class="sourceLine" id="cb8-7" title="7"><span class="co"># Output Folder</span></a>
<a class="sourceLine" id="cb8-8" title="8"><span class="co">####################</span></a>
<a class="sourceLine" id="cb8-9" title="9"><span class="co"># directory</span></a>
<a class="sourceLine" id="cb8-10" title="10"><span class="co"># This is slightly obtuse for vignette building reasons</span></a>
<a class="sourceLine" id="cb8-11" title="11"><span class="co">#  Really, all you need is a base directory, then the repetitions in folders inside that.</span></a>
<a class="sourceLine" id="cb8-12" title="12"><span class="co">#  Here, our base directory is "mgdrive", and the repetition folders are "001","002", etc.</span></a>
<a class="sourceLine" id="cb8-13" title="13"><span class="co">#  So, the final structure is "mgdrive/001","mgdrive/002", etc. </span></a>
<a class="sourceLine" id="cb8-14" title="14">outFolder &lt;-<span class="st"> "mgdrive"</span></a>
<a class="sourceLine" id="cb8-15" title="15"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/files2">dir.create</a></span>(<span class="dt">path =</span> outFolder)</a>
<a class="sourceLine" id="cb8-16" title="16"></a>
<a class="sourceLine" id="cb8-17" title="17"><span class="co">####################</span></a>
<a class="sourceLine" id="cb8-18" title="18"><span class="co"># Simulation Parameters</span></a>
<a class="sourceLine" id="cb8-19" title="19"><span class="co">####################</span></a>
<a class="sourceLine" id="cb8-20" title="20"><span class="co"># days to run the simulation</span></a>
<a class="sourceLine" id="cb8-21" title="21">tMax &lt;-<span class="st"> </span><span class="dv">365</span></a>
<a class="sourceLine" id="cb8-22" title="22"></a>
<a class="sourceLine" id="cb8-23" title="23"><span class="co"># number of Monte Carlo iterations</span></a>
<a class="sourceLine" id="cb8-24" title="24">nRep &lt;-<span class="st"> </span><span class="dv">50</span></a>
<a class="sourceLine" id="cb8-25" title="25"></a>
<a class="sourceLine" id="cb8-26" title="26"><span class="co"># each Monte Carlo iteration gets its own folder</span></a>
<a class="sourceLine" id="cb8-27" title="27">folderNames &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/file.path">file.path</a></span>(outFolder,</a>
<a class="sourceLine" id="cb8-28" title="28">                         <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/formatc">formatC</a></span>(<span class="dt">x =</span> <span class="dv">1</span><span class="op">:</span>nRep, <span class="dt">width =</span> <span class="dv">3</span>, <span class="dt">format =</span> <span class="st">"d"</span>, <span class="dt">flag =</span> <span class="st">"0"</span>))</a>
<a class="sourceLine" id="cb8-29" title="29"></a>
<a class="sourceLine" id="cb8-30" title="30"><span class="co"># entomological parameters</span></a>
<a class="sourceLine" id="cb8-31" title="31">bioParameters &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">betaK=</span><span class="dv">20</span>,<span class="dt">tEgg=</span><span class="dv">5</span>,<span class="dt">tLarva=</span><span class="dv">6</span>,<span class="dt">tPupa=</span><span class="dv">4</span>,<span class="dt">popGrowth=</span><span class="fl">1.175</span>,<span class="dt">muAd=</span><span class="fl">0.09</span>)</a>
<a class="sourceLine" id="cb8-32" title="32"></a>
<a class="sourceLine" id="cb8-33" title="33"><span class="co"># a 1-node network where mosquitoes do not leave</span></a>
<a class="sourceLine" id="cb8-34" title="34">moveMat &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">as.matrix</a></span>(<span class="dv">1</span>)</a>
<a class="sourceLine" id="cb8-35" title="35"></a>
<a class="sourceLine" id="cb8-36" title="36"><span class="co"># parameters of the population equilibrium</span></a>
<a class="sourceLine" id="cb8-37" title="37">adultPopEquilibrium &lt;-<span class="st"> </span><span class="dv">500</span></a>
<a class="sourceLine" id="cb8-38" title="38">sitesNumber &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">nrow</a></span>(moveMat)</a>
<a class="sourceLine" id="cb8-39" title="39">patchPops &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>(adultPopEquilibrium,sitesNumber)</a>
<a class="sourceLine" id="cb8-40" title="40"></a>
<a class="sourceLine" id="cb8-41" title="41"><span class="co">####################</span></a>
<a class="sourceLine" id="cb8-42" title="42"><span class="co"># Basic Inheritance pattern</span></a>
<a class="sourceLine" id="cb8-43" title="43"><span class="co">####################</span></a>
<a class="sourceLine" id="cb8-44" title="44"><span class="co"># Mendelian cube with standard (default) paraameters</span></a>
<a class="sourceLine" id="cb8-45" title="45"></a>
<a class="sourceLine" id="cb8-46" title="46"><span class="co"># This time, lets put fitness cost on the homozygotes, giving the heterozygote </span></a>
<a class="sourceLine" id="cb8-47" title="47"><span class="co">#  an advantage</span></a>
<a class="sourceLine" id="cb8-48" title="48"></a>
<a class="sourceLine" id="cb8-49" title="49"><span class="co"># These genotypes correspond to ones in the cube. Look at a base cube first, </span></a>
<a class="sourceLine" id="cb8-50" title="50"><span class="co">#  then set this.</span></a>
<a class="sourceLine" id="cb8-51" title="51"><span class="co"># homozygotes are 60% as fit as heterozygote over their entire lifetime</span></a>
<a class="sourceLine" id="cb8-52" title="52"><span class="co">#  Since omega is the adult daily death rate, we use the built-in function to </span></a>
<a class="sourceLine" id="cb8-53" title="53"><span class="co">#  calculate our desired lifetime cost as applied daily</span></a>
<a class="sourceLine" id="cb8-54" title="54">dayOmega &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getOmega.html">getOmega</a></span>(<span class="dt">mu =</span> bioParameters<span class="op">$</span>muAd, <span class="dt">lifespanReduction =</span> <span class="fl">0.60</span>)</a>
<a class="sourceLine" id="cb8-55" title="55">omegaNew &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"AA"</span>=dayOmega, <span class="st">"aa"</span>=dayOmega)</a>
<a class="sourceLine" id="cb8-56" title="56"></a>
<a class="sourceLine" id="cb8-57" title="57"><span class="co"># setup cube</span></a>
<a class="sourceLine" id="cb8-58" title="58">cube &lt;-<span class="st"> </span><span class="kw"><a href="../reference/cubeMendelian.html">cubeMendelian</a></span>(<span class="dt">omega =</span> omegaNew)</a>
<a class="sourceLine" id="cb8-59" title="59"></a>
<a class="sourceLine" id="cb8-60" title="60"></a>
<a class="sourceLine" id="cb8-61" title="61"><span class="co">####################</span></a>
<a class="sourceLine" id="cb8-62" title="62"><span class="co"># Setup releases and batch migration</span></a>
<a class="sourceLine" id="cb8-63" title="63"><span class="co">####################</span></a>
<a class="sourceLine" id="cb8-64" title="64"><span class="co"># set up the empty release vector</span></a>
<a class="sourceLine" id="cb8-65" title="65"><span class="co">#  MGDrivE pulls things out by name</span></a>
<a class="sourceLine" id="cb8-66" title="66">patchReleases &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">replicate</a></span>(<span class="dt">n=</span>sitesNumber,</a>
<a class="sourceLine" id="cb8-67" title="67">                           <span class="dt">expr=</span>{<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">maleReleases=</span><span class="ot">NULL</span>,<span class="dt">femaleReleases=</span><span class="ot">NULL</span>)},</a>
<a class="sourceLine" id="cb8-68" title="68">                           <span class="dt">simplify=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb8-69" title="69"></a>
<a class="sourceLine" id="cb8-70" title="70"><span class="co"># choose release parameters</span></a>
<a class="sourceLine" id="cb8-71" title="71"><span class="co">#  Releases start at time 25, occur every day, for 5 days.</span></a>
<a class="sourceLine" id="cb8-72" title="72"><span class="co">#  There are 50 mosquitoes relaeses every time.</span></a>
<a class="sourceLine" id="cb8-73" title="73">releasesParameters &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">releasesStart=</span><span class="dv">25</span>,</a>
<a class="sourceLine" id="cb8-74" title="74">                           <span class="dt">releasesNumber=</span><span class="dv">5</span>,</a>
<a class="sourceLine" id="cb8-75" title="75">                           <span class="dt">releasesInterval=</span><span class="dv">0</span>,</a>
<a class="sourceLine" id="cb8-76" title="76">                           <span class="dt">releaseProportion=</span><span class="dv">50</span>)</a>
<a class="sourceLine" id="cb8-77" title="77"></a>
<a class="sourceLine" id="cb8-78" title="78"><span class="co"># generate male release vector</span></a>
<a class="sourceLine" id="cb8-79" title="79">maleReleasesVector &lt;-<span class="st"> </span><span class="kw"><a href="../reference/generateReleaseVector.html">generateReleaseVector</a></span>(<span class="dt">driveCube=</span>cube, </a>
<a class="sourceLine" id="cb8-80" title="80">                                            <span class="dt">releasesParameters=</span>releasesParameters,</a>
<a class="sourceLine" id="cb8-81" title="81">                                            <span class="dt">sex=</span><span class="st">"M"</span>)</a>
<a class="sourceLine" id="cb8-82" title="82"></a>
<a class="sourceLine" id="cb8-83" title="83"><span class="co"># generate female release vector</span></a>
<a class="sourceLine" id="cb8-84" title="84">femaleReleasesVector &lt;-<span class="st"> </span><span class="kw"><a href="../reference/generateReleaseVector.html">generateReleaseVector</a></span>(<span class="dt">driveCube=</span>cube,</a>
<a class="sourceLine" id="cb8-85" title="85">                                              <span class="dt">releasesParameters=</span>releasesParameters,</a>
<a class="sourceLine" id="cb8-86" title="86">                                              <span class="dt">sex=</span><span class="st">"F"</span>)</a>
<a class="sourceLine" id="cb8-87" title="87"></a>
<a class="sourceLine" id="cb8-88" title="88"><span class="co"># put releases into the proper place in the release list</span></a>
<a class="sourceLine" id="cb8-89" title="89">patchReleases[[<span class="dv">1</span>]]<span class="op">$</span>maleReleases &lt;-<span class="st"> </span>maleReleasesVector</a>
<a class="sourceLine" id="cb8-90" title="90">patchReleases[[<span class="dv">1</span>]]<span class="op">$</span>femaleReleases &lt;-<span class="st"> </span>femaleReleasesVector</a>
<a class="sourceLine" id="cb8-91" title="91"></a>
<a class="sourceLine" id="cb8-92" title="92"></a>
<a class="sourceLine" id="cb8-93" title="93"><span class="co"># batch migration is disabled by setting the probability to 0</span></a>
<a class="sourceLine" id="cb8-94" title="94"><span class="co"># This is required because of the stochastic simulations, but doesn't make sense </span></a>
<a class="sourceLine" id="cb8-95" title="95"><span class="co">#  in a deterministic simulation.</span></a>
<a class="sourceLine" id="cb8-96" title="96">batchMigration &lt;-<span class="st"> </span><span class="kw"><a href="../reference/basicBatchMigration.html">basicBatchMigration</a></span>(<span class="dt">batchProbs=</span><span class="dv">0</span>,</a>
<a class="sourceLine" id="cb8-97" title="97">                                      <span class="dt">sexProbs=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(.<span class="dv">5</span>,.<span class="dv">5</span>),</a>
<a class="sourceLine" id="cb8-98" title="98">                                      <span class="dt">numPatches=</span>sitesNumber)</a>
<a class="sourceLine" id="cb8-99" title="99"></a>
<a class="sourceLine" id="cb8-100" title="100"></a>
<a class="sourceLine" id="cb8-101" title="101"><span class="co">####################</span></a>
<a class="sourceLine" id="cb8-102" title="102"><span class="co"># Combine parameters and run!</span></a>
<a class="sourceLine" id="cb8-103" title="103"><span class="co">####################</span></a>
<a class="sourceLine" id="cb8-104" title="104"><span class="co"># set MGDrivE to run stochastic</span></a>
<a class="sourceLine" id="cb8-105" title="105"><span class="kw"><a href="../reference/setupMGDrivE.html">setupMGDrivE</a></span>(<span class="dt">stochasticityON =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb8-106" title="106"><span class="co">#&gt; initializing MGDrivE</span></a>
<a class="sourceLine" id="cb8-107" title="107"></a>
<a class="sourceLine" id="cb8-108" title="108"><span class="co"># setup parameters for the network. This builds a list of parameters required for</span></a>
<a class="sourceLine" id="cb8-109" title="109"><span class="co">#  every population in the network. In ths case, we havee a network of 1 population.</span></a>
<a class="sourceLine" id="cb8-110" title="110">netPar &lt;-<span class="st"> </span><span class="kw"><a href="../reference/parameterizeMGDrivE.html">parameterizeMGDrivE</a></span>(<span class="dt">runID=</span><span class="dv">1</span>, <span class="dt">simTime=</span>tMax, <span class="dt">nPatch=</span>sitesNumber,</a>
<a class="sourceLine" id="cb8-111" title="111">                              <span class="dt">beta=</span>bioParameters<span class="op">$</span>betaK, <span class="dt">muAd=</span>bioParameters<span class="op">$</span>muAd,</a>
<a class="sourceLine" id="cb8-112" title="112">                              <span class="dt">popGrowth=</span>bioParameters<span class="op">$</span>popGrowth, <span class="dt">tEgg=</span>bioParameters<span class="op">$</span>tEgg,</a>
<a class="sourceLine" id="cb8-113" title="113">                              <span class="dt">tLarva=</span>bioParameters<span class="op">$</span>tLarva, <span class="dt">tPupa=</span>bioParameters<span class="op">$</span>tPupa,</a>
<a class="sourceLine" id="cb8-114" title="114">                              <span class="dt">AdPopEQ=</span>patchPops)</a>
<a class="sourceLine" id="cb8-115" title="115"></a>
<a class="sourceLine" id="cb8-116" title="116"><span class="co"># build network prior to run</span></a>
<a class="sourceLine" id="cb8-117" title="117">MGDrivESim &lt;-<span class="st"> </span>Network<span class="op">$</span><span class="kw">new</span>(<span class="dt">params=</span>netPar,</a>
<a class="sourceLine" id="cb8-118" title="118">                          <span class="dt">driveCube=</span>cube,</a>
<a class="sourceLine" id="cb8-119" title="119">                          <span class="dt">patchReleases=</span>patchReleases,</a>
<a class="sourceLine" id="cb8-120" title="120">                          <span class="dt">migrationMale=</span>moveMat,</a>
<a class="sourceLine" id="cb8-121" title="121">                          <span class="dt">migrationFemale=</span>moveMat,</a>
<a class="sourceLine" id="cb8-122" title="122">                          <span class="dt">migrationBatch=</span>batchMigration,</a>
<a class="sourceLine" id="cb8-123" title="123">                          <span class="dt">directory=</span>folderNames)</a>
<a class="sourceLine" id="cb8-124" title="124"><span class="co">#&gt; initializing patch:  1  of  1</span></a>
<a class="sourceLine" id="cb8-125" title="125"><span class="co"># run simulation</span></a>
<a class="sourceLine" id="cb8-126" title="126">MGDrivESim<span class="op">$</span><span class="kw">multRun</span>(<span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb8-127" title="127"></a>
<a class="sourceLine" id="cb8-128" title="128"></a>
<a class="sourceLine" id="cb8-129" title="129"><span class="co">####################</span></a>
<a class="sourceLine" id="cb8-130" title="130"><span class="co"># Post Analysis</span></a>
<a class="sourceLine" id="cb8-131" title="131"><span class="co">####################</span></a>
<a class="sourceLine" id="cb8-132" title="132"><span class="co"># First, split output by patch</span></a>
<a class="sourceLine" id="cb8-133" title="133"><span class="co"># Second, aggregate females by their mate choice</span></a>
<a class="sourceLine" id="cb8-134" title="134"><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>nRep){</a>
<a class="sourceLine" id="cb8-135" title="135">  <span class="kw"><a href="../reference/splitOutput.html">splitOutput</a></span>(<span class="dt">readDir=</span>folderNames[i],<span class="dt">remFile =</span> <span class="ot">TRUE</span>,<span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb8-136" title="136">  <span class="kw"><a href="../reference/aggregateFemales.html">aggregateFemales</a></span>(<span class="dt">readDir=</span>folderNames[i],<span class="dt">genotypes=</span>cube<span class="op">$</span>genotypesID,</a>
<a class="sourceLine" id="cb8-137" title="137">                   <span class="dt">remFile=</span><span class="ot">TRUE</span>,<span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb8-138" title="138">}</a>
<a class="sourceLine" id="cb8-139" title="139"></a>
<a class="sourceLine" id="cb8-140" title="140"><span class="co"># plot output of first run to see effect</span></a>
<a class="sourceLine" id="cb8-141" title="141"><span class="kw"><a href="../reference/plotMGDrivESingle.html">plotMGDrivESingle</a></span>(<span class="dt">readDir=</span>folderNames[<span class="dv">1</span>],<span class="dt">totalPop =</span> <span class="ot">TRUE</span>,<span class="dt">lwd=</span><span class="fl">3.5</span>,<span class="dt">alpha=</span><span class="dv">1</span>)</a></code></pre></div>
<p><img src="mgdrive_examples_files/figure-html/unnamed-chunk-14-1.png" width="672"></p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" title="1"></a>
<a class="sourceLine" id="cb9-2" title="2"><span class="co"># plot all 50 repetitions together</span></a>
<a class="sourceLine" id="cb9-3" title="3"><span class="kw"><a href="../reference/plotMGDrivEMult.html">plotMGDrivEMult</a></span>(<span class="dt">readDir=</span>outFolder,<span class="dt">lwd=</span><span class="fl">0.35</span>,<span class="dt">alpha=</span><span class="fl">0.75</span>)</a></code></pre></div>
<p><img src="mgdrive_examples_files/figure-html/unnamed-chunk-14-2.png" width="672"></p>
<p>We first plot a single run of our stochastic simulation, and then plot all 50 runs together. We see that the individual run looks similar to the deterministic run from before, all genotypes behaving heuristically the same. However, the dynamics are not as “clean” as the deterministic version, clearly suffering from the randomness inherent in an actual population, and the dynamics are slower than the deterministic version. This is an inherent effect of stochasticity, something that is important in decision making and with smaller population sizes.</p>
</div>
</div>
<div id="mendelian-inheritance-simulations-two-populations" class="section level3">
<h3 class="hasAnchor">
<a href="#mendelian-inheritance-simulations-two-populations" class="anchor"></a>Mendelian Inheritance Simulations, Two Populations</h3>
<p>Single, panmictic populations are simple to analyze but rarely encountered in nature. Often, populations of identical individuals are separated by geography, be it great distances or a road that is difficult to cross. This creates structures within populations that can drastically alter how genes spread. Here, we explore how the results of our first simulation change when we have a structured population where two halves of the population never mix.</p>
<p><em>MGDrivE</em> handles population structure by considering separate, panmictic populations with some migration rate between them. This makes <em>MGDrivE</em> a meta-population model, where individual populations are part of the same graph, connected via some migration or mixture rate. In the following simulation, we use two nodes, or two well-mixed populations, that have no migration between them. Then, we simulate releases in one population, and analyze what happens.</p>
<div id="deterministic-no-migration" class="section level4">
<h4 class="hasAnchor">
<a href="#deterministic-no-migration" class="anchor"></a>Deterministic, No Migration</h4>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" title="1"><span class="co">####################</span></a>
<a class="sourceLine" id="cb10-2" title="2"><span class="co"># Load libraries</span></a>
<a class="sourceLine" id="cb10-3" title="3"><span class="co">####################</span></a>
<a class="sourceLine" id="cb10-4" title="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(MGDrivE)</a>
<a class="sourceLine" id="cb10-5" title="5"></a>
<a class="sourceLine" id="cb10-6" title="6"><span class="co">####################</span></a>
<a class="sourceLine" id="cb10-7" title="7"><span class="co"># Output Folder</span></a>
<a class="sourceLine" id="cb10-8" title="8"><span class="co">####################</span></a>
<a class="sourceLine" id="cb10-9" title="9">outFolder &lt;-<span class="st"> "mgdrive"</span></a>
<a class="sourceLine" id="cb10-10" title="10"></a>
<a class="sourceLine" id="cb10-11" title="11"><span class="co">####################</span></a>
<a class="sourceLine" id="cb10-12" title="12"><span class="co"># Simulation Parameters</span></a>
<a class="sourceLine" id="cb10-13" title="13"><span class="co">####################</span></a>
<a class="sourceLine" id="cb10-14" title="14"><span class="co"># days to run the simulation</span></a>
<a class="sourceLine" id="cb10-15" title="15">tMax &lt;-<span class="st"> </span><span class="dv">365</span></a>
<a class="sourceLine" id="cb10-16" title="16"></a>
<a class="sourceLine" id="cb10-17" title="17"><span class="co"># entomological parameters</span></a>
<a class="sourceLine" id="cb10-18" title="18">bioParameters &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">betaK=</span><span class="dv">20</span>,<span class="dt">tEgg=</span><span class="dv">5</span>,<span class="dt">tLarva=</span><span class="dv">6</span>,<span class="dt">tPupa=</span><span class="dv">4</span>,<span class="dt">popGrowth=</span><span class="fl">1.175</span>,<span class="dt">muAd=</span><span class="fl">0.09</span>)</a>
<a class="sourceLine" id="cb10-19" title="19"></a>
<a class="sourceLine" id="cb10-20" title="20"><span class="co"># a 2-node network where mosquitoes do not leave</span></a>
<a class="sourceLine" id="cb10-21" title="21">moveMat &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="dt">data =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">1</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">1</span>), <span class="dt">nrow =</span> <span class="dv">2</span>, <span class="dt">ncol =</span> <span class="dv">2</span>)</a>
<a class="sourceLine" id="cb10-22" title="22">moveMat</a>
<a class="sourceLine" id="cb10-23" title="23"><span class="co">#&gt;      [,1] [,2]</span></a>
<a class="sourceLine" id="cb10-24" title="24"><span class="co">#&gt; [1,]    1    0</span></a>
<a class="sourceLine" id="cb10-25" title="25"><span class="co">#&gt; [2,]    0    1</span></a>
<a class="sourceLine" id="cb10-26" title="26"></a>
<a class="sourceLine" id="cb10-27" title="27"><span class="co"># parameters of the population equilibrium</span></a>
<a class="sourceLine" id="cb10-28" title="28">adultPopEquilibrium &lt;-<span class="st"> </span><span class="dv">500</span></a>
<a class="sourceLine" id="cb10-29" title="29">sitesNumber &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">nrow</a></span>(moveMat)</a>
<a class="sourceLine" id="cb10-30" title="30">patchPops &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>(adultPopEquilibrium,sitesNumber)</a>
<a class="sourceLine" id="cb10-31" title="31"></a>
<a class="sourceLine" id="cb10-32" title="32"><span class="co">####################</span></a>
<a class="sourceLine" id="cb10-33" title="33"><span class="co"># Basic Inheritance pattern</span></a>
<a class="sourceLine" id="cb10-34" title="34"><span class="co">####################</span></a>
<a class="sourceLine" id="cb10-35" title="35"><span class="co"># Mendelian cube with standard (default) paraameters</span></a>
<a class="sourceLine" id="cb10-36" title="36">cube &lt;-<span class="st"> </span><span class="kw"><a href="../reference/cubeMendelian.html">cubeMendelian</a></span>()</a>
<a class="sourceLine" id="cb10-37" title="37"></a>
<a class="sourceLine" id="cb10-38" title="38"></a>
<a class="sourceLine" id="cb10-39" title="39"><span class="co">####################</span></a>
<a class="sourceLine" id="cb10-40" title="40"><span class="co"># Setup releases and batch migration</span></a>
<a class="sourceLine" id="cb10-41" title="41"><span class="co">####################</span></a>
<a class="sourceLine" id="cb10-42" title="42"><span class="co"># set up the empty release vector</span></a>
<a class="sourceLine" id="cb10-43" title="43"><span class="co">#  MGDrivE pulls things out by name</span></a>
<a class="sourceLine" id="cb10-44" title="44">patchReleases &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">replicate</a></span>(<span class="dt">n=</span>sitesNumber,</a>
<a class="sourceLine" id="cb10-45" title="45">                           <span class="dt">expr=</span>{<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">maleReleases=</span><span class="ot">NULL</span>,<span class="dt">femaleReleases=</span><span class="ot">NULL</span>)},</a>
<a class="sourceLine" id="cb10-46" title="46">                           <span class="dt">simplify=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb10-47" title="47"></a>
<a class="sourceLine" id="cb10-48" title="48"><span class="co"># choose release parameters</span></a>
<a class="sourceLine" id="cb10-49" title="49"><span class="co">#  Releases start at time 25, occur every day, for 5 day.</span></a>
<a class="sourceLine" id="cb10-50" title="50"><span class="co">#  There are 50 mosquitoes relaeses every time.</span></a>
<a class="sourceLine" id="cb10-51" title="51">releasesParameters &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">releasesStart=</span><span class="dv">25</span>,</a>
<a class="sourceLine" id="cb10-52" title="52">                           <span class="dt">releasesNumber=</span><span class="dv">5</span>,</a>
<a class="sourceLine" id="cb10-53" title="53">                           <span class="dt">releasesInterval=</span><span class="dv">0</span>,</a>
<a class="sourceLine" id="cb10-54" title="54">                           <span class="dt">releaseProportion=</span><span class="dv">50</span>)</a>
<a class="sourceLine" id="cb10-55" title="55"></a>
<a class="sourceLine" id="cb10-56" title="56"><span class="co"># generate male release vector</span></a>
<a class="sourceLine" id="cb10-57" title="57">maleReleasesVector &lt;-<span class="st"> </span><span class="kw"><a href="../reference/generateReleaseVector.html">generateReleaseVector</a></span>(<span class="dt">driveCube=</span>cube, </a>
<a class="sourceLine" id="cb10-58" title="58">                                            <span class="dt">releasesParameters=</span>releasesParameters,</a>
<a class="sourceLine" id="cb10-59" title="59">                                            <span class="dt">sex=</span><span class="st">"M"</span>)</a>
<a class="sourceLine" id="cb10-60" title="60"></a>
<a class="sourceLine" id="cb10-61" title="61"><span class="co"># generate female release vector</span></a>
<a class="sourceLine" id="cb10-62" title="62">femaleReleasesVector &lt;-<span class="st"> </span><span class="kw"><a href="../reference/generateReleaseVector.html">generateReleaseVector</a></span>(<span class="dt">driveCube=</span>cube,</a>
<a class="sourceLine" id="cb10-63" title="63">                                              <span class="dt">releasesParameters=</span>releasesParameters,</a>
<a class="sourceLine" id="cb10-64" title="64">                                              <span class="dt">sex=</span><span class="st">"F"</span>)</a>
<a class="sourceLine" id="cb10-65" title="65"></a>
<a class="sourceLine" id="cb10-66" title="66"><span class="co"># put releases into the proper place in the release list</span></a>
<a class="sourceLine" id="cb10-67" title="67">patchReleases[[<span class="dv">1</span>]]<span class="op">$</span>maleReleases &lt;-<span class="st"> </span>maleReleasesVector</a>
<a class="sourceLine" id="cb10-68" title="68">patchReleases[[<span class="dv">1</span>]]<span class="op">$</span>femaleReleases &lt;-<span class="st"> </span>femaleReleasesVector</a>
<a class="sourceLine" id="cb10-69" title="69"></a>
<a class="sourceLine" id="cb10-70" title="70"></a>
<a class="sourceLine" id="cb10-71" title="71"><span class="co"># batch migration is disabled by setting the probability to 0</span></a>
<a class="sourceLine" id="cb10-72" title="72"><span class="co"># This is required because of the stochastic simulations, but doesn't make sense </span></a>
<a class="sourceLine" id="cb10-73" title="73"><span class="co">#  in a deterministic simulation.</span></a>
<a class="sourceLine" id="cb10-74" title="74">batchMigration &lt;-<span class="st"> </span><span class="kw"><a href="../reference/basicBatchMigration.html">basicBatchMigration</a></span>(<span class="dt">batchProbs=</span><span class="dv">0</span>,</a>
<a class="sourceLine" id="cb10-75" title="75">                                      <span class="dt">sexProbs=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(.<span class="dv">5</span>,.<span class="dv">5</span>),</a>
<a class="sourceLine" id="cb10-76" title="76">                                      <span class="dt">numPatches=</span>sitesNumber)</a>
<a class="sourceLine" id="cb10-77" title="77"></a>
<a class="sourceLine" id="cb10-78" title="78"></a>
<a class="sourceLine" id="cb10-79" title="79"><span class="co">####################</span></a>
<a class="sourceLine" id="cb10-80" title="80"><span class="co"># Combine parameters and run!</span></a>
<a class="sourceLine" id="cb10-81" title="81"><span class="co">####################</span></a>
<a class="sourceLine" id="cb10-82" title="82"><span class="co"># set MGDrivE to run deterministic</span></a>
<a class="sourceLine" id="cb10-83" title="83"><span class="kw"><a href="../reference/setupMGDrivE.html">setupMGDrivE</a></span>(<span class="dt">stochasticityON =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb10-84" title="84"><span class="co">#&gt; initializing MGDrivE</span></a>
<a class="sourceLine" id="cb10-85" title="85"></a>
<a class="sourceLine" id="cb10-86" title="86"><span class="co"># setup parameters for the network. This builds a list of parameters required for</span></a>
<a class="sourceLine" id="cb10-87" title="87"><span class="co">#  every population in the network. In ths case, we havee a network of 1 population.</span></a>
<a class="sourceLine" id="cb10-88" title="88">netPar &lt;-<span class="st"> </span><span class="kw"><a href="../reference/parameterizeMGDrivE.html">parameterizeMGDrivE</a></span>(<span class="dt">runID=</span><span class="dv">1</span>, <span class="dt">simTime=</span>tMax, <span class="dt">nPatch=</span>sitesNumber,</a>
<a class="sourceLine" id="cb10-89" title="89">                              <span class="dt">beta=</span>bioParameters<span class="op">$</span>betaK, <span class="dt">muAd=</span>bioParameters<span class="op">$</span>muAd,</a>
<a class="sourceLine" id="cb10-90" title="90">                              <span class="dt">popGrowth=</span>bioParameters<span class="op">$</span>popGrowth, <span class="dt">tEgg=</span>bioParameters<span class="op">$</span>tEgg,</a>
<a class="sourceLine" id="cb10-91" title="91">                              <span class="dt">tLarva=</span>bioParameters<span class="op">$</span>tLarva, <span class="dt">tPupa=</span>bioParameters<span class="op">$</span>tPupa,</a>
<a class="sourceLine" id="cb10-92" title="92">                              <span class="dt">AdPopEQ=</span>patchPops)</a>
<a class="sourceLine" id="cb10-93" title="93"></a>
<a class="sourceLine" id="cb10-94" title="94"><span class="co"># build network prior to run</span></a>
<a class="sourceLine" id="cb10-95" title="95">MGDrivESim &lt;-<span class="st"> </span>Network<span class="op">$</span><span class="kw">new</span>(<span class="dt">params=</span>netPar,</a>
<a class="sourceLine" id="cb10-96" title="96">                          <span class="dt">driveCube=</span>cube,</a>
<a class="sourceLine" id="cb10-97" title="97">                          <span class="dt">patchReleases=</span>patchReleases,</a>
<a class="sourceLine" id="cb10-98" title="98">                          <span class="dt">migrationMale=</span>moveMat,</a>
<a class="sourceLine" id="cb10-99" title="99">                          <span class="dt">migrationFemale=</span>moveMat,</a>
<a class="sourceLine" id="cb10-100" title="100">                          <span class="dt">migrationBatch=</span>batchMigration,</a>
<a class="sourceLine" id="cb10-101" title="101">                          <span class="dt">directory=</span>outFolder)</a>
<a class="sourceLine" id="cb10-102" title="102"><span class="co">#&gt; initializing patch:  1  of  2 </span></a>
<a class="sourceLine" id="cb10-103" title="103"><span class="co">#&gt; initializing patch:  2  of  2</span></a>
<a class="sourceLine" id="cb10-104" title="104"><span class="co"># run simulation</span></a>
<a class="sourceLine" id="cb10-105" title="105">MGDrivESim<span class="op">$</span><span class="kw">oneRun</span>(<span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb10-106" title="106"></a>
<a class="sourceLine" id="cb10-107" title="107"></a>
<a class="sourceLine" id="cb10-108" title="108"><span class="co">####################</span></a>
<a class="sourceLine" id="cb10-109" title="109"><span class="co"># Post Analysis</span></a>
<a class="sourceLine" id="cb10-110" title="110"><span class="co">####################</span></a>
<a class="sourceLine" id="cb10-111" title="111"><span class="co"># split output by patch</span></a>
<a class="sourceLine" id="cb10-112" title="112"><span class="co">#  Required for plotting later</span></a>
<a class="sourceLine" id="cb10-113" title="113"><span class="kw"><a href="../reference/splitOutput.html">splitOutput</a></span>(<span class="dt">readDir=</span>outFolder, <span class="dt">verbose =</span> <span class="ot">FALSE</span>, <span class="dt">remFile =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb10-114" title="114"></a>
<a class="sourceLine" id="cb10-115" title="115"><span class="co"># aggregate females by their mate choice</span></a>
<a class="sourceLine" id="cb10-116" title="116"><span class="co">#  This reduces the female file to have the same columns as the male file</span></a>
<a class="sourceLine" id="cb10-117" title="117"><span class="kw"><a href="../reference/aggregateFemales.html">aggregateFemales</a></span>(<span class="dt">readDir=</span>outFolder,<span class="dt">genotypes=</span>cube<span class="op">$</span>genotypesID,</a>
<a class="sourceLine" id="cb10-118" title="118">                 <span class="dt">remFile=</span><span class="ot">TRUE</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb10-119" title="119"></a>
<a class="sourceLine" id="cb10-120" title="120"><span class="co"># plot output to see effect</span></a>
<a class="sourceLine" id="cb10-121" title="121"><span class="kw"><a href="../reference/plotMGDrivESingle.html">plotMGDrivESingle</a></span>(<span class="dt">readDir=</span>outFolder,<span class="dt">lwd=</span><span class="fl">3.5</span>,<span class="dt">alpha=</span><span class="dv">1</span>)</a></code></pre></div>
<p><img src="mgdrive_examples_files/figure-html/unnamed-chunk-16-1.png" width="672"></p>
<p>The plots are now expanded to include both populations in our simulation, as labeled on the right side of the plot. We performed 5 releases of 50 <em>aa</em> individuals in the first node. However, since there is no migration between the populations, we see <em>Aa</em> individuals appear in patch 1 only, while patch 2 remains completely <em>AA</em>.</p>
</div>
<div id="deterministic-small-migration" class="section level4">
<h4 class="hasAnchor">
<a href="#deterministic-small-migration" class="anchor"></a>Deterministic, Small Migration</h4>
<p>Now, if we increase the migration to 1% per day between the populations, we should see heterozygous individuals appear in patch 2, even though we only perform releases in patch 1. (We repeat the entire code for users’ benefit.)</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" title="1"><span class="co">####################</span></a>
<a class="sourceLine" id="cb11-2" title="2"><span class="co"># Load libraries</span></a>
<a class="sourceLine" id="cb11-3" title="3"><span class="co">####################</span></a>
<a class="sourceLine" id="cb11-4" title="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(MGDrivE)</a>
<a class="sourceLine" id="cb11-5" title="5"></a>
<a class="sourceLine" id="cb11-6" title="6"><span class="co">####################</span></a>
<a class="sourceLine" id="cb11-7" title="7"><span class="co"># Output Folder</span></a>
<a class="sourceLine" id="cb11-8" title="8"><span class="co">####################</span></a>
<a class="sourceLine" id="cb11-9" title="9">outFolder &lt;-<span class="st"> "mgdrive"</span></a>
<a class="sourceLine" id="cb11-10" title="10"></a>
<a class="sourceLine" id="cb11-11" title="11"><span class="co">####################</span></a>
<a class="sourceLine" id="cb11-12" title="12"><span class="co"># Simulation Parameters</span></a>
<a class="sourceLine" id="cb11-13" title="13"><span class="co">####################</span></a>
<a class="sourceLine" id="cb11-14" title="14"><span class="co"># days to run the simulation</span></a>
<a class="sourceLine" id="cb11-15" title="15">tMax &lt;-<span class="st"> </span><span class="dv">365</span></a>
<a class="sourceLine" id="cb11-16" title="16"></a>
<a class="sourceLine" id="cb11-17" title="17"><span class="co"># entomological parameters</span></a>
<a class="sourceLine" id="cb11-18" title="18">bioParameters &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">betaK=</span><span class="dv">20</span>,<span class="dt">tEgg=</span><span class="dv">5</span>,<span class="dt">tLarva=</span><span class="dv">6</span>,<span class="dt">tPupa=</span><span class="dv">4</span>,<span class="dt">popGrowth=</span><span class="fl">1.175</span>,<span class="dt">muAd=</span><span class="fl">0.09</span>)</a>
<a class="sourceLine" id="cb11-19" title="19"></a>
<a class="sourceLine" id="cb11-20" title="20"><span class="co"># a 2-node network with 1% per day migration rate</span></a>
<a class="sourceLine" id="cb11-21" title="21">moveMat &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="dt">data =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="fl">0.99</span>,<span class="fl">0.01</span>,<span class="fl">0.01</span>,<span class="fl">0.99</span>), <span class="dt">nrow =</span> <span class="dv">2</span>, <span class="dt">ncol =</span> <span class="dv">2</span>)</a>
<a class="sourceLine" id="cb11-22" title="22">moveMat</a>
<a class="sourceLine" id="cb11-23" title="23"><span class="co">#&gt;      [,1] [,2]</span></a>
<a class="sourceLine" id="cb11-24" title="24"><span class="co">#&gt; [1,] 0.99 0.01</span></a>
<a class="sourceLine" id="cb11-25" title="25"><span class="co">#&gt; [2,] 0.01 0.99</span></a>
<a class="sourceLine" id="cb11-26" title="26"></a>
<a class="sourceLine" id="cb11-27" title="27"><span class="co"># parameters of the population equilibrium</span></a>
<a class="sourceLine" id="cb11-28" title="28">adultPopEquilibrium &lt;-<span class="st"> </span><span class="dv">500</span></a>
<a class="sourceLine" id="cb11-29" title="29">sitesNumber &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">nrow</a></span>(moveMat)</a>
<a class="sourceLine" id="cb11-30" title="30">patchPops &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>(adultPopEquilibrium,sitesNumber)</a>
<a class="sourceLine" id="cb11-31" title="31"></a>
<a class="sourceLine" id="cb11-32" title="32"><span class="co">####################</span></a>
<a class="sourceLine" id="cb11-33" title="33"><span class="co"># Basic Inheritance pattern</span></a>
<a class="sourceLine" id="cb11-34" title="34"><span class="co">####################</span></a>
<a class="sourceLine" id="cb11-35" title="35"><span class="co"># Mendelian cube with standard (default) paraameters</span></a>
<a class="sourceLine" id="cb11-36" title="36">cube &lt;-<span class="st"> </span><span class="kw"><a href="../reference/cubeMendelian.html">cubeMendelian</a></span>()</a>
<a class="sourceLine" id="cb11-37" title="37"></a>
<a class="sourceLine" id="cb11-38" title="38"></a>
<a class="sourceLine" id="cb11-39" title="39"><span class="co">####################</span></a>
<a class="sourceLine" id="cb11-40" title="40"><span class="co"># Setup releases and batch migration</span></a>
<a class="sourceLine" id="cb11-41" title="41"><span class="co">####################</span></a>
<a class="sourceLine" id="cb11-42" title="42"><span class="co"># set up the empty release vector</span></a>
<a class="sourceLine" id="cb11-43" title="43"><span class="co">#  MGDrivE pulls things out by name</span></a>
<a class="sourceLine" id="cb11-44" title="44">patchReleases &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">replicate</a></span>(<span class="dt">n=</span>sitesNumber,</a>
<a class="sourceLine" id="cb11-45" title="45">                           <span class="dt">expr=</span>{<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">maleReleases=</span><span class="ot">NULL</span>,<span class="dt">femaleReleases=</span><span class="ot">NULL</span>)},</a>
<a class="sourceLine" id="cb11-46" title="46">                           <span class="dt">simplify=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb11-47" title="47"></a>
<a class="sourceLine" id="cb11-48" title="48"><span class="co"># choose release parameters</span></a>
<a class="sourceLine" id="cb11-49" title="49"><span class="co">#  Releases start at time 25, occur every day, for 5 day.</span></a>
<a class="sourceLine" id="cb11-50" title="50"><span class="co">#  There are 50 mosquitoes relaeses every time.</span></a>
<a class="sourceLine" id="cb11-51" title="51">releasesParameters &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">releasesStart=</span><span class="dv">25</span>,</a>
<a class="sourceLine" id="cb11-52" title="52">                           <span class="dt">releasesNumber=</span><span class="dv">5</span>,</a>
<a class="sourceLine" id="cb11-53" title="53">                           <span class="dt">releasesInterval=</span><span class="dv">0</span>,</a>
<a class="sourceLine" id="cb11-54" title="54">                           <span class="dt">releaseProportion=</span><span class="dv">50</span>)</a>
<a class="sourceLine" id="cb11-55" title="55"></a>
<a class="sourceLine" id="cb11-56" title="56"><span class="co"># generate male release vector</span></a>
<a class="sourceLine" id="cb11-57" title="57">maleReleasesVector &lt;-<span class="st"> </span><span class="kw"><a href="../reference/generateReleaseVector.html">generateReleaseVector</a></span>(<span class="dt">driveCube=</span>cube, </a>
<a class="sourceLine" id="cb11-58" title="58">                                            <span class="dt">releasesParameters=</span>releasesParameters,</a>
<a class="sourceLine" id="cb11-59" title="59">                                            <span class="dt">sex=</span><span class="st">"M"</span>)</a>
<a class="sourceLine" id="cb11-60" title="60"></a>
<a class="sourceLine" id="cb11-61" title="61"><span class="co"># generate female release vector</span></a>
<a class="sourceLine" id="cb11-62" title="62">femaleReleasesVector &lt;-<span class="st"> </span><span class="kw"><a href="../reference/generateReleaseVector.html">generateReleaseVector</a></span>(<span class="dt">driveCube=</span>cube,</a>
<a class="sourceLine" id="cb11-63" title="63">                                              <span class="dt">releasesParameters=</span>releasesParameters,</a>
<a class="sourceLine" id="cb11-64" title="64">                                              <span class="dt">sex=</span><span class="st">"F"</span>)</a>
<a class="sourceLine" id="cb11-65" title="65"></a>
<a class="sourceLine" id="cb11-66" title="66"><span class="co"># put releases into the proper place in the release list</span></a>
<a class="sourceLine" id="cb11-67" title="67">patchReleases[[<span class="dv">1</span>]]<span class="op">$</span>maleReleases &lt;-<span class="st"> </span>maleReleasesVector</a>
<a class="sourceLine" id="cb11-68" title="68">patchReleases[[<span class="dv">1</span>]]<span class="op">$</span>femaleReleases &lt;-<span class="st"> </span>femaleReleasesVector</a>
<a class="sourceLine" id="cb11-69" title="69"></a>
<a class="sourceLine" id="cb11-70" title="70"></a>
<a class="sourceLine" id="cb11-71" title="71"><span class="co"># batch migration is disabled by setting the probability to 0</span></a>
<a class="sourceLine" id="cb11-72" title="72"><span class="co"># This is required because of the stochastic simulations, but doesn't make sense </span></a>
<a class="sourceLine" id="cb11-73" title="73"><span class="co">#  in a deterministic simulation.</span></a>
<a class="sourceLine" id="cb11-74" title="74">batchMigration &lt;-<span class="st"> </span><span class="kw"><a href="../reference/basicBatchMigration.html">basicBatchMigration</a></span>(<span class="dt">batchProbs=</span><span class="dv">0</span>,</a>
<a class="sourceLine" id="cb11-75" title="75">                                     <span class="dt">sexProbs=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(.<span class="dv">5</span>,.<span class="dv">5</span>),</a>
<a class="sourceLine" id="cb11-76" title="76">                                     <span class="dt">numPatches=</span>sitesNumber)</a>
<a class="sourceLine" id="cb11-77" title="77"></a>
<a class="sourceLine" id="cb11-78" title="78"></a>
<a class="sourceLine" id="cb11-79" title="79"><span class="co">####################</span></a>
<a class="sourceLine" id="cb11-80" title="80"><span class="co"># Combine parameters and run!</span></a>
<a class="sourceLine" id="cb11-81" title="81"><span class="co">####################</span></a>
<a class="sourceLine" id="cb11-82" title="82"><span class="co"># set MGDrivE to run deterministic</span></a>
<a class="sourceLine" id="cb11-83" title="83"><span class="kw"><a href="../reference/setupMGDrivE.html">setupMGDrivE</a></span>(<span class="dt">stochasticityON =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb11-84" title="84"><span class="co">#&gt; initializing MGDrivE</span></a>
<a class="sourceLine" id="cb11-85" title="85"></a>
<a class="sourceLine" id="cb11-86" title="86"><span class="co"># setup parameters for the network. This builds a list of parameters required for</span></a>
<a class="sourceLine" id="cb11-87" title="87"><span class="co">#  every population in the network. In ths case, we havee a network of 1 population.</span></a>
<a class="sourceLine" id="cb11-88" title="88">netPar &lt;-<span class="st"> </span><span class="kw"><a href="../reference/parameterizeMGDrivE.html">parameterizeMGDrivE</a></span>(<span class="dt">runID=</span><span class="dv">1</span>, <span class="dt">simTime=</span>tMax, <span class="dt">nPatch=</span>sitesNumber,</a>
<a class="sourceLine" id="cb11-89" title="89">                             <span class="dt">beta=</span>bioParameters<span class="op">$</span>betaK, <span class="dt">muAd=</span>bioParameters<span class="op">$</span>muAd,</a>
<a class="sourceLine" id="cb11-90" title="90">                             <span class="dt">popGrowth=</span>bioParameters<span class="op">$</span>popGrowth, <span class="dt">tEgg=</span>bioParameters<span class="op">$</span>tEgg,</a>
<a class="sourceLine" id="cb11-91" title="91">                             <span class="dt">tLarva=</span>bioParameters<span class="op">$</span>tLarva, <span class="dt">tPupa=</span>bioParameters<span class="op">$</span>tPupa,</a>
<a class="sourceLine" id="cb11-92" title="92">                             <span class="dt">AdPopEQ=</span>patchPops)</a>
<a class="sourceLine" id="cb11-93" title="93"></a>
<a class="sourceLine" id="cb11-94" title="94"><span class="co"># build network prior to run</span></a>
<a class="sourceLine" id="cb11-95" title="95">MGDrivESim &lt;-<span class="st"> </span>Network<span class="op">$</span><span class="kw">new</span>(<span class="dt">params=</span>netPar,</a>
<a class="sourceLine" id="cb11-96" title="96">                         <span class="dt">driveCube=</span>cube,</a>
<a class="sourceLine" id="cb11-97" title="97">                         <span class="dt">patchReleases=</span>patchReleases,</a>
<a class="sourceLine" id="cb11-98" title="98">                         <span class="dt">migrationMale=</span>moveMat,</a>
<a class="sourceLine" id="cb11-99" title="99">                         <span class="dt">migrationFemale=</span>moveMat,</a>
<a class="sourceLine" id="cb11-100" title="100">                         <span class="dt">migrationBatch=</span>batchMigration,</a>
<a class="sourceLine" id="cb11-101" title="101">                         <span class="dt">directory=</span>outFolder)</a>
<a class="sourceLine" id="cb11-102" title="102"><span class="co">#&gt; initializing patch:  1  of  2 </span></a>
<a class="sourceLine" id="cb11-103" title="103"><span class="co">#&gt; initializing patch:  2  of  2</span></a>
<a class="sourceLine" id="cb11-104" title="104"><span class="co"># run simulation</span></a>
<a class="sourceLine" id="cb11-105" title="105">MGDrivESim<span class="op">$</span><span class="kw">oneRun</span>(<span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb11-106" title="106"></a>
<a class="sourceLine" id="cb11-107" title="107"></a>
<a class="sourceLine" id="cb11-108" title="108"><span class="co">####################</span></a>
<a class="sourceLine" id="cb11-109" title="109"><span class="co"># Post Analysis</span></a>
<a class="sourceLine" id="cb11-110" title="110"><span class="co">####################</span></a>
<a class="sourceLine" id="cb11-111" title="111"><span class="co"># split output by patch</span></a>
<a class="sourceLine" id="cb11-112" title="112"><span class="co">#  Required for plotting later</span></a>
<a class="sourceLine" id="cb11-113" title="113"><span class="kw"><a href="../reference/splitOutput.html">splitOutput</a></span>(<span class="dt">readDir=</span>outFolder, <span class="dt">verbose =</span> <span class="ot">FALSE</span>, <span class="dt">remFile =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb11-114" title="114"></a>
<a class="sourceLine" id="cb11-115" title="115"><span class="co"># aggregate females by their mate choice</span></a>
<a class="sourceLine" id="cb11-116" title="116"><span class="co">#  This reduces the female file to have the same columns as the male file</span></a>
<a class="sourceLine" id="cb11-117" title="117"><span class="kw"><a href="../reference/aggregateFemales.html">aggregateFemales</a></span>(<span class="dt">readDir=</span>outFolder,<span class="dt">genotypes=</span>cube<span class="op">$</span>genotypesID,</a>
<a class="sourceLine" id="cb11-118" title="118">                <span class="dt">remFile=</span><span class="ot">TRUE</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb11-119" title="119"></a>
<a class="sourceLine" id="cb11-120" title="120"><span class="co"># plot output to see effect</span></a>
<a class="sourceLine" id="cb11-121" title="121"><span class="kw"><a href="../reference/plotMGDrivESingle.html">plotMGDrivESingle</a></span>(<span class="dt">readDir=</span>outFolder,<span class="dt">totalPop =</span> <span class="ot">TRUE</span>,<span class="dt">lwd=</span><span class="fl">3.5</span>,<span class="dt">alpha=</span><span class="dv">1</span>)</a></code></pre></div>
<p><img src="mgdrive_examples_files/figure-html/unnamed-chunk-18-1.png" width="672"></p>
<p>As expected, with a 1% daily migration rate, we see heterozygotes in patch 2 shortly after they begin emerging in patch 1.</p>
</div>
<div id="stochastic-small-migration" class="section level4">
<h4 class="hasAnchor">
<a href="#stochastic-small-migration" class="anchor"></a>Stochastic, Small Migration</h4>
<p>With very small migration rates, it is interesting to see how a scenario changes using a stochastic simulation. Some repetitions should take significantly longer for heterozygotes to migrate from patch 1 to patch 2. So, we keep all of the parameters the same as above, but run 25 stochastic repetitions instead of a deterministic simulation.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" title="1"><span class="co">####################</span></a>
<a class="sourceLine" id="cb12-2" title="2"><span class="co"># Load libraries</span></a>
<a class="sourceLine" id="cb12-3" title="3"><span class="co">####################</span></a>
<a class="sourceLine" id="cb12-4" title="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(MGDrivE)</a>
<a class="sourceLine" id="cb12-5" title="5"></a>
<a class="sourceLine" id="cb12-6" title="6"><span class="co">####################</span></a>
<a class="sourceLine" id="cb12-7" title="7"><span class="co"># Output Folder</span></a>
<a class="sourceLine" id="cb12-8" title="8"><span class="co">####################</span></a>
<a class="sourceLine" id="cb12-9" title="9"><span class="co"># This is slightly obtuse for vignette building reasons</span></a>
<a class="sourceLine" id="cb12-10" title="10"><span class="co">#  Really, all you need is a base directory, then the repetitions in folders inside that.</span></a>
<a class="sourceLine" id="cb12-11" title="11"><span class="co">#  Here, our base directory is "mgdrive", and the repetition folders are "001","002", etc.</span></a>
<a class="sourceLine" id="cb12-12" title="12"><span class="co">#  So, the final structure is "mgdrive/001","mgdrive/002", etc. </span></a>
<a class="sourceLine" id="cb12-13" title="13">outFolder &lt;-<span class="st"> "mgdrive"</span></a>
<a class="sourceLine" id="cb12-14" title="14"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/files2">dir.create</a></span>(<span class="dt">path =</span> outFolder)</a>
<a class="sourceLine" id="cb12-15" title="15"></a>
<a class="sourceLine" id="cb12-16" title="16"><span class="co">####################</span></a>
<a class="sourceLine" id="cb12-17" title="17"><span class="co"># Simulation Parameters</span></a>
<a class="sourceLine" id="cb12-18" title="18"><span class="co">####################</span></a>
<a class="sourceLine" id="cb12-19" title="19"><span class="co"># days to run the simulation</span></a>
<a class="sourceLine" id="cb12-20" title="20">tMax &lt;-<span class="st"> </span><span class="dv">365</span></a>
<a class="sourceLine" id="cb12-21" title="21"></a>
<a class="sourceLine" id="cb12-22" title="22"><span class="co"># number of Monte Carlo iterations</span></a>
<a class="sourceLine" id="cb12-23" title="23">nRep &lt;-<span class="st"> </span><span class="dv">25</span></a>
<a class="sourceLine" id="cb12-24" title="24"></a>
<a class="sourceLine" id="cb12-25" title="25"><span class="co"># each Monte Carlo iteration gets its own folder</span></a>
<a class="sourceLine" id="cb12-26" title="26">folderNames &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/file.path">file.path</a></span>(outFolder,</a>
<a class="sourceLine" id="cb12-27" title="27">                        <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/formatc">formatC</a></span>(<span class="dt">x =</span> <span class="dv">1</span><span class="op">:</span>nRep, <span class="dt">width =</span> <span class="dv">3</span>, <span class="dt">format =</span> <span class="st">"d"</span>, <span class="dt">flag =</span> <span class="st">"0"</span>))</a>
<a class="sourceLine" id="cb12-28" title="28"></a>
<a class="sourceLine" id="cb12-29" title="29"><span class="co"># entomological parameters</span></a>
<a class="sourceLine" id="cb12-30" title="30">bioParameters &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">betaK=</span><span class="dv">20</span>,<span class="dt">tEgg=</span><span class="dv">5</span>,<span class="dt">tLarva=</span><span class="dv">6</span>,<span class="dt">tPupa=</span><span class="dv">4</span>,<span class="dt">popGrowth=</span><span class="fl">1.175</span>,<span class="dt">muAd=</span><span class="fl">0.09</span>)</a>
<a class="sourceLine" id="cb12-31" title="31"></a>
<a class="sourceLine" id="cb12-32" title="32"><span class="co"># a 2-node network with 1% per day migration rate</span></a>
<a class="sourceLine" id="cb12-33" title="33">moveMat &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="dt">data =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="fl">0.99</span>,<span class="fl">0.01</span>,<span class="fl">0.01</span>,<span class="fl">0.99</span>), <span class="dt">nrow =</span> <span class="dv">2</span>, <span class="dt">ncol =</span> <span class="dv">2</span>)</a>
<a class="sourceLine" id="cb12-34" title="34">moveMat</a>
<a class="sourceLine" id="cb12-35" title="35"><span class="co">#&gt;      [,1] [,2]</span></a>
<a class="sourceLine" id="cb12-36" title="36"><span class="co">#&gt; [1,] 0.99 0.01</span></a>
<a class="sourceLine" id="cb12-37" title="37"><span class="co">#&gt; [2,] 0.01 0.99</span></a>
<a class="sourceLine" id="cb12-38" title="38"></a>
<a class="sourceLine" id="cb12-39" title="39"><span class="co"># parameters of the population equilibrium</span></a>
<a class="sourceLine" id="cb12-40" title="40">adultPopEquilibrium &lt;-<span class="st"> </span><span class="dv">500</span></a>
<a class="sourceLine" id="cb12-41" title="41">sitesNumber &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">nrow</a></span>(moveMat)</a>
<a class="sourceLine" id="cb12-42" title="42">patchPops &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>(adultPopEquilibrium,sitesNumber)</a>
<a class="sourceLine" id="cb12-43" title="43"></a>
<a class="sourceLine" id="cb12-44" title="44"><span class="co">####################</span></a>
<a class="sourceLine" id="cb12-45" title="45"><span class="co"># Basic Inheritance pattern</span></a>
<a class="sourceLine" id="cb12-46" title="46"><span class="co">####################</span></a>
<a class="sourceLine" id="cb12-47" title="47"><span class="co"># Mendelian cube with standard (default) paraameters</span></a>
<a class="sourceLine" id="cb12-48" title="48">cube &lt;-<span class="st"> </span><span class="kw"><a href="../reference/cubeMendelian.html">cubeMendelian</a></span>()</a>
<a class="sourceLine" id="cb12-49" title="49"></a>
<a class="sourceLine" id="cb12-50" title="50"></a>
<a class="sourceLine" id="cb12-51" title="51"><span class="co">####################</span></a>
<a class="sourceLine" id="cb12-52" title="52"><span class="co"># Setup releases and batch migration</span></a>
<a class="sourceLine" id="cb12-53" title="53"><span class="co">####################</span></a>
<a class="sourceLine" id="cb12-54" title="54"><span class="co"># set up the empty release vector</span></a>
<a class="sourceLine" id="cb12-55" title="55"><span class="co">#  MGDrivE pulls things out by name</span></a>
<a class="sourceLine" id="cb12-56" title="56">patchReleases &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">replicate</a></span>(<span class="dt">n=</span>sitesNumber,</a>
<a class="sourceLine" id="cb12-57" title="57">                          <span class="dt">expr=</span>{<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">maleReleases=</span><span class="ot">NULL</span>,<span class="dt">femaleReleases=</span><span class="ot">NULL</span>)},</a>
<a class="sourceLine" id="cb12-58" title="58">                          <span class="dt">simplify=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb12-59" title="59"></a>
<a class="sourceLine" id="cb12-60" title="60"><span class="co"># choose release parameters</span></a>
<a class="sourceLine" id="cb12-61" title="61"><span class="co">#  Releases start at time 25, occur every day, for 5 day.</span></a>
<a class="sourceLine" id="cb12-62" title="62"><span class="co">#  There are 50 mosquitoes relaeses every time.</span></a>
<a class="sourceLine" id="cb12-63" title="63">releasesParameters &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">releasesStart=</span><span class="dv">25</span>,</a>
<a class="sourceLine" id="cb12-64" title="64">                          <span class="dt">releasesNumber=</span><span class="dv">5</span>,</a>
<a class="sourceLine" id="cb12-65" title="65">                          <span class="dt">releasesInterval=</span><span class="dv">0</span>,</a>
<a class="sourceLine" id="cb12-66" title="66">                          <span class="dt">releaseProportion=</span><span class="dv">50</span>)</a>
<a class="sourceLine" id="cb12-67" title="67"></a>
<a class="sourceLine" id="cb12-68" title="68"><span class="co"># generate male release vector</span></a>
<a class="sourceLine" id="cb12-69" title="69">maleReleasesVector &lt;-<span class="st"> </span><span class="kw"><a href="../reference/generateReleaseVector.html">generateReleaseVector</a></span>(<span class="dt">driveCube=</span>cube, </a>
<a class="sourceLine" id="cb12-70" title="70">                                           <span class="dt">releasesParameters=</span>releasesParameters,</a>
<a class="sourceLine" id="cb12-71" title="71">                                           <span class="dt">sex=</span><span class="st">"M"</span>)</a>
<a class="sourceLine" id="cb12-72" title="72"></a>
<a class="sourceLine" id="cb12-73" title="73"><span class="co"># generate female release vector</span></a>
<a class="sourceLine" id="cb12-74" title="74">femaleReleasesVector &lt;-<span class="st"> </span><span class="kw"><a href="../reference/generateReleaseVector.html">generateReleaseVector</a></span>(<span class="dt">driveCube=</span>cube,</a>
<a class="sourceLine" id="cb12-75" title="75">                                             <span class="dt">releasesParameters=</span>releasesParameters,</a>
<a class="sourceLine" id="cb12-76" title="76">                                             <span class="dt">sex=</span><span class="st">"F"</span>)</a>
<a class="sourceLine" id="cb12-77" title="77"></a>
<a class="sourceLine" id="cb12-78" title="78"><span class="co"># put releases into the proper place in the release list</span></a>
<a class="sourceLine" id="cb12-79" title="79">patchReleases[[<span class="dv">1</span>]]<span class="op">$</span>maleReleases &lt;-<span class="st"> </span>maleReleasesVector</a>
<a class="sourceLine" id="cb12-80" title="80">patchReleases[[<span class="dv">1</span>]]<span class="op">$</span>femaleReleases &lt;-<span class="st"> </span>femaleReleasesVector</a>
<a class="sourceLine" id="cb12-81" title="81"></a>
<a class="sourceLine" id="cb12-82" title="82"></a>
<a class="sourceLine" id="cb12-83" title="83"><span class="co"># batch migration is disabled by setting the probability to 0</span></a>
<a class="sourceLine" id="cb12-84" title="84"><span class="co"># This is required because of the stochastic simulations, but doesn't make sense </span></a>
<a class="sourceLine" id="cb12-85" title="85"><span class="co">#  in a deterministic simulation.</span></a>
<a class="sourceLine" id="cb12-86" title="86">batchMigration &lt;-<span class="st"> </span><span class="kw"><a href="../reference/basicBatchMigration.html">basicBatchMigration</a></span>(<span class="dt">batchProbs=</span><span class="dv">0</span>,</a>
<a class="sourceLine" id="cb12-87" title="87">                                     <span class="dt">sexProbs=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(.<span class="dv">5</span>,.<span class="dv">5</span>),</a>
<a class="sourceLine" id="cb12-88" title="88">                                     <span class="dt">numPatches=</span>sitesNumber)</a>
<a class="sourceLine" id="cb12-89" title="89"></a>
<a class="sourceLine" id="cb12-90" title="90"></a>
<a class="sourceLine" id="cb12-91" title="91"><span class="co">####################</span></a>
<a class="sourceLine" id="cb12-92" title="92"><span class="co"># Combine parameters and run!</span></a>
<a class="sourceLine" id="cb12-93" title="93"><span class="co">####################</span></a>
<a class="sourceLine" id="cb12-94" title="94"><span class="co"># set MGDrivE to run stochastic</span></a>
<a class="sourceLine" id="cb12-95" title="95"><span class="kw"><a href="../reference/setupMGDrivE.html">setupMGDrivE</a></span>(<span class="dt">stochasticityON =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb12-96" title="96"><span class="co">#&gt; initializing MGDrivE</span></a>
<a class="sourceLine" id="cb12-97" title="97"></a>
<a class="sourceLine" id="cb12-98" title="98"><span class="co"># setup parameters for the network. This builds a list of parameters required for</span></a>
<a class="sourceLine" id="cb12-99" title="99"><span class="co">#  every population in the network. In ths case, we havee a network of 1 population.</span></a>
<a class="sourceLine" id="cb12-100" title="100">netPar &lt;-<span class="st"> </span><span class="kw"><a href="../reference/parameterizeMGDrivE.html">parameterizeMGDrivE</a></span>(<span class="dt">runID=</span><span class="dv">1</span>, <span class="dt">simTime=</span>tMax, <span class="dt">nPatch=</span>sitesNumber,</a>
<a class="sourceLine" id="cb12-101" title="101">                             <span class="dt">beta=</span>bioParameters<span class="op">$</span>betaK, <span class="dt">muAd=</span>bioParameters<span class="op">$</span>muAd,</a>
<a class="sourceLine" id="cb12-102" title="102">                             <span class="dt">popGrowth=</span>bioParameters<span class="op">$</span>popGrowth, <span class="dt">tEgg=</span>bioParameters<span class="op">$</span>tEgg,</a>
<a class="sourceLine" id="cb12-103" title="103">                             <span class="dt">tLarva=</span>bioParameters<span class="op">$</span>tLarva, <span class="dt">tPupa=</span>bioParameters<span class="op">$</span>tPupa,</a>
<a class="sourceLine" id="cb12-104" title="104">                             <span class="dt">AdPopEQ=</span>patchPops)</a>
<a class="sourceLine" id="cb12-105" title="105"></a>
<a class="sourceLine" id="cb12-106" title="106"><span class="co"># build network prior to run</span></a>
<a class="sourceLine" id="cb12-107" title="107">MGDrivESim &lt;-<span class="st"> </span>Network<span class="op">$</span><span class="kw">new</span>(<span class="dt">params=</span>netPar,</a>
<a class="sourceLine" id="cb12-108" title="108">                         <span class="dt">driveCube=</span>cube,</a>
<a class="sourceLine" id="cb12-109" title="109">                         <span class="dt">patchReleases=</span>patchReleases,</a>
<a class="sourceLine" id="cb12-110" title="110">                         <span class="dt">migrationMale=</span>moveMat,</a>
<a class="sourceLine" id="cb12-111" title="111">                         <span class="dt">migrationFemale=</span>moveMat,</a>
<a class="sourceLine" id="cb12-112" title="112">                         <span class="dt">migrationBatch=</span>batchMigration,</a>
<a class="sourceLine" id="cb12-113" title="113">                         <span class="dt">directory=</span>folderNames)</a>
<a class="sourceLine" id="cb12-114" title="114"><span class="co">#&gt; initializing patch:  1  of  2 </span></a>
<a class="sourceLine" id="cb12-115" title="115"><span class="co">#&gt; initializing patch:  2  of  2</span></a>
<a class="sourceLine" id="cb12-116" title="116"><span class="co"># run simulation</span></a>
<a class="sourceLine" id="cb12-117" title="117">MGDrivESim<span class="op">$</span><span class="kw">multRun</span>(<span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb12-118" title="118"></a>
<a class="sourceLine" id="cb12-119" title="119"></a>
<a class="sourceLine" id="cb12-120" title="120"><span class="co">####################</span></a>
<a class="sourceLine" id="cb12-121" title="121"><span class="co"># Post Analysis</span></a>
<a class="sourceLine" id="cb12-122" title="122"><span class="co">####################</span></a>
<a class="sourceLine" id="cb12-123" title="123"><span class="co"># First, split output by patch</span></a>
<a class="sourceLine" id="cb12-124" title="124"><span class="co"># Second, aggregate females by their mate choice</span></a>
<a class="sourceLine" id="cb12-125" title="125"><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>nRep){</a>
<a class="sourceLine" id="cb12-126" title="126"> <span class="kw"><a href="../reference/splitOutput.html">splitOutput</a></span>(<span class="dt">readDir=</span>folderNames[i],<span class="dt">remFile =</span> <span class="ot">TRUE</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb12-127" title="127"> <span class="kw"><a href="../reference/aggregateFemales.html">aggregateFemales</a></span>(<span class="dt">readDir=</span>folderNames[i],<span class="dt">genotypes=</span>cube<span class="op">$</span>genotypesID,</a>
<a class="sourceLine" id="cb12-128" title="128">                  <span class="dt">remFile=</span><span class="ot">TRUE</span>,<span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb12-129" title="129">}</a>
<a class="sourceLine" id="cb12-130" title="130"></a>
<a class="sourceLine" id="cb12-131" title="131"><span class="co"># plot output of first run to see effect</span></a>
<a class="sourceLine" id="cb12-132" title="132"><span class="kw"><a href="../reference/plotMGDrivESingle.html">plotMGDrivESingle</a></span>(<span class="dt">readDir=</span>folderNames[<span class="dv">1</span>],<span class="dt">totalPop =</span> <span class="ot">TRUE</span>,<span class="dt">lwd=</span><span class="fl">3.5</span>,<span class="dt">alpha=</span><span class="dv">1</span>)</a></code></pre></div>
<p><img src="mgdrive_examples_files/figure-html/unnamed-chunk-20-1.png" width="672"></p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" title="1"></a>
<a class="sourceLine" id="cb13-2" title="2"><span class="co"># plot all 25 repetitions together</span></a>
<a class="sourceLine" id="cb13-3" title="3"><span class="kw"><a href="../reference/plotMGDrivEMult.html">plotMGDrivEMult</a></span>(<span class="dt">readDir=</span>outFolder,<span class="dt">lwd=</span><span class="fl">0.35</span>,<span class="dt">alpha=</span><span class="fl">0.75</span>)</a></code></pre></div>
<p><img src="mgdrive_examples_files/figure-html/unnamed-chunk-20-2.png" width="672"></p>
<p>Looking at the first plot, of just one repetition from the 25 that we ran, we see the release at time 25 and then the increase in heterozygotes after that. If we look closely, there are a few <em>aa</em> individuals that migrate to the second patch, but very few. We see how the population of <em>Aa</em> individuals fluctuates near zero the entire simulation, but there are enough that they never completely drop out of the population. The total population fluctuates around 500 individuals, 250 each of males and females, as expected from the parameters provided.</p>
<p>Looking at the second plot with all 25 repetitions provided, we see how the simulations heuristically agree with the deterministic ones performed above. The general analysis remains the same here, but with more noise, accounting for the random fluctuations in real populations. While not particularly enlightening in this setting, the behavior and outcome are the same, it’s easy to see how low-frequency genotypes could be lost from a population.</p>
</div>
<div id="stochastic-small-migration-and-fitness-cost" class="section level4">
<h4 class="hasAnchor">
<a href="#stochastic-small-migration-and-fitness-cost" class="anchor"></a>Stochastic, Small Migration and Fitness Cost</h4>
<p>As a final example, we perform a stochastic, two-population simulation with the same fitness cost explored previously, ie., a 60% reduction in lifetime on <em>AA</em> and <em>aa</em> individuals. We first explored this in a single, panmictic population using a deterministic simulation, then explored the effect of stochasticity on the results. Here, we extend our analysis to a two-population setting, to explore how population structure might affect our results.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" title="1"><span class="co">####################</span></a>
<a class="sourceLine" id="cb14-2" title="2"><span class="co"># Load libraries</span></a>
<a class="sourceLine" id="cb14-3" title="3"><span class="co">####################</span></a>
<a class="sourceLine" id="cb14-4" title="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(MGDrivE)</a>
<a class="sourceLine" id="cb14-5" title="5"></a>
<a class="sourceLine" id="cb14-6" title="6"><span class="co">####################</span></a>
<a class="sourceLine" id="cb14-7" title="7"><span class="co"># Output Folder</span></a>
<a class="sourceLine" id="cb14-8" title="8"><span class="co">####################</span></a>
<a class="sourceLine" id="cb14-9" title="9"><span class="co"># This is slightly obtuse for vignette building reasons</span></a>
<a class="sourceLine" id="cb14-10" title="10"><span class="co">#  Really, all you need is a base directory, then the repetitions in folders inside that.</span></a>
<a class="sourceLine" id="cb14-11" title="11"><span class="co">#  Here, our base directory is "mgdrive", and the repetition folders are "001","002", etc.</span></a>
<a class="sourceLine" id="cb14-12" title="12"><span class="co">#  So, the final structure is "mgdrive/001","mgdrive/002", etc. </span></a>
<a class="sourceLine" id="cb14-13" title="13">outFolder &lt;-<span class="st"> "mgdrive"</span></a>
<a class="sourceLine" id="cb14-14" title="14"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/files2">dir.create</a></span>(<span class="dt">path =</span> outFolder)</a>
<a class="sourceLine" id="cb14-15" title="15"></a>
<a class="sourceLine" id="cb14-16" title="16"><span class="co">####################</span></a>
<a class="sourceLine" id="cb14-17" title="17"><span class="co"># Simulation Parameters</span></a>
<a class="sourceLine" id="cb14-18" title="18"><span class="co">####################</span></a>
<a class="sourceLine" id="cb14-19" title="19"><span class="co"># days to run the simulation</span></a>
<a class="sourceLine" id="cb14-20" title="20">tMax &lt;-<span class="st"> </span><span class="dv">365</span><span class="op">*</span><span class="dv">2</span></a>
<a class="sourceLine" id="cb14-21" title="21"></a>
<a class="sourceLine" id="cb14-22" title="22"><span class="co"># number of Monte Carlo iterations</span></a>
<a class="sourceLine" id="cb14-23" title="23">nRep &lt;-<span class="st"> </span><span class="dv">25</span></a>
<a class="sourceLine" id="cb14-24" title="24"></a>
<a class="sourceLine" id="cb14-25" title="25"><span class="co"># each Monte Carlo iteration gets its own folder</span></a>
<a class="sourceLine" id="cb14-26" title="26">folderNames &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/file.path">file.path</a></span>(outFolder,</a>
<a class="sourceLine" id="cb14-27" title="27">                         <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/formatc">formatC</a></span>(<span class="dt">x =</span> <span class="dv">1</span><span class="op">:</span>nRep, <span class="dt">width =</span> <span class="dv">3</span>, <span class="dt">format =</span> <span class="st">"d"</span>, <span class="dt">flag =</span> <span class="st">"0"</span>))</a>
<a class="sourceLine" id="cb14-28" title="28"></a>
<a class="sourceLine" id="cb14-29" title="29"><span class="co"># entomological parameters</span></a>
<a class="sourceLine" id="cb14-30" title="30">bioParameters &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">betaK=</span><span class="dv">20</span>,<span class="dt">tEgg=</span><span class="dv">5</span>,<span class="dt">tLarva=</span><span class="dv">6</span>,<span class="dt">tPupa=</span><span class="dv">4</span>,<span class="dt">popGrowth=</span><span class="fl">1.175</span>,<span class="dt">muAd=</span><span class="fl">0.09</span>)</a>
<a class="sourceLine" id="cb14-31" title="31"></a>
<a class="sourceLine" id="cb14-32" title="32"><span class="co"># a 2-node network with 1% per day migration rate</span></a>
<a class="sourceLine" id="cb14-33" title="33">moveMat &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="dt">data =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="fl">0.99</span>,<span class="fl">0.01</span>,<span class="fl">0.01</span>,<span class="fl">0.99</span>), <span class="dt">nrow =</span> <span class="dv">2</span>, <span class="dt">ncol =</span> <span class="dv">2</span>)</a>
<a class="sourceLine" id="cb14-34" title="34">moveMat</a>
<a class="sourceLine" id="cb14-35" title="35"><span class="co">#&gt;      [,1] [,2]</span></a>
<a class="sourceLine" id="cb14-36" title="36"><span class="co">#&gt; [1,] 0.99 0.01</span></a>
<a class="sourceLine" id="cb14-37" title="37"><span class="co">#&gt; [2,] 0.01 0.99</span></a>
<a class="sourceLine" id="cb14-38" title="38"></a>
<a class="sourceLine" id="cb14-39" title="39"><span class="co"># parameters of the population equilibrium</span></a>
<a class="sourceLine" id="cb14-40" title="40">adultPopEquilibrium &lt;-<span class="st"> </span><span class="dv">500</span></a>
<a class="sourceLine" id="cb14-41" title="41">sitesNumber &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">nrow</a></span>(moveMat)</a>
<a class="sourceLine" id="cb14-42" title="42">patchPops &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>(adultPopEquilibrium,sitesNumber)</a>
<a class="sourceLine" id="cb14-43" title="43"></a>
<a class="sourceLine" id="cb14-44" title="44"><span class="co">####################</span></a>
<a class="sourceLine" id="cb14-45" title="45"><span class="co"># Basic Inheritance pattern</span></a>
<a class="sourceLine" id="cb14-46" title="46"><span class="co">####################</span></a>
<a class="sourceLine" id="cb14-47" title="47"><span class="co"># Mendelian cube with standard (default) paraameters</span></a>
<a class="sourceLine" id="cb14-48" title="48"></a>
<a class="sourceLine" id="cb14-49" title="49"><span class="co"># This time, lets put fitness cost on the homozygotes, giving the heterozygote </span></a>
<a class="sourceLine" id="cb14-50" title="50"><span class="co">#  an advantage</span></a>
<a class="sourceLine" id="cb14-51" title="51"></a>
<a class="sourceLine" id="cb14-52" title="52"><span class="co"># These genotypes correspond to ones in the cube. Look at a base cube first, </span></a>
<a class="sourceLine" id="cb14-53" title="53"><span class="co">#  then set this.</span></a>
<a class="sourceLine" id="cb14-54" title="54"><span class="co"># homozygotes are 60% as fit as heterozygote over their entire lifetime</span></a>
<a class="sourceLine" id="cb14-55" title="55"><span class="co">#  Since omega is a daily death rate, we use the built-in function to calculate </span></a>
<a class="sourceLine" id="cb14-56" title="56"><span class="co">#  our desired lifetime cost as applied daily</span></a>
<a class="sourceLine" id="cb14-57" title="57">dayOmega &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getOmega.html">getOmega</a></span>(<span class="dt">mu =</span> bioParameters<span class="op">$</span>muAd, <span class="dt">lifespanReduction =</span> <span class="fl">0.60</span>)</a>
<a class="sourceLine" id="cb14-58" title="58">omegaNew &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"AA"</span>=dayOmega, <span class="st">"aa"</span>=dayOmega)</a>
<a class="sourceLine" id="cb14-59" title="59"></a>
<a class="sourceLine" id="cb14-60" title="60"><span class="co"># setup cube</span></a>
<a class="sourceLine" id="cb14-61" title="61">cube &lt;-<span class="st"> </span><span class="kw"><a href="../reference/cubeMendelian.html">cubeMendelian</a></span>(<span class="dt">omega =</span> omegaNew)</a>
<a class="sourceLine" id="cb14-62" title="62"></a>
<a class="sourceLine" id="cb14-63" title="63"></a>
<a class="sourceLine" id="cb14-64" title="64"><span class="co">####################</span></a>
<a class="sourceLine" id="cb14-65" title="65"><span class="co"># Setup releases and batch migration</span></a>
<a class="sourceLine" id="cb14-66" title="66"><span class="co">####################</span></a>
<a class="sourceLine" id="cb14-67" title="67"><span class="co"># set up the empty release vector</span></a>
<a class="sourceLine" id="cb14-68" title="68"><span class="co">#  MGDrivE pulls things out by name</span></a>
<a class="sourceLine" id="cb14-69" title="69">patchReleases &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">replicate</a></span>(<span class="dt">n=</span>sitesNumber,</a>
<a class="sourceLine" id="cb14-70" title="70">                           <span class="dt">expr=</span>{<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">maleReleases=</span><span class="ot">NULL</span>,<span class="dt">femaleReleases=</span><span class="ot">NULL</span>)},</a>
<a class="sourceLine" id="cb14-71" title="71">                           <span class="dt">simplify=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb14-72" title="72"></a>
<a class="sourceLine" id="cb14-73" title="73"><span class="co"># choose release parameters</span></a>
<a class="sourceLine" id="cb14-74" title="74"><span class="co">#  Releases start at time 25, occur every day, for 5 day.</span></a>
<a class="sourceLine" id="cb14-75" title="75"><span class="co">#  There are 50 mosquitoes relaeses every time.</span></a>
<a class="sourceLine" id="cb14-76" title="76">releasesParameters &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">releasesStart=</span><span class="dv">100</span>,</a>
<a class="sourceLine" id="cb14-77" title="77">                           <span class="dt">releasesNumber=</span><span class="dv">5</span>,</a>
<a class="sourceLine" id="cb14-78" title="78">                           <span class="dt">releasesInterval=</span><span class="dv">0</span>,</a>
<a class="sourceLine" id="cb14-79" title="79">                           <span class="dt">releaseProportion=</span><span class="dv">50</span>)</a>
<a class="sourceLine" id="cb14-80" title="80"></a>
<a class="sourceLine" id="cb14-81" title="81"><span class="co"># generate male release vector</span></a>
<a class="sourceLine" id="cb14-82" title="82">maleReleasesVector &lt;-<span class="st"> </span><span class="kw"><a href="../reference/generateReleaseVector.html">generateReleaseVector</a></span>(<span class="dt">driveCube=</span>cube, </a>
<a class="sourceLine" id="cb14-83" title="83">                                            <span class="dt">releasesParameters=</span>releasesParameters,</a>
<a class="sourceLine" id="cb14-84" title="84">                                            <span class="dt">sex=</span><span class="st">"M"</span>)</a>
<a class="sourceLine" id="cb14-85" title="85"></a>
<a class="sourceLine" id="cb14-86" title="86"><span class="co"># generate female release vector</span></a>
<a class="sourceLine" id="cb14-87" title="87">femaleReleasesVector &lt;-<span class="st"> </span><span class="kw"><a href="../reference/generateReleaseVector.html">generateReleaseVector</a></span>(<span class="dt">driveCube=</span>cube,</a>
<a class="sourceLine" id="cb14-88" title="88">                                              <span class="dt">releasesParameters=</span>releasesParameters,</a>
<a class="sourceLine" id="cb14-89" title="89">                                              <span class="dt">sex=</span><span class="st">"F"</span>)</a>
<a class="sourceLine" id="cb14-90" title="90"></a>
<a class="sourceLine" id="cb14-91" title="91"><span class="co"># put releases into the proper place in the release list</span></a>
<a class="sourceLine" id="cb14-92" title="92">patchReleases[[<span class="dv">1</span>]]<span class="op">$</span>maleReleases &lt;-<span class="st"> </span>maleReleasesVector</a>
<a class="sourceLine" id="cb14-93" title="93">patchReleases[[<span class="dv">1</span>]]<span class="op">$</span>femaleReleases &lt;-<span class="st"> </span>femaleReleasesVector</a>
<a class="sourceLine" id="cb14-94" title="94"></a>
<a class="sourceLine" id="cb14-95" title="95"></a>
<a class="sourceLine" id="cb14-96" title="96"><span class="co"># batch migration is disabled by setting the probability to 0</span></a>
<a class="sourceLine" id="cb14-97" title="97"><span class="co"># This is required because of the stochastic simulations, but doesn't make sense </span></a>
<a class="sourceLine" id="cb14-98" title="98"><span class="co">#  in a deterministic simulation.</span></a>
<a class="sourceLine" id="cb14-99" title="99">batchMigration &lt;-<span class="st"> </span><span class="kw"><a href="../reference/basicBatchMigration.html">basicBatchMigration</a></span>(<span class="dt">batchProbs=</span><span class="dv">0</span>,</a>
<a class="sourceLine" id="cb14-100" title="100">                                     <span class="dt">sexProbs=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(.<span class="dv">5</span>,.<span class="dv">5</span>),</a>
<a class="sourceLine" id="cb14-101" title="101">                                     <span class="dt">numPatches=</span>sitesNumber)</a>
<a class="sourceLine" id="cb14-102" title="102"></a>
<a class="sourceLine" id="cb14-103" title="103"></a>
<a class="sourceLine" id="cb14-104" title="104"><span class="co">####################</span></a>
<a class="sourceLine" id="cb14-105" title="105"><span class="co"># Combine parameters and run!</span></a>
<a class="sourceLine" id="cb14-106" title="106"><span class="co">####################</span></a>
<a class="sourceLine" id="cb14-107" title="107"><span class="co"># set MGDrivE to run stochastic</span></a>
<a class="sourceLine" id="cb14-108" title="108"><span class="kw"><a href="../reference/setupMGDrivE.html">setupMGDrivE</a></span>(<span class="dt">stochasticityON =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb14-109" title="109"><span class="co">#&gt; initializing MGDrivE</span></a>
<a class="sourceLine" id="cb14-110" title="110"></a>
<a class="sourceLine" id="cb14-111" title="111"><span class="co"># setup parameters for the network. This builds a list of parameters required for</span></a>
<a class="sourceLine" id="cb14-112" title="112"><span class="co">#  every population in the network. In ths case, we havee a network of 1 population.</span></a>
<a class="sourceLine" id="cb14-113" title="113">netPar &lt;-<span class="st"> </span><span class="kw"><a href="../reference/parameterizeMGDrivE.html">parameterizeMGDrivE</a></span>(<span class="dt">runID=</span><span class="dv">1</span>, <span class="dt">simTime=</span>tMax, <span class="dt">nPatch=</span>sitesNumber,</a>
<a class="sourceLine" id="cb14-114" title="114">                             <span class="dt">beta=</span>bioParameters<span class="op">$</span>betaK, <span class="dt">muAd=</span>bioParameters<span class="op">$</span>muAd,</a>
<a class="sourceLine" id="cb14-115" title="115">                             <span class="dt">popGrowth=</span>bioParameters<span class="op">$</span>popGrowth, <span class="dt">tEgg=</span>bioParameters<span class="op">$</span>tEgg,</a>
<a class="sourceLine" id="cb14-116" title="116">                             <span class="dt">tLarva=</span>bioParameters<span class="op">$</span>tLarva, <span class="dt">tPupa=</span>bioParameters<span class="op">$</span>tPupa,</a>
<a class="sourceLine" id="cb14-117" title="117">                             <span class="dt">AdPopEQ=</span>patchPops)</a>
<a class="sourceLine" id="cb14-118" title="118"></a>
<a class="sourceLine" id="cb14-119" title="119"><span class="co"># build network prior to run</span></a>
<a class="sourceLine" id="cb14-120" title="120">MGDrivESim &lt;-<span class="st"> </span>Network<span class="op">$</span><span class="kw">new</span>(<span class="dt">params=</span>netPar,</a>
<a class="sourceLine" id="cb14-121" title="121">                         <span class="dt">driveCube=</span>cube,</a>
<a class="sourceLine" id="cb14-122" title="122">                         <span class="dt">patchReleases=</span>patchReleases,</a>
<a class="sourceLine" id="cb14-123" title="123">                         <span class="dt">migrationMale=</span>moveMat,</a>
<a class="sourceLine" id="cb14-124" title="124">                         <span class="dt">migrationFemale=</span>moveMat,</a>
<a class="sourceLine" id="cb14-125" title="125">                         <span class="dt">migrationBatch=</span>batchMigration,</a>
<a class="sourceLine" id="cb14-126" title="126">                         <span class="dt">directory=</span>folderNames)</a>
<a class="sourceLine" id="cb14-127" title="127"><span class="co">#&gt; initializing patch:  1  of  2 </span></a>
<a class="sourceLine" id="cb14-128" title="128"><span class="co">#&gt; initializing patch:  2  of  2</span></a>
<a class="sourceLine" id="cb14-129" title="129"><span class="co"># run simulation</span></a>
<a class="sourceLine" id="cb14-130" title="130">MGDrivESim<span class="op">$</span><span class="kw">multRun</span>(<span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb14-131" title="131"></a>
<a class="sourceLine" id="cb14-132" title="132"></a>
<a class="sourceLine" id="cb14-133" title="133"><span class="co">####################</span></a>
<a class="sourceLine" id="cb14-134" title="134"><span class="co"># Post Analysis</span></a>
<a class="sourceLine" id="cb14-135" title="135"><span class="co">####################</span></a>
<a class="sourceLine" id="cb14-136" title="136"><span class="co"># First, split output by patch</span></a>
<a class="sourceLine" id="cb14-137" title="137"><span class="co"># Second, aggregate females by their mate choice</span></a>
<a class="sourceLine" id="cb14-138" title="138"><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>nRep){</a>
<a class="sourceLine" id="cb14-139" title="139"> <span class="kw"><a href="../reference/splitOutput.html">splitOutput</a></span>(<span class="dt">readDir=</span>folderNames[i],<span class="dt">remFile =</span> <span class="ot">TRUE</span>,<span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb14-140" title="140"> <span class="kw"><a href="../reference/aggregateFemales.html">aggregateFemales</a></span>(<span class="dt">readDir=</span>folderNames[i],<span class="dt">genotypes=</span>cube<span class="op">$</span>genotypesID,</a>
<a class="sourceLine" id="cb14-141" title="141">                  <span class="dt">remFile=</span><span class="ot">TRUE</span>,<span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb14-142" title="142">}</a>
<a class="sourceLine" id="cb14-143" title="143"></a>
<a class="sourceLine" id="cb14-144" title="144"><span class="co"># plot output of first run to see effect</span></a>
<a class="sourceLine" id="cb14-145" title="145"><span class="co">#  per the structure above, we are reading "mgdrive/001" for the single plot</span></a>
<a class="sourceLine" id="cb14-146" title="146"><span class="kw"><a href="../reference/plotMGDrivESingle.html">plotMGDrivESingle</a></span>(<span class="dt">readDir=</span>folderNames[<span class="dv">1</span>],<span class="dt">totalPop =</span> <span class="ot">TRUE</span>, <span class="dt">lwd=</span><span class="fl">3.5</span>,<span class="dt">alpha=</span><span class="dv">1</span>)</a></code></pre></div>
<p><img src="mgdrive_examples_files/figure-html/unnamed-chunk-22-1.png" width="672"></p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" title="1"></a>
<a class="sourceLine" id="cb15-2" title="2"><span class="co"># plot all 50 repetitions together</span></a>
<a class="sourceLine" id="cb15-3" title="3"><span class="co">#  Here, we feed the function "mgdrive/", and it finds all repetition folders</span></a>
<a class="sourceLine" id="cb15-4" title="4"><span class="co">#   inside that.</span></a>
<a class="sourceLine" id="cb15-5" title="5"><span class="kw"><a href="../reference/plotMGDrivEMult.html">plotMGDrivEMult</a></span>(<span class="dt">readDir=</span>outFolder,<span class="dt">lwd=</span><span class="fl">0.35</span>,<span class="dt">alpha=</span><span class="fl">0.75</span>)</a></code></pre></div>
<p><img src="mgdrive_examples_files/figure-html/unnamed-chunk-22-2.png" width="672"></p>
<p>Looking at the first plot, a single repetition of our stochastic simulation, we see that the results are very similar. The population size first drops to about 300, 60% of the equilibration population, and then increases to about 400, as expected from the deterministic simulation and analysis performed before. Do notice however, the dynamics are slower in the first population than in the single-population exploration. This is because we performed the same releases, but with a second population, there is a small immigration of <em>AA</em> individuals into the first patch and a small emigration of <em>aa</em> and <em>Aa</em> individuals out of patch 1 into patch 2. Thus, there is effectively twice the population size total, even though there is only a 1% chance of mixing per day.</p>
<p>The second plot shows all 25 repetitions of our simulation. The results clearly follow the one-node, deterministic simulation performed before. There is an initial population drop, as both patches are fully <em>AA</em> until we perform releases, then releases and a small recovery in population size. The general dynamics remain very similar, with the shift from <em>AA</em> to <em>Aa</em> individuals in the population taking a little longer than originally. This is also something we saw in the one-node stochastic simulation, and is a result of stochasticity in the population dynamics, as well as migration to/from the second patch.</p>
</div>
</div>
<div id="reciprocal-translocation-gene-drive-simulations-one-population" class="section level3">
<h3 class="hasAnchor">
<a href="#reciprocal-translocation-gene-drive-simulations-one-population" class="anchor"></a>Reciprocal Translocation Gene Drive Simulations, One Population</h3>
<p><a href="https://pubs.acs.org/doi/10.1021/acssynbio.7b00451">Reciprocal translocations</a> are a classic form of gene drive, utilizing the underdominance effects of gene dosage compensation to either suppress a population or replace a population, depending on fitness costs and release frequencies. The classic reciprocal translocation involves breaking two chromosomes, swapping the broken ends, and reattaching the ends to the opposite chromosomes. Theoretically, this creates a perfect 50% fitness cost on the population, as has been shown previously. However, previous simulations studying reciprocal translocations have been panmictic, deterministic, and ignore life-stages. Thus, we start with two deterministic simulations, to find the critical threshold when a complete life-history is included, and then explore how that threshold is affected by stochasticity.</p>
<div id="deterministic-below-threshold" class="section level4">
<h4 class="hasAnchor">
<a href="#deterministic-below-threshold" class="anchor"></a>Deterministic, Below Threshold</h4>
<p>We maintain the same theoretical fitness cost, namely, that possession of only 1 copy of a reciprocal chromosome is always fatal. We perform 5 releases of transgenic critters, both male and female, every 7 days starting at day 25. This is not enough to make the population 50% transgenic, and therefore the transgenic critters should die out.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" title="1"><span class="co">####################</span></a>
<a class="sourceLine" id="cb16-2" title="2"><span class="co"># Load libraries</span></a>
<a class="sourceLine" id="cb16-3" title="3"><span class="co">####################</span></a>
<a class="sourceLine" id="cb16-4" title="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(MGDrivE)</a>
<a class="sourceLine" id="cb16-5" title="5"></a>
<a class="sourceLine" id="cb16-6" title="6"><span class="co">#################### </span></a>
<a class="sourceLine" id="cb16-7" title="7"><span class="co"># Output Folder</span></a>
<a class="sourceLine" id="cb16-8" title="8"><span class="co">####################</span></a>
<a class="sourceLine" id="cb16-9" title="9">outFolder &lt;-<span class="st"> "mgdrive"</span></a>
<a class="sourceLine" id="cb16-10" title="10"></a>
<a class="sourceLine" id="cb16-11" title="11"><span class="co">####################</span></a>
<a class="sourceLine" id="cb16-12" title="12"><span class="co"># Simulation Parameters</span></a>
<a class="sourceLine" id="cb16-13" title="13"><span class="co">####################</span></a>
<a class="sourceLine" id="cb16-14" title="14"><span class="co"># days to run the simulation, 2 years</span></a>
<a class="sourceLine" id="cb16-15" title="15">tMax &lt;-<span class="st"> </span><span class="dv">365</span><span class="op">*</span><span class="dv">2</span></a>
<a class="sourceLine" id="cb16-16" title="16"></a>
<a class="sourceLine" id="cb16-17" title="17"><span class="co"># entomological parameters</span></a>
<a class="sourceLine" id="cb16-18" title="18">bioParameters &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">betaK=</span><span class="dv">20</span>,<span class="dt">tEgg=</span><span class="dv">5</span>,<span class="dt">tLarva=</span><span class="dv">6</span>,<span class="dt">tPupa=</span><span class="dv">4</span>,<span class="dt">popGrowth=</span><span class="fl">1.175</span>,<span class="dt">muAd=</span><span class="fl">0.09</span>)</a>
<a class="sourceLine" id="cb16-19" title="19"></a>
<a class="sourceLine" id="cb16-20" title="20"><span class="co"># a 1-node network where mosquitoes do not leave</span></a>
<a class="sourceLine" id="cb16-21" title="21">moveMat &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">as.matrix</a></span>(<span class="dv">1</span>)</a>
<a class="sourceLine" id="cb16-22" title="22"></a>
<a class="sourceLine" id="cb16-23" title="23"><span class="co"># parameters of the population equilibrium</span></a>
<a class="sourceLine" id="cb16-24" title="24">adultPopEquilibrium &lt;-<span class="st"> </span><span class="dv">500</span></a>
<a class="sourceLine" id="cb16-25" title="25">sitesNumber &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">nrow</a></span>(moveMat)</a>
<a class="sourceLine" id="cb16-26" title="26">patchPops &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>(adultPopEquilibrium,sitesNumber)</a>
<a class="sourceLine" id="cb16-27" title="27"></a>
<a class="sourceLine" id="cb16-28" title="28"><span class="co">####################</span></a>
<a class="sourceLine" id="cb16-29" title="29"><span class="co"># Basic Inheritance pattern</span></a>
<a class="sourceLine" id="cb16-30" title="30"><span class="co">####################</span></a>
<a class="sourceLine" id="cb16-31" title="31"><span class="co"># Reciprocal translocation cube with standard (default) paraameters</span></a>
<a class="sourceLine" id="cb16-32" title="32">cube &lt;-<span class="st"> </span><span class="kw"><a href="../reference/cubeReciprocalTranslocations.html">cubeReciprocalTranslocations</a></span>()</a>
<a class="sourceLine" id="cb16-33" title="33"></a>
<a class="sourceLine" id="cb16-34" title="34"></a>
<a class="sourceLine" id="cb16-35" title="35"><span class="co">####################</span></a>
<a class="sourceLine" id="cb16-36" title="36"><span class="co"># Setup releases and batch migration</span></a>
<a class="sourceLine" id="cb16-37" title="37"><span class="co">####################</span></a>
<a class="sourceLine" id="cb16-38" title="38"><span class="co"># set up the empty release vector</span></a>
<a class="sourceLine" id="cb16-39" title="39"><span class="co">#  MGDrivE pulls things out by name</span></a>
<a class="sourceLine" id="cb16-40" title="40">patchReleases &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">replicate</a></span>(<span class="dt">n=</span>sitesNumber,</a>
<a class="sourceLine" id="cb16-41" title="41">                          <span class="dt">expr=</span>{<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">maleReleases=</span><span class="ot">NULL</span>,<span class="dt">femaleReleases=</span><span class="ot">NULL</span>)},</a>
<a class="sourceLine" id="cb16-42" title="42">                          <span class="dt">simplify=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb16-43" title="43"></a>
<a class="sourceLine" id="cb16-44" title="44"><span class="co"># choose release parameters</span></a>
<a class="sourceLine" id="cb16-45" title="45"><span class="co">#  Releases start at time 25, occur once a week, for 5 weeks.</span></a>
<a class="sourceLine" id="cb16-46" title="46"><span class="co">#  There are 10 mosquitoes relaeses every time.</span></a>
<a class="sourceLine" id="cb16-47" title="47">releasesParameters &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">releasesStart=</span><span class="dv">25</span>,</a>
<a class="sourceLine" id="cb16-48" title="48">                          <span class="dt">releasesNumber=</span><span class="dv">5</span>,</a>
<a class="sourceLine" id="cb16-49" title="49">                          <span class="dt">releasesInterval=</span><span class="dv">7</span>,</a>
<a class="sourceLine" id="cb16-50" title="50">                          <span class="dt">releaseProportion=</span><span class="dv">100</span>)</a>
<a class="sourceLine" id="cb16-51" title="51"></a>
<a class="sourceLine" id="cb16-52" title="52"><span class="co"># generate male release vector</span></a>
<a class="sourceLine" id="cb16-53" title="53">maleReleasesVector &lt;-<span class="st"> </span><span class="kw"><a href="../reference/generateReleaseVector.html">generateReleaseVector</a></span>(<span class="dt">driveCube=</span>cube, </a>
<a class="sourceLine" id="cb16-54" title="54">                                           <span class="dt">releasesParameters=</span>releasesParameters,</a>
<a class="sourceLine" id="cb16-55" title="55">                                           <span class="dt">sex=</span><span class="st">"M"</span>)</a>
<a class="sourceLine" id="cb16-56" title="56"></a>
<a class="sourceLine" id="cb16-57" title="57"><span class="co"># generate female release vector</span></a>
<a class="sourceLine" id="cb16-58" title="58">femaleReleasesVector &lt;-<span class="st"> </span><span class="kw"><a href="../reference/generateReleaseVector.html">generateReleaseVector</a></span>(<span class="dt">driveCube=</span>cube,</a>
<a class="sourceLine" id="cb16-59" title="59">                                             <span class="dt">releasesParameters=</span>releasesParameters,</a>
<a class="sourceLine" id="cb16-60" title="60">                                             <span class="dt">sex=</span><span class="st">"F"</span>)</a>
<a class="sourceLine" id="cb16-61" title="61"></a>
<a class="sourceLine" id="cb16-62" title="62"><span class="co"># put releases into the proper place in the release list</span></a>
<a class="sourceLine" id="cb16-63" title="63">patchReleases[[<span class="dv">1</span>]]<span class="op">$</span>maleReleases &lt;-<span class="st"> </span>maleReleasesVector</a>
<a class="sourceLine" id="cb16-64" title="64">patchReleases[[<span class="dv">1</span>]]<span class="op">$</span>femaleReleases &lt;-<span class="st"> </span>femaleReleasesVector</a>
<a class="sourceLine" id="cb16-65" title="65"></a>
<a class="sourceLine" id="cb16-66" title="66"></a>
<a class="sourceLine" id="cb16-67" title="67"><span class="co"># batch migration is disabled by setting the probability to 0</span></a>
<a class="sourceLine" id="cb16-68" title="68"><span class="co"># This is required because of the stochastic simulations, but doesn't make sense </span></a>
<a class="sourceLine" id="cb16-69" title="69"><span class="co">#  in a deterministic simulation.</span></a>
<a class="sourceLine" id="cb16-70" title="70">batchMigration &lt;-<span class="st"> </span><span class="kw"><a href="../reference/basicBatchMigration.html">basicBatchMigration</a></span>(<span class="dt">batchProbs=</span><span class="dv">0</span>,</a>
<a class="sourceLine" id="cb16-71" title="71">                                     <span class="dt">sexProbs=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(.<span class="dv">5</span>,.<span class="dv">5</span>),</a>
<a class="sourceLine" id="cb16-72" title="72">                                     <span class="dt">numPatches=</span>sitesNumber)</a>
<a class="sourceLine" id="cb16-73" title="73"></a>
<a class="sourceLine" id="cb16-74" title="74"></a>
<a class="sourceLine" id="cb16-75" title="75"><span class="co">####################</span></a>
<a class="sourceLine" id="cb16-76" title="76"><span class="co"># Combine parameters and run!</span></a>
<a class="sourceLine" id="cb16-77" title="77"><span class="co">####################</span></a>
<a class="sourceLine" id="cb16-78" title="78"><span class="co"># set MGDrivE to run deterministic</span></a>
<a class="sourceLine" id="cb16-79" title="79"><span class="kw"><a href="../reference/setupMGDrivE.html">setupMGDrivE</a></span>(<span class="dt">stochasticityON =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb16-80" title="80"><span class="co">#&gt; initializing MGDrivE</span></a>
<a class="sourceLine" id="cb16-81" title="81"></a>
<a class="sourceLine" id="cb16-82" title="82"><span class="co"># setup parameters for the network. This builds a list of parameters required for</span></a>
<a class="sourceLine" id="cb16-83" title="83"><span class="co">#  every population in the network. In ths case, we havee a network of 1 population.</span></a>
<a class="sourceLine" id="cb16-84" title="84">netPar &lt;-<span class="st"> </span><span class="kw"><a href="../reference/parameterizeMGDrivE.html">parameterizeMGDrivE</a></span>(<span class="dt">runID=</span><span class="dv">1</span>, <span class="dt">simTime=</span>tMax, <span class="dt">nPatch=</span>sitesNumber,</a>
<a class="sourceLine" id="cb16-85" title="85">                             <span class="dt">beta=</span>bioParameters<span class="op">$</span>betaK, <span class="dt">muAd=</span>bioParameters<span class="op">$</span>muAd,</a>
<a class="sourceLine" id="cb16-86" title="86">                             <span class="dt">popGrowth=</span>bioParameters<span class="op">$</span>popGrowth, <span class="dt">tEgg=</span>bioParameters<span class="op">$</span>tEgg,</a>
<a class="sourceLine" id="cb16-87" title="87">                             <span class="dt">tLarva=</span>bioParameters<span class="op">$</span>tLarva, <span class="dt">tPupa=</span>bioParameters<span class="op">$</span>tPupa,</a>
<a class="sourceLine" id="cb16-88" title="88">                             <span class="dt">AdPopEQ=</span>patchPops)</a>
<a class="sourceLine" id="cb16-89" title="89"></a>
<a class="sourceLine" id="cb16-90" title="90"><span class="co"># build network prior to run</span></a>
<a class="sourceLine" id="cb16-91" title="91">MGDrivESim &lt;-<span class="st"> </span>Network<span class="op">$</span><span class="kw">new</span>(<span class="dt">params=</span>netPar,</a>
<a class="sourceLine" id="cb16-92" title="92">                         <span class="dt">driveCube=</span>cube,</a>
<a class="sourceLine" id="cb16-93" title="93">                         <span class="dt">patchReleases=</span>patchReleases,</a>
<a class="sourceLine" id="cb16-94" title="94">                         <span class="dt">migrationMale=</span>moveMat,</a>
<a class="sourceLine" id="cb16-95" title="95">                         <span class="dt">migrationFemale=</span>moveMat,</a>
<a class="sourceLine" id="cb16-96" title="96">                         <span class="dt">migrationBatch=</span>batchMigration,</a>
<a class="sourceLine" id="cb16-97" title="97">                         <span class="dt">directory=</span>outFolder)</a>
<a class="sourceLine" id="cb16-98" title="98"><span class="co">#&gt; initializing patch:  1  of  1</span></a>
<a class="sourceLine" id="cb16-99" title="99"><span class="co"># run simulation</span></a>
<a class="sourceLine" id="cb16-100" title="100">MGDrivESim<span class="op">$</span><span class="kw">oneRun</span>(<span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb16-101" title="101"></a>
<a class="sourceLine" id="cb16-102" title="102"></a>
<a class="sourceLine" id="cb16-103" title="103"><span class="co">####################</span></a>
<a class="sourceLine" id="cb16-104" title="104"><span class="co"># Post Analysis</span></a>
<a class="sourceLine" id="cb16-105" title="105"><span class="co">####################</span></a>
<a class="sourceLine" id="cb16-106" title="106"><span class="co"># split output by patch</span></a>
<a class="sourceLine" id="cb16-107" title="107"><span class="co">#  Required for plotting later</span></a>
<a class="sourceLine" id="cb16-108" title="108"><span class="kw"><a href="../reference/splitOutput.html">splitOutput</a></span>(<span class="dt">readDir=</span>outFolder, <span class="dt">remFile =</span> <span class="ot">TRUE</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb16-109" title="109"></a>
<a class="sourceLine" id="cb16-110" title="110"><span class="co"># aggregate females by their mate choice</span></a>
<a class="sourceLine" id="cb16-111" title="111"><span class="co">#  This reduces the female file to have the same columns as the male file</span></a>
<a class="sourceLine" id="cb16-112" title="112"><span class="kw"><a href="../reference/aggregateFemales.html">aggregateFemales</a></span>(<span class="dt">readDir=</span>outFolder,<span class="dt">genotypes=</span>cube<span class="op">$</span>genotypesID,</a>
<a class="sourceLine" id="cb16-113" title="113">                <span class="dt">remFile=</span><span class="ot">TRUE</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb16-114" title="114"></a>
<a class="sourceLine" id="cb16-115" title="115"><span class="co"># plot output to see effect</span></a>
<a class="sourceLine" id="cb16-116" title="116"><span class="kw"><a href="../reference/plotMGDrivESingle.html">plotMGDrivESingle</a></span>(<span class="dt">readDir=</span>outFolder,<span class="dt">totalPop =</span> <span class="ot">TRUE</span>, <span class="dt">lwd=</span><span class="fl">3.5</span>,<span class="dt">alpha=</span><span class="dv">1</span>)</a></code></pre></div>
<p><img src="mgdrive_examples_files/figure-html/unnamed-chunk-24-1.png" width="672"></p>
<p>We plot the number of female (left) and male (right) critters over a period of two years. The total population is shown in purple, while wild-type (aabb) individuals are in orange and fully transgenic individuals (AABB) are in light blue. We see the total population increase dramatically during the releases, then fall back to a little below the equilibrium amount. This is because of offspring of transgenic and wile-type critters than are non-viable. Eventually, the transgenic individuals completely die out and the population is wild-type again. This is the expected behavior for reciprocal translocations when the 50% critical threshold is not achieved.</p>
</div>
<div id="deterministic-above-threshold" class="section level4">
<h4 class="hasAnchor">
<a href="#deterministic-above-threshold" class="anchor"></a>Deterministic, Above Threshold</h4>
<p>Now, we perform the same simulation as above, but increase the number of releases by one, from 5 releases to 6. This is enough to get over the critical threshold, so we should see the population turn entirely into transgenic individuals.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" title="1"><span class="co">####################</span></a>
<a class="sourceLine" id="cb17-2" title="2"><span class="co"># Load libraries</span></a>
<a class="sourceLine" id="cb17-3" title="3"><span class="co">####################</span></a>
<a class="sourceLine" id="cb17-4" title="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(MGDrivE)</a>
<a class="sourceLine" id="cb17-5" title="5"></a>
<a class="sourceLine" id="cb17-6" title="6"><span class="co">####################</span></a>
<a class="sourceLine" id="cb17-7" title="7"><span class="co"># Output Folder</span></a>
<a class="sourceLine" id="cb17-8" title="8"><span class="co">####################</span></a>
<a class="sourceLine" id="cb17-9" title="9">outFolder &lt;-<span class="st"> "mgdrive"</span></a>
<a class="sourceLine" id="cb17-10" title="10"></a>
<a class="sourceLine" id="cb17-11" title="11"><span class="co">####################</span></a>
<a class="sourceLine" id="cb17-12" title="12"><span class="co"># Simulation Parameters</span></a>
<a class="sourceLine" id="cb17-13" title="13"><span class="co">####################</span></a>
<a class="sourceLine" id="cb17-14" title="14"><span class="co"># days to run the simulation, 2 years</span></a>
<a class="sourceLine" id="cb17-15" title="15">tMax &lt;-<span class="st"> </span><span class="dv">365</span><span class="op">*</span><span class="dv">2</span></a>
<a class="sourceLine" id="cb17-16" title="16"></a>
<a class="sourceLine" id="cb17-17" title="17"><span class="co"># entomological parameters</span></a>
<a class="sourceLine" id="cb17-18" title="18">bioParameters &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">betaK=</span><span class="dv">20</span>,<span class="dt">tEgg=</span><span class="dv">5</span>,<span class="dt">tLarva=</span><span class="dv">6</span>,<span class="dt">tPupa=</span><span class="dv">4</span>,<span class="dt">popGrowth=</span><span class="fl">1.175</span>,<span class="dt">muAd=</span><span class="fl">0.09</span>)</a>
<a class="sourceLine" id="cb17-19" title="19"></a>
<a class="sourceLine" id="cb17-20" title="20"><span class="co"># a 1-node network where mosquitoes do not leave</span></a>
<a class="sourceLine" id="cb17-21" title="21">moveMat &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">as.matrix</a></span>(<span class="dv">1</span>)</a>
<a class="sourceLine" id="cb17-22" title="22"></a>
<a class="sourceLine" id="cb17-23" title="23"><span class="co"># parameters of the population equilibrium</span></a>
<a class="sourceLine" id="cb17-24" title="24">adultPopEquilibrium &lt;-<span class="st"> </span><span class="dv">500</span></a>
<a class="sourceLine" id="cb17-25" title="25">sitesNumber &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">nrow</a></span>(moveMat)</a>
<a class="sourceLine" id="cb17-26" title="26">patchPops &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>(adultPopEquilibrium,sitesNumber)</a>
<a class="sourceLine" id="cb17-27" title="27"></a>
<a class="sourceLine" id="cb17-28" title="28"><span class="co">####################</span></a>
<a class="sourceLine" id="cb17-29" title="29"><span class="co"># Basic Inheritance pattern</span></a>
<a class="sourceLine" id="cb17-30" title="30"><span class="co">####################</span></a>
<a class="sourceLine" id="cb17-31" title="31"><span class="co"># Reciprocal translocation cube with standard (default) paraameters</span></a>
<a class="sourceLine" id="cb17-32" title="32">cube &lt;-<span class="st"> </span><span class="kw"><a href="../reference/cubeReciprocalTranslocations.html">cubeReciprocalTranslocations</a></span>()</a>
<a class="sourceLine" id="cb17-33" title="33"></a>
<a class="sourceLine" id="cb17-34" title="34"></a>
<a class="sourceLine" id="cb17-35" title="35"><span class="co">####################</span></a>
<a class="sourceLine" id="cb17-36" title="36"><span class="co"># Setup releases and batch migration</span></a>
<a class="sourceLine" id="cb17-37" title="37"><span class="co">####################</span></a>
<a class="sourceLine" id="cb17-38" title="38"><span class="co"># set up the empty release vector</span></a>
<a class="sourceLine" id="cb17-39" title="39"><span class="co">#  MGDrivE pulls things out by name</span></a>
<a class="sourceLine" id="cb17-40" title="40">patchReleases &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">replicate</a></span>(<span class="dt">n=</span>sitesNumber,</a>
<a class="sourceLine" id="cb17-41" title="41">                          <span class="dt">expr=</span>{<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">maleReleases=</span><span class="ot">NULL</span>,<span class="dt">femaleReleases=</span><span class="ot">NULL</span>)},</a>
<a class="sourceLine" id="cb17-42" title="42">                          <span class="dt">simplify=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb17-43" title="43"></a>
<a class="sourceLine" id="cb17-44" title="44"><span class="co"># choose release parameters</span></a>
<a class="sourceLine" id="cb17-45" title="45"><span class="co">#  Releases start at time 25, occur once a week, for 6 weeks.</span></a>
<a class="sourceLine" id="cb17-46" title="46"><span class="co">#  There are 10 mosquitoes relaeses every time.</span></a>
<a class="sourceLine" id="cb17-47" title="47">releasesParameters &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">releasesStart=</span><span class="dv">25</span>,</a>
<a class="sourceLine" id="cb17-48" title="48">                          <span class="dt">releasesNumber=</span><span class="dv">6</span>,</a>
<a class="sourceLine" id="cb17-49" title="49">                          <span class="dt">releasesInterval=</span><span class="dv">7</span>,</a>
<a class="sourceLine" id="cb17-50" title="50">                          <span class="dt">releaseProportion=</span><span class="dv">100</span>)</a>
<a class="sourceLine" id="cb17-51" title="51"></a>
<a class="sourceLine" id="cb17-52" title="52"><span class="co"># generate male release vector</span></a>
<a class="sourceLine" id="cb17-53" title="53">maleReleasesVector &lt;-<span class="st"> </span><span class="kw"><a href="../reference/generateReleaseVector.html">generateReleaseVector</a></span>(<span class="dt">driveCube=</span>cube, </a>
<a class="sourceLine" id="cb17-54" title="54">                                           <span class="dt">releasesParameters=</span>releasesParameters,</a>
<a class="sourceLine" id="cb17-55" title="55">                                           <span class="dt">sex=</span><span class="st">"M"</span>)</a>
<a class="sourceLine" id="cb17-56" title="56"></a>
<a class="sourceLine" id="cb17-57" title="57"><span class="co"># generate female release vector</span></a>
<a class="sourceLine" id="cb17-58" title="58">femaleReleasesVector &lt;-<span class="st"> </span><span class="kw"><a href="../reference/generateReleaseVector.html">generateReleaseVector</a></span>(<span class="dt">driveCube=</span>cube,</a>
<a class="sourceLine" id="cb17-59" title="59">                                             <span class="dt">releasesParameters=</span>releasesParameters,</a>
<a class="sourceLine" id="cb17-60" title="60">                                             <span class="dt">sex=</span><span class="st">"F"</span>)</a>
<a class="sourceLine" id="cb17-61" title="61"></a>
<a class="sourceLine" id="cb17-62" title="62"><span class="co"># put releases into the proper place in the release list</span></a>
<a class="sourceLine" id="cb17-63" title="63">patchReleases[[<span class="dv">1</span>]]<span class="op">$</span>maleReleases &lt;-<span class="st"> </span>maleReleasesVector</a>
<a class="sourceLine" id="cb17-64" title="64">patchReleases[[<span class="dv">1</span>]]<span class="op">$</span>femaleReleases &lt;-<span class="st"> </span>femaleReleasesVector</a>
<a class="sourceLine" id="cb17-65" title="65"></a>
<a class="sourceLine" id="cb17-66" title="66"></a>
<a class="sourceLine" id="cb17-67" title="67"><span class="co"># batch migration is disabled by setting the probability to 0</span></a>
<a class="sourceLine" id="cb17-68" title="68"><span class="co"># This is required because of the stochastic simulations, but doesn't make sense </span></a>
<a class="sourceLine" id="cb17-69" title="69"><span class="co">#  in a deterministic simulation.</span></a>
<a class="sourceLine" id="cb17-70" title="70">batchMigration &lt;-<span class="st"> </span><span class="kw"><a href="../reference/basicBatchMigration.html">basicBatchMigration</a></span>(<span class="dt">batchProbs=</span><span class="dv">0</span>,</a>
<a class="sourceLine" id="cb17-71" title="71">                                     <span class="dt">sexProbs=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(.<span class="dv">5</span>,.<span class="dv">5</span>),</a>
<a class="sourceLine" id="cb17-72" title="72">                                     <span class="dt">numPatches=</span>sitesNumber)</a>
<a class="sourceLine" id="cb17-73" title="73"></a>
<a class="sourceLine" id="cb17-74" title="74"></a>
<a class="sourceLine" id="cb17-75" title="75"><span class="co">####################</span></a>
<a class="sourceLine" id="cb17-76" title="76"><span class="co"># Combine parameters and run!</span></a>
<a class="sourceLine" id="cb17-77" title="77"><span class="co">####################</span></a>
<a class="sourceLine" id="cb17-78" title="78"><span class="co"># set MGDrivE to run deterministic</span></a>
<a class="sourceLine" id="cb17-79" title="79"><span class="kw"><a href="../reference/setupMGDrivE.html">setupMGDrivE</a></span>(<span class="dt">stochasticityON =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb17-80" title="80"><span class="co">#&gt; initializing MGDrivE</span></a>
<a class="sourceLine" id="cb17-81" title="81"></a>
<a class="sourceLine" id="cb17-82" title="82"><span class="co"># setup parameters for the network. This builds a list of parameters required for</span></a>
<a class="sourceLine" id="cb17-83" title="83"><span class="co">#  every population in the network. In ths case, we havee a network of 1 population.</span></a>
<a class="sourceLine" id="cb17-84" title="84">netPar &lt;-<span class="st"> </span><span class="kw"><a href="../reference/parameterizeMGDrivE.html">parameterizeMGDrivE</a></span>(<span class="dt">runID=</span><span class="dv">1</span>, <span class="dt">simTime=</span>tMax, <span class="dt">nPatch=</span>sitesNumber,</a>
<a class="sourceLine" id="cb17-85" title="85">                             <span class="dt">beta=</span>bioParameters<span class="op">$</span>betaK, <span class="dt">muAd=</span>bioParameters<span class="op">$</span>muAd,</a>
<a class="sourceLine" id="cb17-86" title="86">                             <span class="dt">popGrowth=</span>bioParameters<span class="op">$</span>popGrowth, <span class="dt">tEgg=</span>bioParameters<span class="op">$</span>tEgg,</a>
<a class="sourceLine" id="cb17-87" title="87">                             <span class="dt">tLarva=</span>bioParameters<span class="op">$</span>tLarva, <span class="dt">tPupa=</span>bioParameters<span class="op">$</span>tPupa,</a>
<a class="sourceLine" id="cb17-88" title="88">                             <span class="dt">AdPopEQ=</span>patchPops)</a>
<a class="sourceLine" id="cb17-89" title="89"></a>
<a class="sourceLine" id="cb17-90" title="90"><span class="co"># build network prior to run</span></a>
<a class="sourceLine" id="cb17-91" title="91">MGDrivESim &lt;-<span class="st"> </span>Network<span class="op">$</span><span class="kw">new</span>(<span class="dt">params=</span>netPar,</a>
<a class="sourceLine" id="cb17-92" title="92">                         <span class="dt">driveCube=</span>cube,</a>
<a class="sourceLine" id="cb17-93" title="93">                         <span class="dt">patchReleases=</span>patchReleases,</a>
<a class="sourceLine" id="cb17-94" title="94">                         <span class="dt">migrationMale=</span>moveMat,</a>
<a class="sourceLine" id="cb17-95" title="95">                         <span class="dt">migrationFemale=</span>moveMat,</a>
<a class="sourceLine" id="cb17-96" title="96">                         <span class="dt">migrationBatch=</span>batchMigration,</a>
<a class="sourceLine" id="cb17-97" title="97">                         <span class="dt">directory=</span>outFolder)</a>
<a class="sourceLine" id="cb17-98" title="98"><span class="co">#&gt; initializing patch:  1  of  1</span></a>
<a class="sourceLine" id="cb17-99" title="99"><span class="co"># run simulation</span></a>
<a class="sourceLine" id="cb17-100" title="100">MGDrivESim<span class="op">$</span><span class="kw">oneRun</span>(<span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb17-101" title="101"></a>
<a class="sourceLine" id="cb17-102" title="102"></a>
<a class="sourceLine" id="cb17-103" title="103"><span class="co">####################</span></a>
<a class="sourceLine" id="cb17-104" title="104"><span class="co"># Post Analysis</span></a>
<a class="sourceLine" id="cb17-105" title="105"><span class="co">####################</span></a>
<a class="sourceLine" id="cb17-106" title="106"><span class="co"># split output by patch</span></a>
<a class="sourceLine" id="cb17-107" title="107"><span class="co">#  Required for plotting later</span></a>
<a class="sourceLine" id="cb17-108" title="108"><span class="kw"><a href="../reference/splitOutput.html">splitOutput</a></span>(<span class="dt">readDir=</span>outFolder, <span class="dt">remFile =</span> <span class="ot">TRUE</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb17-109" title="109"></a>
<a class="sourceLine" id="cb17-110" title="110"><span class="co"># aggregate females by their mate choice</span></a>
<a class="sourceLine" id="cb17-111" title="111"><span class="co">#  This reduces the female file to have the same columns as the male file</span></a>
<a class="sourceLine" id="cb17-112" title="112"><span class="kw"><a href="../reference/aggregateFemales.html">aggregateFemales</a></span>(<span class="dt">readDir=</span>outFolder,<span class="dt">genotypes=</span>cube<span class="op">$</span>genotypesID,</a>
<a class="sourceLine" id="cb17-113" title="113">                <span class="dt">remFile=</span><span class="ot">TRUE</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb17-114" title="114"></a>
<a class="sourceLine" id="cb17-115" title="115"><span class="co"># plot output to see effect</span></a>
<a class="sourceLine" id="cb17-116" title="116"><span class="kw"><a href="../reference/plotMGDrivESingle.html">plotMGDrivESingle</a></span>(<span class="dt">readDir=</span>outFolder,<span class="dt">totalPop =</span> <span class="ot">TRUE</span>,<span class="dt">lwd=</span><span class="fl">3.5</span>,<span class="dt">alpha=</span><span class="dv">1</span>)</a></code></pre></div>
<p><img src="mgdrive_examples_files/figure-html/unnamed-chunk-26-1.png" width="672"></p>
<p>Analyzing the same plots as above, notice one more release and the dramatic change it has on the outcome of the simulation. This time, our releases surpassed the critical threshold and the transgenic critters took over the population. Again though, we see the slight depression in population size initially after releases, while the population is a mix of wild-type and transgenics. However, once the population is entirely transgenic, the total size returns to the equilibrium value.</p>
</div>
<div id="stochastic-above-threshold" class="section level4">
<h4 class="hasAnchor">
<a href="#stochastic-above-threshold" class="anchor"></a>Stochastic, “Above” Threshold</h4>
<p>The preceding two simulations show dramatically different results with the addition of a single release. This implies that we are near the critical threshold but gives a false impression of how distinct or certain we are of the outcome. Stochastic fluctuations can quickly push us across a threshold, even deterministic simulations implied we had passed/not passed it. Using stochastic simulations, we can put a probability on an outcome, given that we are near the threshold. Thus, we perform the same simulation as above, 6 releases, that we think puts us over the critical threshold. However, as we will see below, this is not entirely the case.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" title="1"><span class="co">####################</span></a>
<a class="sourceLine" id="cb18-2" title="2"><span class="co"># Load libraries</span></a>
<a class="sourceLine" id="cb18-3" title="3"><span class="co">####################</span></a>
<a class="sourceLine" id="cb18-4" title="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(MGDrivE)</a>
<a class="sourceLine" id="cb18-5" title="5"></a>
<a class="sourceLine" id="cb18-6" title="6"><span class="co">####################</span></a>
<a class="sourceLine" id="cb18-7" title="7"><span class="co"># Output Folder</span></a>
<a class="sourceLine" id="cb18-8" title="8"><span class="co">####################</span></a>
<a class="sourceLine" id="cb18-9" title="9"><span class="co"># This is slightly obtuse for vignette building reasons</span></a>
<a class="sourceLine" id="cb18-10" title="10"><span class="co">#  Really, all you need is a base directory, then the repetitions in folders inside that.</span></a>
<a class="sourceLine" id="cb18-11" title="11"><span class="co">#  Here, our base directory is "mgdrive", and the repetition folders are "001","002", etc.</span></a>
<a class="sourceLine" id="cb18-12" title="12"><span class="co">#  So, the final structure is "mgdrive/001","mgdrive/002", etc. </span></a>
<a class="sourceLine" id="cb18-13" title="13">outFolder &lt;-<span class="st"> "mgdrive"</span></a>
<a class="sourceLine" id="cb18-14" title="14"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/files2">dir.create</a></span>(<span class="dt">path =</span> outFolder)</a>
<a class="sourceLine" id="cb18-15" title="15"></a>
<a class="sourceLine" id="cb18-16" title="16"><span class="co">####################</span></a>
<a class="sourceLine" id="cb18-17" title="17"><span class="co"># Simulation Parameters</span></a>
<a class="sourceLine" id="cb18-18" title="18"><span class="co">####################</span></a>
<a class="sourceLine" id="cb18-19" title="19"><span class="co"># days to run the simulation, 2 years</span></a>
<a class="sourceLine" id="cb18-20" title="20">tMax &lt;-<span class="st"> </span><span class="dv">365</span><span class="op">*</span><span class="dv">3</span></a>
<a class="sourceLine" id="cb18-21" title="21"></a>
<a class="sourceLine" id="cb18-22" title="22"><span class="co"># number of Monte Carlo iterations</span></a>
<a class="sourceLine" id="cb18-23" title="23">nRep &lt;-<span class="st"> </span><span class="dv">20</span></a>
<a class="sourceLine" id="cb18-24" title="24"></a>
<a class="sourceLine" id="cb18-25" title="25"><span class="co"># each Monte Carlo iteration gets its own folder</span></a>
<a class="sourceLine" id="cb18-26" title="26">folderNames &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/file.path">file.path</a></span>(outFolder,</a>
<a class="sourceLine" id="cb18-27" title="27">                        <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/formatc">formatC</a></span>(<span class="dt">x =</span> <span class="dv">1</span><span class="op">:</span>nRep, <span class="dt">width =</span> <span class="dv">3</span>, <span class="dt">format =</span> <span class="st">"d"</span>, <span class="dt">flag =</span> <span class="st">"0"</span>))</a>
<a class="sourceLine" id="cb18-28" title="28"></a>
<a class="sourceLine" id="cb18-29" title="29"><span class="co"># entomological parameters</span></a>
<a class="sourceLine" id="cb18-30" title="30">bioParameters &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">betaK=</span><span class="dv">20</span>,<span class="dt">tEgg=</span><span class="dv">5</span>,<span class="dt">tLarva=</span><span class="dv">6</span>,<span class="dt">tPupa=</span><span class="dv">4</span>,<span class="dt">popGrowth=</span><span class="fl">1.175</span>,<span class="dt">muAd=</span><span class="fl">0.09</span>)</a>
<a class="sourceLine" id="cb18-31" title="31"></a>
<a class="sourceLine" id="cb18-32" title="32"><span class="co"># a 1-node network where mosquitoes do not leave</span></a>
<a class="sourceLine" id="cb18-33" title="33">moveMat &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">as.matrix</a></span>(<span class="dv">1</span>)</a>
<a class="sourceLine" id="cb18-34" title="34"></a>
<a class="sourceLine" id="cb18-35" title="35"><span class="co"># parameters of the population equilibrium</span></a>
<a class="sourceLine" id="cb18-36" title="36">adultPopEquilibrium &lt;-<span class="st"> </span><span class="dv">500</span></a>
<a class="sourceLine" id="cb18-37" title="37">sitesNumber &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">nrow</a></span>(moveMat)</a>
<a class="sourceLine" id="cb18-38" title="38">patchPops &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>(adultPopEquilibrium,sitesNumber)</a>
<a class="sourceLine" id="cb18-39" title="39"></a>
<a class="sourceLine" id="cb18-40" title="40"><span class="co">####################</span></a>
<a class="sourceLine" id="cb18-41" title="41"><span class="co"># Basic Inheritance pattern</span></a>
<a class="sourceLine" id="cb18-42" title="42"><span class="co">####################</span></a>
<a class="sourceLine" id="cb18-43" title="43"><span class="co"># Reciprocal translocation cube with standard (default) paraameters</span></a>
<a class="sourceLine" id="cb18-44" title="44">cube &lt;-<span class="st"> </span><span class="kw"><a href="../reference/cubeReciprocalTranslocations.html">cubeReciprocalTranslocations</a></span>()</a>
<a class="sourceLine" id="cb18-45" title="45"></a>
<a class="sourceLine" id="cb18-46" title="46"></a>
<a class="sourceLine" id="cb18-47" title="47"><span class="co">####################</span></a>
<a class="sourceLine" id="cb18-48" title="48"><span class="co"># Setup releases and batch migration</span></a>
<a class="sourceLine" id="cb18-49" title="49"><span class="co">####################</span></a>
<a class="sourceLine" id="cb18-50" title="50"><span class="co"># set up the empty release vector</span></a>
<a class="sourceLine" id="cb18-51" title="51"><span class="co">#  MGDrivE pulls things out by name</span></a>
<a class="sourceLine" id="cb18-52" title="52">patchReleases &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">replicate</a></span>(<span class="dt">n=</span>sitesNumber,</a>
<a class="sourceLine" id="cb18-53" title="53">                          <span class="dt">expr=</span>{<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">maleReleases=</span><span class="ot">NULL</span>,<span class="dt">femaleReleases=</span><span class="ot">NULL</span>)},</a>
<a class="sourceLine" id="cb18-54" title="54">                          <span class="dt">simplify=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb18-55" title="55"></a>
<a class="sourceLine" id="cb18-56" title="56"><span class="co"># choose release parameters</span></a>
<a class="sourceLine" id="cb18-57" title="57"><span class="co">#  Releases start at time 25, occur once a week, for 6 weeks.</span></a>
<a class="sourceLine" id="cb18-58" title="58"><span class="co">#  There are 10 mosquitoes relaeses every time.</span></a>
<a class="sourceLine" id="cb18-59" title="59">releasesParameters &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">releasesStart=</span><span class="dv">25</span>,</a>
<a class="sourceLine" id="cb18-60" title="60">                          <span class="dt">releasesNumber=</span><span class="dv">6</span>,</a>
<a class="sourceLine" id="cb18-61" title="61">                          <span class="dt">releasesInterval=</span><span class="dv">7</span>,</a>
<a class="sourceLine" id="cb18-62" title="62">                          <span class="dt">releaseProportion=</span><span class="dv">100</span>)</a>
<a class="sourceLine" id="cb18-63" title="63"></a>
<a class="sourceLine" id="cb18-64" title="64"><span class="co"># generate male release vector</span></a>
<a class="sourceLine" id="cb18-65" title="65">maleReleasesVector &lt;-<span class="st"> </span><span class="kw"><a href="../reference/generateReleaseVector.html">generateReleaseVector</a></span>(<span class="dt">driveCube=</span>cube, </a>
<a class="sourceLine" id="cb18-66" title="66">                                           <span class="dt">releasesParameters=</span>releasesParameters,</a>
<a class="sourceLine" id="cb18-67" title="67">                                           <span class="dt">sex=</span><span class="st">"M"</span>)</a>
<a class="sourceLine" id="cb18-68" title="68"></a>
<a class="sourceLine" id="cb18-69" title="69"><span class="co"># generate female release vector</span></a>
<a class="sourceLine" id="cb18-70" title="70">femaleReleasesVector &lt;-<span class="st"> </span><span class="kw"><a href="../reference/generateReleaseVector.html">generateReleaseVector</a></span>(<span class="dt">driveCube=</span>cube,</a>
<a class="sourceLine" id="cb18-71" title="71">                                             <span class="dt">releasesParameters=</span>releasesParameters,</a>
<a class="sourceLine" id="cb18-72" title="72">                                             <span class="dt">sex=</span><span class="st">"F"</span>)</a>
<a class="sourceLine" id="cb18-73" title="73"></a>
<a class="sourceLine" id="cb18-74" title="74"><span class="co"># put releases into the proper place in the release list</span></a>
<a class="sourceLine" id="cb18-75" title="75">patchReleases[[<span class="dv">1</span>]]<span class="op">$</span>maleReleases &lt;-<span class="st"> </span>maleReleasesVector</a>
<a class="sourceLine" id="cb18-76" title="76">patchReleases[[<span class="dv">1</span>]]<span class="op">$</span>femaleReleases &lt;-<span class="st"> </span>femaleReleasesVector</a>
<a class="sourceLine" id="cb18-77" title="77"></a>
<a class="sourceLine" id="cb18-78" title="78"></a>
<a class="sourceLine" id="cb18-79" title="79"><span class="co"># batch migration is disabled by setting the probability to 0</span></a>
<a class="sourceLine" id="cb18-80" title="80"><span class="co"># This is required because of the stochastic simulations, but doesn't make sense </span></a>
<a class="sourceLine" id="cb18-81" title="81"><span class="co">#  in a deterministic simulation.</span></a>
<a class="sourceLine" id="cb18-82" title="82">batchMigration &lt;-<span class="st"> </span><span class="kw"><a href="../reference/basicBatchMigration.html">basicBatchMigration</a></span>(<span class="dt">batchProbs=</span><span class="dv">0</span>,</a>
<a class="sourceLine" id="cb18-83" title="83">                                     <span class="dt">sexProbs=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(.<span class="dv">5</span>,.<span class="dv">5</span>),</a>
<a class="sourceLine" id="cb18-84" title="84">                                     <span class="dt">numPatches=</span>sitesNumber)</a>
<a class="sourceLine" id="cb18-85" title="85"></a>
<a class="sourceLine" id="cb18-86" title="86"></a>
<a class="sourceLine" id="cb18-87" title="87"><span class="co">####################</span></a>
<a class="sourceLine" id="cb18-88" title="88"><span class="co"># Combine parameters and run!</span></a>
<a class="sourceLine" id="cb18-89" title="89"><span class="co">####################</span></a>
<a class="sourceLine" id="cb18-90" title="90"><span class="co"># set MGDrivE to run stochastic</span></a>
<a class="sourceLine" id="cb18-91" title="91"><span class="kw"><a href="../reference/setupMGDrivE.html">setupMGDrivE</a></span>(<span class="dt">stochasticityON =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb18-92" title="92"><span class="co">#&gt; initializing MGDrivE</span></a>
<a class="sourceLine" id="cb18-93" title="93"></a>
<a class="sourceLine" id="cb18-94" title="94"><span class="co"># setup parameters for the network. This builds a list of parameters required for</span></a>
<a class="sourceLine" id="cb18-95" title="95"><span class="co">#  every population in the network. In ths case, we havee a network of 1 population.</span></a>
<a class="sourceLine" id="cb18-96" title="96">netPar &lt;-<span class="st"> </span><span class="kw"><a href="../reference/parameterizeMGDrivE.html">parameterizeMGDrivE</a></span>(<span class="dt">runID=</span><span class="dv">1</span>, <span class="dt">simTime=</span>tMax, <span class="dt">nPatch=</span>sitesNumber,</a>
<a class="sourceLine" id="cb18-97" title="97">                             <span class="dt">beta=</span>bioParameters<span class="op">$</span>betaK, <span class="dt">muAd=</span>bioParameters<span class="op">$</span>muAd,</a>
<a class="sourceLine" id="cb18-98" title="98">                             <span class="dt">popGrowth=</span>bioParameters<span class="op">$</span>popGrowth, <span class="dt">tEgg=</span>bioParameters<span class="op">$</span>tEgg,</a>
<a class="sourceLine" id="cb18-99" title="99">                             <span class="dt">tLarva=</span>bioParameters<span class="op">$</span>tLarva, <span class="dt">tPupa=</span>bioParameters<span class="op">$</span>tPupa,</a>
<a class="sourceLine" id="cb18-100" title="100">                             <span class="dt">AdPopEQ=</span>patchPops)</a>
<a class="sourceLine" id="cb18-101" title="101"></a>
<a class="sourceLine" id="cb18-102" title="102"><span class="co"># build network prior to run</span></a>
<a class="sourceLine" id="cb18-103" title="103">MGDrivESim &lt;-<span class="st"> </span>Network<span class="op">$</span><span class="kw">new</span>(<span class="dt">params=</span>netPar,</a>
<a class="sourceLine" id="cb18-104" title="104">                         <span class="dt">driveCube=</span>cube,</a>
<a class="sourceLine" id="cb18-105" title="105">                         <span class="dt">patchReleases=</span>patchReleases,</a>
<a class="sourceLine" id="cb18-106" title="106">                         <span class="dt">migrationMale=</span>moveMat,</a>
<a class="sourceLine" id="cb18-107" title="107">                         <span class="dt">migrationFemale=</span>moveMat,</a>
<a class="sourceLine" id="cb18-108" title="108">                         <span class="dt">migrationBatch=</span>batchMigration,</a>
<a class="sourceLine" id="cb18-109" title="109">                         <span class="dt">directory=</span>folderNames)</a>
<a class="sourceLine" id="cb18-110" title="110"><span class="co">#&gt; initializing patch:  1  of  1</span></a>
<a class="sourceLine" id="cb18-111" title="111"><span class="co"># run simulation</span></a>
<a class="sourceLine" id="cb18-112" title="112">MGDrivESim<span class="op">$</span><span class="kw">multRun</span>(<span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb18-113" title="113"></a>
<a class="sourceLine" id="cb18-114" title="114"></a>
<a class="sourceLine" id="cb18-115" title="115"><span class="co">####################</span></a>
<a class="sourceLine" id="cb18-116" title="116"><span class="co"># Post Analysis</span></a>
<a class="sourceLine" id="cb18-117" title="117"><span class="co">####################</span></a>
<a class="sourceLine" id="cb18-118" title="118"><span class="co"># First, split output by patch</span></a>
<a class="sourceLine" id="cb18-119" title="119"><span class="co"># Second, aggregate females by their mate choice</span></a>
<a class="sourceLine" id="cb18-120" title="120"><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>nRep){</a>
<a class="sourceLine" id="cb18-121" title="121"> <span class="kw"><a href="../reference/splitOutput.html">splitOutput</a></span>(<span class="dt">readDir=</span>folderNames[i], <span class="dt">remFile =</span> <span class="ot">TRUE</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb18-122" title="122"> <span class="kw"><a href="../reference/aggregateFemales.html">aggregateFemales</a></span>(<span class="dt">readDir=</span>folderNames[i],<span class="dt">genotypes=</span>cube<span class="op">$</span>genotypesID,</a>
<a class="sourceLine" id="cb18-123" title="123">                  <span class="dt">remFile=</span><span class="ot">TRUE</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb18-124" title="124">}</a>
<a class="sourceLine" id="cb18-125" title="125"></a>
<a class="sourceLine" id="cb18-126" title="126"><span class="co"># plot output of first run to see effect</span></a>
<a class="sourceLine" id="cb18-127" title="127"><span class="kw"><a href="../reference/plotMGDrivESingle.html">plotMGDrivESingle</a></span>(<span class="dt">readDir=</span>folderNames[<span class="dv">1</span>],<span class="dt">lwd=</span><span class="fl">3.5</span>,<span class="dt">alpha=</span><span class="dv">1</span>)</a></code></pre></div>
<p><img src="mgdrive_examples_files/figure-html/unnamed-chunk-28-1.png" width="672"></p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" title="1"></a>
<a class="sourceLine" id="cb19-2" title="2"><span class="co"># plot all 50 repetitions together</span></a>
<a class="sourceLine" id="cb19-3" title="3"><span class="kw"><a href="../reference/plotMGDrivEMult.html">plotMGDrivEMult</a></span>(<span class="dt">readDir=</span>outFolder,<span class="dt">lwd=</span><span class="fl">0.35</span>,<span class="dt">alpha=</span><span class="fl">0.75</span>)</a></code></pre></div>
<p><img src="mgdrive_examples_files/figure-html/unnamed-chunk-28-2.png" width="672"></p>
<p>We perform 20 repetitions of our stochastic simulation. In an experimental setting, we should perform many more repetitions (the number of repetitions is related to the precision of your estimate. 100 repetitions would provide 1% resolution, while our 20 repetitions here only provides 5% resolution), but outcome resolution needs to be balanced against run time.</p>
<p>First we plot a single repetition. Wild-type (aabb) critters are in orange, and transgenic (AABB) critters are in blue. We see that there is some fluctuation, but the final result is the same as the deterministic simulation. This is a good sign, it implies we are well above the critical threshold, but it is only one repetitions.</p>
<p>Plotting all 20 repetitions, we see that the results are not as clear as before. We have extended the simulation to 1095 days to allow all of the repetitions to achieve their final trajectory. First, near the critical threshold, stochastic fluctuations dramatically slow the population dynamics. The deterministic simulations were finished in two years, but stochastic simulations took nearly three years for all of the trajectories to reach equilibrium. Second, we notice that not all of the simulations ended with a fully transgenic population. In fact, only 60% of the repetitions ended with a fully transgenic population. More repetitions would provide a more precise answer, but now we know that this set of parameters is only 60 +/-5% effective.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#landscape-setup">Landscape Setup</a></li>
      <li><a href="#inheritance-simulations">Inheritance Simulations</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="https://chipdelmal.github.io">Héctor Manuel Sánchez Castellanos</a>, <a href="https://github.com/GilChrist19">Jared Bennett</a>, <a href="https://slwu89.github.io">Sean Wu</a>, <a href="https://www.marshalllab.com">John M. Marshall</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
