<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>MGDrivE Run Example • MGDrivE</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="MGDrivE Run Example">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MGDrivE</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/mgdrive_examples.html">Example Simulations</a>
    </li>
    <li>
      <a href="../articles/mgdrive_math.html">Mathematical Description</a>
    </li>
    <li>
      <a href="../articles/mgdrive_run.html">Detailed Run Example</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/MarshallLab/MGDrivE">
    <span class="fa fa-lg fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>MGDrivE Run Example</h1>
            
      
      
      <div class="hidden name"><code>mgdrive_run.Rmd</code></div>

    </div>

    
    
<style type="text/css">
pre {
  white-space: pre !important;
  overflow-y: scroll !important;
  max-height: 25vh !important;
}

# reference
#  https://stackoverflow.com/questions/41135085/how-to-make-vertical-scrollbar-appear-in-rmarkdown-code-chunks-html-view
</style>
<div id="mgdrive-one-run" class="section level2">
<h2 class="hasAnchor">
<a href="#mgdrive-one-run" class="anchor"></a>MGDrivE, One Run</h2>
<p>The other <a href="mgdrive_examples.html">vignette</a> has many examples of calculating movement matrices, different simulations, and some theory for comparison. However, it does not go through each step of setting up, running, and analyzing a simulation. This vignette is designed to go through every step of a simulation, exploring why each step is performed, some thoughts on how they can be modified, and looking at the output (if any).</p>
<div id="mendelian-inheritance-stochastic-multiple-populations" class="section level3">
<h3 class="hasAnchor">
<a href="#mendelian-inheritance-stochastic-multiple-populations" class="anchor"></a>Mendelian Inheritance, Stochastic, Multiple Populations</h3>
<p>For our example, we will take the Stochastic, Small Migration example from the other vignette as a starting point. We will expand it to 5 populations, so that that code is a bit longer and the file structure becomes more obvious.</p>
<div id="full-code" class="section level4">
<h4 class="hasAnchor">
<a href="#full-code" class="anchor"></a>Full Code</h4>
<p>For reference, here is the full code for our example. It will be explained piece-by-piece below.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1"><span class="co">####################</span></a>
<a class="sourceLine" id="cb1-2" title="2"><span class="co"># Load libraries</span></a>
<a class="sourceLine" id="cb1-3" title="3"><span class="co">####################</span></a>
<a class="sourceLine" id="cb1-4" title="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(MGDrivE)</a>
<a class="sourceLine" id="cb1-5" title="5"></a>
<a class="sourceLine" id="cb1-6" title="6"><span class="co">####################</span></a>
<a class="sourceLine" id="cb1-7" title="7"><span class="co"># Output Folder</span></a>
<a class="sourceLine" id="cb1-8" title="8"><span class="co">####################</span></a>
<a class="sourceLine" id="cb1-9" title="9">outFolder &lt;-<span class="st"> "mgdrive"</span></a>
<a class="sourceLine" id="cb1-10" title="10"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/files2">dir.create</a></span>(<span class="dt">path =</span> outFolder)</a>
<a class="sourceLine" id="cb1-11" title="11"></a>
<a class="sourceLine" id="cb1-12" title="12"><span class="co">####################</span></a>
<a class="sourceLine" id="cb1-13" title="13"><span class="co"># Simulation Parameters</span></a>
<a class="sourceLine" id="cb1-14" title="14"><span class="co">####################</span></a>
<a class="sourceLine" id="cb1-15" title="15"><span class="co"># days to run the simulation</span></a>
<a class="sourceLine" id="cb1-16" title="16">tMax &lt;-<span class="st"> </span><span class="dv">365</span></a>
<a class="sourceLine" id="cb1-17" title="17"></a>
<a class="sourceLine" id="cb1-18" title="18"><span class="co"># number of Monte Carlo iterations</span></a>
<a class="sourceLine" id="cb1-19" title="19">nRep &lt;-<span class="st"> </span><span class="dv">5</span></a>
<a class="sourceLine" id="cb1-20" title="20"></a>
<a class="sourceLine" id="cb1-21" title="21"><span class="co"># each Monte Carlo iteration gets its own folder</span></a>
<a class="sourceLine" id="cb1-22" title="22">folderNames &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/file.path">file.path</a></span>(outFolder,</a>
<a class="sourceLine" id="cb1-23" title="23">                        <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/formatc">formatC</a></span>(<span class="dt">x =</span> <span class="dv">1</span><span class="op">:</span>nRep, <span class="dt">width =</span> <span class="dv">3</span>, <span class="dt">format =</span> <span class="st">"d"</span>, <span class="dt">flag =</span> <span class="st">"0"</span>))</a>
<a class="sourceLine" id="cb1-24" title="24"></a>
<a class="sourceLine" id="cb1-25" title="25"><span class="co"># entomological parameters</span></a>
<a class="sourceLine" id="cb1-26" title="26">bioParameters &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">betaK=</span><span class="dv">20</span>,<span class="dt">tEgg=</span><span class="dv">5</span>,<span class="dt">tLarva=</span><span class="dv">6</span>,<span class="dt">tPupa=</span><span class="dv">4</span>,<span class="dt">popGrowth=</span><span class="fl">1.175</span>,<span class="dt">muAd=</span><span class="fl">0.09</span>)</a>
<a class="sourceLine" id="cb1-27" title="27"></a>
<a class="sourceLine" id="cb1-28" title="28"><span class="co"># a 5-node network with 5% per day migration rate</span></a>
<a class="sourceLine" id="cb1-29" title="29">sitesNumber &lt;-<span class="st"> </span><span class="dv">5</span></a>
<a class="sourceLine" id="cb1-30" title="30">adultPopEquilibrium &lt;-<span class="st"> </span><span class="dv">500</span></a>
<a class="sourceLine" id="cb1-31" title="31">patchPops &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>(adultPopEquilibrium,sitesNumber)</a>
<a class="sourceLine" id="cb1-32" title="32"></a>
<a class="sourceLine" id="cb1-33" title="33"><span class="co"># auxiliary function</span></a>
<a class="sourceLine" id="cb1-34" title="34">triDiag &lt;-<span class="st"> </span><span class="cf">function</span>(upper, lower){</a>
<a class="sourceLine" id="cb1-35" title="35">  </a>
<a class="sourceLine" id="cb1-36" title="36">  <span class="co"># return matrix</span></a>
<a class="sourceLine" id="cb1-37" title="37">  retMat &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="dt">data =</span> <span class="dv">0</span>, <span class="dt">nrow =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(upper) <span class="op">+</span><span class="st"> </span><span class="dv">1</span>, <span class="dt">ncol =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(upper) <span class="op">+</span><span class="st"> </span><span class="dv">1</span>)</a>
<a class="sourceLine" id="cb1-38" title="38">  </a>
<a class="sourceLine" id="cb1-39" title="39">  <span class="co"># set index values for upper/lower triangles</span></a>
<a class="sourceLine" id="cb1-40" title="40">  indx &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(upper)</a>
<a class="sourceLine" id="cb1-41" title="41">  </a>
<a class="sourceLine" id="cb1-42" title="42">  <span class="co"># set forward/backward migration using matrix access</span></a>
<a class="sourceLine" id="cb1-43" title="43">  retMat[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cbind">cbind</a></span>(indx<span class="op">+</span><span class="dv">1</span>,indx)] &lt;-<span class="st"> </span>lower</a>
<a class="sourceLine" id="cb1-44" title="44">  retMat[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cbind">cbind</a></span>(indx,indx<span class="op">+</span><span class="dv">1</span>)] &lt;-<span class="st"> </span>upper</a>
<a class="sourceLine" id="cb1-45" title="45">  </a>
<a class="sourceLine" id="cb1-46" title="46">  <span class="co"># set stay probs</span></a>
<a class="sourceLine" id="cb1-47" title="47">  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/diag">diag</a></span>(<span class="dt">x =</span> retMat) &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">-</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colSums">rowSums</a></span>(<span class="dt">x =</span> retMat)</a>
<a class="sourceLine" id="cb1-48" title="48">  </a>
<a class="sourceLine" id="cb1-49" title="49">  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/function">return</a></span>(retMat)</a>
<a class="sourceLine" id="cb1-50" title="50">}</a>
<a class="sourceLine" id="cb1-51" title="51"></a>
<a class="sourceLine" id="cb1-52" title="52"><span class="co"># fill movement matrix</span></a>
<a class="sourceLine" id="cb1-53" title="53"><span class="co">#  Remember, rows need to sum to 1.</span></a>
<a class="sourceLine" id="cb1-54" title="54">moveMat &lt;-<span class="st"> </span><span class="kw">triDiag</span>(<span class="dt">upper =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep.int</a></span>(<span class="dt">x =</span> <span class="fl">0.05</span>, <span class="dt">times =</span> sitesNumber<span class="dv">-1</span>),</a>
<a class="sourceLine" id="cb1-55" title="55">                   <span class="dt">lower =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep.int</a></span>(<span class="dt">x =</span> <span class="fl">0.05</span>, <span class="dt">times =</span> sitesNumber<span class="dv">-1</span>))</a>
<a class="sourceLine" id="cb1-56" title="56"></a>
<a class="sourceLine" id="cb1-57" title="57"><span class="co"># batch migration is disabled by setting the probability to 0</span></a>
<a class="sourceLine" id="cb1-58" title="58">batchMigration &lt;-<span class="st"> </span><span class="kw"><a href="../reference/basicBatchMigration.html">basicBatchMigration</a></span>(<span class="dt">batchProbs=</span><span class="dv">0</span>,</a>
<a class="sourceLine" id="cb1-59" title="59">                                     <span class="dt">sexProbs=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(.<span class="dv">5</span>,.<span class="dv">5</span>),</a>
<a class="sourceLine" id="cb1-60" title="60">                                     <span class="dt">numPatches=</span>sitesNumber)</a>
<a class="sourceLine" id="cb1-61" title="61"></a>
<a class="sourceLine" id="cb1-62" title="62"><span class="co">####################</span></a>
<a class="sourceLine" id="cb1-63" title="63"><span class="co"># Basic Inheritance pattern</span></a>
<a class="sourceLine" id="cb1-64" title="64"><span class="co">####################</span></a>
<a class="sourceLine" id="cb1-65" title="65"><span class="co"># Mendelian cube with standard (default) paraameters</span></a>
<a class="sourceLine" id="cb1-66" title="66">cube &lt;-<span class="st"> </span><span class="kw"><a href="../reference/cubeMendelian.html">cubeMendelian</a></span>()</a>
<a class="sourceLine" id="cb1-67" title="67"></a>
<a class="sourceLine" id="cb1-68" title="68"></a>
<a class="sourceLine" id="cb1-69" title="69"><span class="co">####################</span></a>
<a class="sourceLine" id="cb1-70" title="70"><span class="co"># Setup releases and batch migration</span></a>
<a class="sourceLine" id="cb1-71" title="71"><span class="co">####################</span></a>
<a class="sourceLine" id="cb1-72" title="72"><span class="co"># set up the empty release vector</span></a>
<a class="sourceLine" id="cb1-73" title="73"><span class="co">#  MGDrivE pulls things out by name</span></a>
<a class="sourceLine" id="cb1-74" title="74">patchReleases &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">replicate</a></span>(<span class="dt">n=</span>sitesNumber,</a>
<a class="sourceLine" id="cb1-75" title="75">                          <span class="dt">expr=</span>{<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">maleReleases=</span><span class="ot">NULL</span>,<span class="dt">femaleReleases=</span><span class="ot">NULL</span>)},</a>
<a class="sourceLine" id="cb1-76" title="76">                          <span class="dt">simplify=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb1-77" title="77"></a>
<a class="sourceLine" id="cb1-78" title="78"><span class="co"># choose release parameters</span></a>
<a class="sourceLine" id="cb1-79" title="79"><span class="co">#  Releases start at time 25, occur every day, for 5 day.</span></a>
<a class="sourceLine" id="cb1-80" title="80"><span class="co">#  There are 50 mosquitoes relaeses every time.</span></a>
<a class="sourceLine" id="cb1-81" title="81">releasesParameters &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">releasesStart=</span><span class="dv">25</span>,</a>
<a class="sourceLine" id="cb1-82" title="82">                          <span class="dt">releasesNumber=</span><span class="dv">10</span>,</a>
<a class="sourceLine" id="cb1-83" title="83">                          <span class="dt">releasesInterval=</span><span class="dv">7</span>,</a>
<a class="sourceLine" id="cb1-84" title="84">                          <span class="dt">releaseProportion=</span><span class="dv">100</span>)</a>
<a class="sourceLine" id="cb1-85" title="85"></a>
<a class="sourceLine" id="cb1-86" title="86"><span class="co"># generate male release vector</span></a>
<a class="sourceLine" id="cb1-87" title="87">maleReleasesVector &lt;-<span class="st"> </span><span class="kw"><a href="../reference/generateReleaseVector.html">generateReleaseVector</a></span>(<span class="dt">driveCube=</span>cube, </a>
<a class="sourceLine" id="cb1-88" title="88">                                           <span class="dt">releasesParameters=</span>releasesParameters,</a>
<a class="sourceLine" id="cb1-89" title="89">                                           <span class="dt">sex=</span><span class="st">"M"</span>)</a>
<a class="sourceLine" id="cb1-90" title="90"></a>
<a class="sourceLine" id="cb1-91" title="91"><span class="co"># generate female release vector</span></a>
<a class="sourceLine" id="cb1-92" title="92">femaleReleasesVector &lt;-<span class="st"> </span><span class="kw"><a href="../reference/generateReleaseVector.html">generateReleaseVector</a></span>(<span class="dt">driveCube=</span>cube,</a>
<a class="sourceLine" id="cb1-93" title="93">                                             <span class="dt">releasesParameters=</span>releasesParameters,</a>
<a class="sourceLine" id="cb1-94" title="94">                                             <span class="dt">sex=</span><span class="st">"F"</span>)</a>
<a class="sourceLine" id="cb1-95" title="95"></a>
<a class="sourceLine" id="cb1-96" title="96"><span class="co"># put releases into the proper place in the release list</span></a>
<a class="sourceLine" id="cb1-97" title="97">patchReleases[[<span class="dv">1</span>]]<span class="op">$</span>maleReleases &lt;-<span class="st"> </span>maleReleasesVector</a>
<a class="sourceLine" id="cb1-98" title="98">patchReleases[[<span class="dv">1</span>]]<span class="op">$</span>femaleReleases &lt;-<span class="st"> </span>femaleReleasesVector</a>
<a class="sourceLine" id="cb1-99" title="99"></a>
<a class="sourceLine" id="cb1-100" title="100"></a>
<a class="sourceLine" id="cb1-101" title="101"><span class="co">####################</span></a>
<a class="sourceLine" id="cb1-102" title="102"><span class="co"># Combine parameters and run!</span></a>
<a class="sourceLine" id="cb1-103" title="103"><span class="co">####################</span></a>
<a class="sourceLine" id="cb1-104" title="104"><span class="co"># setup parameters for the network. This builds a list of parameters required for</span></a>
<a class="sourceLine" id="cb1-105" title="105"><span class="co">#  every population in the network.</span></a>
<a class="sourceLine" id="cb1-106" title="106">netPar &lt;-<span class="st"> </span><span class="kw"><a href="../reference/parameterizeMGDrivE.html">parameterizeMGDrivE</a></span>(<span class="dt">runID=</span><span class="dv">1</span>, <span class="dt">simTime=</span>tMax, <span class="dt">nPatch=</span>sitesNumber,</a>
<a class="sourceLine" id="cb1-107" title="107">                             <span class="dt">beta=</span>bioParameters<span class="op">$</span>betaK, <span class="dt">muAd=</span>bioParameters<span class="op">$</span>muAd,</a>
<a class="sourceLine" id="cb1-108" title="108">                             <span class="dt">popGrowth=</span>bioParameters<span class="op">$</span>popGrowth, <span class="dt">tEgg=</span>bioParameters<span class="op">$</span>tEgg,</a>
<a class="sourceLine" id="cb1-109" title="109">                             <span class="dt">tLarva=</span>bioParameters<span class="op">$</span>tLarva, <span class="dt">tPupa=</span>bioParameters<span class="op">$</span>tPupa,</a>
<a class="sourceLine" id="cb1-110" title="110">                             <span class="dt">AdPopEQ=</span>patchPops)</a>
<a class="sourceLine" id="cb1-111" title="111"></a>
<a class="sourceLine" id="cb1-112" title="112"><span class="co"># set MGDrivE to run stochastic</span></a>
<a class="sourceLine" id="cb1-113" title="113"><span class="kw"><a href="../reference/setupMGDrivE.html">setupMGDrivE</a></span>(<span class="dt">stochasticityON =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb1-114" title="114"></a>
<a class="sourceLine" id="cb1-115" title="115"><span class="co"># build network prior to run</span></a>
<a class="sourceLine" id="cb1-116" title="116">MGDrivESim &lt;-<span class="st"> </span>Network<span class="op">$</span><span class="kw">new</span>(<span class="dt">params=</span>netPar,</a>
<a class="sourceLine" id="cb1-117" title="117">                         <span class="dt">driveCube=</span>cube,</a>
<a class="sourceLine" id="cb1-118" title="118">                         <span class="dt">patchReleases=</span>patchReleases,</a>
<a class="sourceLine" id="cb1-119" title="119">                         <span class="dt">migrationMale=</span>moveMat,</a>
<a class="sourceLine" id="cb1-120" title="120">                         <span class="dt">migrationFemale=</span>moveMat,</a>
<a class="sourceLine" id="cb1-121" title="121">                         <span class="dt">migrationBatch=</span>batchMigration,</a>
<a class="sourceLine" id="cb1-122" title="122">                         <span class="dt">directory=</span>folderNames)</a>
<a class="sourceLine" id="cb1-123" title="123"><span class="co"># run simulation</span></a>
<a class="sourceLine" id="cb1-124" title="124">MGDrivESim<span class="op">$</span><span class="kw">multRun</span>(<span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb1-125" title="125"></a>
<a class="sourceLine" id="cb1-126" title="126"></a>
<a class="sourceLine" id="cb1-127" title="127"><span class="co">####################</span></a>
<a class="sourceLine" id="cb1-128" title="128"><span class="co"># Post Analysis</span></a>
<a class="sourceLine" id="cb1-129" title="129"><span class="co">####################</span></a>
<a class="sourceLine" id="cb1-130" title="130"><span class="co"># First, split output by patch</span></a>
<a class="sourceLine" id="cb1-131" title="131"><span class="co"># Second, aggregate females by their mate choice</span></a>
<a class="sourceLine" id="cb1-132" title="132"><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>nRep){</a>
<a class="sourceLine" id="cb1-133" title="133"> <span class="kw"><a href="../reference/splitOutput.html">splitOutput</a></span>(<span class="dt">readDir=</span>folderNames[i],<span class="dt">remFile =</span> <span class="ot">TRUE</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb1-134" title="134"> <span class="kw"><a href="../reference/aggregateFemales.html">aggregateFemales</a></span>(<span class="dt">readDir=</span>folderNames[i],<span class="dt">genotypes=</span>cube<span class="op">$</span>genotypesID,</a>
<a class="sourceLine" id="cb1-135" title="135">                  <span class="dt">remFile=</span><span class="ot">TRUE</span>,<span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb1-136" title="136">}</a>
<a class="sourceLine" id="cb1-137" title="137"></a>
<a class="sourceLine" id="cb1-138" title="138"><span class="co"># plot output of first run to see effect</span></a>
<a class="sourceLine" id="cb1-139" title="139"><span class="kw"><a href="../reference/plotMGDrivESingle.html">plotMGDrivESingle</a></span>(<span class="dt">readDir=</span>folderNames[<span class="dv">1</span>],<span class="dt">totalPop =</span> <span class="ot">TRUE</span>,<span class="dt">lwd=</span><span class="fl">3.5</span>,<span class="dt">alpha=</span><span class="dv">1</span>)</a>
<a class="sourceLine" id="cb1-140" title="140"></a>
<a class="sourceLine" id="cb1-141" title="141"><span class="co"># plot all 5 repetitions together</span></a>
<a class="sourceLine" id="cb1-142" title="142"><span class="kw"><a href="../reference/plotMGDrivEMult.html">plotMGDrivEMult</a></span>(<span class="dt">readDir=</span>outFolder,<span class="dt">lwd=</span><span class="fl">0.35</span>,<span class="dt">alpha=</span><span class="fl">0.75</span>)</a></code></pre></div>
</div>
<div id="load-libraries" class="section level4">
<h4 class="hasAnchor">
<a href="#load-libraries" class="anchor"></a>Load Libraries</h4>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" title="1"><span class="co">####################</span></a>
<a class="sourceLine" id="cb2-2" title="2"><span class="co"># Load libraries</span></a>
<a class="sourceLine" id="cb2-3" title="3"><span class="co">####################</span></a>
<a class="sourceLine" id="cb2-4" title="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(MGDrivE)</a>
<a class="sourceLine" id="cb2-5" title="5"><span class="co">#&gt; Loading required package: data.table</span></a>
<a class="sourceLine" id="cb2-6" title="6"><span class="co">#&gt; Loading MGDrivE: Mosquito Gene Drive Explorer</span></a></code></pre></div>
<p>Step one, load <em>MGDrivE</em>.</p>
</div>
<div id="setup-folder-directory" class="section level4">
<h4 class="hasAnchor">
<a href="#setup-folder-directory" class="anchor"></a>Setup Folder Directory</h4>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" title="1"><span class="co">####################</span></a>
<a class="sourceLine" id="cb3-2" title="2"><span class="co"># Output Folder</span></a>
<a class="sourceLine" id="cb3-3" title="3"><span class="co">####################</span></a>
<a class="sourceLine" id="cb3-4" title="4">outFolder &lt;-<span class="st"> "mgdrive"</span></a>
<a class="sourceLine" id="cb3-5" title="5"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/files2">dir.create</a></span>(<span class="dt">path =</span> outFolder)</a></code></pre></div>
<p>Here, we setup the folder for simulation data. Since this is a local folder, it doesn’t exist, and needs to be created. If you already have a folder prepared, make sure it is empty. <em>MGDrivE</em> gives a warning if there are other things in the folder (to prevent overwriting of data), but will run fine if the folder is not empty.</p>
</div>
<div id="simulation-parameters" class="section level4">
<h4 class="hasAnchor">
<a href="#simulation-parameters" class="anchor"></a>Simulation Parameters</h4>
<p>Next, there are several parameters that need to be set for a simulation. The run time and the number of repetitions are crucial. Once the number of repetitions has been chosen, the folder names can be created.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" title="1"></a>
<a class="sourceLine" id="cb4-2" title="2"><span class="co">####################</span></a>
<a class="sourceLine" id="cb4-3" title="3"><span class="co"># Simulation Parameters</span></a>
<a class="sourceLine" id="cb4-4" title="4"><span class="co">####################</span></a>
<a class="sourceLine" id="cb4-5" title="5"><span class="co"># days to run the simulation</span></a>
<a class="sourceLine" id="cb4-6" title="6">tMax &lt;-<span class="st"> </span><span class="dv">365</span></a>
<a class="sourceLine" id="cb4-7" title="7"></a>
<a class="sourceLine" id="cb4-8" title="8"><span class="co"># number of Monte Carlo iterations</span></a>
<a class="sourceLine" id="cb4-9" title="9">nRep &lt;-<span class="st"> </span><span class="dv">5</span></a>
<a class="sourceLine" id="cb4-10" title="10"></a>
<a class="sourceLine" id="cb4-11" title="11"><span class="co"># each Monte Carlo iteration gets its own folder</span></a>
<a class="sourceLine" id="cb4-12" title="12">folderNames &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/file.path">file.path</a></span>(outFolder,</a>
<a class="sourceLine" id="cb4-13" title="13">                        <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/formatc">formatC</a></span>(<span class="dt">x =</span> <span class="dv">1</span><span class="op">:</span>nRep, <span class="dt">width =</span> <span class="dv">3</span>, <span class="dt">format =</span> <span class="st">"d"</span>, <span class="dt">flag =</span> <span class="st">"0"</span>))</a>
<a class="sourceLine" id="cb4-14" title="14"></a>
<a class="sourceLine" id="cb4-15" title="15">folderNames</a>
<a class="sourceLine" id="cb4-16" title="16"><span class="co">#&gt; [1] "mgdrive/001" "mgdrive/002" "mgdrive/003" "mgdrive/004" "mgdrive/005"</span></a></code></pre></div>
<p>Here, folder names reflect their purpose - holding each repetition of our stochastic simulation. The precise names don’t matter later on, but they do need to be unique (or data gets overwritten, <em>MGDrivE</em> will not throw a warning about this), and in a location that your user account has permission to create directories.</p>
<p>The biological parameters were chosen for <em>Aedes aegypti</em> in our original simulations. These parameters represent:</p>
<ul>
<li>
<code>betaK</code>: Daily number of eggs laid per female mosquitoe</li>
<li>
<code>tEgg</code>: Number of days spent in the egg stage</li>
<li>
<code>tLarva</code>: Number of days spent in the larval stage</li>
<li>
<code>tPupa</code>: Number of days spent in the pupa stage</li>
<li>
<code>popGrowth</code>: Population growth per generation<br>
</li>
<li>
<code>muAD</code>: Daily death rate for adult mosquitoes</li>
</ul>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" title="1"><span class="co"># entomological parameters</span></a>
<a class="sourceLine" id="cb5-2" title="2">bioParameters &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">betaK=</span><span class="dv">20</span>,<span class="dt">tEgg=</span><span class="dv">5</span>,<span class="dt">tLarva=</span><span class="dv">6</span>,<span class="dt">tPupa=</span><span class="dv">4</span>,<span class="dt">popGrowth=</span><span class="fl">1.175</span>,<span class="dt">muAd=</span><span class="fl">0.09</span>)</a></code></pre></div>
<p>Next, we set the number of populations in the simulation. Here, we chose 5 populations, each containing 500 mosquitoes. After that, we setup the landscape.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" title="1"><span class="co"># a 5-node network with 5% per day migration rate</span></a>
<a class="sourceLine" id="cb6-2" title="2">sitesNumber &lt;-<span class="st"> </span><span class="dv">5</span></a>
<a class="sourceLine" id="cb6-3" title="3">adultPopEquilibrium &lt;-<span class="st"> </span><span class="dv">500</span></a>
<a class="sourceLine" id="cb6-4" title="4">patchPops &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>(adultPopEquilibrium,sitesNumber)</a>
<a class="sourceLine" id="cb6-5" title="5"></a>
<a class="sourceLine" id="cb6-6" title="6"><span class="co"># landscape</span></a>
<a class="sourceLine" id="cb6-7" title="7"><span class="co"># auxiliary function</span></a>
<a class="sourceLine" id="cb6-8" title="8">triDiag &lt;-<span class="st"> </span><span class="cf">function</span>(upper, lower){</a>
<a class="sourceLine" id="cb6-9" title="9">  </a>
<a class="sourceLine" id="cb6-10" title="10">  <span class="co"># return matrix</span></a>
<a class="sourceLine" id="cb6-11" title="11">  retMat &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="dt">data =</span> <span class="dv">0</span>, <span class="dt">nrow =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(upper) <span class="op">+</span><span class="st"> </span><span class="dv">1</span>, <span class="dt">ncol =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(upper) <span class="op">+</span><span class="st"> </span><span class="dv">1</span>)</a>
<a class="sourceLine" id="cb6-12" title="12">  </a>
<a class="sourceLine" id="cb6-13" title="13">  <span class="co"># set index values for upper/lower triangles</span></a>
<a class="sourceLine" id="cb6-14" title="14">  indx &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(upper)</a>
<a class="sourceLine" id="cb6-15" title="15">  </a>
<a class="sourceLine" id="cb6-16" title="16">  <span class="co"># set forward/backward migration using matrix access</span></a>
<a class="sourceLine" id="cb6-17" title="17">  retMat[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cbind">cbind</a></span>(indx<span class="op">+</span><span class="dv">1</span>,indx)] &lt;-<span class="st"> </span>lower</a>
<a class="sourceLine" id="cb6-18" title="18">  retMat[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cbind">cbind</a></span>(indx,indx<span class="op">+</span><span class="dv">1</span>)] &lt;-<span class="st"> </span>upper</a>
<a class="sourceLine" id="cb6-19" title="19">  </a>
<a class="sourceLine" id="cb6-20" title="20">  <span class="co"># set stay probs</span></a>
<a class="sourceLine" id="cb6-21" title="21">  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/diag">diag</a></span>(<span class="dt">x =</span> retMat) &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">-</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colSums">rowSums</a></span>(<span class="dt">x =</span> retMat)</a>
<a class="sourceLine" id="cb6-22" title="22">  </a>
<a class="sourceLine" id="cb6-23" title="23">  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/function">return</a></span>(retMat)</a>
<a class="sourceLine" id="cb6-24" title="24">}</a>
<a class="sourceLine" id="cb6-25" title="25"></a>
<a class="sourceLine" id="cb6-26" title="26"><span class="co"># fill movement matrix</span></a>
<a class="sourceLine" id="cb6-27" title="27"><span class="co">#  Remember, rows need to sum to 1.</span></a>
<a class="sourceLine" id="cb6-28" title="28">moveMat &lt;-<span class="st"> </span><span class="kw">triDiag</span>(<span class="dt">upper =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep.int</a></span>(<span class="dt">x =</span> <span class="fl">0.05</span>, <span class="dt">times =</span> sitesNumber<span class="dv">-1</span>),</a>
<a class="sourceLine" id="cb6-29" title="29">                   <span class="dt">lower =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep.int</a></span>(<span class="dt">x =</span> <span class="fl">0.05</span>, <span class="dt">times =</span> sitesNumber<span class="dv">-1</span>))</a>
<a class="sourceLine" id="cb6-30" title="30"></a>
<a class="sourceLine" id="cb6-31" title="31">moveMat</a>
<a class="sourceLine" id="cb6-32" title="32"><span class="co">#&gt;      [,1] [,2] [,3] [,4] [,5]</span></a>
<a class="sourceLine" id="cb6-33" title="33"><span class="co">#&gt; [1,] 0.95 0.05 0.00 0.00 0.00</span></a>
<a class="sourceLine" id="cb6-34" title="34"><span class="co">#&gt; [2,] 0.05 0.90 0.05 0.00 0.00</span></a>
<a class="sourceLine" id="cb6-35" title="35"><span class="co">#&gt; [3,] 0.00 0.05 0.90 0.05 0.00</span></a>
<a class="sourceLine" id="cb6-36" title="36"><span class="co">#&gt; [4,] 0.00 0.00 0.05 0.90 0.05</span></a>
<a class="sourceLine" id="cb6-37" title="37"><span class="co">#&gt; [5,] 0.00 0.00 0.00 0.05 0.95</span></a></code></pre></div>
<p>Our landscape will be a line, with each population only connected to the populations on either side. We set a 5% per day migration rate. There are several examples in the other <a href="../doc/mgdrive_examples.html">vignette</a> for setting up different landscapes.</p>
<p>Finally, there is the possibility of large-group movement facilitated by humans (using <em>Aedes</em> for example, this could be boats or water containers in truck beds, etc.). <em>MGDrivE</em> handles this through <code><a href="../reference/basicBatchMigration.html">basicBatchMigration()</a></code>. For simplicity, we are ignoring this possibility (by setting the probability of it to zero).</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" title="1"><span class="co"># batch migration is disabled by setting the probability to 0</span></a>
<a class="sourceLine" id="cb7-2" title="2">batchMigration &lt;-<span class="st"> </span><span class="kw"><a href="../reference/basicBatchMigration.html">basicBatchMigration</a></span>(<span class="dt">batchProbs=</span><span class="dv">0</span>, <span class="dt">numPatches=</span>sitesNumber)</a></code></pre></div>
</div>
<div id="inheritance-pattern" class="section level4">
<h4 class="hasAnchor">
<a href="#inheritance-pattern" class="anchor"></a>Inheritance Pattern</h4>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" title="1"><span class="co">####################</span></a>
<a class="sourceLine" id="cb8-2" title="2"><span class="co"># Basic Inheritance pattern</span></a>
<a class="sourceLine" id="cb8-3" title="3"><span class="co">####################</span></a>
<a class="sourceLine" id="cb8-4" title="4"><span class="co"># Mendelian cube with standard (default) paraameters</span></a>
<a class="sourceLine" id="cb8-5" title="5">cube &lt;-<span class="st"> </span><span class="kw"><a href="../reference/cubeMendelian.html">cubeMendelian</a></span>()</a></code></pre></div>
<p>For this experiment, we use simple Mendelian inheritance. <em>MGDrivE</em> has several other inheritance patterns, which include:</p>
<ul>
<li>
<code><a href="../reference/cubeReciprocalTranslocations.html">cubeReciprocalTranslocations()</a></code>: Reciprocal Translocations<br>
</li>
<li>
<code><a href="../reference/cubeHomingDrive.html">cubeHomingDrive()</a></code>: CRISPR-based homing drive, with two resistance alleles<br>
</li>
<li>
<code><a href="../reference/cubeMEDEA.html">cubeMEDEA()</a></code>: Maternal Effect Dominant Embryonic Arrest</li>
</ul>
<p>While we use the default parameters, there are several fitness costs that can be applied. Every cost is applied in a genotype-specific manner. The optional fitness costs are:</p>
<ul>
<li>
<span class="math inline">\(\eta\)</span>: Male mating fitness<br>
</li>
<li>
<span class="math inline">\(\phi\)</span>: Sex-ratio at emergence<br>
</li>
<li>
<span class="math inline">\(\omega\)</span>: Multiplicative modifier of adult mortality (males and females)<br>
</li>
<li>
<span class="math inline">\(\xi_F\)</span>: Female pupatory success (applies after <span class="math inline">\(\phi\)</span>)<br>
</li>
<li>
<span class="math inline">\(\xi_M\)</span>: Male pupatory success (applies after <span class="math inline">\(\phi\)</span>)<br>
</li>
<li>
<span class="math inline">\(s\)</span>: Fractional increase/decrease in fertility</li>
</ul>
</div>
<div id="releases" class="section level4">
<h4 class="hasAnchor">
<a href="#releases" class="anchor"></a>Releases</h4>
<p><em>MGDrivE</em> uses a list specifying releases of males or females to paramtrize the release schedule. The initial list is <code>NULL</code>, and only if there are releases to be performed is anything added.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" title="1"><span class="co">####################</span></a>
<a class="sourceLine" id="cb9-2" title="2"><span class="co"># Setup releases and batch migration</span></a>
<a class="sourceLine" id="cb9-3" title="3"><span class="co">####################</span></a>
<a class="sourceLine" id="cb9-4" title="4"><span class="co"># set up the empty release vector</span></a>
<a class="sourceLine" id="cb9-5" title="5"><span class="co">#  MGDrivE pulls things out by name</span></a>
<a class="sourceLine" id="cb9-6" title="6">patchReleases &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">replicate</a></span>(<span class="dt">n=</span>sitesNumber,</a>
<a class="sourceLine" id="cb9-7" title="7">                          <span class="dt">expr=</span>{<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">maleReleases=</span><span class="ot">NULL</span>,<span class="dt">femaleReleases=</span><span class="ot">NULL</span>)},</a>
<a class="sourceLine" id="cb9-8" title="8">                          <span class="dt">simplify=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb9-9" title="9"></a>
<a class="sourceLine" id="cb9-10" title="10"><span class="co"># choose release parameters</span></a>
<a class="sourceLine" id="cb9-11" title="11"><span class="co">#  Releases start at time 25, occur every day, for 5 day.</span></a>
<a class="sourceLine" id="cb9-12" title="12"><span class="co">#  There are 50 mosquitoes relaeses every time.</span></a>
<a class="sourceLine" id="cb9-13" title="13">releasesParameters &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">releasesStart=</span><span class="dv">25</span>,</a>
<a class="sourceLine" id="cb9-14" title="14">                          <span class="dt">releasesNumber=</span><span class="dv">10</span>,</a>
<a class="sourceLine" id="cb9-15" title="15">                          <span class="dt">releasesInterval=</span><span class="dv">7</span>,</a>
<a class="sourceLine" id="cb9-16" title="16">                          <span class="dt">releaseProportion=</span><span class="dv">100</span>)</a></code></pre></div>
<p>Here, we are performing 10 releases of 100 individuals. These releases start on day 25 and occur every 7 days. We now generate what this looks like for a male release.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" title="1"><span class="co"># generate male release vector</span></a>
<a class="sourceLine" id="cb10-2" title="2">maleReleasesVector &lt;-<span class="st"> </span><span class="kw"><a href="../reference/generateReleaseVector.html">generateReleaseVector</a></span>(<span class="dt">driveCube=</span>cube, </a>
<a class="sourceLine" id="cb10-3" title="3">                                           <span class="dt">releasesParameters=</span>releasesParameters,</a>
<a class="sourceLine" id="cb10-4" title="4">                                           <span class="dt">sex=</span><span class="st">"M"</span>)</a>
<a class="sourceLine" id="cb10-5" title="5">maleReleasesVector[[<span class="dv">1</span>]]</a>
<a class="sourceLine" id="cb10-6" title="6"><span class="co">#&gt; $nuM</span></a>
<a class="sourceLine" id="cb10-7" title="7"><span class="co">#&gt;  AA  Aa  aa </span></a>
<a class="sourceLine" id="cb10-8" title="8"><span class="co">#&gt;   0   0 100 </span></a>
<a class="sourceLine" id="cb10-9" title="9"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb10-10" title="10"><span class="co">#&gt; $tRelease</span></a>
<a class="sourceLine" id="cb10-11" title="11"><span class="co">#&gt; [1] 25</span></a></code></pre></div>
<p>The release vector generated is 10 elements long, because we have 10 releases, and each element contains the number and genotype of individuals to release and the time of that release. We now generate the same thing for female releases, and insert the releases into our release list so that releases are only performed in the first patch.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" title="1"><span class="co"># generate female release vector</span></a>
<a class="sourceLine" id="cb11-2" title="2">femaleReleasesVector &lt;-<span class="st"> </span><span class="kw"><a href="../reference/generateReleaseVector.html">generateReleaseVector</a></span>(<span class="dt">driveCube=</span>cube,</a>
<a class="sourceLine" id="cb11-3" title="3">                                             <span class="dt">releasesParameters=</span>releasesParameters,</a>
<a class="sourceLine" id="cb11-4" title="4">                                             <span class="dt">sex=</span><span class="st">"F"</span>)</a>
<a class="sourceLine" id="cb11-5" title="5"></a>
<a class="sourceLine" id="cb11-6" title="6"><span class="co"># put releases into the proper place in the release list</span></a>
<a class="sourceLine" id="cb11-7" title="7">patchReleases[[<span class="dv">1</span>]]<span class="op">$</span>maleReleases &lt;-<span class="st"> </span>maleReleasesVector</a>
<a class="sourceLine" id="cb11-8" title="8">patchReleases[[<span class="dv">1</span>]]<span class="op">$</span>femaleReleases &lt;-<span class="st"> </span>femaleReleasesVector</a></code></pre></div>
</div>
<div id="setup-and-run" class="section level4">
<h4 class="hasAnchor">
<a href="#setup-and-run" class="anchor"></a>Setup and Run</h4>
<p>The final steps before running our simulation! All of the parameters we have setup are combined using <code><a href="../reference/parameterizeMGDrivE.html">parameterizeMGDrivE()</a></code>. Additionally, this function calculates equilibrium values for aquatic populations, death rates, and density dependence for every population. See the documentation for all of the possible options.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" title="1"><span class="co">####################</span></a>
<a class="sourceLine" id="cb12-2" title="2"><span class="co"># Combine parameters and run!</span></a>
<a class="sourceLine" id="cb12-3" title="3"><span class="co">####################</span></a>
<a class="sourceLine" id="cb12-4" title="4"><span class="co"># setup parameters for the network. This builds a list of parameters required for</span></a>
<a class="sourceLine" id="cb12-5" title="5"><span class="co">#  every population in the network.</span></a>
<a class="sourceLine" id="cb12-6" title="6">netPar &lt;-<span class="st"> </span><span class="kw"><a href="../reference/parameterizeMGDrivE.html">parameterizeMGDrivE</a></span>(<span class="dt">runID=</span><span class="dv">1</span>, <span class="dt">simTime=</span>tMax, <span class="dt">nPatch=</span>sitesNumber,</a>
<a class="sourceLine" id="cb12-7" title="7">                             <span class="dt">beta=</span>bioParameters<span class="op">$</span>betaK, <span class="dt">muAd=</span>bioParameters<span class="op">$</span>muAd,</a>
<a class="sourceLine" id="cb12-8" title="8">                             <span class="dt">popGrowth=</span>bioParameters<span class="op">$</span>popGrowth, <span class="dt">tEgg=</span>bioParameters<span class="op">$</span>tEgg,</a>
<a class="sourceLine" id="cb12-9" title="9">                             <span class="dt">tLarva=</span>bioParameters<span class="op">$</span>tLarva, <span class="dt">tPupa=</span>bioParameters<span class="op">$</span>tPupa,</a>
<a class="sourceLine" id="cb12-10" title="10">                             <span class="dt">AdPopEQ=</span>patchPops)</a></code></pre></div>
<p><em>MGDrivE</em> can run in either a deterministic or stochastic setting, so we set it to stochastic for these simulations.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" title="1"><span class="co"># set MGDrivE to run stochastic</span></a>
<a class="sourceLine" id="cb13-2" title="2"><span class="kw"><a href="../reference/setupMGDrivE.html">setupMGDrivE</a></span>(<span class="dt">stochasticityON =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb13-3" title="3"><span class="co">#&gt; initializing MGDrivE</span></a></code></pre></div>
<p>Finaly, we setup the complete network to run our simulation.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" title="1"><span class="co"># build network prior to run</span></a>
<a class="sourceLine" id="cb14-2" title="2">MGDrivESim &lt;-<span class="st"> </span>Network<span class="op">$</span><span class="kw">new</span>(<span class="dt">params=</span>netPar,</a>
<a class="sourceLine" id="cb14-3" title="3">                         <span class="dt">driveCube=</span>cube,</a>
<a class="sourceLine" id="cb14-4" title="4">                         <span class="dt">patchReleases=</span>patchReleases,</a>
<a class="sourceLine" id="cb14-5" title="5">                         <span class="dt">migrationMale=</span>moveMat,</a>
<a class="sourceLine" id="cb14-6" title="6">                         <span class="dt">migrationFemale=</span>moveMat,</a>
<a class="sourceLine" id="cb14-7" title="7">                         <span class="dt">migrationBatch=</span>batchMigration,</a>
<a class="sourceLine" id="cb14-8" title="8">                         <span class="dt">directory=</span>folderNames)</a>
<a class="sourceLine" id="cb14-9" title="9"><span class="co">#&gt; initializing patch:  1  of  5 </span></a>
<a class="sourceLine" id="cb14-10" title="10"><span class="co">#&gt; initializing patch:  2  of  5 </span></a>
<a class="sourceLine" id="cb14-11" title="11"><span class="co">#&gt; initializing patch:  3  of  5 </span></a>
<a class="sourceLine" id="cb14-12" title="12"><span class="co">#&gt; initializing patch:  4  of  5 </span></a>
<a class="sourceLine" id="cb14-13" title="13"><span class="co">#&gt; initializing patch:  5  of  5</span></a>
<a class="sourceLine" id="cb14-14" title="14"></a>
<a class="sourceLine" id="cb14-15" title="15"><span class="co"># list folders to show that they have been created</span></a>
<a class="sourceLine" id="cb14-16" title="16"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list.files">list.files</a></span>(<span class="dt">path =</span> outFolder)</a>
<a class="sourceLine" id="cb14-17" title="17"><span class="co">#&gt; [1] "001" "002" "003" "004" "005"</span></a></code></pre></div>
<p>The <code>Network</code> object represents the metapopulation of <em>MGDrivE</em>. It can be setup to hold 1 patch, and therefore be a single population, or several patches, and be a true metapopulation. Documentation can be found by running <code><a href="../reference/Network.html">?Network</a></code>. The different notation is an attribute of the <code>R6</code> structure that <em>MGDrivE</em> is built on (Shoutout to <a href="https://github.com/r-lib/R6">R6</a>). The <code>params</code> argument takes the network parameters created by <code><a href="../reference/parameterizeMGDrivE.html">parameterizeMGDrivE()</a></code>. The inheritance pattern is supplied in the <code>driveCube</code> option, which we chose to be Mendelian. Migration can be different for males and females, so even though here we use the same migration matrix, users can supply different ones if desired. The <code>migrationBatch</code> option must always be filled in, but can be set to zero, as done here, if it isn’t desired or for deterministic simulations (where it doesn’t make sense). Finally, the <code>directory</code> parameter takes the vector of folder names, and creates them for use during the simulation.</p>
<p>Finally, <em>MGDrivE</em> can be run once using <code>MGDrivESim$oneRun()</code> or several times using <code>MGDrivESim$multRun()</code>. The benefit of using the built-in multiple repetition function is that memory is not released between runs, just cleared, so run-time is reduced. Additionally, each of these functions can produce a progress bar, which is suppressed in all of our vignettes.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" title="1"><span class="co"># run simulation</span></a>
<a class="sourceLine" id="cb15-2" title="2">MGDrivESim<span class="op">$</span><span class="kw">multRun</span>(<span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
</div>
<div id="post-processing" class="section level4">
<h4 class="hasAnchor">
<a href="#post-processing" class="anchor"></a>Post Processing</h4>
<p><em>MGDrivE</em> outputs 2 *.csv files every simulation. These are stored as <code>{F|M}_RunXXX.csv</code> for female and male adults. In them are the population counts for every genotype at every day.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" title="1"><span class="co"># first and last repetitions</span></a>
<a class="sourceLine" id="cb16-2" title="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list.files">list.files</a></span>(<span class="dt">path =</span> outFolder)[<span class="dv">1</span>]</a>
<a class="sourceLine" id="cb16-3" title="3"><span class="co">#&gt; [1] "001"</span></a>
<a class="sourceLine" id="cb16-4" title="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list.files">list.files</a></span>(<span class="dt">path =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list.files">list.files</a></span>(<span class="dt">path =</span> outFolder, <span class="dt">full.names =</span> <span class="ot">TRUE</span>)[<span class="dv">1</span>], <span class="dt">recursive =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb16-5" title="5"><span class="co">#&gt; [1] "F_Run001.csv" "M_Run001.csv"</span></a>
<a class="sourceLine" id="cb16-6" title="6"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list.files">list.files</a></span>(<span class="dt">path =</span> outFolder)[<span class="dv">5</span>]</a>
<a class="sourceLine" id="cb16-7" title="7"><span class="co">#&gt; [1] "005"</span></a>
<a class="sourceLine" id="cb16-8" title="8"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list.files">list.files</a></span>(<span class="dt">path =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list.files">list.files</a></span>(<span class="dt">path =</span> outFolder, <span class="dt">full.names =</span> <span class="ot">TRUE</span>)[<span class="dv">5</span>], <span class="dt">recursive =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb16-9" title="9"><span class="co">#&gt; [1] "F_Run005.csv" "M_Run005.csv"</span></a></code></pre></div>
<p>This is the raw output from <em>MGDrivE</em>. The files contain all of the information from the simulation.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" title="1"><span class="co"># read in male and female files</span></a>
<a class="sourceLine" id="cb17-2" title="2">mMat &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">as.matrix</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/read.table">read.csv</a></span>(<span class="dt">file =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list.files">list.files</a></span>(<span class="dt">path =</span> outFolder, <span class="dt">full.names =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb17-3" title="3">                                             <span class="dt">recursive =</span> <span class="ot">TRUE</span>)[<span class="dv">2</span>],</a>
<a class="sourceLine" id="cb17-4" title="4">                           <span class="dt">header =</span> <span class="ot">TRUE</span>, <span class="dt">sep =</span> <span class="st">","</span>))</a>
<a class="sourceLine" id="cb17-5" title="5">fMat &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">as.matrix</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/read.table">read.csv</a></span>(<span class="dt">file =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list.files">list.files</a></span>(<span class="dt">path =</span> outFolder, <span class="dt">full.names =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb17-6" title="6">                                             <span class="dt">recursive =</span> <span class="ot">TRUE</span>)[<span class="dv">1</span>],</a>
<a class="sourceLine" id="cb17-7" title="7">                           <span class="dt">header =</span> <span class="ot">TRUE</span>, <span class="dt">sep =</span> <span class="st">","</span>))</a>
<a class="sourceLine" id="cb17-8" title="8"></a>
<a class="sourceLine" id="cb17-9" title="9"><span class="co"># look at male file header</span></a>
<a class="sourceLine" id="cb17-10" title="10"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames</a></span>(mMat)</a>
<a class="sourceLine" id="cb17-11" title="11"><span class="co">#&gt; [1] "Time"  "Patch" "AA"    "Aa"    "aa"</span></a></code></pre></div>
<p>The header for male files (<code>M_RunXXX.csv</code>) is shown above. They always containst the time, patch, and every possible genotype in the simulation. For the Mendelian inheritance here, there are three possible genotypes, AA, Aa, aa. The Time and Patch columns will always be the same, but the number of genotypes changes depending on the inheritance pattern used. For example, CRISPR-based homing inheritance has 10, so there would be 12 columns in the male files. Below, we see how each patch is designated in the file.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" title="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(<span class="dt">x =</span> mMat, <span class="dt">n =</span> <span class="dv">2</span><span class="op">*</span>sitesNumber)</a>
<a class="sourceLine" id="cb18-2" title="2"><span class="co">#&gt;       Time Patch  AA Aa aa</span></a>
<a class="sourceLine" id="cb18-3" title="3"><span class="co">#&gt;  [1,]    1     1 250  0  0</span></a>
<a class="sourceLine" id="cb18-4" title="4"><span class="co">#&gt;  [2,]    1     2 250  0  0</span></a>
<a class="sourceLine" id="cb18-5" title="5"><span class="co">#&gt;  [3,]    1     3 250  0  0</span></a>
<a class="sourceLine" id="cb18-6" title="6"><span class="co">#&gt;  [4,]    1     4 250  0  0</span></a>
<a class="sourceLine" id="cb18-7" title="7"><span class="co">#&gt;  [5,]    1     5 250  0  0</span></a>
<a class="sourceLine" id="cb18-8" title="8"><span class="co">#&gt;  [6,]    2     1 241  0  0</span></a>
<a class="sourceLine" id="cb18-9" title="9"><span class="co">#&gt;  [7,]    2     2 243  0  0</span></a>
<a class="sourceLine" id="cb18-10" title="10"><span class="co">#&gt;  [8,]    2     3 249  0  0</span></a>
<a class="sourceLine" id="cb18-11" title="11"><span class="co">#&gt;  [9,]    2     4 268  0  0</span></a>
<a class="sourceLine" id="cb18-12" title="12"><span class="co">#&gt; [10,]    2     5 249  0  0</span></a></code></pre></div>
<p>Notice how each patch is printed for every time point, then the time increases and every patch is printed again. Time = 0 is at equilibrium, where we start the simulations, then the patches begin to diverge, staying around equilibrium but no longer being exactly the same.</p>
<p>The female output files (<code>F_RunXXX.csv</code>) are slightly more complicated. Since <em>MGDrivE</em> keeps track of females and their mates, the output maintains that as well. Thus, the header in the female files will be <span class="math inline">\(2 + \text{numGenotypes}^2\)</span> long. For Mendelian inheritance, this looks like:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" title="1"><span class="co"># look at female file header</span></a>
<a class="sourceLine" id="cb19-2" title="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames</a></span>(fMat)</a>
<a class="sourceLine" id="cb19-3" title="3"><span class="co">#&gt;  [1] "Time"  "Patch" "AAAA"  "AAAa"  "AAaa"  "AaAA"  "AaAa"  "Aaaa" </span></a>
<a class="sourceLine" id="cb19-4" title="4"><span class="co">#&gt;  [9] "aaAA"  "aaAa"  "aaaa"</span></a></code></pre></div>
<p>The Time and Patch columns are the same as in male files, but the “genotypes” are now a combination of the female genotype and her mate’s genotype. The length of the header is 11, consistent with having 3 genotypes, times the number of mate genotypes, plus the Time and Patch columns. The header “genotypes” are printed female genotype first, male (mate) second. So, the <em>aaAA</em> column represents females that are <em>aa</em> mated to males that are <em>AA</em>.<br>
The Time/Patch pattern is the same as in the male files.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" title="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(<span class="dt">x =</span> fMat, <span class="dt">n =</span> <span class="dv">2</span><span class="op">*</span>sitesNumber)</a>
<a class="sourceLine" id="cb20-2" title="2"><span class="co">#&gt;       Time Patch AAAA AAAa AAaa AaAA AaAa Aaaa aaAA aaAa aaaa</span></a>
<a class="sourceLine" id="cb20-3" title="3"><span class="co">#&gt;  [1,]    1     1  250    0    0    0    0    0    0    0    0</span></a>
<a class="sourceLine" id="cb20-4" title="4"><span class="co">#&gt;  [2,]    1     2  250    0    0    0    0    0    0    0    0</span></a>
<a class="sourceLine" id="cb20-5" title="5"><span class="co">#&gt;  [3,]    1     3  250    0    0    0    0    0    0    0    0</span></a>
<a class="sourceLine" id="cb20-6" title="6"><span class="co">#&gt;  [4,]    1     4  250    0    0    0    0    0    0    0    0</span></a>
<a class="sourceLine" id="cb20-7" title="7"><span class="co">#&gt;  [5,]    1     5  250    0    0    0    0    0    0    0    0</span></a>
<a class="sourceLine" id="cb20-8" title="8"><span class="co">#&gt;  [6,]    2     1  246    0    0    0    0    0    0    0    0</span></a>
<a class="sourceLine" id="cb20-9" title="9"><span class="co">#&gt;  [7,]    2     2  251    0    0    0    0    0    0    0    0</span></a>
<a class="sourceLine" id="cb20-10" title="10"><span class="co">#&gt;  [8,]    2     3  254    0    0    0    0    0    0    0    0</span></a>
<a class="sourceLine" id="cb20-11" title="11"><span class="co">#&gt;  [9,]    2     4  258    0    0    0    0    0    0    0    0</span></a>
<a class="sourceLine" id="cb20-12" title="12"><span class="co">#&gt; [10,]    2     5  248    0    0    0    0    0    0    0    0</span></a></code></pre></div>
<p>There are several things that we can do from here. We can work with the raw output directly, but this gets difficult as the simulations get larger. Therefor, we have provided functions to split the raw output by patches, generating a new file for every patch.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" title="1"><span class="co"># First, split output by patch</span></a>
<a class="sourceLine" id="cb21-2" title="2"><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>nRep){</a>
<a class="sourceLine" id="cb21-3" title="3"> <span class="kw"><a href="../reference/splitOutput.html">splitOutput</a></span>(<span class="dt">readDir=</span>folderNames[i],<span class="dt">remFile =</span> <span class="ot">TRUE</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb21-4" title="4">}</a></code></pre></div>
<p><code>splitOutput</code> splits the raw output by patch, generating one *.csv for every patch in our simulation. We loop over each repetition folder, removing the original files, and replacing them with individual patch files. Inside each repetition folder, we now see the files:</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" title="1"><span class="co"># first and last repetitions</span></a>
<a class="sourceLine" id="cb22-2" title="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list.files">list.files</a></span>(<span class="dt">path =</span> outFolder)[<span class="dv">1</span>]</a>
<a class="sourceLine" id="cb22-3" title="3"><span class="co">#&gt; [1] "001"</span></a>
<a class="sourceLine" id="cb22-4" title="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list.files">list.files</a></span>(<span class="dt">path =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list.files">list.files</a></span>(<span class="dt">path =</span> outFolder, <span class="dt">full.names =</span> <span class="ot">TRUE</span>)[<span class="dv">1</span>], <span class="dt">recursive =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb22-5" title="5"><span class="co">#&gt;  [1] "F_Run001_Patch001.csv" "F_Run001_Patch002.csv"</span></a>
<a class="sourceLine" id="cb22-6" title="6"><span class="co">#&gt;  [3] "F_Run001_Patch003.csv" "F_Run001_Patch004.csv"</span></a>
<a class="sourceLine" id="cb22-7" title="7"><span class="co">#&gt;  [5] "F_Run001_Patch005.csv" "M_Run001_Patch001.csv"</span></a>
<a class="sourceLine" id="cb22-8" title="8"><span class="co">#&gt;  [7] "M_Run001_Patch002.csv" "M_Run001_Patch003.csv"</span></a>
<a class="sourceLine" id="cb22-9" title="9"><span class="co">#&gt;  [9] "M_Run001_Patch004.csv" "M_Run001_Patch005.csv"</span></a>
<a class="sourceLine" id="cb22-10" title="10"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list.files">list.files</a></span>(<span class="dt">path =</span> outFolder)[<span class="dv">5</span>]</a>
<a class="sourceLine" id="cb22-11" title="11"><span class="co">#&gt; [1] "005"</span></a>
<a class="sourceLine" id="cb22-12" title="12"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list.files">list.files</a></span>(<span class="dt">path =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list.files">list.files</a></span>(<span class="dt">path =</span> outFolder, <span class="dt">full.names =</span> <span class="ot">TRUE</span>)[<span class="dv">5</span>], <span class="dt">recursive =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb22-13" title="13"><span class="co">#&gt;  [1] "F_Run005_Patch001.csv" "F_Run005_Patch002.csv"</span></a>
<a class="sourceLine" id="cb22-14" title="14"><span class="co">#&gt;  [3] "F_Run005_Patch003.csv" "F_Run005_Patch004.csv"</span></a>
<a class="sourceLine" id="cb22-15" title="15"><span class="co">#&gt;  [5] "F_Run005_Patch005.csv" "M_Run005_Patch001.csv"</span></a>
<a class="sourceLine" id="cb22-16" title="16"><span class="co">#&gt;  [7] "M_Run005_Patch002.csv" "M_Run005_Patch003.csv"</span></a>
<a class="sourceLine" id="cb22-17" title="17"><span class="co">#&gt;  [9] "M_Run005_Patch004.csv" "M_Run005_Patch005.csv"</span></a>
<a class="sourceLine" id="cb22-18" title="18"></a>
<a class="sourceLine" id="cb22-19" title="19"><span class="co"># read in examples of new male/female files</span></a>
<a class="sourceLine" id="cb22-20" title="20">twoFiles &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list.files">list.files</a></span>(<span class="dt">path =</span> outFolder, <span class="dt">full.names =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb22-21" title="21">                       <span class="dt">recursive =</span> <span class="ot">TRUE</span>)[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">1</span>,<span class="dv">6</span>)]</a>
<a class="sourceLine" id="cb22-22" title="22"></a>
<a class="sourceLine" id="cb22-23" title="23"><span class="co"># read in male and female files</span></a>
<a class="sourceLine" id="cb22-24" title="24">mMat &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">as.matrix</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/read.table">read.csv</a></span>(<span class="dt">file =</span> twoFiles[<span class="dv">2</span>], <span class="dt">header =</span> <span class="ot">TRUE</span>, <span class="dt">sep =</span> <span class="st">","</span>))</a>
<a class="sourceLine" id="cb22-25" title="25">fMat &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">as.matrix</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/read.table">read.csv</a></span>(<span class="dt">file =</span> twoFiles[<span class="dv">1</span>], <span class="dt">header =</span> <span class="ot">TRUE</span>, <span class="dt">sep =</span> <span class="st">","</span>))</a></code></pre></div>
<p>We again look at a male file, examining the header and the first few rows.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" title="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(<span class="dt">x =</span> mMat, <span class="dt">n =</span> <span class="dv">5</span>)</a>
<a class="sourceLine" id="cb23-2" title="2"><span class="co">#&gt;      Time  AA Aa aa</span></a>
<a class="sourceLine" id="cb23-3" title="3"><span class="co">#&gt; [1,]    1 250  0  0</span></a>
<a class="sourceLine" id="cb23-4" title="4"><span class="co">#&gt; [2,]    2 241  0  0</span></a>
<a class="sourceLine" id="cb23-5" title="5"><span class="co">#&gt; [3,]    3 250  0  0</span></a>
<a class="sourceLine" id="cb23-6" title="6"><span class="co">#&gt; [4,]    4 233  0  0</span></a>
<a class="sourceLine" id="cb23-7" title="7"><span class="co">#&gt; [5,]    5 241  0  0</span></a></code></pre></div>
<p>The Patch column has been removed, it is now part of the file name, and the Time column increases every row. This is the first repetition, first male file.</p>
<p>Female files are still a little more complicated. They have been split by patch, but still retain the mate information.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" title="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(<span class="dt">x =</span> fMat, <span class="dt">n =</span> <span class="dv">5</span>)</a>
<a class="sourceLine" id="cb24-2" title="2"><span class="co">#&gt;      Time AAAA AAAa AAaa AaAA AaAa Aaaa aaAA aaAa aaaa</span></a>
<a class="sourceLine" id="cb24-3" title="3"><span class="co">#&gt; [1,]    1  250    0    0    0    0    0    0    0    0</span></a>
<a class="sourceLine" id="cb24-4" title="4"><span class="co">#&gt; [2,]    2  246    0    0    0    0    0    0    0    0</span></a>
<a class="sourceLine" id="cb24-5" title="5"><span class="co">#&gt; [3,]    3  253    0    0    0    0    0    0    0    0</span></a>
<a class="sourceLine" id="cb24-6" title="6"><span class="co">#&gt; [4,]    4  235    0    0    0    0    0    0    0    0</span></a>
<a class="sourceLine" id="cb24-7" title="7"><span class="co">#&gt; [5,]    5  223    0    0    0    0    0    0    0    0</span></a></code></pre></div>
<p>The Patch column is gone, the Time column counts as expected, and we still have all of the mate information.</p>
<p>In the event that we don’t want/need the mate information, we can remove it using the <code><a href="../reference/aggregateFemales.html">aggregateFemales()</a></code> function.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb25-1" title="1"><span class="co"># Second, aggregate females by their mate choice</span></a>
<a class="sourceLine" id="cb25-2" title="2"><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>nRep){</a>
<a class="sourceLine" id="cb25-3" title="3"> <span class="kw"><a href="../reference/aggregateFemales.html">aggregateFemales</a></span>(<span class="dt">readDir=</span>folderNames[i],<span class="dt">genotypes=</span>cube<span class="op">$</span>genotypesID,</a>
<a class="sourceLine" id="cb25-4" title="4">                  <span class="dt">remFile=</span><span class="ot">TRUE</span>,<span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb25-5" title="5">}</a></code></pre></div>
<p>This collects all of the females for each genotype, summing over the possible mates.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb26-1" title="1"><span class="co"># changed female files in first repetition</span></a>
<a class="sourceLine" id="cb26-2" title="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list.files">list.files</a></span>(<span class="dt">path =</span> outFolder)[<span class="dv">1</span>]</a>
<a class="sourceLine" id="cb26-3" title="3"><span class="co">#&gt; [1] "001"</span></a>
<a class="sourceLine" id="cb26-4" title="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list.files">list.files</a></span>(<span class="dt">path =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list.files">list.files</a></span>(<span class="dt">path =</span> outFolder, <span class="dt">full.names =</span> <span class="ot">TRUE</span>)[<span class="dv">1</span>], <span class="dt">recursive =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb26-5" title="5"><span class="co">#&gt;  [1] "F_Aggregate_Run001_Patch001.csv" "F_Aggregate_Run001_Patch002.csv"</span></a>
<a class="sourceLine" id="cb26-6" title="6"><span class="co">#&gt;  [3] "F_Aggregate_Run001_Patch003.csv" "F_Aggregate_Run001_Patch004.csv"</span></a>
<a class="sourceLine" id="cb26-7" title="7"><span class="co">#&gt;  [5] "F_Aggregate_Run001_Patch005.csv" "M_Run001_Patch001.csv"          </span></a>
<a class="sourceLine" id="cb26-8" title="8"><span class="co">#&gt;  [7] "M_Run001_Patch002.csv"           "M_Run001_Patch003.csv"          </span></a>
<a class="sourceLine" id="cb26-9" title="9"><span class="co">#&gt;  [9] "M_Run001_Patch004.csv"           "M_Run001_Patch005.csv"</span></a>
<a class="sourceLine" id="cb26-10" title="10"></a>
<a class="sourceLine" id="cb26-11" title="11"><span class="co"># read in examples of new female file</span></a>
<a class="sourceLine" id="cb26-12" title="12">fMat2 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">as.matrix</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/read.table">read.csv</a></span>(<span class="dt">file =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list.files">list.files</a></span>(<span class="dt">path =</span> outFolder,</a>
<a class="sourceLine" id="cb26-13" title="13">                                              <span class="dt">recursive =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb26-14" title="14">                                              <span class="dt">full.names =</span> <span class="ot">TRUE</span>)[<span class="dv">1</span>],</a>
<a class="sourceLine" id="cb26-15" title="15">                            <span class="dt">header =</span> <span class="ot">TRUE</span>, <span class="dt">sep =</span> <span class="st">","</span>))</a></code></pre></div>
<p>We see that all of the female files have been updated to denote that they are aggregated over their mate genotypes. Internally, the files now resemble the male files.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb27-1" title="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(<span class="dt">x =</span> fMat2, <span class="dt">n =</span> <span class="dv">5</span>)</a>
<a class="sourceLine" id="cb27-2" title="2"><span class="co">#&gt;      Time  AA Aa aa</span></a>
<a class="sourceLine" id="cb27-3" title="3"><span class="co">#&gt; [1,]    1 250  0  0</span></a>
<a class="sourceLine" id="cb27-4" title="4"><span class="co">#&gt; [2,]    2 246  0  0</span></a>
<a class="sourceLine" id="cb27-5" title="5"><span class="co">#&gt; [3,]    3 253  0  0</span></a>
<a class="sourceLine" id="cb27-6" title="6"><span class="co">#&gt; [4,]    4 235  0  0</span></a>
<a class="sourceLine" id="cb27-7" title="7"><span class="co">#&gt; [5,]    5 223  0  0</span></a></code></pre></div>
<p>This doesn’t show much change, but if we compare the female file split by patch to the aggregated female file at the time of the release, we can see that the genotypes have been combined.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb28-1" title="1"><span class="co"># show non-aggregated female file split by patch</span></a>
<a class="sourceLine" id="cb28-2" title="2"><span class="co">#  This is for patch number 1</span></a>
<a class="sourceLine" id="cb28-3" title="3">fMat[(releasesParameters<span class="op">$</span>releasesStart<span class="dv">-2</span>)<span class="op">:</span>(releasesParameters<span class="op">$</span>releasesStart<span class="op">+</span><span class="dv">2</span>), ]</a>
<a class="sourceLine" id="cb28-4" title="4"><span class="co">#&gt;      Time AAAA AAAa AAaa AaAA AaAa Aaaa aaAA aaAa aaaa</span></a>
<a class="sourceLine" id="cb28-5" title="5"><span class="co">#&gt; [1,]   23  266    0    0    0    0    0    0    0    0</span></a>
<a class="sourceLine" id="cb28-6" title="6"><span class="co">#&gt; [2,]   24  263    0    0    0    0    0    0    0    0</span></a>
<a class="sourceLine" id="cb28-7" title="7"><span class="co">#&gt; [3,]   25  263    0    5    0    0    0   64    0   31</span></a>
<a class="sourceLine" id="cb28-8" title="8"><span class="co">#&gt; [4,]   26  265    0    5    0    0    0   58    0   27</span></a>
<a class="sourceLine" id="cb28-9" title="9"><span class="co">#&gt; [5,]   27  254    0    6    0    0    0   46    0   23</span></a>
<a class="sourceLine" id="cb28-10" title="10"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cat">cat</a></span>(<span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</a>
<a class="sourceLine" id="cb28-11" title="11"></a>
<a class="sourceLine" id="cb28-12" title="12"><span class="co"># show aggregated female file</span></a>
<a class="sourceLine" id="cb28-13" title="13"><span class="co">#  This is for patch number 1</span></a>
<a class="sourceLine" id="cb28-14" title="14">fMat2[(releasesParameters<span class="op">$</span>releasesStart<span class="dv">-2</span>)<span class="op">:</span>(releasesParameters<span class="op">$</span>releasesStart<span class="op">+</span><span class="dv">2</span>), ]</a>
<a class="sourceLine" id="cb28-15" title="15"><span class="co">#&gt;      Time  AA Aa aa</span></a>
<a class="sourceLine" id="cb28-16" title="16"><span class="co">#&gt; [1,]   23 266  0  0</span></a>
<a class="sourceLine" id="cb28-17" title="17"><span class="co">#&gt; [2,]   24 263  0  0</span></a>
<a class="sourceLine" id="cb28-18" title="18"><span class="co">#&gt; [3,]   25 268  0 95</span></a>
<a class="sourceLine" id="cb28-19" title="19"><span class="co">#&gt; [4,]   26 270  0 85</span></a>
<a class="sourceLine" id="cb28-20" title="20"><span class="co">#&gt; [5,]   27 260  0 69</span></a></code></pre></div>
<p>Looking closely, we see the “AAAA” column combined wiht the “AAaa” column in the first output sum to the value in the “AA” column in the second.<br>
The same goes for the “aaAA” and “aaaa” columns in the first and the “aa” column in the second.</p>
</div>
<div id="plotting" class="section level4">
<h4 class="hasAnchor">
<a href="#plotting" class="anchor"></a>Plotting</h4>
<p><em>MGDrivE</em> comes with two built-in plot functions, <code><a href="../reference/plotMGDrivESingle.html">plotMGDrivESingle()</a></code> and <code><a href="../reference/plotMGDrivEMult.html">plotMGDrivEMult()</a></code>.<br><code><a href="../reference/plotMGDrivESingle.html">plotMGDrivESingle()</a></code> takes the name of one repetition folder, and plots all of the patches in it for both males and females.</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb29-1" title="1"><span class="co"># plot the first repetition</span></a>
<a class="sourceLine" id="cb29-2" title="2"><span class="kw"><a href="../reference/plotMGDrivESingle.html">plotMGDrivESingle</a></span>(<span class="dt">readDir=</span>folderNames[<span class="dv">1</span>],<span class="dt">totalPop =</span> <span class="ot">TRUE</span>,<span class="dt">lwd=</span><span class="fl">3.5</span>,<span class="dt">alpha=</span><span class="dv">1</span>)</a></code></pre></div>
<p><img src="mgdrive_run_files/figure-html/unnamed-chunk-31-1.png" width="672"></p>
<p>The left Y-axis is population size, the right Y-axis is patch number, and the X-axis is time. We can see the releases in the first patch, equal for males and females, and follow their migration down the plots from patch 1 to patch 5. The total population remains around 250 individuals for males and females, which gives the expected equilibrium value of 500.<br>
We can also plot all of our repetitions using <code><a href="../reference/plotMGDrivEMult.html">plotMGDrivEMult()</a></code>.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb30-1" title="1"><span class="co"># plot all 5 repetitions together</span></a>
<a class="sourceLine" id="cb30-2" title="2"><span class="kw"><a href="../reference/plotMGDrivEMult.html">plotMGDrivEMult</a></span>(<span class="dt">readDir=</span>outFolder,<span class="dt">lwd=</span><span class="fl">0.35</span>,<span class="dt">alpha=</span><span class="fl">0.75</span>)</a></code></pre></div>
<p><img src="mgdrive_run_files/figure-html/unnamed-chunk-32-1.png" width="672"></p>
<p><code><a href="../reference/plotMGDrivEMult.html">plotMGDrivEMult()</a></code> takes the name of the main directory, and finds all of the repetitions inside of it. It then plots males and females by patch, the same as <code><a href="../reference/plotMGDrivESingle.html">plotMGDrivESingle()</a></code>, but overlays a trace of every repetition for each patch.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#mgdrive-one-run">MGDrivE, One Run</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="https://chipdelmal.github.io">Héctor Manuel Sánchez Castellanos</a>, <a href="https://github.com/GilChrist19">Jared Bennett</a>, <a href="https://slwu89.github.io">Sean Wu</a>, <a href="https://www.marshalllab.com">John M. Marshall</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
